// @ebay/locale-formatter-ebay/lib/formats/currencyFormat.js
"use strict";

var formatsReader = require('../reader/formatsReader'),
	currPatterns = require('../formatUtil').currencyPatterns,
	Cache = require('lru-backup-cache'),
	currFormCache = new Cache({
		"enabled": true,
        "max": 20000,          //Cache size
        "maxAge": 1800
	});

function CurrencyFormat (locale, currencyCode, options) {
	locale = locale && locale.replace('-', '_');
	this._locale = locale;
	this._currencyCode = currencyCode;
	this._formats = formatsReader.load(locale, options);
	this._positiveCurrencyPattern = this._formats && this._formats.getPositiveCurrencyPattern(currencyCode);
	this._negativeCurrencyPattern = this._formats && this._formats.getNegativeCurrencyPattern(currencyCode);
	//Find the Currency Pattern Matcher Based on Positive and Negative Currency Patterns
	this._currencyPattern = this._findPattern(this._positiveCurrencyPattern, this._negativeCurrencyPattern);
	//Currency Symbols
	this._standardCurrencySymbol = this._formats && this._formats.getStandardCurrencySymbol(currencyCode);
	this._simpleCurrencySymbol = this._formats && this._formats.getSimpleCurrencySymbol(currencyCode);
	this._worldCurrencySymbol = this._formats && this._formats.getWorldCurrencySymbol(currencyCode);
	this._fancyCurrencySymbol = this._formats && this._formats.getFancyCurrencySymbol(currencyCode);
	this._numCurrencyDecPoints = this._formats && this._formats.getNumCurrencyDecPoints(currencyCode);
	this._maxCurrencyVal = this._formats && this._formats.getMaxCurrencyVal(currencyCode);
	// this._currencySingularName = this._formats && this._formats.getCurrencySingularName(currencyCode);
	// this._currencyPluralName = this._formats && this._formats.getCurrencyPluralName(currencyCode);
}

CurrencyFormat.prototype = {
	_findPattern : function (pos, neg) {
		if (pos === "sn" && neg === "s-n") {
			return currPatterns.sN;
        }
	    if (pos === "s n" && neg === "s -n") {
			return currPatterns.seN;
		}
		if (pos === "sn" && neg === "-sn") {
			return currPatterns.Sn;
		}
		if (pos === "s n" && neg === "-s n") {
			return currPatterns.Sen;
		}
		if (pos === "ns" && neg === "-ns") {
			return currPatterns.Ns;
		}
		if (pos === "n s" && neg === "-n s") {
			return currPatterns.Nes;
		}
		return currPatterns.Sn;
	},
	getLocale : function () {
		return this._locale;
	},
	getCurrencyCode : function () {
		return this._currencyCode;
	},
	getPositiveCurrencyPattern : function () {
		return this._positiveCurrencyPattern;
	},
	getNegativeCurrencyPattern : function () {
		return this._negativeCurrencyPattern;
	},
	getStandardCurrencySymbol : function () {
		return this._standardCurrencySymbol;
	},
	getSimpleCurrencySymbol : function () {
		return this._simpleCurrencySymbol;
	},
	getWorldCurrencySymbol : function () {
		return this._worldCurrencySymbol;
	},
	getFancyCurrencySymbol : function () {
		return this._fancyCurrencySymbol;
	},
	getNumCurrencyDecPoints : function () {
		return this._numCurrencyDecPoints;
	},
	getMaxCurrencyVal : function () {
		return this._maxCurrencyVal;
	},
	//From Content formats
	// getCurrencySingularName : function () {
	// 	return this._currencySingularName;
	// },
	// getCurrencyPluralName : function () {
	// 	return this._currencyPluralName;
	// },
	getCurrencyPattern : function () {
		return this._currencyPattern;
	}
};

function getFormat(locale, currencyCode, options) {
	// TODO Check for locale null condition
	options = options || {};
	var key = locale + "_" + currencyCode;
	var format = currFormCache.get(key);
	if (!format) {
		format = new CurrencyFormat(locale, currencyCode, options);
		currFormCache.set(key, format);

	}
	return format;
}

exports = module.exports = getFormat;
