// @ebay/locale-formatter-ebay/lib/formats/dateTimeFormat.js
"use strict";

var formatsReader = require('../reader/formatsReader'),
	Cache = require('lru-backup-cache'),
	dtFormCache = new Cache({
		"enabled": true,
        "max": 20000,          //Cache size
        "maxAge": 1800
	});

function DateTimeFormat (locale, options) {
	locale = locale && locale.replace('-', '_');
	var nD = this._normalizeDateFormat,
		nT = this._normalizeTimeFormat;

	this._locale = locale;
	this._formats = formatsReader.load(locale, options);
	//Date
	this._shortDate = this._formats && nD(this._formats.getShortDateFormat());
	this._shortCleanDate = this._formats && nD(this._formats.getShortCleanDateFormat());
	this._mediumDate = this._formats && nD(this._formats.getMediumDateFormat());
	this._numericDate = this._formats && nD(this._formats.getNumericDateFormat());
	this._numericNoDayDate = this._formats && nD(this._formats.getNumericNoDayDateFormat());
	this._mediumNoDayDate = this._formats && nD(this._formats.getMediumNoDayDateFormat());
	this._fullDate = this._formats && nD(this._formats.getFullDateFormat());
	this._fullCompleteDate = this._fullDate && this._fullDate.replace(/EEE/g, "EEEE").replace(/MMM/g, "MMMM");
	this._fullNoYearDate = this._formats && nD(this._formats.getFullNoYearDateFormat());
	this._longDate = this._fullDate && this._convertToLongDate(this._fullDate);
	//Time
	this._shortTime = this._formats && nT(this._formats.getShortTimeFormat());
	this._shortAmPmTime = this._formats && nT(this._formats.getShortAmPmTimeFormat());
	this._shortZoneTime = this._formats && nT(this._formats.getShortZoneTimeFormat());
	this._shortAmPmZoneTime = this._formats && nT(this._formats.getShortAmPmZoneTimeFormat());
	this._mediumTime = this._formats && nT(this._formats.getMediumTimeFormat());
	this._longTime = this._formats && nT(this._formats.getLongTimeFormat());
	this._longAmPmTime = this._formats && nT(this._formats.getLongAmPmTimeFormat());
}

DateTimeFormat.prototype = {
	/**
	 * This is a HACK. Forced to write this because the properties files are not properly
	 * configured to use the industry standard format characters and guidelines for Date and Time.
	 * Changing the properties files would result in unseen errors.
	 */
	_normalizeDateFormat : function (format) {
		/* moment.js alllows escaping literals with [] we should not replace any escaped charcters */

		var inEsacpedPart = false;
		var normalizedFormat = new Array(format.length);

		for (var i=0; i < format.length; i++) {
			if (inEsacpedPart) {
				inEsacpedPart = format[i] !== ']';
				normalizedFormat.push(format[i]);
				continue;
			}

			switch(format[i]) {
				case '[': normalizedFormat.push('[');
						  inEsacpedPart = true; break;
				case 'M': !inEsacpedPart && normalizedFormat.push('MMM'); break;
				case 'D': !inEsacpedPart && normalizedFormat.push('EEE'); break;
				case 'd': !inEsacpedPart && normalizedFormat.push('DD'); break;
				case 'Y': !inEsacpedPart && normalizedFormat.push('YYYY'); break;
				case 'y': !inEsacpedPart && normalizedFormat.push('YY'); break;
				case 'm': !inEsacpedPart && normalizedFormat.push('MM'); break;
				default: normalizedFormat.push(format[i]);
			}
		}

		return normalizedFormat.join('');
	},
	/**
	 * This is a HACK. Forced to write this because the properties files are not properly
	 * configured to use the industry standard format characters and guidelines for Date and Time.
	 * Changing the properties files would result in unseen errors.
	 */
	_normalizeTimeFormat : function (format) {
		var hr = "hh";
		if (format.indexOf('a') === -1) {
			hr = "HH";
		}

		return format.replace("h", hr).
			replace("m","mm").
			replace("d","dd").
			replace("s","ss");
	},
	_convertToLongDate : function (format) {
		var tokenIndex = format.indexOf("EEE");
        if (tokenIndex === -1) {
            return format;
        }

        var endIndex = format.indexOf(" ", tokenIndex) + 1;
        if (endIndex < tokenIndex + 3) {
            endIndex = tokenIndex + 3;
        }
		if (tokenIndex === 0) {
			return format.substring(endIndex);
		}

		var startIndex = format.lastIndexOf(" ", tokenIndex);
		if (startIndex === -1) {
			startIndex = tokenIndex;
		}

		if (startIndex > format.length - 5) {
			return format.substring(0, startIndex);
		}

		return format.substring(0, startIndex) +
			format.substring(endIndex);
	},
	getLocale : function () {
		return this._locale;
	},
	// Date formats
	getShortDate : function () {
		return this._shortDate;
	},
	getShortCleanDate : function () {
		return this._shortCleanDate;
	},
	getMediumDate : function () {
		return this._mediumDate;
	},
	getNumericDate : function () {
		return this._numericDate;
	},
	getMediumNoDayDate : function () {
		return this._mediumNoDayDate;
	},
	getNumericNoDayDate : function () {
		return this._numericNoDayDate;
	},
	getLongDate : function () {

		return this._longDate;
	},
	getFullDate : function () {

		return this._fullDate;
	},
	getFullCompleteDate : function () {

		return this._fullCompleteDate;
	},
	getFullNoYearDate : function () {

		return this._fullNoYearDate;
	},
	// Time formats
	getShortTime : function () {

		return this._shortTime;
	},
	getShortAmPmTime : function () {

		return this._shortAmPmTime;
	},
	getShortAmPmZoneTime : function () {

		return this._shortAmPmZoneTime;
	},
	getShortZoneTime : function () {

		return this._shortZoneTime;
	},
	getMediumTime : function () {

		return this._mediumTime;
	},
	getLongTime : function () {

		return this._longTime;
	},
	getLongAmPmTime : function () {

		return this._longAmPmTime;
	}
};

function getFormat(locale, options) {
	// TODO Check for locale null condition
	options = options || {};
	var format = dtFormCache.get(locale);
	if (!format) {
		format = new DateTimeFormat(locale, options);
		dtFormCache.set(locale, format);

	}
	return format;
}

exports = module.exports = getFormat;
