// @ebay/locale-formatter-ebay/lib/formats/decimalFormat.js
"use strict";

var formatsReader = require('../reader/formatsReader'),
	Cache = require('lru-backup-cache'),
	decFormCache = new Cache({
		"enabled": true,
        "max": 20000,          //Cache size
        "maxAge": 1800
	});

function DecimalFormat (locale, options) {
	locale = locale && locale.replace('-', '_');
	this._locale = locale;
	this._formats = formatsReader.load(locale, options);
	this._decimalChar = this._formats && this._formats.getDecimalChar();
	this._decimalDigits = this._formats && this._formats.getDecimalDigits();
	this._groupingChar = this._formats && this._formats.getGroupingChar();
	this._groupingDigits = this._formats && this._formats.getGroupingDigits();
}

DecimalFormat.prototype = {
	getLocale : function () {
		return this._locale;
	},
	getDecimalChar : function () {
		return this._decimalChar;
	},
	getDecimalDigits : function () {
		return this._decimalDigits;
	},
	getGroupingChar : function () {
		return this._groupingChar;
	},
	getGroupingDigits : function () {
		return this._groupingDigits;
	},
	toString : function () {
		return "Locale = " + this._locale + " | Decimal Char = " + this._decimalChar + " | Decimal digits = " + this._decimalDigits +
				" | Grouping Char = " + this._groupingChar + " | Grouping digits = " + this._groupingDigits;
	}
};

function getFormat(locale, options) {
	// TODO Check for locale null condition
	options = options || {};
	var format = decFormCache.get(locale);
	if (!format) {
		format = new DecimalFormat(locale, options);
		decFormCache.set(locale, format);

	}
	return format;
}

exports = module.exports = getFormat;
