// @ebay/validate-internals-ebay/lib/index.js
'use strict';

var Assert = require('assert');
var componentStatus = require('./component-status.js');
var parentApp;
var viApp = require('./app.js');
var ViUtils = require('./utils');
var tryRequire = require('try-require');
var express = tryRequire('express');
let configBean = require('@ebay/config-bean-ebay');
let url = require('url');
let utils = require('@ebay/utils-ebay');
let logger = require('@ebay/logging-inc').logger('validate-internals-ebay');
const IPMod = require('ip');
const sdEbay = require('@ebay/service-discovery-ebay');
const secureViConfig = require('./secure-vi-redirect-config');
const SECURE_VI_PORT = 10083;
let viServer;

var viRedirectionBean = {
    'configbean': true,
    'id': 'nodejs.config.validate-internals-redirection',
    'alias': 'nodejs.config.validate-internals-redirection',
    'group': 'nodejs.config',
    'desc': 'VI 8080->8083 Redirection bean',
    'attributes': [{
        'name': 'enabled',
        'value': 'true',
        'desc': 'VI 8080->8083 Redirection bean',
        'readable': 'true',
        'writable': 'true'
    }]
};

function secureRedirectMiddleware(req, res, next) {
    (async () => {
        // Check if this is a request to view the ServeTraffic configuration
        const reqURL = url.parse(req.url);
        if (shouldSecureRedirect(req)) {
            const secureRedirectUrl = await getSecureRedirectUrl(req);
            if (secureRedirectUrl) {
                res.redirect(secureRedirectUrl);
            } else {
                next();
            }
        } else {
            next();
        }
    })();
}

/**
 * Override parent express app's listen to start dedicated vi listener only when parentApp listener is invoked
 * @param parentApp
 */
async function launchDedicatedVI(parentApp) {
    let oldListen = parentApp.listen;
    parentApp.listen = function () {
        let appServer;
        if (viServer) {
            appServer = oldListen.apply(parentApp, arguments);
        } else if (!viServer) {
            let dedicatedVI = express();
            // This is for http -> https redirection
            dedicatedVI.use('/admin', secureRedirectMiddleware, viApp);

            let VIPORT = process.env.VIPORT || 8083;
            if (process.env.NODE_ENV === 'test' && !process.env.VIPORT) {
                VIPORT = 0;       //Override for test environment, so that multiple apps can execute on local/dev without explicit configuration changes
            }

            viServer = dedicatedVI.listen(VIPORT, function (err) {
                if (err) {
                    logger.error(`Failed to start ValidateInternals on ${VIPORT}`, err);
                    return;
                }
                process.env.VIPORT = this.address().port;
                if (process.env.NODE_ENV === 'test') {
                    process.env.VIREDIRECTPORT = process.env.VIREDIRECTPORT || process.env.VIPORT;
                }
                logger.debug(`ValidateInternals now running on ${process.env.VIPORT}`);
                process.emit('vistart');
            });

            appServer = oldListen.apply(parentApp, arguments);
        }
        let oldClose = appServer.close;
        appServer.close = function () {
            viServer && viServer.close();
            viServer = null;
            return oldClose.apply(appServer, arguments);
        };
        return appServer;
    };
}

function createVIRedirectionBean() {
    !configBean.getBeanById(viRedirectionBean.id) && configBean.define(viRedirectionBean);
}

// This function is now managed by secureViConfig module
function createSecureVIRedirectionBean() {
    secureViConfig.createSecureVIRedirectionBean();
}

function isRedirectionEnabled(req) {
    const cfg = configBean.getBeanById(viRedirectionBean.id);

    const isRedirect = (cfg && cfg.get('enabled')) || (req.headers['X-EBAY-FORCE-VI-PORT-REDIRECTION'.toLowerCase()] === 'true') || false;

    if (isRedirect) {
        //Skip Port redirect in Dev & Test env, also if the request is from localhost
        let env = process.env.NODE_ENV;

        let ip = ViUtils.isEnvoyEnabled() ? req.headers['x-client-ip'] : req.ip;

        if(!env || env === 'test' || env === 'development' || IPMod.isLoopback(ip)) {
            return false;
        }
        return true;
    }
    return false;
}

function isSecureRedirectionEnabled() {
    return secureViConfig.isEnabled();
}

function shouldSecureRedirect(req) {
     // Skip the redirect if it from localhost or if the environment is test or development
     const env = process.env.NODE_ENV;
     const ip = ViUtils.isEnvoyEnabled() ? req.headers['x-client-ip'] : req.ip;
     if(!env || env === 'development' || IPMod.isLoopback(ip)) {
         return false;
     }
    const port = parseInt(req.headers.host.split(':')[1]);
    return port === 8083 && isSecureRedirectionEnabled();
}

// Sample: curl http://10.43.129.17:8083/admin/v3console/ValidateInternals -i
// HTTP/1.1 302 Found
// location: https://adminproxy.vip.qa.ebay.com/proxy/10.43.129.17/admin/v3console/ValidateInternals
// Where the adminproxy url is determined by service-discovery-ebay lookup function, if not found, it will fallback to the original request and return undefined
async function getSecureRedirectUrl(req) {
    const reqURL = url.parse(req.url);
    const host = req.headers.host.split(':')[0];
    try {
        // If we are able to find admin proxy host from config, then use it
        const cachedAdminProxyUrl = secureViConfig.getAdminProxyUrl();
        if (cachedAdminProxyUrl) {
            return cachedAdminProxyUrl + `/proxy/${host}/admin${reqURL.path}`;
        }

        // Otherwise look it up from service discovery
        const discoverData = await sdEbay.fetchDiscoveryData(true);
        const adminProxyHost = discoverData['urn:REST:adminproxy'][0].serviceEndpoint;
        logger.debug(`adminProxyHost: ${adminProxyHost}`);
        
        if (!adminProxyHost) {
            return undefined;
        }
        
        // Cache the admin proxy host for future use
        secureViConfig.setAdminProxyUrl(adminProxyHost);
        return adminProxyHost + `/proxy/${host}/admin${reqURL.path}`;
    }
    catch (e) {
        logger.error('Failed to lookup admin proxy host', e);
        return undefined;
    }
}

// Cleanup function for stopping refresh timer and other resources
function cleanup() {
    secureViConfig.stopRefreshTimer();
    if (viServer) {
        viServer.close();
        viServer = null;
    }
}

// Handle process shutdown
process.on('SIGINT', cleanup);
process.on('SIGTERM', cleanup);

// Called from brogan-ebay to establish VI as a subapp
module.exports = {
    // TODO: this should be deprecated in favor of brogan setup
    middleware: function middleware(app) {
        app = app ? app :
            express ? express() : null;
        Assert.ok(app, 'You must provide parent app');

        function redirectionMiddleware(req, res, next) {
            const port = req.headers.host.split(':')[1];
        
            // This is for 8080 -> 8083 redirection
            if (isRedirectionEnabled(req) && port !== SECURE_VI_PORT) {
                let reqURL = url.parse(req.url);
        
                let protocol = 'http';
                let redirectPort = process.env.VIREDIRECTPORT || 8083;
                let rurl = `${protocol}://${reqURL.hostname || utils.getServerName(req)}:${redirectPort}/admin${reqURL.path}`;
                /\/v3console\/ViewConfigCategory(Xml)?\?id=ebay\.kernel\.ServeTraffic/.test(reqURL.path) ? next() : res.redirect(rurl);
            } else {
                next();
            }
        }
        
        app.use('/admin', redirectionMiddleware, viApp);

        // Capture parentApp so we can later get routes/middleware stack for meta
        parentApp = app;
        launchDedicatedVI(parentApp);

        // expose worker port for worker specific requests, used for example to profile specific worker
        app.once('start', () => {

            const workerPort = ViUtils.getWorkerPort();
            const workerApp = express();

            createVIRedirectionBean();
            createSecureVIRedirectionBean();
            
            // Initialize secure VI configuration with refresh timer
            secureViConfig.initialize();

            // add 404 terminator for VI
            viApp.use((req, res, next) => {
                res.status(404).end();
            });
            workerApp.use('/admin', viApp);
            ViUtils.isPortFree(workerPort, free => {
                if (!free) {
                    logger.error(`Failed to connect to worker port ${workerPort}, it is already taken`);
                    return;
                }
                workerApp.listen(workerPort, err => {
                    if (err) {
                        logger.error(`Failed to connect to worker port ${workerPort}`, err);
                        return;
                    }

                    logger.debug(`Listening on worker port ${workerPort}`);
                });
            });

        });
        return app;
    },
    getParentApp: function getParentApp() {
        return parentApp;
    },
    app: viApp, // TODO: brogan ebay should hook this to /admin path explicitly
    register: componentStatus.register,
    componentStatus: componentStatus,
    enableComponentStatus: componentStatus.enableComponentStatus,
    isRedirectionEnabled: isRedirectionEnabled,
    isSecureRedirectionEnabled: isSecureRedirectionEnabled,
    shouldSecureRedirect: shouldSecureRedirect,
    getSecureRedirectUrl: getSecureRedirectUrl,
    secureViRedirectionBean: secureViConfig.secureViRedirectionBean,
    launchDedicatedVI: launchDedicatedVI
};
