// @ebay/validate-internals-ebay/lib/secure-vi-redirect-config.js
'use strict';

const Httpunch = require('@ebay/httpunch');
const configBean = require('@ebay/config-bean-ebay');
const appContext = require('@ebay/app-context-ebay');
const logger = require('@ebay/logging-inc').logger('validate-internals-ebay/secure-vi-redirect');
const ACCESS_CONTROL_LIST_HOSTS = require('./access-control-list-hosts');
const utils = require('./utils');
const ACL_PATH = '/authzsvc/api/v1/securevi/config';
// Get the appropriate host based on environment
// Convert paasRealm to lowercase for consistent lookup (e.g., 'staging', 'production')
const rawEnv = appContext.paasRealm || 'staging';
const normalizedEnv = rawEnv.toLowerCase();
// Use normalized environment for lookup in ACCESS_CONTROL_LIST_HOSTS
const host = ACCESS_CONTROL_LIST_HOSTS[normalizedEnv] || ACCESS_CONTROL_LIST_HOSTS['staging']; // Fallback to staging if env not found

const SECURE_VI_BEAN_ID = 'nodejs.config.secure-validate-internals-redirection';
const REFRESH_INTERVAL = 30 * 60 * 1000; // 30 minutes in milliseconds

// Enhanced configuration bean with additional attributes
const secureViRedirectionBean = {
    'configbean': true,
    'id': SECURE_VI_BEAN_ID,
    'alias': SECURE_VI_BEAN_ID,
    'group': 'nodejs.config',
    'desc': 'VI 8083-> adminproxy 10083 Redirection bean',
    'attributes': [
        {
            'name': 'enabled',
            'value': 'false',
            'desc': 'VI *-> adminproxy 10083 Redirection bean',
            'readable': 'true',
            'writable': 'true'
        },
        {
            'name': 'forceEnable',
            'value': '',
            'desc': 'Force enable secure VI redirection regardless of other settings (true/false/undefined, default: undefined)',
            'readable': 'true',
            'writable': 'true'
        },
        {
            'name': 'remoteConfigEnabled',
            'value': '',
            'desc': 'Enable fetching configuration from remote endpoint',
            'readable': 'true',
            'writable': 'true'
        },
        {
            'name': 'lastRefreshed',
            'value': '',
            'desc': 'Timestamp of last configuration refresh',
            'readable': 'true',
            'writable': 'true'
        },
        {
            'name': 'endpoint',
            'value': 'https://' + host + ACL_PATH,
            'desc': 'Endpoint for remote secure VI redirection',
            'readable': 'true',
            'writable': 'true'
        },
        {
            'name': 'adminproxy_url',
            'value': '',
            'desc': 'URL of the admin proxy service',
            'readable': 'true',
            'writable': 'true'
        }
    ]
};

let refreshTimer;

/**
 * Initialize the secure VI redirection configuration
 */
function initialize() {
    createSecureVIRedirectionBean();
    startRefreshTimer();
}

/**
 * Create the secure VI redirection configuration bean if it doesn't exist
 */
function createSecureVIRedirectionBean() {
    if (!configBean.getBeanById(SECURE_VI_BEAN_ID)) {
        configBean.define(secureViRedirectionBean);
        logger.debug('Secure VI redirection bean created');
    }
}

/**
 * Start the timer to periodically refresh the configuration
 */
function startRefreshTimer() {
    if (!refreshTimer) {
        // Initial fetch without delay
        fetchRemoteConfig();
        
        // Setup periodic refresh
        refreshTimer = setInterval(fetchRemoteConfig, REFRESH_INTERVAL);
        refreshTimer.unref(); // Allow the process to exit even if the timer is running
        
        logger.debug(`Secure VI redirection config refresh timer started (${REFRESH_INTERVAL / 60000} minutes)`);
    }
}

/**
 * Stop the refresh timer
 */
function stopRefreshTimer() {
    if (refreshTimer) {
        clearInterval(refreshTimer);
        refreshTimer = null;
        logger.debug('Secure VI redirection config refresh timer stopped');
    }
}

/**
 * Fetch configuration from the remote endpoint
 */
async function fetchRemoteConfig() {
    const bean = configBean.getBeanById(SECURE_VI_BEAN_ID);
    
    if (!bean || bean.get('remoteConfigEnabled') === false) {
        return;
    }
    
    try {
        const appName = appContext.appName || '';
        // Apply capitalizeFirstLetter to paasRealm to ensure consistent casing in remote API calls
        const paasRealm = utils.capitalizeFirstLetter(appContext.paasRealm) || 'Staging';
        const vpc = appContext.vpc || '';
        
        const baseUrl = bean.get('endpoint') || `https://${host}${ACL_PATH}`;
        const url = `${baseUrl}?appName=${encodeURIComponent(appName)}&paasRealm=${encodeURIComponent(paasRealm)}&vpc=${encodeURIComponent(vpc)}`;
        
        logger.debug(`Fetching secure VI redirection config from: ${url}`);
        
        const response = await new Promise((resolve, reject) => {
            Httpunch.get({
                url: url,
                method: 'GET',
                rejectUnauthorized: false,
                timeout: 5000 // 5 second timeout
            }, (err, res) => {
                if (err) {
                    return reject(err);
                }
                return resolve(res);
            });
        });
        
        if (response.statusCode !== 200) {
            throw new Error(`Received status code ${response.statusCode}`);
        }
        
        const configData = JSON.parse(response.body.toString());
        updateBeanFromRemoteConfig(configData);
        
        // Update last refreshed timestamp
        bean.set('lastRefreshed', new Date().toISOString());
        
        logger.debug('Secure VI redirection config successfully refreshed from remote endpoint');
    } catch (err) {
        logger.error('Failed to fetch secure VI redirection config from remote endpoint', err);
        // Fallback to default (false) is handled by isEnabled() method
    }
}

/**
 * Update the configuration bean with data from the remote config
 * @param {Object} configData - Data from the remote config endpoint
 */
function updateBeanFromRemoteConfig(configData) {
    const bean = configBean.getBeanById(SECURE_VI_BEAN_ID);
    
    if (!bean) {
        return;
    }
    
    logger.debug(`Received remote config: ${JSON.stringify(configData)}`);
    
    // Handle the server response format: { "appName": "...", "option": "..." }
    if (configData && configData.option) {
        switch (configData.option) {
            case 'REDIRECT':
                // set remoteConfigEnabled to true
                bean.set('remoteConfigEnabled', 'true');
                logger.debug(`SecureViEnforceOption: REDIRECT - Setting remoteConfigEnabled to true for ${configData.appName}`);
                break;
                
            case 'NO_REDIRECT':
                // set remoteConfigEnabled to false
                bean.set('remoteConfigEnabled', 'false');
                logger.debug(`SecureViEnforceOption: NO_REDIRECT - Setting remoteConfigEnabled to false for ${configData.appName}`);
                break;
                
            case 'DEFAULT':
                // Default should not change remoteConfigEnabled, but we can set it based on ADMIN_VI_REDIRECT
                logger.debug(`SecureViEnforceOption: DEFAULT - Using environment variable ADMIN_VI_REDIRECT for ${configData.appName}`);
                break;
                
            default:
                logger.warn(`Unknown SecureViEnforceOption: ${configData.option} for ${configData.appName}`);
        }
    } else {
        logger.warn('Missing option in configData, falling back to default behavior');
    }
}

/**
 * Check if secure VI redirection is enabled
 * @returns {boolean} True if redirection is enabled
 */
function isEnabled() {
    // First check if this is a Container PDLC pool
    // Secure VI redirection should not happen if the pool is non-CPDLC
    const isContainerPdlc = process.env.CONTAINER_PDLC === 'true' || process.env.CONTAINER_PDLC === true;
    if (!isContainerPdlc) {
        logger.debug('Secure VI redirection disabled: not a Container PDLC pool (CONTAINER_PDLC != true)');
        return false;
    }
    
    const bean = configBean.getBeanById(SECURE_VI_BEAN_ID);
    const adminViRedirect = process.env.ADMIN_VI_REDIRECT === 'true' || process.env.ADMIN_VI_REDIRECT === true;
    
    // Fallback to default environment variable if bean doesn't exist
    if (!bean) {
        return adminViRedirect || false;
    }
    
    // 1. Always check forceEnable first if provided
    const forceEnableValue = bean.get('forceEnable');
    if (forceEnableValue === 'true' || forceEnableValue === true) {
        return true;
    }
    if (forceEnableValue === 'false' || forceEnableValue === false) {
        return false;
    }
    
    // 2. Check remote configuration if it was fetched (has lastRefreshed)
    const lastRefreshed = bean.get('lastRefreshed');
    const remoteConfigEnabled = bean.get('remoteConfigEnabled');
    if (lastRefreshed && (remoteConfigEnabled === 'true' || remoteConfigEnabled === true || remoteConfigEnabled === 'false' || remoteConfigEnabled === false)) {
        return remoteConfigEnabled === 'true' || remoteConfigEnabled === true;
    }
    
    // 3. Fallback to environment variable
    if (adminViRedirect) {
        return true;
    }
    
    // 4. Finally fallback to the bean's enabled attribute
    const beanEnabled = bean.get('enabled') === 'true' || bean.get('enabled') === true;
    return beanEnabled || false;
}

/**
 * Get the admin proxy URL from the configuration
 * @returns {string|undefined} The admin proxy URL or undefined if not set
 */
function getAdminProxyUrl() {
    const bean = configBean.getBeanById(SECURE_VI_BEAN_ID);
    
    if (!bean) {
        return undefined;
    }
    
    return bean.get('adminproxy_url') || undefined;
}

/**
 * Set the admin proxy URL in the configuration
 * @param {string} url - The admin proxy URL
 */
function setAdminProxyUrl(url) {
    const bean = configBean.getBeanById(SECURE_VI_BEAN_ID);
    
    if (bean && url) {
        bean.set('adminproxy_url', url);
    }
}

module.exports = {
    initialize,
    createSecureVIRedirectionBean,
    isEnabled,
    getAdminProxyUrl,
    setAdminProxyUrl,
    startRefreshTimer,
    stopRefreshTimer,
    secureViRedirectionBean,
    fetchRemoteConfig
};
