// @ebay/cookies-ebay/lib/utils/utils.js
/*global escape:false, unescape:false*/

'use strict';

var EBAY_COM = '.ebay.com',
    TOP_DOMAINS = {
        '.ebay.com': 1,
        '.ebay.de': 1,
        '.ebay.co.uk': 1
    },
    OWN_DOMAINS = {
        '.ebay.com.au': 1,
        '.ebay.compuserve.com': 1,
        '.ebay.com.cn': 1,
        '.ebay.com.hk': 1,
        '.ebay.com.sg': 1,
        '.ebay.com.my': 1
    };

function extractRootDomainName(domain) {
    if (domain.indexOf(EBAY_COM) === 0 && !OWN_DOMAINS[domain]) {
        return EBAY_COM;
    }
    return domain;
}

/**
 * True if val contains an invalid field-vchar
 *  field-value    = *( field-content / obs-fold )
 *  field-content  = field-vchar [ 1*( SP / HTAB ) field-vchar ]
 *  field-vchar    = VCHAR / obs-text
 *
 * checkInvalidHeaderChar() is currently designed to be inlinable by v8,
 * so take care when making changes to the implementation so that the source
 * code size does not exceed v8's default max_inlined_source_size setting.
 **/
 // Taken from nodejs source code: https://github.com/nodejs/node/blob/v6.x/lib/_http_common.js
function checkInvalidHeaderChar(val) {
    val += '';
    if (val.length < 1) {
        return false;
    }
    var c = val.charCodeAt(0);
    if ((c <= 31 && c !== 9) || c > 255 || c === 127) {
        return true;
    }
    if (val.length < 2) {
        return false;
    }
    c = val.charCodeAt(1);
    if ((c <= 31 && c !== 9) || c > 255 || c === 127) {
        return true;
    }
    if (val.length < 3) {
        return false;
    }
    c = val.charCodeAt(2);
    if ((c <= 31 && c !== 9) || c > 255 || c === 127) {
        return true;
    }
    for (var i = 3; i < val.length; ++i) {
        c = val.charCodeAt(i);
        if ((c <= 31 && c !== 9) || c > 255 || c === 127) {
            return true;
        }
    }
    return false;
}

module.exports = {
    values: function (o) {
        var values = [];
        for (var k in o) {
            if (o.hasOwnProperty(k)) {
                values.push(o[k]);
            }
        }
        return values;
    },
    encode: function (str) {
        return escape(str).replace(/\%20/g, '+');
    },
    decode: function (str) {
        return unescape(str.replace(/[+]/g, ' '));
    },
    extractDomainName: function (serverName) {
        if (!serverName) {
            return EBAY_COM;
        }
        var pos = serverName.indexOf('.ebay');
        if (pos === -1) {
            pos = serverName.indexOf('.');
            if (pos === -1) {
                return EBAY_COM;
            }
        } else {
            var pos1 = pos;
            while ((pos1 = serverName.indexOf('.ebay', pos + 1)) > -1) {
                pos = pos1;
            }
        }
        var tmpDomain = serverName.substring(pos);
        pos = tmpDomain.indexOf('/');
        if (pos !== -1) {
            tmpDomain = tmpDomain.substring(0, pos);
        }
        // make an exception for top domains which make up more than 90% of
        // traffic and return them immediately
        if (TOP_DOMAINS[tmpDomain]) {
            return tmpDomain;
        }
        if (serverName.indexOf('catalogservice.vip') === 0) {
            return '.' + serverName;
        }
        return extractRootDomainName(tmpDomain);
    },

    tryCatch: function tryCatch(fn, log) {
        return function _tryCatchWrapper() {
            try {
                return fn.apply(this, arguments);
            }
            catch (err) {
                log && log(err);
            }
        };
    },

    checkInvalidHeaderChar: checkInvalidHeaderChar

};
