// @ebay/cal-publishing-inc/lib/config.js
'use strict';

var events = require('events');
var objutil = require('objutil');
var configer = require('@ebay/module-config-inc');
var configuration;
var instances = [];
var TOPIC = 'cal-event-args';
var LOG_LEVELS = module.exports.LOG_LEVELS = ['debug', 'info', 'warn', 'transaction', 'error', 'fatal'];

module.exports = new events.EventEmitter();

var configureDefaults = module.exports.configureDefaults = function(callback) {
    configuration = null;
    configer(module, function(err, config) {
        config = module.exports.configure(config.get('cal-publishing-inc'));
        callback && callback(null, config);
    });
};

module.exports.configure = function(config, merge) {
    configuration = merge && objutil.mixin(config, configuration || {}) || config;
    configureAppenders(configuration);
    module.exports.emit('change', configuration);
    return configuration;
};

function configureAppenders(config) {
	(instances || []).forEach(function(instance) {
		instance.destroy();
	});

	instances = [];
    for(var key in config.appenders || {}) {
        var appender = config.appenders[key];
		if (appender.enabled && appender.module) {
			var m = require(appender.module);
			var consumeFn = m.configure(appender.config || {}, config);
            if (consumeFn) {
                consumeFn = appender.loglevel &&
                    configureAppenderConsumeFunc(appender, consumeFn) || consumeFn;
                var inst = register(consumeFn);
                inst.name = key;
                instances.push(inst);
            }
		}
	}
}

function configureAppenderConsumeFunc(config, consumeFunc) {
    var levelSettings = {};
    var appendersLoglevels = refreshLogLevels(config);
    LOG_LEVELS.forEach(function(level, index) {
        var rules = appendersLoglevels[level];
        if (rules) {
            levelSettings[level] = function isEnabled(name) {
                return !rules.skips.some(function(re) {
                    return !!re.test(name);
                }) && rules.names.some(function(re) {
                    return !!re.test(name);
                });
            };
        }
        else if (index) {
            levelSettings[level] = levelSettings[LOG_LEVELS[index - 1]];
        }
    });

    if (Object.keys(levelSettings).length) {
        return function consumeWrapper(evt) {
            var level = evt.loglevel || 'transaction';
            var isEnabled = levelSettings[level];
            if (isEnabled && isEnabled(evt.logger)) {
                consumeFunc(evt);
            }
        };
    }

    return consumeFunc;
}

function parseLogLevel(levelValue) {
    var skips = [];
    var names = [];
    (levelValue || '')
        .split(/[\s,]+/)
        .forEach(function(name){
            name = name.replace('*', '.*?');
            if (name[0] === '-') {
                skips.push(new RegExp('^' + name.substr(1) + '$'));
            } else {
                names.push(new RegExp('^' + name + '$'));
            }
        });

    return {
        skips: skips,
        names: names
    };
}

function refreshLogLevels(config) {
    var loglevel = {};
    config = config && Object.keys(config.loglevel || {}).length && config.loglevel || {
        warn: '*'
    };
    LOG_LEVELS.forEach(function(level) {
        var value = config[level] || '';
        if (process.env.LOG_LEVEL) {
            value += ' ' + process.env[level];
        }
        if (!value) {
            return;
        }
        loglevel[level] = parseLogLevel(value);
    });
    return loglevel;
}

function register(consumeFn) {

    function wrapper() {
        if (wrapper.enabled) {
            consumeFn.apply(null, arguments);
        }
    }
    wrapper.enabled = true;
    process.on(TOPIC, wrapper);

    return {
        destroy: function() {
            process.removeListener(TOPIC, wrapper);
            wrapper.enabled = false;
        }
    };
}

configureDefaults();
