// @ebay/openfeature-ebay/dist/openfeature-context/ServiceContext.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.ServiceContext = void 0;
const tslib_1 = require("tslib");
const header_multi_value_ebay_1 = tslib_1.__importDefault(require("@ebay/header-multi-value-ebay"));
const Context_1 = require("./Context");
const CHANNEL = 'CHANNEL';
const OPTIN_FLAG = 'EP_OPTIN';
const COOKILET_HEADER_OPTIN_FLAG = '321';
class ServiceContext extends Context_1.Context {
    constructor(req, res, openfeatureGlobalCache) {
        super(req, res, openfeatureGlobalCache);
        this.epCOSHdr = this.toHMV(this.req.get('x-ebay-c-exp'));
    }
    getOptedIntoFlag() {
        return (this.optinFlag || (this.epCOSHdr && this.epCOSHdr.get(OPTIN_FLAG)));
    }
    isEmailChannel() {
        return this.epCOSHdr && this.epCOSHdr.get(CHANNEL) === 6;
    }
    setExperimentTreatements(experimentationCookieletHeader) {
        if (experimentationCookieletHeader && !this.res.headersSent) {
            this.optinFlag = this.toHMV(experimentationCookieletHeader).get(COOKILET_HEADER_OPTIN_FLAG);
            /* Send the X-EBAY-SVC-EP-COOKIELET to web app so it can persist in cookie */
            this.res.set('X-EBAY-SVC-EP-COOKIELET', experimentationCookieletHeader);
        }
    }
    getClientIP() {
        return this.ebay.getEndUserContext().getIPAddress();
    }
    getReferer() {
        return this.ebay.getEndUserContext().getReferer();
    }
    // Not applicable for service ctx so providing a default function
    getCookiesAndCookielets() {
        return null;
    }
    toHMV(httpHeader) {
        httpHeader = httpHeader || '';
        try {
            return header_multi_value_ebay_1.default.parse(httpHeader);
        }
        catch (e) {
            console.error('Unable to parse Header into HMV objet: ', httpHeader, e);
            return header_multi_value_ebay_1.default.parse('');
        }
    }
    getCountryCode() {
        return this.ebay.getEndUserContext().getUserContextualLocation().getCountry();
    }
}
exports.ServiceContext = ServiceContext;
