// @ebay/openfeature-ebay/dist/openfeature-context/WebContext.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.WebContext = void 0;
const Context_1 = require("./Context");
const header_multi_value_ebay_1 = require("@ebay/header-multi-value-ebay");
const OPTIN_FLAG = '321';
class WebContext extends Context_1.Context {
    constructor(req, res, openfeatureGlobalCache) {
        super(req, res, openfeatureGlobalCache);
    }
    getOptedIntoFlag() {
        return (this.ebay &&
            this.ebay.cookies &&
            this.ebay.cookies.getCookieValue('EXPERIMENT_TREATMENTS'));
    }
    setExperimentTreatements(experienceServiceCookieletHeader) {
        if (experienceServiceCookieletHeader) {
            this.ebay &&
                this.ebay.cookies &&
                this.ebay.cookies.setCookieValue('EXPERIMENT_TREATMENTS', this.extractOptedIntoFlag(experienceServiceCookieletHeader));
        }
    }
    getCookiesAndCookielets() {
        const cookiesAndCookielets = {
            cookies: {},
            cookielets: {},
        };
        cookiesAndCookielets.cookies = Object.keys(cookiesAndCookielets.cookies).reduce((acc, key) => {
            const val = encodeURIComponent(cookiesAndCookielets.cookies[key]).replace(/\%20/g, '+');
            return (acc += key + '=' + val + ';');
        }, '');
        cookiesAndCookielets.cookielets = Object.keys(cookiesAndCookielets.cookielets).reduce((acc, key) => {
            const val = encodeURIComponent(cookiesAndCookielets.cookielets[key]).replace(/\%20/g, '+');
            return (acc += key + '=' + val + ';');
        }, '');
        return cookiesAndCookielets;
    }
    getClientIP() {
        return this.ebay.getClientIP();
    }
    getReferer() {
        return (this.req.headers &&
            this.req.headers.referer &&
            this.req.headers.referer.trim());
    }
    isEmailChannel() {
        return this.req.emailChannel;
    }
    extractOptedIntoFlag(experienceServiceCookieletHeader) {
        if (experienceServiceCookieletHeader) {
            try {
                const parsed = (0, header_multi_value_ebay_1.parse)(experienceServiceCookieletHeader);
                return parsed && parsed.get(OPTIN_FLAG);
            }
            catch (e) {
                console.error('Unable to get EP Cookielet Value from Header: ', experienceServiceCookieletHeader, e);
            }
        }
    }
    getCountryCode() {
        return this.req && this.req.locality && this.req.locality.country;
    }
}
exports.WebContext = WebContext;
