// @ebay/gatekeeper-ebay/lib/req-utils.js
'use strict';

var Utils = require('util');
var cookie = require('cookie');
var QueryString = require('querystring');
var ip = require('ip');
var appContext = require('@ebay/app-context-ebay');
var _SENSITIVE_KEYS = ['authorization'];

function getSensitiveValue() {
    return '*****';
}

function tryGetGuid(req) {
    try {
        return req.ebay && req.ebay.getGuid && req.ebay.getGuid();
    }
    catch (err) {}
}

function tryGetRlogId(req) {
    try {
        return req.ebay && req.ebay.getRlogId && req.ebay.getRlogId();
    }
    catch (err) {}
}

module.exports = {
    getInfo: function(req) {
        if (req) {
            return {
                guid: tryGetGuid(req),
                rlogid: tryGetRlogId(req),
                hostname: req.hostname,
                ip: req.ip,
                url: req.url,
                baseUrl: req.baseUrl,
                path: req.path,
                params: req.params,
                qs: req.query,
                headers: req.headers
            };
        }
    },
    formatText: function(options) {

        var formatted = [];
        formatted.push('==== ENVIRONMENT ====');
        formatted.push(Utils.format('deployed environment: %s', options.req.env));
        formatted.push(Utils.format('Logs: http://%s:8080/admin/v3console/logs', /.+\.ebay(c3)?\.com/.test(appContext.machineName) ? appContext.machineName : ip.address()));
        formatted.push('==== ERROR ====');
        formatted.push(Utils.format('message: %s', options.err.message));
        formatted.push(Utils.format('stack: %s', options.err.stack));
        formatted.push('');

        formatted.push('==== REQUEST ====');
        formatted.push(Utils.format('guid: %s', options.req.guid));
        formatted.push(Utils.format('rlogid: %s', options.req.rlogid));
        formatted.push(Utils.format('host: %s', options.req.hostname));
        formatted.push(Utils.format('ip: %s', options.req.ip));
        formatted.push(Utils.format('url: %s', options.req.url));
        formatted.push(Utils.format('baseurl: %s', options.req.baseUrl));
        formatted.push(Utils.format('params: %s', options.req.params ? JSON.stringify(options.req.params) : ''));
        formatted.push(Utils.format('query: %s', options.req.query ? QueryString.stringify(options.req.query) : ''));
        formatted.push('');

        if (options.req.headers) {

            formatted.push('==== Headers ====');

            Object.keys(options.req.headers).forEach(function(key) {
                //Don't Add Cookie to emails
                if(key === 'cookie') {
                    try {
                        var cookieObj = cookie.parse(options.req.headers.cookie);
                        var maskedCookie = '';
                        Object.keys(cookieObj).forEach((cookieKey) => {
                            maskedCookie += `${cookieKey}=${getSensitiveValue()};`;
                        });
                        formatted.push(`${key}:${maskedCookie}`);
                    } catch(err) {
                        console.error('Error while parsing cookie', err);
                    }
                } else {
                    formatted.push(key + ': ' + (_SENSITIVE_KEYS.indexOf(key) >= 0 ? getSensitiveValue() : options.req.headers[key]));
                }
            });

        }

        return formatted.join('\n');
    },
    getBaseHost: function(hostname) {
        var ebayHost = hostname || '.ebay.com';

        var matches = ebayHost.match(/(([^\.]+\.paradise\.qa\.ebay|qa\.ebay|latest\.ebay|ebay)\.[^\/:]+)/);
        if (matches) {
            ebayHost = matches[0];
        }
        return ebayHost;
    }
};
