// @ebay/openfeature-ebay/dist/tracing/FeatureFlagSpanProcessorManager.js
"use strict";
/**
 * Global manager for FeatureFlagSpanProcessor
 *
 * This singleton provides global access to the FeatureFlagSpanProcessor instance,
 * allowing middleware and other components to access trace error information.
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.FeatureFlagSpanProcessorManager = exports.featureFlagSpanManager = void 0;
class FeatureFlagSpanProcessorManager {
    constructor() {
        this.processor = null;
    }
    static getInstance() {
        if (!FeatureFlagSpanProcessorManager.instance) {
            FeatureFlagSpanProcessorManager.instance = new FeatureFlagSpanProcessorManager();
        }
        return FeatureFlagSpanProcessorManager.instance;
    }
    /**
     * Register the FeatureFlagSpanProcessor instance globally
     */
    setProcessor(processor) {
        this.processor = processor;
    }
    /**
     * Get the registered FeatureFlagSpanProcessor instance
     */
    getProcessor() {
        return this.processor;
    }
    /**
     * Get logging context error from traceId
     * It counts up the number of errors by each error name.
     */
    getLoggingContextError(traceId) {
        const loggingContextError = {
            errorCountMap: {},
            errorCount: 0
        };
        // Add null safety check for processor
        if (!this.processor) {
            return loggingContextError;
        }
        const traceIdMap = this.processor.getTraceIdMap();
        if (traceIdMap != null && traceId != null) {
            const errorInfo = traceIdMap.get(traceId);
            if (errorInfo != null && errorInfo.getErrorEvents() != null) {
                for (const [key, value] of errorInfo.getErrorEvents()) {
                    loggingContextError.errorCountMap[key] = value;
                }
                loggingContextError.errorCount = errorInfo.getTotalErrorCount();
            }
        }
        return loggingContextError;
    }
    /**
     * Removes the trace from trace id map
     * @param traceId
     */
    removeTrace(traceId) {
        if (this.processor) {
            this.processor.removeTrace(traceId);
        }
    }
}
exports.FeatureFlagSpanProcessorManager = FeatureFlagSpanProcessorManager;
exports.featureFlagSpanManager = FeatureFlagSpanProcessorManager.getInstance();
