// @ebay/openfeature-ebay/dist/tracing/TracingDataHolder.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.TracingDataHolder = void 0;
/**
 * The `TracingDataHolder` class is responsible for managing a mapping between trace IDs
 * and collections of `FeatureFlagRawEvent` objects. It provides methods to add, retrieve,
 * and remove data associated with specific trace IDs.
 */
class TracingDataHolderClass {
    constructor() {
        this.traceIdToCalData = new Map();
    }
    /**
     * Adds feature flag raw events to the collection for a given trace ID.
     * If the trace ID doesn't exist, creates a new array.
     *
     * @param traceId - The trace identifier
     * @param featureFlagRawEvents - Collection of feature flag raw events to add
     */
    addEntry(traceId, featureFlagRawEvents) {
        const existingEvents = this.traceIdToCalData.get(traceId) || [];
        const updatedEvents = [...existingEvents, ...featureFlagRawEvents];
        this.traceIdToCalData.set(traceId, updatedEvents);
    }
    /**
     * Retrieves the feature flag raw events for a given trace ID.
     * Returns a copy of the array to prevent external modification.
     *
     * @param traceId - The trace identifier
     * @returns Array of feature flag raw events, or empty array if not found
     */
    getCalDataListById(traceId) {
        const events = this.traceIdToCalData.get(traceId);
        return events ? [...events] : []; // Return a copy to maintain immutability
    }
    /**
     * Returns a copy of the entire trace ID to events mapping.
     * The returned map is a shallow copy to prevent external modification.
     *
     * @returns Map of trace IDs to their associated feature flag raw events
     */
    getCalDataList() {
        const result = new Map();
        this.traceIdToCalData.forEach((events, traceId) => {
            result.set(traceId, [...events]); // Copy each array
        });
        return result;
    }
    /**
     * Removes all feature flag raw events associated with a specific trace ID.
     *
     * @param traceId - The trace identifier to remove
     */
    removeByTraceId(traceId) {
        this.traceIdToCalData.delete(traceId);
    }
    /**
     * Clears all trace ID mappings and their associated data.
     */
    removeTraceIdMap() {
        this.traceIdToCalData.clear();
    }
}
exports.TracingDataHolder = new TracingDataHolderClass();
