// @ebay/openfeature-ebay/src/openfeature-lib/tracing/FeatureFlagEventUtils.ts
import { FlagEvalDetail } from '../types/FlagEvalDetail';
import { FeatureFlagRawEvent } from '../types/FeatureFlagRawEvent';
import { FeatureFlagCalEvent } from '../types/FeatureFlagCalEvent';
import { TracingDataHolder } from '.';
import { Context } from '../openfeature-context/Context';

const logger = require('@ebay/logging-inc').logger(
    'openfeature-ebay:FeatureFlagEventUtils',
);

/**
 * Constructs metrics data from evaluation results and flag keys.
 * Maps flag keys to their variant IDs, with null for missing or unavailable flags.
 *
 * @param evaluationResults - Map of flag evaluation results
 * @param flagKeys - Set of flag keys to process
 * @returns Map of flag keys to variant IDs (or null if not available)
 */
export function constructMetricsData(
    evaluationResults: Map<string, FlagEvalDetail>, 
    flagKeys: Set<string>
): Map<string, number | null> {
    const flagMetricsMap = new Map<string, number | null>();

    if (evaluationResults) {
        evaluationResults.forEach((flagDetails, flagName) => {
            // only add to flagMetricsMap if the flag was invoked and stored in flagKeys
            if (flagKeys.has(flagName)) {
                const variantId = flagDetails?.details?.variantId || null;
                flagMetricsMap.set(flagName, variantId);
            }
        });
    }
    return flagMetricsMap;
}

/**
 * Constructs feature flag CAL events from variant ID map and error context.
 * The command name should always be present at this point because it is referencing req.route.name, otherwise default is 'unset'
 * The feature flag key should always have a value because the feature flag key is iterated through by the active flag keys used in this request.
 * The feature flag variant id may be null, and will be properly handled to avoid string "null" values.
 * 
 * @param fFlagVariantIdMap - Map of flag names to variant IDs
 * @param commandName - The command/route name
 * @param loggingContextError - Error context containing error counts and error count map
 * @returns Array of FeatureFlagRawEvent objects
 */
export function constructFeatureFlagEvent(
    fFlagVariantIdMap: Map<string, number | null>, 
    commandName: string, 
    loggingContextError: any
): FeatureFlagRawEvent[] {
    const events: FeatureFlagRawEvent[] = [];

    if (fFlagVariantIdMap) {
        fFlagVariantIdMap.forEach((variantId, flagName) => {
            const calEvent: FeatureFlagCalEvent = {
                commandName: commandName, 
                featureFlagKey: flagName,
                featureFlagErrorCount: loggingContextError.errorCount,
                variantId: String(variantId) !== "null" ? String(variantId) : null,
            };
            
            const event: FeatureFlagRawEvent = {
                calErrorCountMap: loggingContextError.errorCountMap,
                calEvent
            };
            events.push(event);
        });
    }
    return events;
}

/**
 * Sets error metrics to the tracing data holder for later CAL publishing.
 * This function orchestrates the construction of metrics data and feature flag events,
 * then stores them in the TracingDataHolder for the given trace ID.
 *
 * @param context - The OpenFeature context containing evaluation results and flag keys
 * @param compositeKey - The trace ID used as a key for storing the data
 * @param loggingContextError - Error context containing error counts and error count map
 * @param commandName - The command/route name
 */
export function setErrorMetricsToTracingDataHolder(
    context: Context, 
    compositeKey: string, 
    loggingContextError: any, 
    commandName: string
): void {
    const fFlagVariantIdMap = constructMetricsData(
        context.getEvaluationResults(), 
        context.getFeatureFlagKeys()
    );
    
    const featureFlagRawEvents: FeatureFlagRawEvent[] = constructFeatureFlagEvent(
        fFlagVariantIdMap, 
        commandName, 
        loggingContextError
    );

    logger.info("Size of feature flag raw events:" + featureFlagRawEvents.length);
    TracingDataHolder.addEntry(compositeKey, featureFlagRawEvents);
}
