// @ebay/openfeature-ebay/src/openfeature-lib/tracing/FeatureFlagSpanProcessorManager.ts
/**
 * Global manager for FeatureFlagSpanProcessor
 * 
 * This singleton provides global access to the FeatureFlagSpanProcessor instance,
 * allowing middleware and other components to access trace error information.
 */

import { FeatureFlagErrorInfo } from '../types/FeatureFlagErrorInfo';
import { FeatureFlagSpanProcessor } from './FeatureFlagSpanProcessor';

class FeatureFlagSpanProcessorManager {
  private static instance: FeatureFlagSpanProcessorManager;
  private processor: FeatureFlagSpanProcessor | null = null;

  private constructor() {}

  public static getInstance(): FeatureFlagSpanProcessorManager {
    if (!FeatureFlagSpanProcessorManager.instance) {
      FeatureFlagSpanProcessorManager.instance = new FeatureFlagSpanProcessorManager();
    }
    return FeatureFlagSpanProcessorManager.instance;
  }

  /**
   * Register the FeatureFlagSpanProcessor instance globally
   */
  public setProcessor(processor: FeatureFlagSpanProcessor): void {
    this.processor = processor;
  }

  /**
   * Get the registered FeatureFlagSpanProcessor instance
   */
  public getProcessor(): FeatureFlagSpanProcessor | null {
    return this.processor;
  }

  /**
   * Get logging context error from traceId
   * It counts up the number of errors by each error name.
   */
  public getLoggingContextError(traceId: string) {
    const loggingContextError: {
      errorCountMap: Record<string, number>;
      errorCount: number;
    } = {
      errorCountMap: {},
      errorCount: 0
    };
    
    // Add null safety check for processor
    if (!this.processor) {
      return loggingContextError;
    }
    
    const traceIdMap: Map<string, FeatureFlagErrorInfo> = this.processor.getTraceIdMap();

    if (traceIdMap != null && traceId != null) {
      const errorInfo: FeatureFlagErrorInfo = traceIdMap.get(traceId);
      if (errorInfo != null && errorInfo.getErrorEvents() != null) {
        for (const [key, value] of errorInfo.getErrorEvents()) {
          loggingContextError.errorCountMap[key] = value;
        }
        loggingContextError.errorCount = errorInfo.getTotalErrorCount();
      }
    }
    return loggingContextError;
  }

  /**
   * Removes the trace from trace id map
   * @param traceId
   */
  public removeTrace(traceId: string): void {
    if (this.processor) {
      this.processor.removeTrace(traceId);
    }
  }
}

export const featureFlagSpanManager = FeatureFlagSpanProcessorManager.getInstance();
export { FeatureFlagSpanProcessorManager };
