// @ebay/commons-ebay/lib/cos/c-request-id.js
'use strict';

var GUIDGenerator = require('../guid-generator');

var guidGenerator = new GUIDGenerator();

var hmvUtils = require('../header-multi-value'),
    CONST = {
        X_EBAY_C_REQUEST_ID: 'X-EBAY-C-REQUEST-ID',
        REQUEST_ID: 'ri',
        CORRELATION_REQUEST_ID: 'rci',
        NODE_ID: 'node_id'
    };

function CRequestId(req) {
    var ctx = hmvUtils.parse(req.headers[CONST.X_EBAY_C_REQUEST_ID] ||
        req.headers[CONST.X_EBAY_C_REQUEST_ID.toLowerCase()], true, true);

    //ri, rci, node_id specs
    // https://wiki.vip.corp.ebay.com/pages/viewpage.action?spaceKey=GlobalPF&title=Distributed+Tracing+Across+Framework
    // https://docs.google.com/document/d/1nfWcDP10LOwvb3rzjpDQcU4K1g2OteyiD6grbeEw854/edit#

    this.requestId = ctx.get(CONST.REQUEST_ID);
    this.nodeId = guidGenerator.nextPaddedGUID();    
    this.corrRequestId = ctx.get(CONST.CORRELATION_REQUEST_ID) || guidGenerator.nextPaddedGUID();
}

CRequestId.prototype = {
    buildHeader: function buildHeader() {
        console.warn('Deprecated CRequestId.buildHeader()');
    },

    update: function update() {
        console.warn('Deprecated CRequestId.update()');
    },

    getRequestId: function getRequestId() {
        return this.requestId;
    },

    getCorrelationRequestId: function getCorrelationRequestId() {
        return this.corrRequestId;
    },

    getNodeId: function getNodeId() {
        return this.nodeId;
    }

};

module.exports = CRequestId;

module.exports.create = function(req) {
    return new CRequestId(req);
};
