// @ebay/commons-ebay/lib/cos/c-tracking.js
'use strict';

var hmvUtils = require('../header-multi-value'),
    CONST = {
        X_EBAY_C_TRACKING: 'X-EBAY-C-TRACKING',
        GUID: 'guid',
        CORRELATION_GUID: 'cguid',
        TRACKING_GUID: 'tguid',
        BEST_USER_ID: 'buid',
        PAGE_ID: 'pageid',
        ROVER_MEDIAPLEX: 'roverMplx',
        PARTNER_ID: 'cobrandId',
        CACHE_KEY: 'cacheKey',
        CACHE_MODE: 'cacheMode',
        CACHE_TTL: 'cacheTTL',
        DEVICE_TIMESTAMP: 'devicetimestamp',
        EBAY_DAA_OPTOUT: 'eBayDAAOptout'
    };

function extractEbayDAAOptout(encodedString) {
    try {
        // Decode the base64url encoded string - RFC 4648 sec 5.
        const decoded = Buffer.from(encodedString, 'base64');

        if (decoded.length < 7) {
            return false;
        }

        const decodedByteString = Array.from(decoded)
            .map(byte => byte.toString(2).padStart(8, '0'))
            .join('');

        // Skip first 38 bits and extract the next 4 bits
        const globalOptOutBits = decodedByteString.slice(38, 42);
        if (globalOptOutBits === '0000') {
            return true;
        }

        // Extract the next 12 bits for number of participants
        const numberOfParticipantsBits = decodedByteString.slice(42, 54);
        const numberOfParticipants = parseInt(numberOfParticipantsBits, 2);

        var currentIndex = 54;
        for (var idx = 0; idx < numberOfParticipants; idx++) {
            // Extract 12 bits for participant code
            const participantCodeBits = decodedByteString.slice(currentIndex, currentIndex + 12);
            const participantCode = parseInt(participantCodeBits, 2);
            currentIndex += 12;

            // Check if participant code is eBAY - 00681
            if (participantCode === 681) {
                // Extract next 4 bits
                const participantOptOutBits = decodedByteString.slice(currentIndex, currentIndex + 4);
                if (participantOptOutBits === '0000') {
                    return true;
                }
            }
            currentIndex += 4;
        }

        return false;
    } catch (error) {
        return false;
    }
}

function CTracking(req) {

    var ctx = hmvUtils.parse(req.headers[CONST.X_EBAY_C_TRACKING] ||
        req.headers[CONST.X_EBAY_C_TRACKING.toLowerCase()], false, true);

    this.req = req;
    this.guid = ctx.get(CONST.GUID);
    this.cguid = ctx.get(CONST.CORRELATION_GUID);
    this.tguid = ctx.get(CONST.TRACKING_GUID);
    this.buid = ctx.get(CONST.BEST_USER_ID);
    this.pageid = ctx.get(CONST.PAGE_ID);
    this.roverMplx = ctx.get(CONST.ROVER_MEDIAPLEX);
    this.cobrandId = ctx.get(CONST.PARTNER_ID);
    this.cacheKey = ctx.get(CONST.CACHE_KEY);
    
    if (ctx.get(CONST.CACHE_MODE)) {
        this.cacheMode = ctx.get(CONST.CACHE_MODE);
    } else if (this.isCacheRequest()) {
        this.cacheMode = 'cache';
    }

    if(ctx.get(CONST.DEVICE_TIMESTAMP)) {
        this.devicetimestamp = ctx.get(CONST.DEVICE_TIMESTAMP);
    }

    var headers = this.req && this.req.headers;
    if (headers) {
        var adChoiceHeader = headers['x-adchoices'] || headers['cookie2'];
        var adChoicesFromTrackingHeader = ctx.get(CONST.EBAY_DAA_OPTOUT);
        var daaOptout = false;
        if (adChoiceHeader) {
            daaOptout = extractEbayDAAOptout(adChoiceHeader); 
        } else if (adChoicesFromTrackingHeader) {
            daaOptout = adChoicesFromTrackingHeader === 'true';
        }
        this.eBayDAAOptout = daaOptout;
    }

    this.cacheTTL = ctx.get(CONST.CACHE_TTL);
}

CTracking.prototype = {
    buildHeader: function buildHeader() {
        var headerVal = hmvUtils.create();
        var ebay = this.req.ebay;
        headerVal.set(CONST.GUID, this.guid || ebay.getGuid());
        headerVal.set(CONST.CORRELATION_GUID, this.cguid || ebay.getDCCorrelationGuid());
        headerVal.set(CONST.TRACKING_GUID, this.tguid || ebay.getTrackingGuid());
        headerVal.set(CONST.BEST_USER_ID, this.buid || ebay.getBestUserId());
        headerVal.set(CONST.PAGE_ID, this.pageid || ebay.getPageId());
        headerVal.set(CONST.ROVER_MEDIAPLEX, this.roverMplx || ebay.getRoverMediaplex());
        headerVal.set(CONST.PARTNER_ID, this.cobrandId || ebay.getPartnerId());
        headerVal.set(CONST.CACHE_KEY, this.cacheKey);
        headerVal.set(CONST.CACHE_MODE, this.cacheMode);
        headerVal.set(CONST.CACHE_TTL, this.cacheTTL);

        if(this.devicetimestamp) {
            headerVal.set(CONST.DEVICE_TIMESTAMP, this.devicetimestamp);
        }

        if(this.eBayDAAOptout && this.eBayDAAOptout === true) {
            headerVal.set(CONST.EBAY_DAA_OPTOUT, 'true');
        }

        return headerVal.toString();
    },

    isCacheRequest() {
        const req = this.req;
        return req.headers && req.headers['ufes-nonuser-context'] === 'true';
    },

    update: function update(outboundContext) {
        outboundContext[CONST.X_EBAY_C_TRACKING] = this.buildHeader();
    },

    getGuid: function getGuid() {
        return this.guid;
    },

    getCorrelationGuid: function getCorrelationGuid() {
        return this.cguid;
    },

    getTrackingGuid: function getTrackingGuid() {
        return this.tguid;
    },

    getBestUserId: function getBestUserId() {
        return this.buid;
    },

    getPageId: function getPageId() {
        return this.pageid;
    },

    getRoverMediaplex: function getRoverMediaplex() {
        return this.roverMplx;
    },

    getPartnerId: function getPartnerId() {
        return this.cobrandId;
    },

    getCacheKey: function getCacheKey() {
        return this.cacheKey;
    },

    getCacheMode: function getCacheMode() {
        return this.cacheMode;
    },

    getCacheTTL: function getCacheTTL() {
        return this.cacheTTL;
    },
    getDeviceTimeStamp: function getDeviceTimeStamp() {
        return this.devicetimestamp;
    },
    getEBayDAAOptout: function getEBayDAAOptout() {
        return this.eBayDAAOptout;
    }
};

module.exports = CTracking;

module.exports.create = function(req) {
    return new CTracking(req);
};

module.exports.extractEbayDAAOptout = extractEbayDAAOptout;
