// @ebay/commons-ebay/lib/cos/edp.js
'use strict';
const Cal = require('@ebay/cal');

var hmvUtils = require('../header-multi-value'),
    CONST = {
        CORR_ID: 'corr_id',
        CHANNEL: 'channel',
        HEADER_NAME: 'x-ebay-edp'
    };

function tryGetGuid(webRequest) {
    return webRequest.ebay?.guidGenerator?.nextPaddedGUID();
}

class Edp {
    constructor(webRequest) {
        const value = hmvUtils.parse(webRequest.headers[CONST.HEADER_NAME], false, true);
        this.corrId = value.get(CONST.CORR_ID);
        this.channel = value.get(CONST.CHANNEL);
        this.webRequest = webRequest;
    }

    buildHeader() {
        const headerVal = hmvUtils.create();
        headerVal.set(CONST.CORR_ID, this.corrId || tryGetGuid(this.webRequest));
        headerVal.set(CONST.CHANNEL, this.channel ||
            (this.webRequest.ebay?.isService ?
                this.webRequest.security?.getTokenAttributes()?.attributes?.type === 'ajax' && 'WEB' || 'NATIVE' :
                'WEB')
        );
        return headerVal.toString();
    }

    update(outboundContext) {
        const headerStr = this.buildHeader();
        Cal.createEvent('edp', 'header').addData(headerStr).complete();
        outboundContext[CONST.HEADER_NAME] = headerStr;
    }
}


module.exports.Edp = Edp;

module.exports.create = function (webRequest) {
    return new Edp(webRequest);
};
