// @ebay/commons-ebay/lib/cos/location.js
'use strict';

const hmvUtils = require('../header-multi-value');
const _ = require('underscore');
const CONST = {
    LOCATION_COUNTRY: 'country',
    LOCATION_STATE: 'state',
    LOCATION_ZIP: 'zip',
    LOCATION_LATITUDE: 'latitude',
    LOCATION_LONGITUDE: 'longitude'
};
const countryCodes = require('../constants/country-codes');

function toUpperCase(val) {
    if (val) {
        val = val.toUpperCase();
    }
    return val;
}

class LocationContext {
    constructor(multiValue, req) {
        const value = hmvUtils.parse(multiValue, true, true);

        this.country = toUpperCase(value.get(CONST.LOCATION_COUNTRY));
        this.state = toUpperCase(value.get(CONST.LOCATION_STATE));
        this.zip = value.get(CONST.LOCATION_ZIP);
        this.latitude = parseFloat(value.get(CONST.LOCATION_LATITUDE));
        this.longitude = parseFloat(value.get(CONST.LOCATION_LONGITUDE));
    
        this.req = req;
        this.validateOrDefault();
    }

    getState() {
        return this.state;
    }

    setState(state) {
        this.state = state;
    }

    getCountry() {
        return this.country;
    }

    setCountry(country) {
        if (!isValidCountry(country)) {
            throw new Error('Invalid country, country: ' + country);
        }
        this.country = country.toUpperCase();
    }

    getZip() {
        return this.zip;
    }

    setZip(zip) {
        this.zip = zip;
    }

    getLatitude() {
        return this.latitude;
    }

    setLatitude(latitude) {
        if (!isValidLatitude(latitude)) {
            throw new Error('Invalid latitude value');
        }
        this.latitude = latitude;
    }

    getLongitude() {
        return this.longitude;
    }

    setLongitude(longitude) {
        if (!isValidLongitude(longitude)) {
            throw new Error('Invalid longitude value');
        }
        this.longitude = longitude;
    }

    validateOrDefault() {
        if (!isValidCountry(this.country)) {
            if (this.state || this.zip) {
                throw new Error('Invalid country detected, state: ', this.state, ', zip: ', this.zip);
            }
            this.country = this.req.locality && this.req.locality.country || 'US';
        }

        if (!isValidLatitude(this.latitude) || !isValidLongitude(this.longitude)) {
            this.latitude = this.longitude = null;
        }
    }

    buildHeader() {
        var headerVal = hmvUtils.create();
        headerVal.set(CONST.LOCATION_COUNTRY, this.country);
        headerVal.set(CONST.LOCATION_STATE, this.state);
        headerVal.set(CONST.LOCATION_ZIP, this.zip);
        headerVal.set(CONST.LOCATION_LATITUDE, this.latitude);
        headerVal.set(CONST.LOCATION_LONGITUDE, this.longitude);

        return headerVal.toString();
    }    
}

function isValidCountry(country) {
    return country && countryCodes.valueOf(country.toUpperCase()) !== undefined;
}

function isValidLongitude(longitude) {
    if (!_.isNumber(longitude)) {
        longitude = parseFloat(longitude);
    }
    return longitude > -180 && longitude < 180;
}

function isValidLatitude(latitude) {
    if (!_.isNumber(latitude)) {
        latitude = parseFloat(latitude);
    }
    return latitude > -90 && latitude < 90;
}

module.exports = {
    LocationContext
};
