// @ebay/commons-ebay/lib/cos/shipping-awareness.js
'use strict';

var CONST = {

        EBAY_ETS_SHIP_TO_LOCATION: 'ebay-ets-ship-to-location',
        PERSISTENT_DESTINATION_POSTALCODE: 'PERSISTENT_DESTINATION_POSTALCODE',
        COUNTRYCODE: 'countryCode',
        POSTALCODE: 'postalCode',
        COUNTRY_CODE_MAPPING: {
             'USA' : 'US',
             'CAN' : 'CA',
             'GBR' : 'GB',
             'AFG' : 'AF',
             'ALB' : 'AL',
             'DZA' : 'DZ',
             'ASM' : 'AS',
             'AND' : 'AD',
             'AGO' : 'AO',
             'AIA' : 'AI',
             'ATG' : 'AG',
             'ARG' : 'AR',
             'ARM' : 'AM',
             'ABW' : 'AW',
             'AUS' : 'AU',
             'AUT' : 'AT',
             'AZE' : 'AZ',
             'BHS' : 'BS',
             'BHR' : 'BH',
             'BGD' : 'BD',
             'BRB' : 'BB',
             'BLR' : 'BY',
             'BEL' : 'BE',
             'BLZ' : 'BZ',
             'BEN' : 'BJ',
             'BMU' : 'BM',
             'BTN' : 'BT',
             'BOL' : 'BO',
             'BIH' : 'BA',
             'BWA' : 'BW',
             'BRA' : 'BR',
             'VGB' : 'VG',
             'BRN' : 'BN',
             'BGR' : 'BG',
             'BFA' : 'BF',
             'MMR' : 'MM',
             'BDI' : 'BI',
             'KHM' : 'KH',
             'CMR' : 'CM',
             'CPV' : 'CV',
             'CYM' : 'KY',
             'CAF' : 'CF',
             'TCD' : 'TD',
             'CHL' : 'CL',
             'CHN' : 'CN',
             'COL' : 'CO',
             'COM' : 'KM',
             'COD' : 'CD',
             'COG' : 'CG',
             'COK' : 'CK',
             'CRI' : 'CR',
             'CIV' : 'CI',
             'HRV' : 'HR',
             'CUB' : 'CU',
             'CYP' : 'CY',
             'CZE' : 'CZ',
             'DNK' : 'DK',
             'DJI' : 'DJ',
             'DMA' : 'DM',
             'DOM' : 'DO',
             'ECU' : 'EC',
             'EGY' : 'EG',
             'SLV' : 'SV',
             'GNQ' : 'GQ',
             'ERI' : 'ER',
             'EST' : 'EE',
             'ETH' : 'ET',
             'FLK' : 'FK',
             'FJI' : 'FJ',
             'FIN' : 'FI',
             'FRA' : 'FR',
             'GUF' : 'GF',
             'PYF' : 'PF',
             'GAB' : 'GA',
             'GMB' : 'GM',
             'GEO' : 'GE',
             'DEU' : 'DE',
             'GHA' : 'GH',
             'GIB' : 'GI',
             'GRC' : 'GR',
             'GRL' : 'GL',
             'GRD' : 'GD',
             'GLP' : 'GP',
             'GUM' : 'GU',
             'GTM' : 'GT',
             'GGY' : 'GG',
             'GIN' : 'GN',
             'GNB' : 'GW',
             'GUY' : 'GY',
             'HTI' : 'HT',
             'HND' : 'HN',
             'HKG' : 'HK',
             'HUN' : 'HU',
             'ISL' : 'IS',
             'IND' : 'IN',
             'IDN' : 'ID',
             'IRN' : 'IR',
             'IRQ' : 'IQ',
             'IRL' : 'IE',
             'ISR' : 'IL',
             'ITA' : 'IT',
             'JAM' : 'JM',
             'JPN' : 'JP',
             'JEY' : 'JE',
             'JOR' : 'JO',
             'KAZ' : 'KZ',
             'KEN' : 'KE',
             'KIR' : 'KI',
             'PRK' : 'KP',
             'KOR' : 'KR',
             'KWT' : 'KW',
             'KGZ' : 'KG',
             'LAO' : 'LA',
             'LVA' : 'LV',
             'LBN' : 'LB',
             'LSO' : 'LS',
             'LBR' : 'LR',
             'LBY' : 'LY',
             'LIE' : 'LI',
             'LTU' : 'LT',
             'LUX' : 'LU',
             'MAC' : 'MO',
             'MKD' : 'MK',
             'MDG' : 'MG',
             'MWI' : 'MW',
             'MYS' : 'MY',
             'MDV' : 'MV',
             'MLI' : 'ML',
             'MLT' : 'MT',
             'MHL' : 'MH',
             'MTQ' : 'MQ',
             'MRT' : 'MR',
             'MUS' : 'MU',
             'MYT' : 'YT',
             'MEX' : 'MX',
             'MDA' : 'MD',
             'MCO' : 'MC',
             'MNG' : 'MN',
             'MSR' : 'MS',
             'MAR' : 'MA',
             'MOZ' : 'MZ',
             'NAM' : 'NA',
             'NRU' : 'NR',
             'NPL' : 'NP',
             'NLD' : 'NL',
             'ANT' : 'AN',
             'NCL' : 'NC',
             'NZL' : 'NZ',
             'NIC' : 'NI',
             'NER' : 'NE',
             'NGA' : 'NG',
             'NIU' : 'NU',
             'NOR' : 'NO',
             'OMN' : 'OM',
             'PAK' : 'PK',
             'PLW' : 'PW',
             'PAN' : 'PA',
             'PNG' : 'PG',
             'PRY' : 'PY',
             'PER' : 'PE',
             'PHL' : 'PH',
             'POL' : 'PL',
             'PRT' : 'PT',
             'PRI' : 'PR',
             'QAT' : 'QA',
             'ROU' : 'RO',
             'RUS' : 'RU',
             'RWA' : 'RW',
             'SHN' : 'SH',
             'KNA' : 'KN',
             'LCA' : 'LC',
             'SPM' : 'PM',
             'VCT' : 'VC',
             'SMR' : 'SM',
             'SAU' : 'SA',
             'SEN' : 'SN',
             'SCG' : 'CS',
             'SRB' : 'RS',
             'MNE' : 'ME',
             'SYC' : 'SC',
             'SLE' : 'SL',
             'SGP' : 'SG',
             'SVK' : 'SK',
             'SVN' : 'SI',
             'SLB' : 'SB',
             'SOM' : 'SO',
             'ZAF' : 'ZA',
             'ESP' : 'ES',
             'LKA' : 'LK',
             'SDN' : 'SD',
             'SUR' : 'SR',
             'SJM' : 'SJ',
             'SWZ' : 'SZ',
             'SWE' : 'SE',
             'CHE' : 'CH',
             'SYR' : 'SY',
             'TWN' : 'TW',
             'TJK' : 'TJ',
             'TZA' : 'TZ',
             'THA' : 'TH',
             'TGO' : 'TG',
             'TON' : 'TO',
             'TTO' : 'TT',
             'TUN' : 'TN',
             'TUR' : 'TR',
             'TKM' : 'TM',
             'TCA' : 'TC',
             'TUV' : 'TV',
             'UGA' : 'UG',
             'UKR' : 'UA',
             'ARE' : 'AE',
             'URU' : 'UY',
             'UZB' : 'UZ',
             'VUT' : 'VU',
             'VAT' : 'VA',
             'VEN' : 'VE',
             'VNM' : 'VN',
             'VIR' : 'VI',
             'WLF' : 'WF',
             'ESH' : 'EH',
             'WSM' : 'WS',
             'YEM' : 'YE',
             'ZMB' : 'ZM',
             'ZWE' : 'ZW',
             'FSM' : 'FM',
             'REU' : 'RE'
        }
    };
CONST.COUNTRY_CODE_REVERSE_MAPPING = Object.keys(CONST.COUNTRY_CODE_MAPPING).reduce((memo, key) => {
             memo[CONST.COUNTRY_CODE_MAPPING[key]] = key;
             return memo;
}, {});
var hmvUtils = require('../header-multi-value');

function ShippingAwareness(req) {
     if(req.ebay.isService) {
        const code = req && req.headers &&  
            req.headers[CONST.EBAY_ETS_SHIP_TO_LOCATION];
        if(code) {
            this.ctx = hmvUtils.create(code, false);
            this.countryCode = this.ctx.get(CONST.COUNTRYCODE) ;
            this.postalCode = this.ctx.get(CONST.POSTALCODE);
         } 
     }
    else {
         const code = req && req.ebay && req.ebay.cookies &&
            req.ebay.cookies.getCookieValue(CONST.PERSISTENT_DESTINATION_POSTALCODE) ;
            if(code) {
              if(code.indexOf(',') !== -1) {
                    var arr = code.split(',');
                    this.setCountryCode(arr[1]);
                    if(this.isPostalCode(arr[0])) {
                        this.postalCode = arr[0];
                    }
              }
              else {
                    if(this.isPostalCode(code)) {
                        this.postalCode = code;
                    }
              }
          }
    }
 }

ShippingAwareness.prototype = {
    buildHeader: function buildHeader() {
        var headerVal = hmvUtils.create();    
        var flag = false ;
        if(this.countryCode) {
            flag = true ;
            headerVal.set(CONST.COUNTRYCODE, this.countryCode);
        }

        if(this.postalCode) {
            flag = true ;
            headerVal.set(CONST.POSTALCODE, this.postalCode);
        }
        if(flag) {
            return  headerVal.toString();
              
        } 
        return ;
    },

    getCountryCode: function() {

        return this.countryCode;
    },
    
    setCountryCode: function(code) {
        if(this.isCountry(code)) {
            if(code.length === 3) {
                this.countryCode =  CONST.COUNTRY_CODE_MAPPING[code];
            }
            else if(code.length === 2) {
                if(CONST.COUNTRY_CODE_REVERSE_MAPPING[code]) {
                    this.countryCode = code ;
                }
            }
        }
    },

    getPostalCode:function() {
        
        return this.postalCode;
    },

    setPostalCode:function(code) {
        if(this.isPostalCode(code)) {
            this.postalCode = code;
        } 
    },

    update:function(outboundContext) {
        const val = this.buildHeader() ;
         if(val) {
            outboundContext[CONST.EBAY_ETS_SHIP_TO_LOCATION] = val;
        }
    },

    isPostalCode:function(val) {
        return !/^$|\s+/.test(val);
    },

    isCountry:function(val) {
        return /^[A-Z]+$/.test(val);
    }
};

module.exports = ShippingAwareness;
module.exports.create = function(req) {
    return new ShippingAwareness(req);
};
