// @ebay/commons-ebay/lib/cos/user-id.js
'use strict';
var hmvUtils = require('../header-multi-value');
const ACCT_ID = 'acctId',
    ACCT_ROLE = 'acctRole',
    ACCT_NAME = 'acctName',
    LOGIN_ID = 'loginId',
    LOGIN_NAME = 'loginName';


class UserId {
    constructor(multiHeader, req) {
        this.ctx = hmvUtils.create(multiHeader, false);
        if(req.ebay.isService) {
            this.acctId = this.ctx.get(ACCT_ID); 
            this.acctRole = this.ctx.get(ACCT_ROLE);
            this.acctName = this.ctx.get(ACCT_NAME);
            this.loginId = this.ctx.get(LOGIN_ID);
            this.loginName = this.ctx.get(LOGIN_NAME);
        } else {
            let multiAccountLoginCtx = req.ebay.multiAccountLoginCtx;
            if(multiAccountLoginCtx && multiAccountLoginCtx.isUserAuthorized()) {
                this.acctId = multiAccountLoginCtx.getAcctId();
                this.acctName = multiAccountLoginCtx.getAcctName();
                this.loginId = multiAccountLoginCtx.getLoginId();
                this.loginName = multiAccountLoginCtx.getLoginName();
            } else {
                this.acctId = req.ebay.getPersistentVisitorAccountId(); 
                this.acctRole = req.ebay.isEmailVisitor() ? 'visitor': undefined;
            }
        }
    }

    buildHeader() {
        if(this.acctId && this.loginId){
            let headermv = hmvUtils.create();
            headermv.set(ACCT_ID, this.acctId);
            if(this.acctName) {
                headermv.set(ACCT_NAME, this.acctName);
            }
            headermv.set(LOGIN_ID, this.loginId);
            if(this.loginName) {
                headermv.set(LOGIN_NAME, this.loginName);
            }
            return headermv.toString();
        } else if(this.acctRole && this.acctId) {
            let headermv = hmvUtils.create();
            headermv.set(ACCT_ID, this.acctId);
            headermv.set(ACCT_ROLE, this.acctRole);
            return headermv.toString();
        }        
    }
}

module.exports.create = function(multiHeader, req) {
    return new UserId(multiHeader, req);
};
