// @ebay/commons-ebay/lib/cos/user-cultural-pref-context.js
'use strict';

var svcClient = require('@ebay/service-client-ebay'),
    hmvUtils = require('../header-multi-value'),
    currenyList = require('../../resources/currency-list.json').list,
    ONE_DAY = 86400000,
    lastUpdated,
    CONST = {
        USER_CULTURAL_PREF_CURRENCY: 'currency',
        USER_CULTURAL_PREF_TIMEZONE: 'timezone',
        USER_CULTURAL_PREF_UNITS: 'units',
        X_EBAY_C_CULTURAL_PREF: 'X-EBAY-C-CULTURAL-PREF'
    },
    UNITS = {
        UNITS_US: 'US',
        UNITS_METRIC: 'METRIC'
    };

/*
 * Header is a request header value of 'X-EBAY-C-CULTURAL-PREF'
 * The context should be resolved only after GEO context as well as siteid/country has been resolved
 */
function UserCulturalPrefContext(req) {
    this.req = req;
}

function isValidUnits(units) {
    units = units && units.toUpperCase();
    return units && (UNITS.UNITS_US === units || UNITS.UNITS_METRIC === units);
}

function validateOrDefault(ctx) {
    if (!isValidUnits(ctx.units)) {
        ctx.units = null;
    }

    if (!ctx.getTimezone() && ctx.req.locality) {
        ctx.setTimezone(ctx.req.locality.timeZone);
    }
}

UserCulturalPrefContext.prototype = {
    getTimezone: function() {
        return this.timezone || this.req.locality && this.req.locality.timeZone;
    },

    getUnits: function() {
        return this.units;
    },

    getCurrency: function() {
        return this.currency;
    },

    setTimezone: function(timezone) {
        if (timezone) {
            this.timezone = timezone;
        }
    },

    setUnits: function(units) {
        if (isValidUnits(units)) {
            this.units = units.toUpperCase();
        }
    },

    setCurrency: function(currency) {
        this.currency = currency;
    },

    buildHeader: function() {
        validateOrDefault(this);
        var headerVal = hmvUtils.create();
        headerVal.set(CONST.USER_CULTURAL_PREF_CURRENCY, this.currency && this.currency.code);
        headerVal.set(CONST.USER_CULTURAL_PREF_UNITS, this.getUnits());
        headerVal.set(CONST.USER_CULTURAL_PREF_TIMEZONE, this.getTimezone());
        return headerVal.toString();
    },

    update: function(outboundContext) {
        outboundContext[CONST.X_EBAY_C_CULTURAL_PREF] = this.buildHeader();
    }
};

function getCurrencyList(req, callback) {
    var client;
    if (arguments.length === 1) {
        callback = arguments[arguments.length - 1];
        console.log(new Error('WARN: request object is missing, please update getCurrencyList call'));
        client = svcClient.getClient('currency-list-svc');
    }
    else {
        client = svcClient.context(req).getClient('currency-list-svc');
    }
    client.get().end(function(err, data) {
        callback(err, data ? data.body : data);
    });
}

function assignCurrencyList(err, data) {
    if (!err && data && data.list) {
        currenyList = data.list;
    }
}

function create(req) {
    if (!lastUpdated || (lastUpdated + ONE_DAY < Date.now())) {
        lastUpdated = Date.now();
        setImmediate(getCurrencyList.bind(null, req, assignCurrencyList));
    }
    var ucpCtx = new UserCulturalPrefContext(req);

    ucpCtx.reqCtx = req.ebay;

    var currencyValue;
    if (req.ebay.isService) {
        var ctx = hmvUtils.parse(req.headers[CONST.X_EBAY_C_CULTURAL_PREF] ||
            req.headers[CONST.X_EBAY_C_CULTURAL_PREF.toLowerCase()], true, true);
        ucpCtx.setTimezone(ctx.get(CONST.USER_CULTURAL_PREF_TIMEZONE));
        ucpCtx.setUnits(ctx.get(CONST.USER_CULTURAL_PREF_UNITS));

        currencyValue = ctx.get(CONST.USER_CULTURAL_PREF_CURRENCY);
    }
    else {
        // web flow, read from request
        if (req.locality) {
            ucpCtx.setTimezone(req.locality.timeZone);
            currencyValue = req.locality.isoCurrencyCode;
        }
    }
    if (currencyValue) {
        var currency = currenyList[currencyValue.toUpperCase()];
        ucpCtx.setCurrency(currency);
    }
    return ucpCtx;
}

module.exports.UserCulturalPrefContext = UserCulturalPrefContext;

module.exports.create = create;

module.exports.getCurrencyList = getCurrencyList;
