// @ebay/cookies-ebay/lib/config.js
'use strict';

const { promisify } = require('util');
const loadConfig = promisify(require('@ebay/module-config-inc'));
const Provider = require('@ebay/service-client-ebay');
const configbean = require('@ebay/config-bean-ebay');
const logger = require('@ebay/logging-inc').logger('cookies-ebay/config');

function loadAuthzCookieConfig() {
    return new Promise((resolve, reject) => {
        Provider.context({}).getClient('authzCookies')
            .get().end((err, response) => {
                if (err) {
                    return reject(err);
                }
                resolve(response.body);
            });
    });
}

function persistToConfigBean(config) {
    const bean = configbean.getBeanById('nodejs.config.cookies-ebay.common');

    if (bean) {
        const keys = Object.keys(config);
        keys.forEach(key => {
            let val = config[key];
            if (key === 'blackListedCookies' || key === 'httpOnlyCookies' || key === 'unmanagedCookiesAllowList') {
                val = (val || '').split(',').map(item => item.trim());
            }
            bean.setPersist(`cookies-ebay:${key}`, val);
        });

    }
}

function createCookieConfigProvider(overrideInterval) {
    let interval;
    let cookieConfig;
    let timer;
    const defaultInterval = 15 * 60 * 1000; // 15 mins
    const refreshCookieConfig = async () => {
        try {
            cookieConfig = cookieConfig || await internals
                .loadConfig(module).then(cfg => cfg.get('cookies-ebay'));
            const authzConfig = await internals.loadAuthzCookieConfig();
            // we are going to persist authz config to
            // the config bean of our configuration
            // if skipUpdate is true, do not persist the cookie
            if (cookieConfig.skipUpdate !== true)
            {
                persistToConfigBean(authzConfig);
            }
        }
        catch (err) {
            logger.error(`Failed to refresh cookies config`, err);
        }
        finally {
            interval = overrideInterval || cookieConfig && cookieConfig.refreshInterval || defaultInterval;
            timer && clearTimeout(timer);
            timer = setTimeout(refreshCookieConfig, interval).unref();
        }
    };

    return {
        config() {
            if (!timer) {
                console.error(`[ERROR] The config load cycle must be started`);
            }
            return cookieConfig || {};
        },

        start() {
            return !this.isStarted() && refreshCookieConfig();
        },

        stop() {
            timer && clearTimeout(timer);
            timer = undefined;
        },

        isStarted() {
            return !!timer;
        }
    };
}

module.exports = createCookieConfigProvider;
const internals = module.exports[Symbol.for('internals')] = {
    loadAuthzCookieConfig,
    loadConfig
};
