// @ebay/cookies-ebay/lib/EbayCookiesContext.js
'use strict';

var EbayCookieManager = require('./ebay-cookies/EbayCookieManager');
var metadata = require('./ebay-cookies/metadata');
var CookieAuditInstrumentation = require('./CookieAuditInstrumentation');
var cookieAuditInstrumentation = new CookieAuditInstrumentation();

function EbayCookiesContext(req, res) {
    this._request = req;
    this._response = res;
    this._cookieManager = null;
}

EbayCookiesContext.prototype = {
    getCookieManager: function() {
        if (this._cookieManager === null) {
            this._cookieManager = new EbayCookieManager(this._request, this._response);
        }
        return this._cookieManager;
    },
    setCookieValue: function (cookieId, value) {
        cookieAuditInstrumentation.updateMetrics(cookieId, 'send');
        this.getCookieManager().setCookieValue(cookieId, value);
    },
    getCookieBitNameFromId: function (cookieId) {
        return this.getCookieManager().getCookieBitNameFromId(cookieId);
    },
    getCookieValue: function (cookieId) {
        cookieAuditInstrumentation.updateMetrics(cookieId, 'read');
        return this.getCookieManager().getCookieValue(cookieId);
    },
    setFlag: function (flag, value) {
        return this.getCookieManager().setFlag(flag, value);
    },
    hasFlag: function(flag) {
        return this.getCookieManager().hasFlag(flag);
    },
    removeCookie: function (cookieId) {
        cookieAuditInstrumentation.updateMetrics(cookieId, 'delete');
        this.getCookieManager().removeCookie(cookieId);
    },
    get Cookies() {
        return metadata.Cookies;
    },
    get Constants() {
        return metadata.Constants;
    }
};

EbayCookiesContext.prototype.set = EbayCookiesContext.prototype.setCookieValue;
EbayCookiesContext.prototype.get = EbayCookiesContext.prototype.getCookieValue;
EbayCookiesContext.prototype.remove = EbayCookiesContext.prototype.removeCookie;

module.exports = EbayCookiesContext;
