// @ebay/cookies-ebay/lib/middleware.js
'use strict';
const onHeaders = require('on-headers');

const cal = require('@ebay/cal');
const util = require('@ebay/utils-ebay');

const sessionKey = require('./session/session-key');
const EbayCookiesContext = require('./EbayCookiesContext');

const modConfig = require('@ebay/module-config-inc');
const createConfigProvider = require('./config');
// will hold and start configProvider when requested
const configProvider = createConfigProvider();

module.exports = function () {
    return function cookiesEbay(req, res, next) {

        /*if (!isCookiesDevMounted) {
            //TODO need to fix this
            isCookiesDevMounted = true;
            if (req.app && deployEnv.isNotLive()) {
                require('./dev')(req.app);
            }
        }*/

        req.ebay = req.ebay || {};
        var cookies = req.ebay.cookies = new EbayCookiesContext(req, res);
        var ebayCookieManager = cookies.getCookieManager();

        cal.createTransaction('middleware', 'Cookie', function (err, tx) {
            var writingHeaders = false;
            tx.flush();

            if (!configProvider.isStarted()) {
                // this will take care of syncing config from 
                configProvider.start();
            }

            modConfig(module, (_, config) => {
                ebayCookieManager.setConfig(config.get('cookies-ebay'));
                sessionKey.warmUpSessionKeyCache(2, function (error, sessionKeyCache) {
                    //ignore the results
                    ebayCookieManager.validateAllCookies(function (_, results) {
                        // error will never happen, results will have true/false flag
                        if (req.query && req.query.redirect && req.query.redirect === 'mobile') {
                            cookies.setCookieValue('_fsom', {
                                path: '/',
                                value: 'y',
                                httpOnly: true,
                                domain: '.' + util.getBaseHost(req.headers.host),
                                maxAge: 60 * 60 * 24 * 7 // 7 days
                            });
                        }

                        onHeaders(res, function () {
                            if (!writingHeaders) {
                                writingHeaders = true;
                                ebayCookieManager.commitCookies();
                            }
                        });
                        tx.complete();
                        next();
                    });
                });
            });
        });
    };
};

Object.defineProperty(module.exports, Symbol.for('configProvider'), {
    get() {
        return configProvider;
    }
});
