// @ebay/openapi-ebay/lib/index.js
const Assert = require('assert');
const ExpressRoutes = require('./express-routes');
const Parser = require('@apidevtools/swagger-parser');
const { merge } = require('merge-object-files');
const RouteBuilder = require('./route-builder');
const createModuleResolver = require('./module-resolver').create;
const express = require('express');

async function configure({
    schema, basedir, handlers, securityProviders,
    preMiddleware, postMiddleware, // <<< deprecated attributes
    preControllerMiddleware, postControllerMiddleware, controllerMiddleware,
    docspath, refResolver,
    handlerRuntimeFactory }) {
    Assert.ok(schema, 'Expected an api definition.');

    schema = await resolveSchema(schema, basedir, refResolver);
    if (handlers && typeof handlers === 'string') {
        handlers = await merge(handlers, ['js']);
    }

    preControllerMiddleware = preControllerMiddleware || preMiddleware;
    postControllerMiddleware = postControllerMiddleware || postMiddleware;

    securityProviders = resolveSecurityProviders(securityProviders);

    let routes = RouteBuilder.build({
        schema,
        handlers,
        securityProviders,
        handlerRuntimeFactory
    });

    routes = ExpressRoutes.create({
        schema,
        routes,
        preControllerMiddleware,
        postControllerMiddleware,
        controllerMiddleware,
        docspath
    });

    const app = express();
    routes.forEach(route => {
        const args = [];

        if (route.path) {
            args.push(route.path);
        }
        args.push(route.handler);
        const methods = route.method ? [route.method] : route.methods;
        methods.forEach(method => {
            app[method](...args);
        });
    });

    Object.defineProperty(app, 'api', {
        value: {
            schema,
            routes
        }
    });

    return app;
}

function resolveSecurityProviders(securityProviders) {
    if (securityProviders) {
        if (typeof securityProviders === 'string') {
            return require(securityProviders);
        }
        return securityProviders;
    }
}

function resolveSchema(api, basedir, refResolver) {
    const args = [];
    if (typeof api === 'string') {
        args.push(api);
    }
    else {
        args.push(basedir);
        args.push(api);
    }
    const options = {
        resolve: {
            module: createModuleResolver(refResolver)
        },
        dereference: {
            circular: 'ignore'
        }
    };
    return Parser.dereference(...args, options)
        .then(bundledApi => basedir ?
            Parser.validate(basedir, bundledApi, options) : Parser.validate(bundledApi, options));
}

module.exports = {
    configure,
    resolveSchema,
    createModuleResolver
};
