// @ebay/utils-ebay/lib/guid-generator.js
'use strict';

const os = require('os'),
    ip = require('ip'),
    crypto = require('crypto');

const NON_PADDED_DELIMETER = '.',
    ISO_DELIMETER = '-';


function getIPAddress() {
    const ifaces = os.networkInterfaces();
    for (let dev in ifaces) {
        for (let i = 0; i < ifaces[dev].length; i++) {
            const details = ifaces[dev][i];
            if (details.family === 'IPv4' && !details.internal) {
                return details.address;
            }
        }
    }

    return null;
}

function genHash(len) {
    let shasum = crypto.createHash('sha1');
    shasum.update(process.id + '');
    shasum.update(Date.now() + '');
    let hash = shasum.digest('hex');
    if (hash.length > len) {
        hash = hash.substring(0, len);
    }
    return hash;
}

function getTimeHex(len) {
    let timeHex = Date.now().toString(16);
    if (timeHex.length > len) {
        timeHex = timeHex.substr(timeHex.length - len, len);
    }
    return timeHex;
}

function pad(unpadded, len) {
    unpadded = unpadded + '';
    if (unpadded.length >= len) {
        return unpadded;
    }
    let padded = '';
    for (let i = unpadded.length; i < len; i++) {
        padded += '0';
    }
    padded += unpadded;
    return padded;
}

function getAddrHex(serverAddr) {
    let inetAddrHex = ip.toLong(serverAddr || '');

    return inetAddrHex.toString(16);
}

function GUIDGenerator() {
    let hash = genHash(5);
    let addrHex = getAddrHex(getIPAddress());
    this.staticGUID = NON_PADDED_DELIMETER + addrHex + NON_PADDED_DELIMETER + hash + NON_PADDED_DELIMETER;
    this.staticPaddedGUID = pad(addrHex, 8) + pad(hash, 5);
    this.staticISOGUID = this.staticPaddedGUID.substring(0, 1) + ISO_DELIMETER + this.staticPaddedGUID.substring(1, 5) + ISO_DELIMETER + this.staticPaddedGUID.substring(5, 9) + ISO_DELIMETER + this.staticPaddedGUID.substring(9, this.staticPaddedGUID.length);
    this.counter = 0xffffffff;
}

function parseDate(dateHex) {
    return new Date(parseInt(dateHex, 16));
}

function parseIP(ipHex) {
    let addr = parseInt(ipHex, 16);
    let addrs = [];
    for (var i = 3; i >= 0; i--) {
        addrs[i] = addr & 0xff;
        addr >>>= 8;
    }
    return addrs.join('.');
}

function parseCounter(counterHex) {
    // to keep consistent with Java counter, for Java the counter starts from -1, since Java has 32-bit integer and its hex will be 0xffffffff
    return -(~parseInt(counterHex, 16) + 1);
}

GUIDGenerator.prototype.nextCounter = function () {
    let next = this.counter--;
    if (this.counter <= 0) {
        this.counter = 0xffffffff;
    }
    return next;
};

GUIDGenerator.prototype.nextGUID = function () {
    const timeHex = getTimeHex(11);
    return pad(timeHex, 11) + this.staticGUID + this.nextCounter().toString(16);
};

GUIDGenerator.prototype.nextPaddedGUID = function () {
    let timeHex = getTimeHex(11);
    timeHex = pad(timeHex, 11);
    return timeHex.substring(3, timeHex.length) + timeHex.substring(0, 3) + this.staticPaddedGUID + this.nextCounter().toString(16);
};

GUIDGenerator.prototype.nextISOGUID = function () {
    let timeHex = getTimeHex(11);
    timeHex = pad(timeHex, 11);
    return timeHex.substring(0, 8) + ISO_DELIMETER + timeHex.substring(8, timeHex.length) + this.staticISOGUID + this.nextCounter().toString(16);
};

GUIDGenerator.prototype.parseGUID = function (guid) {
    if (!guid) {
        return null;
    }
    const components = guid.split(NON_PADDED_DELIMETER);
    if (!components || components.length !== 4) {
        return null;
    }
    const data = {};
    data.date = parseDate(components[0]);
    data.ip = parseIP(components[1]);
    data.hash = parseInt(components[2], 16);
    data.counter = parseCounter(components[3]);
    return data;
};

GUIDGenerator.prototype.parsePaddedGUID = function (guid) {
    if (!guid || guid.length !== 32) {
        return null;
    }
    const data = {};
    data.date = parseDate(guid.substring(8, 11) + guid.substring(0, 8));
    data.ip = parseIP(guid.substring(11, 19));
    data.hash = parseInt(guid.substring(19, 24), 16);
    data.counter = parseCounter(guid.substring(24, 32));
    return data;
};

GUIDGenerator.prototype.parseISOGUID = function (guid) {
    if (!guid || guid.length !== 36) {
        return null;
    }
    const data = {};
    data.date = parseDate(guid.substring(0, 8) + guid.substring(9, 12));
    data.ip = parseIP(guid.substring(12, 13) + guid.substring(14, 18) + guid.substring(19, 22));
    data.hash = parseInt(guid.substring(22, 23) + guid.substring(24, 28), 16);
    data.counter = parseCounter(guid.substring(28, 36));
    return data;
};

module.exports.GUIDGenerator = GUIDGenerator;
