// @ebay/utils-ebay/lib/request-utils.js
'use strict';
const net = require('net');
const httpUtils = require('../http');
const path = require('path');
const url = require('url');
const util = require('util');
const os = require('os');
const ip = require('ip');

function getRemoteAddr(req, trustedOnly) {
    //Akamai headers check
    var clientIP = getIPFromHeader(req, 'x-ebay-akamai-9');
    if(!clientIP) {
        //X-eBay-Client-IP IP
        clientIP = getIPFromHeader(req, 'x-ebay-client-ip');
    }
    if(!clientIP) {
        //Read x-client-ip for Envoy
        clientIP = getIPFromHeader(req, 'x-client-ip');
    }

    if(!clientIP && !trustedOnly) {
        //Read x-forwarded-for as Akamai & ebay-client-ip doesn't exists
        clientIP = getIPFromHeader(req, 'x-forwarded-for');
    }

    var ip = clientIP || req.ip;

    // handling IPv4-mapped IPv6 addresses - http://en.wikipedia.org/wiki/IPv6#IPv4-mapped_IPv6_addresses
    if (/^::ffff:(\d{1,3}\.){3,3}\d{1,3}$/.test(ip)) {
      ip = ip.replace(/^::ffff:/, '');
    }
    return ip && ip.trim();
}

function getRemoteAddrFromTrustedHeader(req) {
    return getRemoteAddr(req, true);
}

function getClientIP(req) {
    var clientIP = getIPFromHeader(req, 'x-forwarded-for');
    var ip = clientIP || req.ip;

    // handling IPv4-mapped IPv6 addresses - http://en.wikipedia.org/wiki/IPv6#IPv4-mapped_IPv6_addresses
    if (/^::ffff:(\d{1,3}\.){3,3}\d{1,3}$/.test(ip)) {
      ip = ip.replace(/^::ffff:/, '');
    }
    return ip && ip.trim();
}

function isIPFromAkamaiHeader(req) {
    var clientIP = getIPFromHeader(req, 'x-ebay-akamai-9');

    if(clientIP) {
        return true;
    }

    return false;
}

function getIPFromHeader(req, headerName) {
    var headerValue = req.headers && req.headers[headerName] && req.headers[headerName].split(',');
    headerValue = headerValue && headerValue[0] && headerValue[0].trim();
    return headerValue;
}

function getServerName(req) {
    var host = req.headers.host || 'ebay.com';

    if(host.startsWith('http')) {
        host = url.parse(host).hostname || 'ebay.com';
    }

    var portStart = host.indexOf(':');
    if (portStart !== -1) {
        host = host.substring(0, portStart);
    }

    return host;
}

function getStaticUrl(req, relativePath, secure) {
    let baseHost = 'pages.' + httpUtils.getBaseHost(req.headers.host);
    let protocol = secure ? 'https://': 'http://';
    let localePath = '/';
    if(req.locality && req.locality.isGeoCountry) {
        localePath = req.locality.country+'/'+req.locality.locale;
    }
    return protocol + path.join(baseHost, localePath, relativePath);
}

/*
  Method provide a formated string to be logged to CAL
  @maskedPayload can be generated using security-ebay module
*/
function getClientInfo(req, maskedPayload) {
    const headers = req.headers || {};
    maskedPayload = url.parse(maskedPayload).pathname;
    const scriptPath = '/' + maskedPayload.split('/')[1];

    var formatString = 'RemoteIP=%s&Server=%s&Script=%s&Encoding=%s&Agent=%s';
    var params = [getRemoteAddr(req),
        'www.ebay.com',
        scriptPath,
        headers['accept-encoding'],
        headers['user-agent']
    ];

    if (headers.referer && headers.referer.trim()) {
        formatString += '&Referer=%s';
        params.push(headers.referer);
    }

    if(headers['x-forwarded-for']) {
        formatString += '&ForwardedFor=%s';
        params.push(headers['x-forwarded-for']);
    }

    params.unshift(formatString);
    return util.format.apply(util, params);
}

function getIpAddress() {
    var interfaces = os.networkInterfaces();

    // If no specific network interface has been named,
    // check for 'eth0' interface
    if (interfaces.eth0) {
        var result = interfaces.eth0.filter((details) => {
            var itemFamily = _normalizeFamily(details.family);
            return itemFamily === 'ipv4';
        });
        if (result.length !== 0) {
            return result[0].address;
        }
    }
}

function _normalizeFamily(family) {
    if (family === 4) {
        return 'ipv4';
    }
    if (family === 6) {
        return 'ipv6';
    }
    return family ? family.toLowerCase() : 'ipv4';
}

module.exports = {
    getRemoteAddr: getRemoteAddr,
    getClientIP: getClientIP,
    getServerName: getServerName,
    isIPFromAkamaiHeader: isIPFromAkamaiHeader,
    getStaticUrl: getStaticUrl,
    getClientInfo: getClientInfo,
    getRemoteAddrFromTrustedHeader,
    getIpAddress: getIpAddress,
    _normalizeFamily: _normalizeFamily
};
