// @ebay/service-errors-ebay/lib/formatter.js
'use strict';

const Env = require('@ebay/environment-ebay');

const formatError = module.exports.formatError = (err, options) => {
    let category = err.category || 'SYSTEM';
    let domain = typeof err.domain === 'string' ? err.domain : 'framework';
    if (err.code === 'INVALID_PARAMETER') {
        domain = 'app';
        category = 'USER';
    }

    const allowProperties = options && options.allowProperties;
    const message = err.details && `${err.message}:${err.details}` || err.message || 'Unknown';
    const longMessage = err.longMessage || (!Env.isNotDev() && err.stack) || message;
    return Object.assign({
        errorId: err.code,
        domain,
        subdomain: err.subdomain,
        category,
        message,
        longMessage,
        statusCode: err.statusCode || err.status || 500
    }, pick(err, allowProperties));
};

function pick(err, allowProperties) {
    return err && Object.keys(err).filter(
        key => allowProperties && allowProperties.indexOf(key) !== -1).reduce((memo, key) => {
        memo[key] = err[key];
        return memo;
    }, {});
}

module.exports.format = (err, options) => {
    // first check if errors are already formatted and use them
    const errData = err.body && err.body.errors || [formatError(err, options)];

    return {
        errorMessage: {
            error: errData
        }
    };
};
