// @ebay/app-context-ebay/lib/index.js
'use strict';

var os = require('os'),
    path = require('path'),
    tryRequire = require('try-require'),
    pkg = tryRequire(path.resolve(process.cwd(), 'package.json'));

//Return contextual data about the application
var appContext = {
    appName: pkg && pkg.name || 'nodeapp',
    experienceService: pkg && pkg.gpaas && pkg.gpaas['experience-service'],
    service: pkg && pkg.gpaas && pkg.gpaas.service,
    machineName: process.env.MACHINE_NAME || os.hostname(),
    consumerId: pkg && pkg.gpaas && pkg.gpaas['consumer-id'],
    teamDL: getOwnerAndTeamDL()
};

// separate pre-production and lnp environment from the rest
var suffix = /pre-production|lnp|sandbox|xstage/i.test(process.env.NODE_ENV) ?
    ('-' + process.env.NODE_ENV.toLowerCase()) : '';

var manifest = tryRequire(path.join(process.cwd(), 'manifest.json'));

appContext.buildId = manifest && manifest.version ? manifest.version.rpmName : undefined;
appContext.nuggetName = process.env.NUGGET_NAME || appContext.appName;
appContext.poolType = process.env.POOL_ARCHITECTURE || 'r1';
appContext.poolName = (process.env.POOL_NAME || appContext.poolType + appContext.nuggetName) + suffix;
appContext.odbMoName = process.env.ODBMO_NAME;
appContext.baseDir = process.env.BASE_DIR || process.cwd();
appContext.paasRealm = process.env.PAAS_REALM;
appContext.classOfService = process.env.CLASS_OF_SERVICE;
appContext.vpc = process.env.VPC;
appContext.useFileBeatsForCAL = process.env.USE_FILEBEATS_FOR_CAL || 'false';
appContext.calPoolName = process.env.CAL_POOL_NAME;

if (appContext.experienceService) {
    appContext.defaultExpSvc = Array.isArray(appContext.experienceService) ? appContext.experienceService[0] : appContext.experienceService;
}

function getOwnerAndTeamDL() {
    var ownerAndTeamDL = [];
    if (pkg) {
        var ebayProps = pkg.gpaas || pkg.ebay;
        if (ebayProps) {
            ebayProps.owner && ownerAndTeamDL.push(ebayProps.owner);
            ebayProps["team-dl"] && ownerAndTeamDL.push(ebayProps["team-dl"]);
        }
    }
    return ownerAndTeamDL;
}

module.exports = appContext;
