// @ebay/tracking-ebay/lib/tracking-context/service-context/index.js
'use strict';

const onHeaders = require('on-headers');
const Hoek = require('hoek');

const Context = require('../context');
const trackingBookKeeping = require('./bookkeeping');
const SojournerWriter = require('../../soj/sojourner-writer.js');
const SojournerEventEnum = require('../../soj/sojourner-event-enums.js');
const Tags = require('../../entities/tags.js');
const { PageSourceInfoCookielet } = require('../../tracking2/page-source-info');
const logger = require('@ebay/logging-inc').logger('tracking-ebay:tracking-context:service-context');

const X_EBAY_SVC_TRACKING_DATA_SIZE_LIMIT = 4096;

const EXCLUDE_PARAMS = {
    n: 'n',
    p: 'p',
    g: 'g'
};

class ServiceContext extends Context {
    constructor(req, res) {
        super(req, res);
        this.isService = true;

        // TODO: should we control it similar to web???
        // RaptorTrackingGingerServiceAdaptor doesn't disable tracking
        this.enabled = true;
    }

    getEventAction() {
        if (this.req.ebay.isExpSvc(this.req)) {
            return 'EXPC';
        } else if (this.req.ebay.isExpSvcModuleProvider(this.req)) {
            return 'EXPM';
        }

        return;
    }

    setSendSOJDataInResFromExpSvc(value) {
        this.sendSOJDataInResFromExpSvc = value;
    }

    instrumentResponse() {
        const _this = this;
        const req = this.req;
        const res = this.res;
        const _end = res.end;

        trackingBookKeeping.track(this.req, this.res);
        this.requestScopeTracker = this.getRequestScopeTracker({noAutoFlush: true});

        res.end = Hoek.once(function () {
            if (res.statusCode === 200) {
                _this.requestScopeTracker.getPulsarContext((err, pulsarContext) => {

                    if (!pulsarContext.getEventAction()) {
                        pulsarContext.setEventAction(_this.getEventAction());
                    }

                    _this.requestScopeTracker.flush(err => {
                        err && logger.error('Error happened while flushing to Pulsar', err);
                    });
                });
            }

            _end.apply(res, arguments);
        });

        /**
         * Is this send when response status != 200 ?
         */
        onHeaders(res,  () => {

            // TOOD: for experience service for mobile device use case, soj logging must be performed based on the same logic describe here https://github.corp.ebay.com/Tracking/RaptorTrackingClientLibrary/blob/1b82bb972026b1ce5e420e95424d427f6cbe9b93/RaptorTrackingAdaptor/src/main/java/com/ebay/raptor/tracking/handler/RaptorTrackingHandler.java#L526-L551
            // while the logic should use a better architecture based on openapi schema defintion to inject data into the response instead of a hack used in the above logic. https://jirap.corp.ebay.com/browse/NODE-2065

            if (req.ebay.isExpSvc() || req.ebay.isExpSvcModuleProvider()) {
                /**
                 * As per experience service spec redirect tag is explicitly dropped by the app.
                 * https://wiki.vip.corp.ebay.com/pages/viewpage.action?spaceKey=EXPSVCARCH&title=Redirect+Tracking+in+Experience+Service
                 */
                _this.requestScopeTracker.getPulsarContext((err, context) => {
                    if (err) {
                        logger.error('Failed to get pulsar context', err);
                        return;
                    }
                    let headerVal;
                    //ciid needs an A to know the version
                    if (res.statusCode === 200) {
                        const pageid = context.getPageID();
                        const ciid = PageSourceInfoCookielet.formatCookie(context.getCIID());
                        headerVal = '<a>ciid=' + ciid  + '&operationId=' + pageid + '</a>';
                    } else if (res.statusCode > 299 && res.statusCode < 400) {
                        headerVal = '<a>rdt=1&g=' + (req && req.ebay && req.ebay.getGuid()) + '</a>';
                    }

                    headerVal && res.set('X-EBAY-SVC-TRACKING-DATA', headerVal);
                });

                if (!this.sendSOJDataInResFromExpSvc) {
                    return;
                }
            }

            // from this point we handle non-experience service use-case
            /**
             * https://github.corp.ebay.com/Tracking/RaptorTrackingClientLibrary/blob/1b82bb972026b1ce5e420e95424d427f6cbe9b93/RaptorTrackingAdaptor/src/main/java/com/ebay/raptor/tracking/handler/RaptorTrackingHandler.java#L265-L273
             * in nodejs case we have web and service context split and ignore image and script flushing
             * SSE is only enabled for experience service, so we can ignore SSE logic in relation to soj tracking
             */

            const sojournerContext = this.sojournerContext;

            if (res.statusCode === 300 || res.statusCode === 301 || res.statusCode === 302) {
                //set redirect indicator to soj
                sojournerContext.addContext(Tags.RDT, '1');
            }
    
            // TODO: should we log it for service or not as it may change later due to error?
            // log response code to soj
            sojournerContext.addData(SojournerEventEnum.APPLICATION,
                Tags.HTTPResponseCode, res.statusCode);
    
            // for single chunk, write everything to headers
            if (/application\/json/.test(res.getHeader('content-type'))) {
    
                var headerVal = SojournerWriter.writeToHeader(sojournerContext);
    
                if (Buffer.byteLength(headerVal) > X_EBAY_SVC_TRACKING_DATA_SIZE_LIMIT) {
                    return logger.error(new Error('Tracking data cannot be logged as the size of X_EBAY_SVC_TRACKING_DATA exceeds 4K.'));
                }
                res.set('X-EBAY-SVC-TRACKING-DATA', headerVal);
            }
            else if (/text\/event-stream|multipart\/json/.test(res.getHeader('content-type'))) {
                SojournerWriter.writeChunk(sojournerContext, res);
            }
        });
    }

    handleResponseHeaders(headers) {
        if (this.enabled === false || !headers) {
            return;
        }

        const trackingData = headers['x-ebay-svc-tracking-data'] || headers['X-EBAY-SVC-TRACKING-DATA'];

        logger.debug('trackingData:', trackingData);

        if (!trackingData) {
            return;
        }

        const aMatch = /(<a>(.+)<\/a>)/.exec(trackingData);
        const aEvent = aMatch ? aMatch[2] : null;

        if (!aEvent){
            return;
        }

        aEvent.split('&').forEach((nv) => {
            const kv = nv.split('=');

            if (kv && kv.length >1 && !EXCLUDE_PARAMS[kv[0]]) {
                this.requestScopeTracker.sharedTracker.addToSvcMap(kv[0], kv[1]);
            }
        });
    }
}

module.exports = ServiceContext;
