// @ebay/openfeature-ebay/src/openfeature-lib/Api.ts
export const CHANNEL_ENUM = {
    WEB: 1,
    MOBILE_WEB: 2,
    TABLET_WEB: 3,
    ANDROID_APP: 4,
    IOS_APP: 5,
    EMAIL: 6,
    PLACEHOLDER1: 7,
    PLACEHOLDER2: 8,
    OTHER: 99,
};

export function getChannelId(deviceInfo) {
    if (!deviceInfo) {
        return CHANNEL_ENUM.OTHER;
    }

    if (deviceInfo.isFSOM) {
        return CHANNEL_ENUM.WEB;
    }

    if (deviceInfo.requestIsWeb && deviceInfo.requestIsDesktop) {
        return CHANNEL_ENUM.WEB;
    }

    if (deviceInfo.requestIsWeb && deviceInfo.requestIsMobileWeb) {
        return CHANNEL_ENUM.MOBILE_WEB;
    }

    if (deviceInfo.requestIsWeb && /ios/i.test(deviceInfo.deviceOS)) {
        return CHANNEL_ENUM.WEB;
    }

    if (deviceInfo.requestIsWeb && /android/i.test(deviceInfo.deviceOS)) {
        return CHANNEL_ENUM.WEB;
    }

    if (!deviceInfo.requestIsWeb && /ios/i.test(deviceInfo.deviceOS)) {
        return CHANNEL_ENUM.IOS_APP;
    }

    if (!deviceInfo.requestIsWeb && /android/i.test(deviceInfo.deviceOS)) {
        return CHANNEL_ENUM.ANDROID_APP;
    }

    return CHANNEL_ENUM.OTHER;
}
