// @ebay/openfeature-ebay/src/openfeature-lib/Services.ts
import serviceClientEbay from '@ebay/service-client-ebay';
import { EvaluationContext } from '@openfeature/js-sdk';
import translateContext from './openfeature-context/translateContext';
import { EvaluationResults, EvaluationResultsMap } from './types/FlagEvalDetail';
import { EvalRequest } from './types/EvalRequest';
import { CHANNEL_ENUM } from './Api';

/**
 * This class is used for handling all services that may be used in the middleware.
 * Currently it is calling epsvc and it does not require any authentication (hence empty req)
 */
export default class Services {
    client;

    constructor() {
        this.client = serviceClientEbay.context({}).getClient('featureflag');
    }

    public async invokeInternalSearchApiByFlagKey(
        evalContext: EvaluationContext,
        flagKey: string,
    ): Promise<EvaluationResultsMap> {
        const request = translateContext(evalContext, flagKey);
        const response = await this.call(request);
        return convertToEvaluationResultsMap(response);
    }

    public async invokeInternalSearchApiByFlagList(
        evalContext: EvaluationContext,
        flagsList: Set<string>,
    ): Promise<EvaluationResultsMap> {
        const request = translateContext(evalContext, null, Array.from(flagsList));
        const response = await this.call(request);
        return convertToEvaluationResultsMap(response);
    }

    public async invokeInternalSearchApiByLabel(
        evalContext: EvaluationContext,
        labelForEval: Set<string>,
    ): Promise<EvaluationResultsMap> {
        evalContext.labelForEval = Array.from(labelForEval);
        const request = translateContext(evalContext);
        const response = await this.call(request);
        return convertToEvaluationResultsMap(response);
    }

    private async call(data: EvalRequest): Promise<EvaluationResults> {
        return new Promise((resolve, reject) => {
            // ChannelEnum.OTHER is a default/invalid channel id=99 and will never qualify for any flags
            // filter out bots by using bot-handler-ebay. If the traffic is bot it will be including req.ebay.bot object which is passed to evalMetadata
            if (data.evalCriteria.channelId == CHANNEL_ENUM.OTHER || data.evalCriteria.evalMetadata?.IS_BOT) {
                resolve(getEmptyEvaluationResults());
                return;
            }

            this.client
                .post({
                    body: JSON.stringify(data),
                    qs: {
                        detail: 'true',
                    },
                })
                .end((error, response) => {
                    if (error) {
                        reject(new Error(error));
                        return;
                    }
                    resolve(response.body);
                });
        });
    }

    
}

export function convertToEvaluationResultsMap(
    evaluationResults: EvaluationResults,
) {
    const evaluationResultsMap = new Map();
    evaluationResults.evaluationResults.forEach((result) => {
        evaluationResultsMap.set(result.featureFlagKey, result);
    });
    return {
        evaluationResultsMap,
        epTagsEvent: evaluationResults.epTagsEvent,
    };
}

function getEmptyEvaluationResults(): EvaluationResults {
    return {
        evaluationResults: [],
        epTagsEvent: null,
    };
}
