// @ebay/openfeature-ebay/dist/types/CalResponseCode.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.CalResponseCode = exports.CalResponseCodeUtils = void 0;
/**
 * Enum representing different categories of CAL (Centralized Application Logging) response codes.
 * Each enum constant corresponds to a specific category of HTTP response codes.
 */
var CalResponseCode;
(function (CalResponseCode) {
    /**
     * Represents the category of successful HTTP response codes (2XX).
     */
    CalResponseCode["SUCCESS_2XX"] = "2XX";
    /**
     * Represents the category of client error HTTP response codes (4XX).
     */
    CalResponseCode["CLIENT_ERROR_4XX"] = "4XX";
    /**
     * Represents the category of server error HTTP response codes (5XX).
     */
    CalResponseCode["SERVER_ERROR_5XX"] = "5XX";
    /**
     * Represents any other response codes that do not fall into the above categories.
     */
    CalResponseCode["OTHER_ERROR_CODE"] = "999";
})(CalResponseCode || (exports.CalResponseCode = CalResponseCode = {}));
/**
 * Utility functions for CalResponseCode
 */
class CalResponseCodeUtils {
    /**
     * Determines the response code category based on the given HTTP response code.
     * @param respCode The HTTP response code as a number.
     * @returns The corresponding response code category as a string.
     */
    static getResponseCodeByCode(respCode) {
        const respCodeStr = String(respCode);
        if (respCodeStr.startsWith("2")) {
            return CalResponseCode.SUCCESS_2XX;
        }
        else if (respCodeStr.startsWith("4")) {
            return CalResponseCode.CLIENT_ERROR_4XX;
        }
        else if (respCodeStr.startsWith("5")) {
            return CalResponseCode.SERVER_ERROR_5XX;
        }
        return CalResponseCode.OTHER_ERROR_CODE;
    }
}
exports.CalResponseCodeUtils = CalResponseCodeUtils;
