// @ebay/openfeature-ebay/dist/types/DiagnosticsData.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.ParticipationDebugMessages = exports.DebugCodes = exports.DiagnosticsData = void 0;
class DiagnosticsData {
    constructor() {
        this.platformData = {};
        this.qualificationInput = {};
        this.apiLogList = [];
        this.qualificationOutputMap = {};
    }
}
exports.DiagnosticsData = DiagnosticsData;
;
var DebugCodes;
(function (DebugCodes) {
    DebugCodes[DebugCodes["BOT_REQUEST"] = 1] = "BOT_REQUEST";
    DebugCodes[DebugCodes["NULL_DOMAIN_REQUEST_DATA"] = 2] = "NULL_DOMAIN_REQUEST_DATA";
    DebugCodes[DebugCodes["NULL_SITE"] = 6] = "NULL_SITE";
    DebugCodes[DebugCodes["NULL_CHANNEL"] = 7] = "NULL_CHANNEL";
})(DebugCodes || (exports.DebugCodes = DebugCodes = {}));
var ParticipationDebugMessages;
(function (ParticipationDebugMessages) {
    ParticipationDebugMessages["BOT_REQUEST"] = "The incoming request is detected as a bot traffic. No EP qualification was ran.";
    ParticipationDebugMessages["NULL_DOMAIN_REQUEST_DATA"] = "Domain request data is null. Ensure the domain-request-filter and X-EBAY-C-MARKETPLACE-ID are included.";
    ParticipationDebugMessages["NULL_SITE"] = "Not able to get site id. Ensure the X-EBAY-C-MARKETPLACE-ID header is included or site id is included in the request.";
    ParticipationDebugMessages["NULL_CHANNEL"] = "Not able to get channel id. Ensure the X-EBAY-C-ENDUSERCTX header is included or channel id included in the request.";
})(ParticipationDebugMessages || (exports.ParticipationDebugMessages = ParticipationDebugMessages = {}));
