// @ebay/openfeature-ebay/dist/types/FeatureFlagErrorInfo.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.FeatureFlagErrorInfo = void 0;
class FeatureFlagErrorInfo {
    constructor() {
        this.errorEvents = new Map();
        this.errorsLogged = [];
    }
    /**
     * Counts up the total count of errors from each error name in errorEvents.
     * @returns
     */
    getTotalErrorCount() {
        if (this.errorEvents.size == 0) {
            return 0;
        }
        let total = 0;
        for (const [name, count] of this.errorEvents) {
            total += count;
        }
        return total;
    }
    addErrorEvent(name) {
        this.countOne(name);
        try {
            this.getErrorsList().push(name);
        }
        catch (e) {
            // EXPRMT-17397: avoid ArrayIndexOutOfBoundsException throw - do nothing since it is not critical
        }
    }
    getErrorsList() {
        return this.errorsLogged;
    }
    getErrorEvents() {
        return this.errorEvents;
    }
    countOne(key) {
        const currentCount = this.errorEvents.get(key) || 0;
        this.errorEvents.set(key, currentCount + 1);
    }
}
exports.FeatureFlagErrorInfo = FeatureFlagErrorInfo;
