// @ebay/trust-fabric-ebay/lib/logger.js
const path = require('path');
const pino = require('pino');
const fs = require('fs');
const env = require('@ebay/environment-ebay');

class Logger {
    constructor() {
        if (!env.isDev() && !env.isTest()) {
            const logsDir = path.join('.', 'logs');
            if (!fs.existsSync(logsDir)) {
                fs.mkdirSync(logsDir);
            }
            const pm2id = process.env.pm_id ? `-${process.env.pm_id}` : '';
            const auditLogPath = path.join(logsDir, `tfaudit${pm2id}.log`);
            this.logger = pino({
                prettyPrint: {
                    levelFirst: false
                },
                prettifier: this.customPrettifier
            },
            pino.destination({
                dest: auditLogPath
            }));
        }
    }

    log(data) {
        this.logger && this.logger.info(data);
    }

    customPrettifier(options) {
        return function prettifier(inputData) {
            delete inputData.time;
            delete inputData.pid;
            delete inputData.hostname;
            delete inputData.level;
            return `${JSON.stringify(inputData)}\n`;
        };
    }
}

module.exports = new Logger();
