// @ebay/openfeature-ebay/dist/hooks/MetricsPublisherHook.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.MetricsPublisherHook = void 0;
const logging_inc_1 = require("@ebay/logging-inc");
/**
 * A hook that is used to set the EvaluationContext with required values
 * from request before flag evaluation.
 */
const NANOSECONDS_PER_SECOND = 1000000000;
class MetricsPublisherHook {
    constructor(metricsPublisher) {
        this.metricsPublisher = metricsPublisher;
    }
    before() {
        this.startTimeInNanoSeconds = process.hrtime.bigint();
        this.logger = (0, logging_inc_1.getLogger)('openfeature-ebay');
    }
    after(hookContext, evaluationDetails) {
        this.metricsPublisher.usageCounter.inc({ apiName: 'evaluation', status: 'success', techStack: 'nodejs' }, 1);
        const currentTimeInNanoSeconds = process.hrtime.bigint(); // returns time in nanoseconds
        const nsDuration = currentTimeInNanoSeconds - this.startTimeInNanoSeconds;
        // converting to seconds to make it same as raptor and raptorio
        const secondsDuration = Number(nsDuration) / NANOSECONDS_PER_SECOND;
        this.metricsPublisher.epEvaluationLatencySla
            .labels('nodejs')
            .observe(secondsDuration);
    }
    finally() { }
}
exports.MetricsPublisherHook = MetricsPublisherHook;
