// @ebay/tracking-ebay/lib/tracking2/common-utils.js
'use strict';
/** Test IP address work */

const os = require('os');
const logger = require('@ebay/logging-inc').logger('tracking-ebay/common-utils');
var headerMultiValue = require('./header-multi-value');
var MAX_FLAG = 1023;

function encode(bs) {
    if(!bs) {
        return null;
    }
    var keys = Object.keys(bs);
    var len = 0;
    var i = 0;
    var key = 0;
    for(i=0; i<keys.length; i++) {
        key = parseInt(keys[i], 10);
        if(key + 1 > len) {
            len = key + 1;
        }
    }
    // if(keys.length === 0) {
    //     return null;
    // }
    // truncate if len is greater than maxPos and maxPos > 0
    if(len > MAX_FLAG) {
        len = MAX_FLAG;
    }
    var byteArrSize = Math.floor(len/8);
    if( (byteArrSize === 0) || (len%8 !== 0) ) {
        byteArrSize++;
    }
    var bytes = new Buffer(byteArrSize);
    bytes.fill(0);
    i = 0;
    key = 0;
    for(i=0; i<keys.length; i++) {
        key = parseInt(keys[i], 10);
        if(key<0 || key > len) {
            continue;
        }

        if(bs[key]) {
            var bytePos = Math.floor(key / 8);
            var bitPos = key % 8;
            var value = bytes[bytePos];
            value = value | (1 << (7 -bitPos) );
            bytes[bytePos] = value;
        }
    }
    // console.log(JSON.stringify(bytes));

    var base64Str = bytes.toString('base64');
    base64Str = base64Str.replace(/=/g, '*');
    return base64Str;
}

function getContextDataFrom20Header(key, cosHeader, req) {
    if(!req) {
        return null;
    }
    if(!req.headers) {
        return null;
    }
    if(!cosHeader) {
        return null;
    }
    if (req._trkCorrSessionHmv) {
        return req._trkCorrSessionHmv.get(key);
    }
    var header = req.headers[cosHeader.toLowerCase()] || req.headers[cosHeader.toUpperCase()];
    if(!header) {
        return null;
    }
    if(!key) {
        return null;
    }
    try {
        var hmv = headerMultiValue.create(header, true, true);
        // keep cache for faster access per request
        req._trkCorrSessionHmv = hmv;
        return hmv.get(key);    
    }
    catch (err) {
        logger.error(`Invalid multi-value header detected name: "${cosHeader}", key: "${key}"`, err);
    }
}

function getContextDataFromHeader(key, req) {
    if(!req) {
        return null;
    }
    if(!key) {
        return null;
    }
    if(!req.headers) {
        return null;
    }
    var cosHeader = 'X-EBAY-C-TRACKING';
    var header = req.headers[cosHeader.toLowerCase()] || req.headers[cosHeader.toUpperCase()];
    if(!header) {
        return null;
    }
    var hmv = headerMultiValue.create(header, true, true);
    return hmv.get(key);
}

function getGuidFrom20Header(req) {
    return getContextDataFrom20Header('si', 'X-EBAY-C-CORRELATION-SESSION', req);
}

function getCGuidFrom20Header(req) {
    var gci = getContextDataFrom20Header('gci', 'X-EBAY-C-CORRELATION-SESSION', req);

    if((!gci || gci === null) && req && req.ebay && typeof req.ebay.getDCCorrelationGuid === 'function'){
        gci =  req.ebay.getDCCorrelationGuid();
    }

    return gci;
}

function getGuidFrom10Header(req) {
    return getContextDataFromHeader('guid', req);
}

function getCGuidFrom10Header(req) {
    return getContextDataFromHeader('cguid', req);
}

function djb2Hash(str) {
    str = '' + str;
    var hash = 5381;
    var i=0;
    var ch = ' ';
    for (i = 0; i < str.length; i++) {
        ch = str.charCodeAt(i);
        hash = ((hash << 5) + hash) + ch; /* hash * 33 + c */
    }
    return hash;
}

function ipToHex(ipaddress) {
    let ip = 0;

    ipaddress.split('.').forEach(octet => {
        ip <<= 8;
        ip += parseInt(octet);
    });

    return ip.toString(16);
}

let HOST_ADDR;

function getIPAddress() {
    if (HOST_ADDR) {
        return HOST_ADDR;
    }

    let ifaces = os.networkInterfaces();

    for (let dev in ifaces) {
        for (let i = 0; i < ifaces[dev].length; i++) {
            let details = ifaces[dev][i];
            if (details.family === 'IPv4' && !details.internal) {
                HOST_ADDR = details.address;
                return HOST_ADDR;
            }
        }
    }
}

module.exports = {
    encode,
    ipToHex,
    getIPAddress,
    hash: djb2Hash,
    getGuidFrom20Header,
    getCGuidFrom20Header,
    getGuidFrom10Header,
    getCGuidFrom10Header
};
