// @ebay/tracking-ebay/lib/tracking2/event-data.js
'use strict';

function EventData() {
    this.data = {
        "et": null,
        "rq": null,
        "ai": null,
        "rc": null,
        "ei": null,
        "exa": {},
        "exe": {},
        "exg": {},
        "exs": {},
        "exz": {},
        "gf": null,
        "ef": null,
        "lc": null,
        "la": null,
        "lb": null,
        "rf": null,
        "ch": null,
        "tp": null,
        "tc": null,
        "ts": null,
        "tm": null,
        "ta": null,
        "ki": null,
        "kw": null,
        "qc": null,
        "qt": null,
        "qi": null,
        "ct": null,
        "pr": null
    };
}

var pro = EventData.prototype;

pro.getCaptureTimeZone = function getCaptureTimeZone() {
    return this.data.ct;
};

pro.setCaptureTimeZone = function setCaptureTimeZone(captureTimeZone) {
    this.data.ct = captureTimeZone;
};

pro.getRequestCorrelationID = function getRequestCorrelationID() {
    return this.data.rq;
};

pro.setRequestCorrelationID = function setRequestCorrelationID(requestCorrelationID) {
    this.data.rq = requestCorrelationID;
};

pro.getEventType = function getEventType() {
    return this.data.et;
};

pro.setEventType = function setEventType(eventType) {
    this.data.et = eventType;
};

pro.getEventID = function getEventID() {
    return this.data.ei;
};

pro.setEventID = function setEventID(eventID) {
    this.data.ei = eventID;
};

pro.getUrl = function getUrl() {
    return this.data.url;
};

pro.setUrl = function setUrl(url) {
    this.data.url = url;
};

pro.getActorDetails = function getActorDetails() {
    return this.data.ai;
};

pro.setActorDetails = function setActorDetails(actorDetails) {
    this.data.ai = actorDetails;
};

pro.getAccountDetails = function getAccountDetails() {
    return this.data.ac;
};

pro.setAccountDetails = function setAccountDetails(accountDetails) {
    this.data.ac = accountDetails;
};

pro.getLastAccountID = function getLastAccountID() {
    return this.data.lac;
};

pro.setLastAccountID = function setLastAccountID(lastAccountID) {
    this.data.lac = lastAccountID;
};

pro.getSourceID = function getSourceID() {
    return this.data.exs;
};

pro.setSourceID = function setSourceID(sourceID) {
    this.data.exs = sourceID;
};

pro.getNameValuePairs = function getNameValuePairs() {
    return this.data.exa;
};

pro.setNameValuePairs = function setNameValuePairs(nameValuePairs) {
    this.data.exa = nameValuePairs;
};

pro.getEventFlags = function getEventFlags() {
    return this.data.ef;
};

pro.setEventFlags = function setEventFlags(eventFlags) {
    this.data.ef = eventFlags;
};

pro.getPrEventID = function getPrEventID() {
    return this.data.pr;
};

pro.setPrEventID = function setPrEventID(prEventID) {
    this.data.pr = prEventID;
};

pro.getGlobalFlags = function getGlobalFlags() {
    return this.data.gf;
};

pro.setGlobalFlags = function setGlobalFlags(globalFlags) {
    this.data.gf = globalFlags;
};

pro.getGlobalTags = function getGlobalTags() {
    return this.data.exg;
};

pro.setGlobalTags = function setGlobalTags(globalTags) {
    this.data.exg = globalTags;
};

pro.getResponseCode = function getResponseCode() {
    return this.data.rc;
};

pro.setResponseCode = function setResponseCode(responseCode) {
    this.data.rc = responseCode;
};

pro.getGeoDetails = function getGeoDetails() {
    return this.data.lc;
};

pro.setGeoDetails = function setGeoDetails(geoDetails) {
    this.data.lc = geoDetails;
};

pro.getAltitude = function getAltitude() {
    return this.data.la;
};

pro.setAltitude = function setAltitude(altitude) {
    this.data.la = altitude;
};

pro.getSpeedCourse = function getSpeedCourse() {
    return this.data.lb;
};

pro.setSpeedCourse = function setSpeedCourse(speedCourse) {
    this.data.lb = speedCourse;
};

pro.getReferralUrl = function getReferralUrl() {
    return this.data.rf;
};

pro.setReferralUrl = function setReferralUrl(referralUrl) {
    this.data.rf = referralUrl;
};

pro.getChannelID = function getChannelID() {
    return this.data.ch;
};

pro.setChannelID = function setChannelID(channelID) {
    this.data.ch = channelID;
};

pro.getPlacementID = function getPlacementID() {
    return this.data.tp;
};

pro.setPlacementID = function setPlacementID(placementID) {
    this.data.tp = placementID;
};

pro.getCampaignID = function() {
    return this.data.tc;
};

pro.setCampaignID = function(campaignID) {
    this.data.tc = campaignID;
};

pro.getSegmentID = function() {
    return this.data.ts;
};

pro.setSegmentID = function(segmentID) {
    this.data.ts = segmentID;
};

pro.getMessageID = function() {
    return this.data.tm;
};

pro.setMessageID = function(messageID) {
    this.data.tm = messageID;
};

pro.getCreativeID = function() {
    return this.data.ta;
};

pro.setCreativeID = function(creativeID) {
    this.data.ta = creativeID;
};

pro.getKeywordID = function() {
    return this.data.ki;
};

pro.setKeywordID = function(keywordID) {
    this.data.ki = keywordID;
};

pro.getKeyword = function() {
    return this.data.kw;
};

pro.setKeyword = function(keyword) {
    this.data.kw = keyword;
};

pro.getLegacyPayload = function() {
    return this.data.exz;
};

pro.setLegacyPayload = function(legacyPayload) {
    this.data.exz = legacyPayload;
};

pro.getExpChannelId = function() {
    return this.data.qc;
};

pro.setExpChannelId = function(expChannelId) {
    this.data.qc = expChannelId;
};

pro.getQualified = function() {
    return this.data.qt;
};

pro.setQualified = function(qualified) {
    this.data.qt = qualified;
};

pro.isOptIn = function() {
    return this.data.qi;
};

pro.setOptIn = function(optIn) {
    this.data.qi = optIn;
};

/**
 * Experimentation tag. TRK-3172 EP support in Pulsar for implemenation in
 * service and client library
 */

pro.getExperimentTags = function() {
    return this.data.exe;
};

pro.setExperimentTags = function(experimentTags) {
    this.data.exe = experimentTags;
};



pro.toJson = function toJson() {
    return this.data;
};


function create() {
    return new EventData();
}

module.exports.create = create;
