// @ebay/tracking-ebay/lib/tracking2/page-source-info.js
"use strict";

const VERSION_A = 'A';
const VERSION_SIZE = 1;
const SIZE_IN_BYTES = 8;
const CURRENT_VERSION = VERSION_A;
const KNOWN_VERSIONS = [VERSION_A];
const BYTES_IN_USE = SIZE_IN_BYTES + VERSION_SIZE;
const BYTES_IN_USE_FOR_LAYERS = VERSION_SIZE + SIZE_IN_BYTES +  VERSION_SIZE + SIZE_IN_BYTES;

class PageSourceInfoCookielet {
    constructor (cookielet) {
        if(cookielet == null || cookielet.length < BYTES_IN_USE) {
            return ;
        }

        if(!KNOWN_VERSIONS.includes(cookielet[0])){
            return;
        }

        this.version = cookielet[0];
        this.impressionID = cookielet.substr(1,8);

        //Check if the cookie length is equal to the layer length
        if (cookielet.length === BYTES_IN_USE_FOR_LAYERS) {
			if (!KNOWN_VERSIONS.includes(cookielet[9])) {
				return;
			}

            this.layerVersion = cookielet[9];
            this.layerImpressionID = cookielet.substr(10, 8);
        }
    }

    static formatCookie(impressionId) {
    	return CURRENT_VERSION + impressionId;
    }

    getVersion() {
        return this.version ;
    }

    getLayerVersion() {
        return this.layerVersion ;
    }

    getImpressionID() {
        return this.impressionID ;
    }

    getCookieletValueWithVersion() {
        if(!this.impressionID) {
            return null ;
        }

        return (this.version + this.impressionID);
    }

    getCookieletValueWithoutVersion() {
        return this.impressionID;
    }

    getLayerCookieletValueWithoutVersion() {
        return this.layerImpressionID;
    }
}

module.exports = {
    PageSourceInfoCookielet
};
