// @ebay/tracking-ebay/lib/tracking2/pulsar-context-manager.js
'use strict';

var requestScopeTracker = require('./request-scope-tracker');
var pulsarEventTracker = require('./pulsar-event-tracker');
var pulsarMetadataManager = require('./pulsar-metadata-cache-manager');


function PulsarContextManager(req, res) {
    this.req = req;
    this.res = res;
    this.scopeTracker = null;

    this.pulsarMetadataManagerInstance = pulsarMetadataManager.getInstance(this.req);

    req.ebay.tracking = req.ebay.tracking || {};
    req.ebay.tracking.metadataManager = this.pulsarMetadataManagerInstance;

    //request share
    this.scopeTracker = null;
}

var pro = PulsarContextManager.prototype;

pro.getRequestScopeTracker = function (options) {
    options = options || {};
    if(this.scopeTracker) {
        return this.scopeTracker;
    }
    this.scopeTracker = requestScopeTracker.getInstance(this.req, this.res, options);
    return this.scopeTracker;
};

pro.getPulsarEventTracker = function (options) {
    options = options || {};
    return pulsarEventTracker.getInstance(this.req, options);
};

function getInstance(req, res) {
    req = req || {};
    if(req && req.ebay && req.ebay.tracking && req.ebay.tracking.pulsarContextManager) {
        return req.ebay.tracking.pulsarContextManager;
    }
    req.ebay = req.ebay || {};
    req.ebay.tracking = req.ebay.tracking || {};
    var instance = new PulsarContextManager(req, res);
    req.ebay.tracking.pulsarContextManager = instance;
    return instance;
}

module.exports = {
    getInstance: getInstance
};
