// @ebay/tracking-ebay/lib/tracking2/pulsar-metadata-cache-manager.js
'use strict';

const path = require('path');
const utils = require('../utils');
const util = require('./common-utils');
const metadataSvc = require('../services/tracking20-metadata-service');
const eventMetaData = require('../../resources/tracking2-event-metadata.json');

let metadata = utils.tryRequire(path.resolve(__dirname, '../../resources/tracking2-event-metadata.json'));

class PulsarMetadataCacheManager {
    constructor(req) {
        this.metadataLoadCounter = 0;
        this.refreshInterval = 1000 * 60 * 60; // 1 hour

        const defaultInterval = 1000 * 60 * 60 * 24; //1day;
        const interval = req && req.ebayPulsarMetadataCacheManagerInterval || defaultInterval;

        /* Use the cached version while current data is loaded from the service */
        this.metadata = eventMetaData;

        // start refresh
        this.loadMetadataCache();

        // schedule refresh once a day
        setInterval(() => {
            this.loadMetadataCache();
        }, interval); //1 day
    }

    isInSampling(pulsarCtx) {
        if (!pulsarCtx) {
            return false;
        }

        const eventMeta = this.metadata[getEventMetaKey(pulsarCtx)];
        const threshold = eventMeta && eventMeta.sampling;

        if (!threshold) {
            return false;
        }

        if (threshold === 100) {
            return true;
        }

        if (!pulsarCtx.getStreamID()) {
            return false;
        }

        const hash = util.hash(pulsarCtx.getStreamID());
        return (hash % 100) < threshold;
    }

    loadMetadataCache(callback) {
        this.metadataLoadCounter = (this.metadataLoadCounter + 1) % 100;

        metadataSvc.getResourceInstancesByEventType((err, res) => {
            if (err) {
                return callback && callback(err);
            }
            if (res && res.body && res.body.eventTypes) {
                this.metadata = res.body.eventTypes;
            }
            callback && callback(err, !err && metadata);
        });

    }
}

let instanceCache = null;

function getEventMetaKey(pulsarCtx) {
    if(!pulsarCtx) {
        return null;
    }

    return pulsarCtx.getEventAction() + ":" + pulsarCtx.getEventFamily() + ":" + Number.parseFloat(pulsarCtx.getEventVersion()).toFixed(1) + ":" + pulsarCtx.getTenant();
}

function getRefreshInstance(req) {
    instanceCache = null;
    return getInstance(req);
}

function getInstance(req) {
    if (instanceCache) {
        return instanceCache;
    }
    instanceCache = new PulsarMetadataCacheManager(req);
    return instanceCache;
}

function postInstall(callback) {
    getInstance().loadMetadataCache(callback);
}

module.exports = {
    getRefreshInstance: getRefreshInstance,
    getInstance: getInstance,
    postInstall: postInstall,
    reset: function reset() {
        instanceCache = undefined;
    }
};
