// @ebay/tracking-ebay/lib/tracking2/pulsar-publisher.js
'use strict';

const net = require('net');
const assert = require('assert');
const moment = require('moment');
const deployEnv = require('@ebay/environment-ebay');
const logger = require('@ebay/logging-inc').logger('tracking-ebay:pulsar-publisher');

const EventData = require('./event-data');
const CommonUtils = require('./common-utils');

const TrackingRequest = require('./tracking-request');
const tracking20Service = require('../services/tracking20-service');

const Tags = require('../entities/tags.js');
const SojournerEventEnum = require('../soj/sojourner-event-enums.js');

const EXPERIMENTATION_TAGS = {};

function defineEPTag(name, sourceName) {
    Object.defineProperty(EXPERIMENTATION_TAGS, name, {
        enumerable: true,
        get() {
            return Tags.getTagByName(sourceName);
        }
    });
}

defineEPTag('experimentationSite', 'es');
defineEPTag('experimentationError', 'epec');
defineEPTag('experimentationChannel', 'ec');
defineEPTag('UserOverrideTreatmentIds', 'ot');
defineEPTag('userExperiencedTreatments', 'xt');
defineEPTag('encodedQualifiedTreatments', 'nqt');
defineEPTag('userIdentificationTimestamp', 'uit');
defineEPTag('userOverrideQualificationSources', 'os');
defineEPTag('encodedContextQualifiedTreatments', 'nqc');
defineEPTag('experimentationTrafficLossTreatments', 'tlt');

function defaultCallback(err, res) {
    var msg = [];
    if(err || !res) {
        logger.warn('trackng20 error', err);
        return;
    }

    if(res) {
        res.setEncoding('utf8');
        var body = res.body;
        var ev = body && body.evResp && body.evResp[0];
        if(body && body.ackValue === 'SUCCESS') {
            msg.push('tracking20 success. ');
            if(ev) {
                msg.push(' eventId: ');
                msg.push( ev.eventID );
                msg.push(' streamID: ');
                msg.push(ev.streamID );
            }
            logger.info.apply(logger, msg);
            return;
        }
        logger.warn('tracking20 warn:', ev && ev.errorMessage || 'Unknown', ev);

        return;
    }
}

/*
function isEventImpression(eventAction) {
    if(!eventAction) {
        return false;
    }
    eventAction = eventAction.trim().toUpperCase();
    var elements = [ "SERV", "OPEN", "VIEW" ];
    for(var i =0; i < elements.length; i++) {
        if(elements[i] === eventAction) {
            return true;
        }
    }
    return false;
}
*/

function getEventData(pulsarContext, headerDataDelegator) {
    const eventData = EventData.create();
    eventData.setActorDetails(headerDataDelegator.getActorInfo());
    eventData.setAltitude(headerDataDelegator.getAltitude());
    eventData.setLastAccountID(headerDataDelegator.getBestGuessUserId());

    //changing date to reflect the precision to milliseconds level
    var dateStr = moment().format("YYYY-MM-DDTHH:mm:ss:SSS Z");
    eventData.setCaptureTimeZone(dateStr);

    eventData.setChannelID(headerDataDelegator.getChannelID());
	eventData.setCreativeID(headerDataDelegator.getCreativeID());
    eventData.setEventID(pulsarContext.getEventId());
	eventData.setEventType(pulsarContext.getEventAction() + ":" + pulsarContext.getEventFamily() + ":" +
        pulsarContext.getEventVersion());
    eventData.setGeoDetails(headerDataDelegator.getGeoDetails());

    if (pulsarContext.ctx && pulsarContext.ctx.locality) {
        pulsarContext.addGlobal('uc', pulsarContext.ctx.locality.countryId);
        pulsarContext.addGlobal('bs', pulsarContext.ctx.locality.siteId);
        pulsarContext.addGlobal('ul', pulsarContext.ctx.locality.locale);
        pulsarContext.ctx.locality.isGeoCountry && pulsarContext.ctx.tracking.pulsarContext.setGlobalFlag(15, true);
    }

    let trkgFlgs = headerDataDelegator.getTrackingGlobalFlags() && headerDataDelegator.getTrackingGlobalFlags().trim();

    if (trkgFlgs) {
        eventData.setGlobalFlags(trkgFlgs);
    } else {
        eventData.setGlobalFlags(CommonUtils.encode(pulsarContext.getGlobalFlags()));
    }

    eventData.setEventFlags(CommonUtils.encode( pulsarContext.getEventFlags()));
    eventData.setGlobalTags(pulsarContext.getGlobalTags());
    eventData.setResponseCode(pulsarContext.getResponseCode());
    eventData.setExperimentTags(getExperimentTags(pulsarContext.tracking.sojournerContext));

    /**
     *  This code has been removed from Java Pulsar Publisher
     *
     *  if(isEventImpression(pulsarContext.getEventAction())) {
     *     pulsarContext.setGlobalFlag(PulsarFlagEnum.IMP.bitPosition, true);
     *  }
     */

     /**
      * TODO: Implment AVRO event payloads as in currently support only simple key-value pairs.
      * com.ebay.pulsar.client.library.PulsarPublisher.extractComplexProperties
      */
    eventData.setUrl(headerDataDelegator.getCurrentURL());
    eventData.setLegacyPayload(null);

    eventData.setNameValuePairs(pulsarContext.getEventPayload());
    eventData.setAccountDetails(headerDataDelegator.getAccountDetails());

    eventData.setReferralUrl(headerDataDelegator.getReferralUrl());
    eventData.setRequestCorrelationID(pulsarContext.getRequestCorrelationId());

    eventData.setSourceID(pulsarContext.getSourceDetails());
    eventData.setSpeedCourse(headerDataDelegator.getSpeedCourse());

    return [eventData];
}

function getExperimentTags(sojContext) {
    let experimentTags = {};

    let tagVal = sojContext.getDataAsString(SojournerEventEnum.APPLICATION, EXPERIMENTATION_TAGS.experimentationSite);
    let tag = parseInt(tagVal);

    if (tag) {
        experimentTags[EXPERIMENTATION_TAGS.experimentationSite.sojournerName] = tag;
    } else {
        logger.warn(`Logging experiment site it is either an invalid integer or not present, actual val:${tagVal}`);
    }

    tagVal = sojContext.getDataAsString(SojournerEventEnum.APPLICATION, EXPERIMENTATION_TAGS.experimentationChannel);
    tag = parseInt(tagVal);

    if (tag) {
        experimentTags[EXPERIMENTATION_TAGS.experimentationChannel.sojournerName] = tag;
    } else {
        logger.warn(`Logging experiment channel it is either an invalid integer or not present, actual val:${tagVal}`);
    }

    tag = sojContext.getDataAsString(SojournerEventEnum.APPLICATION, EXPERIMENTATION_TAGS.encodedQualifiedTreatments);
    tag && (experimentTags[EXPERIMENTATION_TAGS.encodedQualifiedTreatments.sojournerName] = tag);

    tag = sojContext.getDataAsString(SojournerEventEnum.APPLICATION, EXPERIMENTATION_TAGS.encodedContextQualifiedTreatments);
    tag && (experimentTags[EXPERIMENTATION_TAGS.encodedContextQualifiedTreatments.sojournerName] = tag);

    tag = sojContext.getDataAsString(SojournerEventEnum.APPLICATION, EXPERIMENTATION_TAGS.UserOverrideTreatmentIds);
    tag && (experimentTags[EXPERIMENTATION_TAGS.UserOverrideTreatmentIds.sojournerName] = tag);

    tag = sojContext.getDataAsString(SojournerEventEnum.APPLICATION, EXPERIMENTATION_TAGS.userOverrideQualificationSources);
    tag && (experimentTags[EXPERIMENTATION_TAGS.userOverrideQualificationSources.sojournerName] = tag);

    tag = sojContext.getDataAsString(SojournerEventEnum.APPLICATION, EXPERIMENTATION_TAGS.userExperiencedTreatments);
    tag && (experimentTags[EXPERIMENTATION_TAGS.userExperiencedTreatments.sojournerName] = tag);

    tag = sojContext.getDataAsString(SojournerEventEnum.APPLICATION, EXPERIMENTATION_TAGS.userIdentificationTimestamp);
    tag && (experimentTags[EXPERIMENTATION_TAGS.userIdentificationTimestamp.sojournerName] = tag);

    tag = sojContext.getDataAsString(SojournerEventEnum.APPLICATION, EXPERIMENTATION_TAGS.experimentationTrafficLossTreatments);
    tag && (experimentTags[EXPERIMENTATION_TAGS.experimentationTrafficLossTreatments.sojournerName] = tag);

    tag = sojContext.getDataAsString(SojournerEventEnum.APPLICATION, EXPERIMENTATION_TAGS.experimentationError);
    tag && (experimentTags[EXPERIMENTATION_TAGS.experimentationError.sojournerName] = tag);

    logger.isLogLevelEnabled('debug') && logger.debug('Pulasar tracking event tags: ', experimentTags);
    return experimentTags;
}

function createTrackingRequest(ctx) {
    const request = TrackingRequest.create();
    const headerDataDelegator = ctx.headerDelegator;

    const ip = headerDataDelegator.getIP();
    if (ip) {
        //FIXME: Check if both IPV4 and IPV6 can exist. If so then let the application set these values.
        if (net.isIPv6(ip)) {
            request.setIPv6(ip);
        } else {
            request.setIPv4(ip);
        }
    }

    request.setDeviceID(ctx.getDeviceId());
    request.setOriginDetails(headerDataDelegator.getOrigin());
    request.setStreamID(headerDataDelegator.getStreamId());
    request.setTenantID(headerDataDelegator.getTenantId());
    request.setUserAgent(headerDataDelegator.getUserAgent());

    request.setPoolName(headerDataDelegator.getPoolName());
    request.setServerName(headerDataDelegator.getServerName());
    request.setPublisherIP(headerDataDelegator.getPublisherIP());

    request.setGlobalCorrelationID(headerDataDelegator.getGlobalSessionCorrelationId());

    const eventDataList = getEventData(ctx, headerDataDelegator);
    request.eventData = eventDataList;

    if(deployEnv.isDev() ) {
        request.setJunitDebug(true);
    }

    return request;
}

function publish(ctx, callback) {
    assert.ok(ctx);
    var requestBody = createTrackingRequest(ctx);
  
    var options = {
        body: requestBody.toJson()
    };
    tracking20Service.getTracker(ctx.req).track(options, callback || defaultCallback);
}

module.exports.publish = publish;
