// @ebay/tracking-ebay/lib/tracking2/pulsar-header-data-delegator.js
'use strict';

const os = require('os');
const utils = require('./common-utils');
const UtilsEbay = require('@ebay/utils-ebay');

class PulsarHeaderDataDelegator {
    constructor(req) {
        this.req = req;
        this.ctx = req && req.ebay;
        this.euc = this.ctx && this.ctx.getEndUserContext();
        this.correlationTracking = this.ctx && this.ctx.getCorrelationTracking();
        const corrRequestIdCtx = this.ctx && this.ctx.getCorrelationRequestId();
        this.correlationRequestId = corrRequestIdCtx && corrRequestIdCtx.corrRequestId;
        this.correlationSession = this.ctx && this.ctx.tracking && this.ctx.tracking.getCorrelationSession();
    }

    getStreamId() {
        return this.correlationSession && this.correlationSession.getStreamId() || this.ctx && this.ctx.getGuid();
    }

    getRequestCorrelationId() {
        if (this.ctx.isService) {
            return this.correlationRequestId && this.correlationRequestId;
        }

        return this.ctx && this.ctx.getCorrelationGuid() || this.ctx && this.ctx.getGuid() || null;
    }

    getGlobalSessionCorrelationId() {
        let globalSessionCorrelationId = this.correlationSession && this.correlationSession.getGlobalSessionCorrelationId();

        /* Changing the order as in the Raptor Java pulsar context */
        if (!globalSessionCorrelationId) {
            globalSessionCorrelationId = this.ctx && this.ctx.getCorrelationGuid();
        }
        if (!globalSessionCorrelationId) {
            globalSessionCorrelationId = this.correlationTracking && this.correlationTracking.getCorrelationGuid();
        }
        return globalSessionCorrelationId;
    }

    getAccountName() {
        if (!this.ctx) {
            return null;
        }

        if (!this.ctx.isService) {
            return this.ctx.getUserId() || this.ctx.getJSPersistentUserId();
        }

        return this.ctx.getOriginalUserId();
    }

    getAccountId() {
        if (!this.ctx) {
            return null;
        }

        if (!this.ctx.isService) {
            return this.ctx.getAccountId() || this.ctx.getPersistentAccountId() || getUserIdFromHeader(this.req) || 0;
        }

        return this.ctx.getAccountId() || this.ctx.getOriginalAccountId();
    }

    getBestGuessUserId() {
        return this.ctx && this.ctx.getBestUserId();
    }

    getUserAgent() {
        return this.ctx && this.ctx.getUserAgent();
    }

    getAccountDetails() {
        let acctId = this.getAccountId();
        if (acctId === undefined || acctId === -1) {
            return null;
        } else {
            return '' + acctId;
        }
    }

    getIP() {
        return this.ctx.getRemoteAddr();
    }

    getDeviceInfo() {
        return this.euc && this.euc.getDeviceId();
    }

    getNextStreamId() {
        return this.ctx && this.ctx.getGuid() || null;
    }

    getOrigin() {
        const appName = this.getAppName();
        const appType = this.getAppType();

        if (!appName || !appType) {
            return null;
        }
        return appType + ':' + appName;
    }

    getTenantId() {
        return 'MP:' + (this.ctx && this.ctx.getSiteId());
    }

    getSamplingParamater() {
        return this.getStreamId();
    }

    getCurrentURL() {
            return this.euc && this.euc.getStartingPointURI() || this.req.ebay.requestURL;
    }

    getReferralUrl() {
        return this.euc && this.euc.getReferer() || this.req.headers.Referer || this.req.headers.referer;
    }

    getLegacyPayload() {
        if (!this.ctx) {
            return null;
        }

        if (this.ctx.isService) {
            return null;
        }

        if (!this.ctx.cookies) {
            return null;
        }

        const sid = this.ctx.cookies.getCookieValue('SOJOURNER_TRACKING');
        if (sid && sid.trim().length > 0) {
            return 'sid:' + sid;
        }
    }

    getAppName() {
        return this.ctx && this.ctx.appContext && this.ctx.appContext.appName;
    }

    getPoolName() {
        return this.ctx && this.ctx.appContext && this.ctx.appContext.poolName;
    }

    getHostName() {
        return os.hostname();
    }

    getPublisherIP() {
        return utils.getIPAddress();
    }
    setAppType(appType) {
        this.appType = appType;
    }

    getAppType() {
        return this.ctx.appContext && this.ctx.appContext.service ? 'svc' : 'app';
    }

    /**
    Logic to read value from experince service header or ebay Cookie Value for Sojourner Tracking
    moved to correlation session
    **/
    getSOJTrackingId() {
        /**andes calls should not read sid from cookie, andes may read from _trksid(rover clicks or impsressions)**/
        if (this.req && this.req.appName === 'andes') {
            return undefined;
        }
        return this.correlationSession && this.correlationSession.getSID();
    }


    /**
    Logic to return value from experince service header or ebay Cookie Value for Page Source Id
    moved to correlation session
    **/
    getPageSourceId() {
        /**andes calls should not read sid from cookie, andes may read from _trksid(rover clicks or impsressions)**/
        if (this.req && this.req.appName === 'andes') {
            return undefined;
        }
        return this.correlationSession && this.correlationSession.getSIID();
    }


    /**
    Logic to return value from experince service header or ebay Cookie Value for Click Tracking CTAG
    moved to correlation session
    **/
    getCTag() {
        return this.correlationSession && this.correlationSession.getCTag();
    }

    /**
     * Evalute this lzazily so any added SOJ context params make it to the header in webapps.
     */
    getCorrelationSessionHeader(clientSide) {
        return this.ctx && this.ctx.tracking && this.ctx.tracking.getCorrelationSession(clientSide);
    }

    getTrkParams() {
        return this.correlationSession && this.correlationSession.getTrackingParameters();
    }

    getOperationId() {
        return this.correlationSession && this.correlationSession.getOperationId();
    }

    getTrackingGlobalFlags() {
        return this.correlationSession && this.correlationSession.getTrackingGlobalFlags();
    }

    getServerName() {
        return UtilsEbay.getServerName(this.req);
    }

    getExpChannelId() {
        return 0;
    }

    isOptIn() {
        return false;
    }

    getQualified() {
        return null;
    }

    getActorInfo() {
        return null;
    }

    getGeoDetails() {
        return null;
    }

    getAltitude() {
        return null;
    }

    getSpeedCourse() {
        return null;
    }

    getChannelID() {
        // TODO Auto-generated method stub
        return 0;
    }

    getPlacementID() {
        // TODO Auto-generated method stub
        return 0;
    }

    getCampaignID() {
        // TODO Auto-generated method stub
        return 0;
    }


    getSegmentID() {
        // TODO Auto-generated method stub
        return 0;
    }


    getMessageID() {
        // TODO Auto-generated method stub
        return 0;
    }


    getCreativeID() {
        // TODO Auto-generated method stub
        return 0;
    }

    getKeywordID() {
        // TODO Auto-generated method stub
        return 0;
    }

    getKeyword() {
        // TODO Auto-generated method stub
        return null;
    }
}

function create(req) {
    return new PulsarHeaderDataDelegator(req);
}

function getUserIdFromHeader(req) {
    var header = req.headers.uid;
    if (header) {
        if (Object.prototype.toString(header) === '[object String]') {
            var actId = parseInt(header, 10);
            if (actId) {
                return actId;
            }
        }
    }
}

module.exports.create = create;
