// @ebay/tracking-ebay/lib/tracking2/pulsar-context.js
'use strict';

const IIDGenerator = require('./impression-id-generator');
const pulsarHeaderDataDelegator = require('./pulsar-header-data-delegator');
const logger = require('@ebay/logging-inc').logger('tracking-ebay:pulsar-context');
const EP_PAGE_ID = 2555220;

class PulsarContext {
    constructor(tracker, taasContext) {
        this.tsPayload = {};
        this.eventFlags = {};
        this.globalTags = {};
        this.globalFlags = {};
        this.experimentTags = {};
        this.eventPayloadMap = {};

        this.responseCode = 0;
        this.eventAction = null;
        this.eventFamily = null;
        this.eventVersion = null;

        this.req = null;
        this.tracker = null;
        this.tracking = null;

        if (tracker) {
            this.tracker = tracker;

            this.svcMap = tracker.svcMap;
            this.globalTags = tracker.globalTags;

            this.eventFlags = tracker.eventFlags;
            this.eventAction = tracker.eventAction;
            this.eventFamily = tracker.eventFamily;
            this.eventVersion = tracker.eventVersion;
            this.eventPayloadMap = tracker.eventPayloadMap;

            this.req = tracker.req;
            this.ctx = this.req && this.req.ebay;
            this.tracking = this.ctx && this.ctx.tracking;
        }

        if (!this.tracking || !(this.tracking.isService || this.tracking.getTaasContext)) {
            logger.warn('Failed to set Tracker, req.ebay.tracking is not available');
            return;
        }

        this.headerDelegator = pulsarHeaderDataDelegator.create(this.req);

        this.actorId = this.headerDelegator.getActorInfo();

        // Stream ID and GUID are same
        this.streamId = this.headerDelegator.getStreamId() || this.headerDelegator.getNextStreamId();
        this.initStreamId = this.streamId;

        this.deviceId = this.headerDelegator.getDeviceInfo();
        this.requestCorrelationId = this.headerDelegator.getRequestCorrelationId();
        this.globalSessionCorrelationId = this.headerDelegator.getGlobalSessionCorrelationId();

        this.userId = this.headerDelegator.getAccountName();
        this.accountId = this.headerDelegator.getAccountId();
        this.bestGuessUserId = this.headerDelegator.getBestGuessUserId();

        this.eventId = (Math.floor(Math.random() * 9007199254740992) + 1);

        // This code originally overwrites the eventPayloadMap page id with "this.ctx.getPageId()"
        // This prevents setting a custom page id on an event tracker. We don't understand why this code is here.
        // To be safe, we are allowing the EP_PAGE_ID to be set manually on the event tracker. We may consider allowing
        // any custom page id to be used instead of the one provided by this.ctx.getPageId().
        this.pageId = EP_PAGE_ID === this.eventPayloadMap.p ? EP_PAGE_ID : this.ctx.getPageId();
        this.pageId && this.pageId !== -1 && (this.eventPayloadMap.p = 0 | this.pageId);

        if (this.ctx.isExpSvc && this.ctx.isExpSvc()){
            this.operationId = this.pageId + '';
        }

        if (!this.operationId){
            this.operationId = this.headerDelegator.getOperationId();
        }

        this.operationId && this.operationId !== '-1' && (this.eventPayloadMap.cp = 0 | this.operationId);
        this.headerDelegator.getTrkParams() && (this.tsPayload.trkp = this.headerDelegator.getTrkParams());

        /* For Experience solution
        * all the service will emit siid,ciid,sid  in the pulsar event
        * siid,sid are from header
        * ciid is from the svcs context manager
        */
        this.sid = this.headerDelegator.getSOJTrackingId();
        this.sid && (this.tsPayload.sid = this.sid);
        this.sourceId = this.headerDelegator.getPageSourceId();
        this.sourceId && (this.tsPayload.siid = this.sourceId);

        // TODO: this condition may need to be relaxed to req.ebay.service if tracking team cofirms that ciid should be used for
        // non-exp/non-provider services
        if (this.ctx.isExpSvc && this.ctx.isExpSvc() || this.ctx.isExpSvcModuleProvider && this.ctx.isExpSvcModuleProvider()) {
            this.ciid = IIDGenerator.newImpressionID();
            this.ciid && (this.tsPayload.ciid = this.ciid);
            this.addGlobal('c', this.headerDelegator.getCTag());
        } else {
            /** TODO: ??? https://github.corp.ebay.com/Tracking/RaptorTrackingClientLibrary/blob/0c7948cb49e4caec4674a447b64da03be26848b4/PulsarTracking/src/main/java/com/ebay/pulsar/client/library/PulsarContextHelper.java
             * When running in web app following is performed in Java Raptor Stack do we have to do this too ?
             * 		    //copy soj flags to pulsar flags
             * 			for (int i=0; i < PulsarContextHelper.GLOBAL_FLAGS_LENGTH;i++){
             * 				if (SojournerContextManager.getInstance().isContextFlagSet((GlobalFlags)GlobalFlags.getEnum(GlobalFlags.class, i))){
             * 					ctx.setGlobalFlag(i, true);
             * 				}
             * 			}
             */
            taasContext && (this.tsPayload.ciid = taasContext.ciid);
        }

        const dnt = this.req.get('DNT');
        dnt && this.addGlobal('dnt', dnt);

    }

    add(k, v) {
        this.eventPayloadMap[k] = v;
    }

    addGlobal(k, v) {
        this.globalTags[k] = v;
    }

    addToSvcMap(k, v) {
        if (this.svcMap[k]) {
            return;
        }

        this.svcMap[k] = v;
    }

    setEventAction(eventAction) {
        this.eventAction = eventAction;
        this.tracker && this,this.tracker.setEventAction(eventAction);
    }
    getDeviceId() {
        return this.deviceId;
    }

    getBestGuessUserId() {
        return this.bestGuessUserId;
    }

    getEventId() {
        return this.eventId;
    }

    getEventAction() {
        return  this.eventAction || (this.tracker && this.tracker.eventAction);
    }

    getEventFamily() {
        return  this.eventFamily || (this.tracker && this.tracker.eventFamily) ;
    }

    getEventVersion() {
        return  this.eventVersion || (this.tracker && this.tracker.eventVersion );
    }

    setGlobalFlag(bitpos, value) {
        this.globalFlags[bitpos] = value;
    }

    getGlobalFlags() {
        return this.globalFlags;
    }

    getEventFlags() {
        return this.eventFlags;
    }

    getGlobalTags() {
        return this.globalTags;
    }

    getExperimentTags() {
        return this.experimentTags;
    }

    getEventPayload() {
        return this.eventPayloadMap;
    }

    getRequestCorrelationId() {
        return this.requestCorrelationId;
    }

    getSourceDetails() {
        return this.tsPayload;
    }

    setEventFlag(bitPos, booleanVal) {
        this.eventFlags[bitPos] = booleanVal;
    }

    isEventFlagSet(bitPos) {
        return this.eventFlags[bitPos];
    }

    getCIID() {
        return this.ciid;
    }

    getPageID() {
        return this.pageId;
    }

    getTenant() {
        return this.tenant || 'MP';
    }

    setTenant(tenant) {
        this.tenant = tenant;
    }

    getStreamID() {
        return this.streamId;
    }

    getResponseCode() {
        return this.responseCode;
    }
}

function createContext(tracker, callback) {
    if (typeof tracker === 'function') {
        callback = tracker;
        tracker = undefined;
    }

    const req = tracker && tracker.req;
    const ctx = req && req.ebay;
    const tracking = ctx && ctx.tracking;

    if (!tracking) {
        return callback(new Error('Cannot create pulsar context due to req.ebay.tracking not found'));
    }

    if (ctx.isService){
        callback(null, new PulsarContext(tracker));
        return;
    }

    if (!tracking.getTaasContext) {
        return callback(new Error('Cannot create pulsar context due to missing taas context. Make sure you have tracking middleware in the inbound pipeline'));
    }

    tracking.getTaasContext((err, taasContext) => {
        if (err) {
            return callback(err);
        }

        callback(null, new PulsarContext(tracker, taasContext));
    });
}

module.exports = {
    createContext : createContext
};
