// @ebay/tracking-ebay/lib/tracking2/pulsar-event-tracker.js
'use strict';

var pulsarPublisher = require('./pulsar-publisher');
var pulsarContext = require('./pulsar-context');
var assert = require('assert');
var logger = require('@ebay/logging-inc').logger('tracking-ebay:pulsar-event-tracker');

const X_CDN = 'x-cdn';
const X_REQUEST_CATEGORY = 'x-request-category';
const BOTTYPE_TAG = 'bottype';
const CDN_TAG = 'cdn';

var defaultCallback = function(err, res) {
    var msg = [];
    if(err || !res) {
        logger.info('trackng20 error', err);
        return;
    }
    if(res) {
        res.setEncoding('utf8');
        var body = res.body;
        var ev = body && body.evResp && body.evResp[0];
        if(body && body.ackValue === 'SUCCESS') {
            msg.push('tracking20 success. ');
            if(ev) {
                msg.push(' eventId: ');
                msg.push(ev.eventID);
                msg.push(' streamID: ');
                msg.push(ev.streamID);
            }
            logger.info.apply(logger, msg);
            return;
        }
        logger.warn('tracking20 warn:', ev && ev.errorMessage || 'Unknown', err);
        return;
    }
};

function PulsarEventTracker (req, options) {
    options = options || {};

    this.req = req;
    this.eventFamily = null;
    this.eventAction = null;
    this.eventVersion = null;
    this.eventPayloadMap = {};
    this.eventFlags = {};
    this.globalTags = {};
    this.svcMap = {};
    this.eventFamily = options.eventFamily;
    this.eventAction = options.eventAction;
    this.eventVersion = options.eventVersion;

    if (req.headers && req.headers[X_CDN] && req.headers[X_CDN].length < 10) {
        this.addData(CDN_TAG,  req.headers[X_CDN]);
    }

    if (req.headers && req.headers[X_REQUEST_CATEGORY] && req.headers[X_REQUEST_CATEGORY].length < 40) {
        this.addData(BOTTYPE_TAG,  req.headers[X_REQUEST_CATEGORY]);
    }
}

PulsarEventTracker.prototype.setEventType = function setEventType(eventAction, eventFamily,  version) {
    this.eventFamily = eventFamily;
    this.eventAction = eventAction;
    this.eventVersion = version;
};

PulsarEventTracker.prototype.setEventAction = function setEventAction(action) {
    this.eventAction = action;
};

PulsarEventTracker.prototype.setEventFamily = function setEventFamily(family) {
    this.eventFamily = family;
};

PulsarEventTracker.prototype.setEventVersion = function setEventVersion(version) {
    this.eventVersion = version;
};

PulsarEventTracker.prototype.addData = function addData(k, v) {
    this.eventPayloadMap[k] = v;
    this.hasNewData = true;
};

PulsarEventTracker.prototype.addGlobal = function(k, v) {
    this.globalTags[k] = v;
    this.hasNewData = true;
};

PulsarEventTracker.prototype.addToSvcMap = function(k, v) {
    if (this.svcMap[k]) {
        return;
    }
    this.svcMap[k] = v;
    this.hasNewData = true;
};

PulsarEventTracker.prototype.setEventFlag = function setEventFlag(key, value) {
    this.eventFlags[key] = value;
    this.hasNewData = true;
};


PulsarEventTracker.prototype.getPulsarContext = function(cb) {
    if (this.pulsarContext) {
        return cb(null, this.pulsarContext);
    }

    return pulsarContext.createContext(this, (err, context) => {
        this.pulsarContext = context;
        cb(err, context);
    });
};

PulsarEventTracker.prototype.flush = function flush(callback) {

    /** Validate Event */
    // public boolean isValidEventType(String family, String action, String version)
	// 		throws PulsarException {
	// 	return family != null && !family.isEmpty() && action != null
	// 			&& !action.isEmpty() && version != null && !version.isEmpty();
    // }

    /**
     * Validate si is not empty or null
     */


    /**
     * Validate page id is not 0 or -1
     */

    callback = callback || defaultCallback;

    if(!this.hasNewData) {
        logger.warn('Puslar: No New Data for Flushing!');
        return callback();
    }

    if (this.pulsarContext) {
        this.publish(this.pulsarContext, callback);
        return;
    }

    pulsarContext.createContext(this, (err, context) => {
        if (err) {
            logger.error(`Cannot create pulsar context due to ${err.message}`, err);
            callback(err);
            return;
        }

        this.pulsarContext = context;
        this.publish(this.pulsarContext, callback);
        return;
    });
};

PulsarEventTracker.prototype.publish = function publish(context, cb) {
    const metaManager = this.req.ebay && this.req.ebay.tracking && this.req.ebay.tracking.metadataManager;
    const sampling = metaManager ? metaManager.isInSampling(context) : false;

    this.hasNewData = false;

    if(sampling) {
        pulsarPublisher.publish(context, function (err, response) {
            // if error, then it is already logged by the client
            // we only want to anaylize and log partial failures
            // coming as part of the response
            err = err || validateResponse(response);

            cb(err, response);
        });
    }
};

function validateResponse(response) {
    if (response && response.body && response.body.ackValue !== 'SUCCESS') {
        const formattedErrors = [];
        const evResp = response.body.evResp;
        const errors = evResp && Array.isArray(evResp) ? evResp : [];

        errors.forEach(error => {
            if (error.errorMessage && Array.isArray(error.errorMessage.error)) {
                formattedErrors.push(JSON.stringify(error.errorMessage.error));
            }
        });

        const generalError = response.body.errorMessage;
        if (generalError && generalError.error) {
            generalError.error.forEach(error => {
                formattedErrors.push(JSON.stringify(error));
            });
        }

        const err = new Error(`Pulsar.publish error ask:${response.body.ackValue}, status:${response.body.httpStatusCode || 500}, message:${formattedErrors.join(';')}`);
        err.httpStatusCode = response.body.httpStatusCode;
        err.askValue = response.body.ackValue;

        logger.warn(err);

        return err;
    }
}

function getInstance(req, options) {
    assert.ok(req);
    options = options || {};
    var tracker = new PulsarEventTracker(req, options);
    return tracker;
}

module.exports = {
    getInstance : getInstance,
    Utils: {
        validateResponse: validateResponse
    }
};
