// @ebay/tracking-ebay/lib/tracking2/request-scope-tracker.js
'use strict';

var pulsarEventTracker = require('./pulsar-event-tracker');
var assert = require('assert');

class RequestScopeTracker {
    constructor(req, res, options) {
        assert.ok(req);
        options = options || {};

        this.sharedTracker = pulsarEventTracker.getInstance(req, options);
        this.flushFlag = false;
        var self = this;

        if (options.noAutoFlush) {
            return;
        }

        //flush at the end of the response
        if(res && res.once) {
            res.once('finish', function() {
                self.flushFlag = true;
                if(self.sharedTracker) {
                    self.sharedTracker.flush();
                }
            });
        }
    }

    setEventType(eventAction, eventFamily,  version) {
        this.sharedTracker.setEventType(eventAction, eventFamily,  version);
    }

    addData(key, value) {
        this.sharedTracker.addData(key, value);
    }

    setEventFlag(bitPos, booleanValue) {
        this.sharedTracker.setEventFlag(bitPos, booleanValue);
    }

    flush(cb) {
        this.sharedTracker.flush(cb);
    }

    getPulsarContext(cb) {
        this.sharedTracker.getPulsarContext(cb);
    }

    static getInstance(req, res, options) {
        assert.ok(req);
        options = options || {};
        var tracker = new RequestScopeTracker(req, res, options);
        return tracker;
    }
}

module.exports.getInstance = RequestScopeTracker.getInstance;
