// @ebay/tracking-ebay/lib/tracking2/source-id-builder.js
'use strict';

var Base64Coder = require('./base64coder').TrakingBase64;

var SOTR_LENGTH = 12;
var BASE_VERSION = 'a';
var PLACEHOLDER = 'z';
var VERSION_POSITION = [0];
var PAGEID_POSITION = [1, 2, 3, 4];
var CONFIG_POSITION = [5];
var MODULE_POSITION = [6, 7, 8];
var LINK_POSITION = [9, 10, 11];
var CURRENT_VERSION = 'b'; // enable new version
var MAX_VERSION = CURRENT_VERSION;

var MAX_ID = [16777215, 63, 10000000, 262143, 16777215];
var MIN_ID = [0, 0, 1, 1, 1];
var PAGE = 0;
var MODULE = 2;
var LINK = 3;

var SIZES = {
    pageId: 4,
    configId: 1,
    moduleId: 3,
    linkId: 3
};

var START = {
    pageId: 1,
    configId: 5,
    moduleId: 6,
    linkId: 9
};

function getIdFromBaseSotrCookie(cookieValue, pos) {
    var rb = '';
    for (var idx = 0; idx < pos.length; idx++) {
        if (cookieValue.charAt(pos[idx]) !== PLACEHOLDER) {
            rb += cookieValue.charAt(pos[idx]);
        }
    }
    return rb || undefined;
}

function isValidSotrCookie(cookieValue) {
    if (!cookieValue ||
        //check length of sotr
        cookieValue.length !== SOTR_LENGTH) {
        return false;
    }

    // the version is always base64 due to client side script
    // https://github.corp.ebay.com/CLEARCASE/v4ebaypres/blob/0358d9056128dcf553e2e8e11e2bab69da0839a7/DarwinTrackingSharedComponents/src/com/ebay/darwin/tracking/util/TrackingBaseDecoderImpl.java
    // so we just validate
    var version = cookieValue.charAt(VERSION_POSITION[0]);
    return version >= BASE_VERSION && version <= MAX_VERSION;
}

function isValidPageId(pageId) {
    return pageId >= MIN_ID[PAGE] && pageId <= MAX_ID[PAGE];
}

function isValidLinkId(linkId) {
    return linkId >= MIN_ID[LINK] && linkId <= MAX_ID[LINK];
}

function isValidModuleId(id) {
    return (id >= MIN_ID[MODULE] && id <= MAX_ID[MODULE]);
}

module.exports.format = function format(decodedValue) {
    return Object.keys(decodedValue).reduce(function reduce(memo, key) {
        var val = decodedValue[key];
        if (val !== undefined) {
            if (memo) {
                memo += '.';
            }
            memo += key.charAt(0) + val;
        }
        return memo;
    }, '');
};

module.exports.decode = function decode(cookie) {
    var pageId;
    var configId;
    var moduleId;
    var linkId;

    if (isValidSotrCookie(cookie)) {
        //extract pageid
        pageId = getIdFromBaseSotrCookie(cookie, PAGEID_POSITION);
        if (pageId !== undefined) {
            pageId = Base64Coder.decode(pageId);
            if (pageId === undefined) {
                return; // bad format
            }
            pageId = isValidPageId(pageId) ? pageId : undefined;
        }
        //extract configid
        configId = getIdFromBaseSotrCookie(cookie, CONFIG_POSITION);
        if (configId !== undefined) {
            configId = Base64Coder.decode(configId);
            if (configId === undefined) {
                return; // bad format
            }
        }
        //extract moduleId
        moduleId = getIdFromBaseSotrCookie(cookie, MODULE_POSITION);
        if (moduleId !== undefined) {
            moduleId = Base64Coder.decode(moduleId);
            if (moduleId === undefined) {
                return; // bad format
            }
            moduleId = isValidModuleId(moduleId) ? moduleId : undefined;
        }
        //extract linkId
        linkId = getIdFromBaseSotrCookie(cookie, LINK_POSITION);
        if (linkId !== undefined) {
            linkId = Base64Coder.decode(linkId);
            if (linkId === undefined) {
                return; // bad format
            }
            linkId = isValidLinkId(linkId) ? linkId : undefined;
        }

        return {
            pageId: pageId,
            configId: configId,
            moduleId: moduleId,
            linkId: linkId
        };
    }
};

module.exports.encode = function encode(decodedValue) {

    var chars = Object.keys(decodedValue).reduce(function reduce(memo, key) {
        var val = decodedValue[key];
        if (val !== undefined) {

            var size = SIZES[key];
            var start = START[key];

            var encoded = Base64Coder.encode(val).concat('zzzz').substring(0, size);
            encoded = encoded.split('');
            for (var sdx = 0; sdx < size; sdx++) {
                memo[start + sdx] = encoded[sdx];
            }
        }
        return memo;
    }, 'bzzzzzzzzzzz'.split(''));

    return chars.join('');
};
