// @ebay/tracking-ebay/lib/tracking2/tracking-request.js
'use strict';

class TrackingRequest {
    constructor() {
        this.data = {
            tn: null,
            or: null,
            ua: null,
            ipv4: null,
            ipv6: null,
            si: null,
            pn: null,
            sn: null,
            pip: null,
            // hostIP: null,
            gci: null,
            di: null,
            ev: []
        };
        this.eventData = [];
    }

    getGlobalCorrelationID() {
        return this.data.gci;
    }

    setGlobalCorrelationID(globalCorrelationID) {
        this.data.gci = globalCorrelationID;
    }

    getStreamID() {
        return this.data.si;
    }

    setStreamID(streamID) {
        this.data.si = streamID;
    }

    getOriginDetails() {
        return this.data.or;
    }

    setOriginDetails(originDetails) {
        this.data.or = originDetails;
    }

    getDeviceID() {
        return this.data.di;
    }

    setDeviceID(deviceID) {
        this.data.di = deviceID;
    }

    getUserAgent() {
        return this.data.ua;
    }

    setUserAgent(userAgent) {
        this.data.ua = userAgent;
    }

    getIPv4() {
        return this.data.ipv4;
    }

    setIPv4(ipv4) {
        this.data.ipv4 = ipv4;
    }

    getIPv6() {
        return this.data.ipv6;
    }

    setIPv6(ipv6) {
        this.data.ipv6 = ipv6;
    }

    getTenantID() {
        return this.data.tn;
    }

    setTenantID(tenantID) {
        this.data.tn = tenantID;
    }

    getEventData() {
        return this.eventData;
    }

    setEventData(eventData) {
        this.eventData = eventData;
    }

    isJunitDebug() {
        return this.data.jdg;
    }

    setJunitDebug(junitDebug) {
        this.data.jdg = junitDebug;
    }

    setPoolName (poolname) {
        this.data.pn = poolname;
    }

    setServerName(serverName) {
        this.data.sn = serverName;
    }

    getServerName() {
        return this.data.sn;
    }

    setPublisherIP(publisherIP) {
        this.data.pip = publisherIP;
    }

    getPublisherIP() {
        return this.data.pip;
    }

    setHostIP(hostIP) {
        this.data.hip = hostIP;
    }

    getHostIP() {
        return this.data.hip;
    }

    toJson() {
        var i = 0;
        var event = {};
        this.data.ev = [];
        for (i = 0; i < this.eventData.length; i++) {
            event = this.eventData[i];
            this.data.ev.push(event.toJson());
        }
        return this.data;
    }


    static create() {
        return new TrackingRequest();
    }
}
module.exports = {
    create: TrackingRequest.create
};
