// @ebay/tracking-ebay/lib/tracking2/user-action-id.js
'use strict';

const assert = require('assert');

const DELIMITER = 'S';
const TOTAL_FIELDS = 2;
const INITIAL_COUNT = 0;

class UserActionId {
	static isValidUserActionId(userActionIdString) {
		let userActionId = UserActionId.parse(userActionIdString);
		return userActionId == null ? false : true;
	}

	static parse(userActionIdString) {
		if (!userActionIdString) {
			return null;
		}

		const fields = userActionIdString.split(DELIMITER);
		if (fields.length !== TOTAL_FIELDS || !fields[0]) {
			return null;
		}

		const uuid = fields[0];
		const count = parseInt(fields[1]);

		if (!count) {
            return null;
        }

		try {
			return new UserActionId(uuid, count);
		} catch (e) {
			return null;
		}
	}

	constructor(uuid, count) {
		assert(uuid && uuid.trim(), 'UUID cannot be null or empty');

		this.uuid = uuid;
        this.count = count || INITIAL_COUNT;
	}

	getCount() {
		return this.count;
	}

	getUUID() {
		return this.uuid;
	}

	increment() {
		this.count++;
	}

	reset(uuid) {
		assert(uuid && uuid.trim(), 'UUID cannot be null or empty');
		this.uuid = uuid;
		this.count = INITIAL_COUNT;
	}

	setCount(count) {
		this.count = count;
	}

	toString() { // jshint ignore:line
		return this.uuid + DELIMITER + this.count;
	}
}

module.exports = UserActionId;
