// @ebay/geo-service-ebay/lib/geo-ebay.js
'use strict';

var Cache = require('lru-backup-cache'),
    GeoService = require('./GeoService'),
    moduleConfig = require('@ebay/module-config-inc'),
    debug = require('debug')('geoservice');


var geoserviceModuleConfig,
    modCache;

var DEFAULT_COUNTRY_CODE = 'US',
    DEFAULT_MARKETPLACE_ID = 'EBAY-US';

var HEADER = {
    MARKETPLACE_ID: 'X-EBAY-C-MARKETPLACE-ID',
    MARKETPLACE_ID_L: 'x-ebay-c-marketplace-id',
    ACCEPT_LANGUAGE: 'ACCEPT-LANGUAGE',
    ACCEPT_LANGUAGE_L: 'accept-language'
};



function initConfig(callback) {
    moduleConfig(module, function (err, config) {
        geoserviceModuleConfig = config.get('geoservice');
        modCache = modCache || new Cache(geoserviceModuleConfig.cache);
        callback();
        return;
    });
}


function _getGeoByCountryCode(countryCode, headers, cacheKey, retry, callback) {
    initConfig(function onConfig() {
        GeoService.getGeoByCountryCode(countryCode, headers, function (error, body) {
            if (error && retry) {
                headers[HEADER.MARKETPLACE_ID] = DEFAULT_MARKETPLACE_ID;
                _getGeoByCountryCode(DEFAULT_COUNTRY_CODE, headers, cacheKey, false, callback);
                return;
            }
            if (body) {
                modCache.set(cacheKey, body);
            }
            callback && callback(error, body);
        });
    });
}


function _getGeoCountries(headers, cacheKey, callback) {

    initConfig(function onConfig() {
        GeoService.getGeoCountries(headers, function (error, body) {
            if (body) {
                modCache.set(cacheKey, body);
            }
            callback && callback(error, body);
        });
    });

}

function getGeoByCountryCode(countryCode, headers, callback) {
    var cacheKey,
        cachedCopy;

    countryCode = countryCode || DEFAULT_COUNTRY_CODE;
    headers = headers || {};
    headers[HEADER.MARKETPLACE_ID] = headers[HEADER.MARKETPLACE_ID] ||
        headers[HEADER.MARKETPLACE_ID_L] ||
        DEFAULT_MARKETPLACE_ID;

    cacheKey = 'geo_' + headers[HEADER.MARKETPLACE_ID] + '_' + countryCode;
    debug(cacheKey);

    cachedCopy = modCache && modCache.get(cacheKey);

    if (cachedCopy) {
        callback(null, cachedCopy);
    } else {
        cachedCopy = modCache && modCache.getBackup(cacheKey);
        if (cachedCopy) {
            modCache.set(cacheKey, cachedCopy); //Cache response;
            debug('using backup copy');
            callback(null, cachedCopy);
            setImmediate(_getGeoByCountryCode.bind(null, countryCode, headers, cacheKey, true, null));
            return;
        }
        _getGeoByCountryCode(countryCode, headers, cacheKey, true, callback);
    }
}


function getGeoCountries(headers, callback) {
    var cacheKey,
        cachedCopy;

    headers = headers || {};
    headers[HEADER.MARKETPLACE_ID] = headers[HEADER.MARKETPLACE_ID] ||
        headers[HEADER.MARKETPLACE_ID_L] ||
        'EBAY-US';

    headers[HEADER.ACCEPT_LANGUAGE] = headers[HEADER.ACCEPT_LANGUAGE] ||
        headers[HEADER.ACCEPT_LANGUAGE_L] ||
        'en-US';

    cacheKey = 'all_' + headers[HEADER.MARKETPLACE_ID] + '_' + headers[HEADER.ACCEPT_LANGUAGE];
    debug(cacheKey);

    cachedCopy = modCache && modCache.get(cacheKey);

    if (cachedCopy) {
        callback(null, cachedCopy);
    } else {
        cachedCopy = modCache && modCache.getBackup(cacheKey);
        if (cachedCopy) {
            modCache.set(cacheKey, cachedCopy); //Cache response;
            debug('using backup copy');
            callback(null, cachedCopy);
            _getGeoCountries(headers, cacheKey, null);
            return;
        }
        _getGeoCountries(headers, cacheKey, callback);
    }
}

module.exports = {
    getGeoByCountryCode: getGeoByCountryCode,
    getGeoCountries: getGeoCountries
};
