// @ebay/geo-service-ebay/lib/GeoService.js
'use strict';

var svcClient = require('@ebay/service-client-ebay');

function svcClientGet(options, callback) {
    var client = svcClient.context({}).getClient('geoservice');
    return client.get().options(options).end(callback);
}

module.exports = {


    /**
     * getGeoByCountryCode
     * @param countryCode
     * @param headers
     * @param callback
     */
    getGeoByCountryCode: function getCountry(countryCode, headers, callback) {
        return svcClientGet({
            path: '/country/' + countryCode,
            headers: headers
        }, function (err, response) {

            if (err || !(response && response.body && response.body.ack === 'SUCCESS')) {
                err = err || response.body && response.body.errorMessage ||
                    new Error('fail to get location:/country for ' + countryCode);
                callback(err, null);
            } else {
                callback(err, response.body);
            }

        });

    },

    /**
     * getGeoCountries
     * @param headers
     * @param callback
     */
    getGeoCountries: function getCountries(headers, callback) {
        return svcClientGet({
            path: '/geocountries',
            headers: headers
        }, function (err, response) {

            if (err || !(response && response.body && response.body.ack === 'SUCCESS')) {
                err = err || response.body && response.body.errorMessage ||
                    new Error('fail to get location:/geocountries');
                callback(err, null);
            } else {
                callback(err, response.body);
            }
        });
    }


};
