// @ebay/openfeature-ebay/src/openfeature-lib/tracking/utils.ts
import { EPTag, EPTags, TrackingTagTypes } from '../types/EpTagsEventMap';

export const tagTypeLookup = {
    es: TrackingTagTypes.SCALAR,
    ec: TrackingTagTypes.SCALAR,
    os: TrackingTagTypes.VECTOR,
    ot: TrackingTagTypes.VECTOR,
    mdbreftime: TrackingTagTypes.SCALAR,
    epcalenv: TrackingTagTypes.SCALAR,
    eprlogid: TrackingTagTypes.SCALAR,
};

const EP_TRACKING_PAGEID = 2555220;
const EP_TRACKING_EVENT_ACTION = 'TAGXT';
const EP_TRACKING_EVENT_FAMILY = 'EXPRMT';

export {
    EP_TRACKING_PAGEID,
    EP_TRACKING_EVENT_ACTION,
    EP_TRACKING_EVENT_FAMILY,
};

/**
 * Translates the string ep tags map from service to qTags data structure.
 *
 * @param epTagsMap
 * @returns
 */
export function translateTags(epTagsMap: string): EPTags {
    const epTags: EPTags = {
        qTags: [],
    };

    if (epTagsMap) {
        const epTagsEventMap = epTagsMap;
        const epTagsMapArray = epTagsEventMap.split('&');

        epTagsMapArray.forEach((str) => {
            const keyVal = str.split('=');
            if (keyVal != null && keyVal.length == 2) {
                const tagName: string = keyVal[0].replace('!', '');
                const tagValue: string = keyVal[1];

                //add only valid tag names to qTags
                if (tagTypeLookup[tagName]) {
                    let tag: EPTag = {
                        key: tagName,
                        value: tagValue,
                        type: tagTypeLookup[tagName],
                    };
                    epTags.qTags.push(tag);
                }
            }
        });
    }
    return epTags;
}

export function addQtagsToPulsarEventTracker(
    pulsarEventTracker,
    data,
) {
    const { qTags = [] } = data || {};

    if (pulsarEventTracker) {
        pulsarEventTracker.setEventType(
            EP_TRACKING_EVENT_ACTION,
            EP_TRACKING_EVENT_FAMILY,
            `1.0`,
        );
        pulsarEventTracker.addData(`p`, EP_TRACKING_PAGEID);
        qTags.forEach((tag) => {
            if (tagTypeLookup.hasOwnProperty(tag.key)) {
                pulsarEventTracker.addData(tag.key, tag.value);
            }
        });
    }
}
/**
 * This helper function adds the list of treatment ids that were triggered to the xt tag using pulsar event tracker
 * @param pulsarEventTracker
 * @param treatmentIds 
 */
export function addXtTagsToPulsarEventTracker(
    pulsarEventTracker,
    treatmentIds
) {
    pulsarEventTracker.addData(`!xt`, treatmentIds);
}

/**
 * This helper function adds generic tag data to pulsarEventTracker
 * @param pulsarEventTracker
 * @param dataKey
 * @param dataValue 
 */
export function addTagToPulsarEventTracker(
    pulsarEventTracker,
    dataKey,
    dataValue
) {
    pulsarEventTracker.addData(dataKey, dataValue);
}
