// @ebay/opentelemetry-ebay/middleware.js
const utils = require('./lib/utils');
const otelEbay = require('.');
const utilsEbay = require('@ebay/utils-ebay');
const { SemanticAttributes } = require('@opentelemetry/semantic-conventions');
const {
    ROOT_CONTEXT,
    trace,
    context,
    SpanKind,
    SpanStatusCode,
    propagation
} = require('@opentelemetry/api');
const {
    suppressTracing
} = require('@opentelemetry/core');
const Hoek = require('hoek');
const HostIPChecker = require('@ebay/hostip-checker-ebay');
const UtilsEbay = require('@ebay/utils-ebay');

function isQualyscan(req) {
    req.isQualyscan = req.isQualyscan !== undefined ? req.isQualyscan :
        (HostIPChecker.isQualysRequest(req) ||
        HostIPChecker.isQualyscanIP(UtilsEbay.getRemoteAddr(req)));
    return req.isQualyscan;
}

function getRootContext(req) {
    return propagation.extract(ROOT_CONTEXT, req.headers);
}

module.exports = () =>
    async (req, res, next) => {
        const tracer = otelEbay.getTracer();

        // if trace is undefined, it means it is disabled
        if (!tracer) {
            return next();
        }

        const rootCtx = isQualyscan(req) ?
            suppressTracing(getRootContext(req)) :
            getRootContext(req);

        const span = tracer.startSpan('handleRequest', {
            attributes: {
                [SemanticAttributes.HTTP_METHOD]: req.method,
                [SemanticAttributes.HTTP_FLAVOR]: req.httpVersion,
                [SemanticAttributes.HTTP_SCHEME]: req.protocol,
                [SemanticAttributes.HTTP_URL]: utils.formatRequestUrl(req),
                [SemanticAttributes.NET_PEER_IP]: req.connection.remoteAddress,
                // we need to determine a separate attribute key to indicate 'client ip' instead of peer addr
                // [SemanticAttributes.NET_PEER_NAME]: req.headers && utilsEbay.getServerName(req),
                [SemanticAttributes.NET_PEER_PORT]: req.connection.remotePort,
                [SemanticAttributes.HTTP_USER_AGENT]: req.headers && req.headers['user-agent'],
                [SemanticAttributes.HTTP_REQUEST_CONTENT_LENGTH]: req.headers && req.headers['content-length'],
                [SemanticAttributes.HTTP_HOST]: req.headers && utilsEbay.getServerName(req),
                [SemanticAttributes.HTTP_TARGET]: req.originalUrl
            },
            kind: SpanKind.SERVER
        }, rootCtx);

        const ctx = trace.setSpan(rootCtx, span);
        const traceId = span.spanContext().traceId;
        const spanId = span.spanContext().spanId;
        const isRecording = span.isRecording();
        if (isRecording) {
            res.setHeader('traceid', traceId);
        }
        req.otel = {
            traceId: traceId,
            spanId: spanId,
            isRecording: isRecording
        };
        const closeSpan = Hoek.once((spanStatus, message) => {
            const pageName = req.ebay?.calPageName || utils.getPageName(req) || 'handleRequest';
            span.updateName(pageName);
            const spanStatusObj = { code: spanStatus };
            if (message) {
                spanStatusObj.message = message;
            }

            if (req.ebay?.getRlogId) {
                span.setAttribute('rlogid', req.ebay.getRlogId());
            }
            span.setAttribute('commandName', pageName);
            span.setAttribute(SemanticAttributes.HTTP_STATUS_CODE, res.statusCode);
            if (res.statusCode >= 400) {
                span.setAttribute(SemanticAttributes.EXCEPTION_MESSAGE, res.statusMessage);
                span.setAttribute(SemanticAttributes.EXCEPTION_STACKTRACE, res.stackTrace);
            }
            span.setStatus(spanStatusObj);
            span.end();
        });

        res.once('finish', () => {
            closeSpan(SpanStatusCode.OK);
        });

        res.once('close', () => {
            closeSpan(SpanStatusCode.ERROR, 'ECONNRESET');
        });

        res.once('error', (err) => {
            closeSpan(
                SpanStatusCode.ERROR,
                err.message
            );
        });

        context.with(ctx, next);
    };
