// @ebay/appmeta-ebay/lib/AppMeta.js
'use strict';

/**
 * Module dependencies.
 */
var env = require('@ebay/environment-ebay'),
    logger = require('@ebay/logging-inc').logger('appmeta-ebay:AppMeta'),
    debug = require('debug')('appmeta-ebay:AppMeta');

function getParamsMap(params) {
    var paramsMap = {};
    if (Array.isArray(params)) {
        params.forEach(function(param) {
            var pKey = param['cb:key'];
            var pValue = param.$t;
            paramsMap[pKey] = pValue;
        });
    } else {
        paramsMap[params['cb:key']] = params.$t;
    }
    return paramsMap;
}

function getKeyNameForAppSecret() {
    if (env.isProd() || env.isPreProd()) {
        return 'productionSecret';
    }

    if (env.isSandbox()) {
        return 'sandboxSecret';
    }

    return 'stagingSecret';
}

function AppMeta(rawMeta) {
    debug('rawMeta:', rawMeta);

    /* Keep the raw metadata around to make serialization easier */
    this._pagesByName = {};
    this._rawMeta = rawMeta;

    Object.assign(this, rawMeta);

    if (this['page-Map']) {
        var pages = this['page-Map'].page;

        if (!Array.isArray(pages)) {
            pages = [pages];
        }

        pages.forEach(function(pageMeta) {
            this._pagesByName[pageMeta.name] = pageMeta;
            var params = pageMeta['parameters-Map'].parameters;
            pageMeta.paramsMap = params ? getParamsMap(params) : {};
        }, this);
    }

    if (this['parameters-Map']) {
        var params = this['parameters-Map'].parameters;
        this._paramsMap = getParamsMap(params);
    }

    if (this._paramsMap) {
        const key = getKeyNameForAppSecret();
        this._secret = this._paramsMap[key] ? new Buffer(this._paramsMap[key], 'base64').toString() : null;
    }
}

AppMeta.serialize = function(appMeta) {
    return JSON.stringify(appMeta._rawMeta);
};

AppMeta.deserialize = function(reader) {
    return new Promise((resolve, reject) => {
        var json = '';

        reader()
            .on('data', function (data) {
                json += data;
            })
            .on('end', function () {
                var rawMeta = JSON.parse(json);
                var appMeta = new AppMeta(rawMeta);
                resolve(appMeta);
            })
            .on('error', function (err) {
                reject(err);
            });
    });
};

AppMeta.NO_CHANNEL_FOUND ='NO_CHANNEL_FOUND';
AppMeta.NO_CHANNEL_ACCESS ='NO_CHANNEL_ACCESS';


AppMeta.prototype = {
    getPageId: function(pageName) {
        var pageMeta = this.getPageMeta(pageName);

        return pageMeta ? pageMeta.id : null;
    },

    getPageMeta: function(pageName) {
        var pageMeta = this._pagesByName[pageName];
        if (pageMeta && pageMeta.lifecycleState === 'Deprecated' && !pageMeta._warned) {
            // Warn about pages that are marked as deprecated
            var err = `Detected a pageName:${pageMeta.name} that has been marked as deprecated, appmeta will continue using it, but you better mark it back to Deployed`;
            logger.error(new Error(err));
            pageMeta._warned = true;
        }
        return pageMeta ? pageMeta : null;
    },

    needIAFToken: function(pageName) {
        var pageMeta = this.getPageMeta(pageName);
        return pageMeta ? !!pageMeta.paramsMap.needsIAFToken : false;
    },

    getConsumerId: function() {
        return this.id;
    },

    getAppSecret: function() {
        if (this._secret === AppMeta.NO_CHANNEL_ACCESS || this._secret === AppMeta.NO_CHANNEL_FOUND) {
            return undefined;
        }

        return this._secret;
    },

    setAppSecret: function(value) {
        this._secret = value;
    },

    getEbayAppId: function() {
        var key;
        if (env.isProd() || env.isPreProd()) {
            key = 'prod-ebayAppId';
        } else if (env.isSandbox()) {
            key = 'sandbox-ebayAppId';
        } else if (env.isStage() || env.isQA()) {
            key = 'staging-ebayAppId';
        } else {
            key = 'dev-ebayAppId';
        }
        return this._paramsMap[key];
    },

    getEbayCertId: function() {
        var key;
        if (env.isProd() || env.isPreProd()) {
            key = 'prod-ebayCertId';
        } else if (env.isSandbox()) {
            key = 'sandbox-ebayCertId';
        } else if (env.isStage() || env.isQA()) {
            key = 'staging-ebayCertId';
        } else {
            key = 'dev-ebayCertId';
        }
        return this._paramsMap[key];
    },

    getEbayDevId: function() {
        var key;
        if (env.isProd() || env.isPreProd()) {
            key = 'prod-ebayDevId';
        } else if (env.isSandbox()) {
            key = 'sandbox-ebayDevId';
        } else if (env.isStage() || env.isQA()) {
            key = 'staging-ebayDevId';
        } else {
            key = 'dev-ebayDevId';
        }
        return this._paramsMap[key];
    },

    isSecretEncodedOnly: function() {
        return !!this._paramsMap.SECRET_ENCODED_ONLY || (env.isSandbox() && !!this._paramMap.SANDBOX_SECRET_ENCODED_ONLY);
    }
};

module.exports = AppMeta;
