// @ebay/appmeta-ebay/lib/service.js
'use strict';

var fs = require('fs'),
    path = require('path');

var Async = require('async'),
    xml2js = require('xml2js'),
    tryRequire = require('try-require');

var Provider = require('@ebay/service-client-ebay'),
    debug = require('debug')('appmeta-ebay:service'),
    esams = require('@ebay/esams-consumer-ebay');

var manifest = tryRequire(path.join(process.cwd(), 'manifest.json'));

var parser = new xml2js.Parser({
    mergeAttrs: true,
    explicitArray: false,
    charkey: '$t'
});

function getMetadataFromService(path, callback) {
    Provider.context({}).getClient('appmeta-rest-service').get().path(path).end(function(err, response) {
        if (err) {
            debug(err);
            return callback(err);
        }
        var result = response.body;

        // return as failed if body.error not empty
        if (result.error) {
            debug(result);
            return callback(result);
        }

        var bodyStr = result.metadata;
        if (bodyStr) {

            if (bodyStr.indexOf('xml version') === -1) {
                if(bodyStr.indexOf('application.metadata.raptor.ebay.com') === -1) {
                    var errMsg = 'No Xml Metadata! and Not Valid Metadata Payload ' + bodyStr;
                    debug(errMsg);
                    err = new Error(errMsg);
                    return callback(err, bodyStr);
                }
                bodyStr = '<?xml version="1.0" encoding="UTF-8" standalone="no"?>' + bodyStr;
            }
            debug(bodyStr);
            var len = bodyStr.length;
            var start = 0;
            if (bodyStr.charAt(0) === '"') {
                start = 1;
            }
            var end = len;
            if (bodyStr.charAt(len - 1) === '"') {
                end = len - 1;
            }
            bodyStr = bodyStr.substring(start, end);
            bodyStr = bodyStr.replace(/\\\"/g, '"');

            parser.parseString(bodyStr, function(err, data) {
                return callback(err, data);
            });

        } else {
            debug(err);
            return callback(err, bodyStr);
        }
    });
}

function findMetadata(appName, callback) {
    if (appName === getRunningAppName()) {
        Async.waterfall([
            function(cb) {
                var buildId = getBuildId();
                if(buildId) {
                    debug('buildId is', buildId);
                    return cb(null, buildId);
                }
                debug('buildId not found, fallback to get latest metadata');
                return cb(new Error('buildId not found, fallback to get latest metadata'));
            }, function(buildId, cb) {
                getMetadataFromService('/metadata_id/' + buildId, function(err, result) {
                    cb(err, result);
                });
            }
        ], function(err, results) {
            if(!err && results) {
                debug('got metadata from buildId');
                return callback(null, results);
            }
            debug('falling back to get latest metadata');
            getMetadataFromService('/name/' + appName, function(err, data) {
                callback(err, data);
            });
        });
    } else {
        // skip the build id if query metadata for other apps
        getMetadataFromService('/name/' + appName, function(err, data) {
            callback(err, data);
        });
    }
}

function getBuildId() {
    var cronus = loadCronusProp();
    var buildId;

    if (manifest && manifest.name && cronus && cronus.version) {
        buildId = manifest.name + '-' + cronus.version;
    }
    return buildId;
}

function getRunningAppName() {
    var appPkg = tryRequire(path.join(process.cwd(), 'package.json')) || {};
    var ebayProps = appPkg.gpaas || appPkg.ebay || {};
    return ebayProps['short-app-name'] || ebayProps.nugget || ebayProps.shortAppName || appPkg.name;
}

function loadCronusProp() {
    try{
        return JSON.parse(fs.readFileSync(path.join(process.cwd(), '../cronus.prop'), "utf8"));
    } catch(err) {
        return null;
    }
}

function getAppSecret(appName, callback) {
    var channelName = 'site.' + appName.toLowerCase() + '.tokensecret.nonkey';

    esams.getEsamsConsumer().getNonKey(channelName, '-1', function handleESAMSResponse(err, nonkey) {
        callback(err, nonkey && nonkey.value);
    });
}

module.exports.getAppSecret = getAppSecret;
module.exports.findMetadata = findMetadata;
