// @ebay/tracking-ebay/lib/tracking-context/web-context/bookkeeping.js
'use strict';

var eBayFlags = require('../../entities/flags.js');
var eBayTags = require('../../entities/tags.js');
var SojournerEventEnum = require('../../soj/sojourner-event-enums.js');
var botDetector = require('../../bot/BotDetector.js');
var utils = require('../../utils');

/**
 * This is a rewrite to com.ebay.raptor.tracking.bookkeeping.SojournerBookKeeping &
 * com.ebay.raptor.tracking.bookkeeping.TrackingBookKeeping called in RaptorTrackingHandler.handleRequest
 */
module.exports = {
    updateTrackingCookiesBookKeeping: function (req) {
        //guid
        this.handleGuid(req);
        //cguid
        this.handleCGuid(req);
        //click count
        this.handleClickCount(req);
        //user loginin
        this.handleUserLogin(req);
        //user registration
        this.handleUserReg(req);
        //js logging
        this.handleJsLogging(req);
        return this;
    },
    updateSojournerBookKeeping: function (req) {
        var ebay = req.ebay;
        var _tracking = req.ebay.tracking;
        var sojournerContext = _tracking.sojournerContext;

        // perform this action once per request
        // this action can happen early in case of service request
        if (_tracking._updateSojournerBookKeeping) {
            return this;
        }
        _tracking._updateSojournerBookKeeping = true;

        //referer
        var referer = req.get('referrer');
        if (referer) {
            sojournerContext.addContext(eBayTags.R, utils.hashCode(referer));
        }
        sojournerContext.addContext(eBayTags.T, ebay.getSiteId());
        sojournerContext.addContext(eBayTags.PN, ebay.getPartnerId());
        //moz pre-fetch
        var mozPre = req.get('X-moz');
        if (mozPre && mozPre.toLowerCase().indexOf('prefetch') >= 0) {
            sojournerContext.setContextFlag(eBayFlags.GlobalFlags.MOZILLA_PREFETCH, true);
        }
        //safari pre-fetch
        var safariPre = req.get('X-Purpose');
        if (safariPre && safariPre.toLowerCase().indexOf('preview') >= 0) {
            sojournerContext.setContextFlag(eBayFlags.GlobalFlags.SAFARI_PREVIEW, true);
            sojournerContext.setContextFlag(eBayFlags.GlobalFlags.NON_USER_REQ, true);
        }
        sojournerContext.addContext(eBayTags.P, ebay.getPageId());
        var lowVolumn = _tracking.isLowVolume;
        if (lowVolumn) {
            sojournerContext.addData(SojournerEventEnum.APPLICATION, eBayTags.LOW_VOLUME, '1');
        }
        var appliationId = _tracking.appliationId;
        if (appliationId) {
            sojournerContext.addData(SojournerEventEnum.APPLICATION, eBayTags.APP_ID, appliationId);
        }
        //trkparams
        var trkParams = req.query._trkparms;
        if (trkParams) {
            sojournerContext.addData(SojournerEventEnum.CLICK, eBayTags.TRACK_PARMS, trkParams);
        }

        var isBot = req.ebay.bot;
        if (isBot) {
            // keep this as we moved this tracking from bot
            req.ebay.tracking.trackTag('bot_provider', JSON.stringify(req.ebay.bot));
            // commented this out till we get rid of tracking TEV list
            // sojournerContext.setAppFlag(eBayFlags.GlobalFlags.DETECTED_BOT, true);
            // sojournerContext.setContextFlag(eBayFlags.GlobalFlags.DETECTED_BOT, true);
        }

        var userAgent = ebay.getUserAgent();
        var ip = ebay.getRemoteAddr();
        botDetector.load(function (err, detector) {
            if (err) {
                return;
            }
            var isBot = detector.isBotAgent(userAgent) || detector.isBotIp(ip);
            if (isBot) {
                sojournerContext.setAppFlag(eBayFlags.GlobalFlags.DETECTED_BOT, true);
                sojournerContext.setContextFlag(eBayFlags.GlobalFlags.DETECTED_BOT, true);
            }
        });

        return this;
    },
    handleGuid: function (req) {
        var ebay = req.ebay;
        var _tracking = req.ebay.tracking;
        var sojournerContext = _tracking.sojournerContext;

        var guidContext = ebay.getGuidContext();

        if (guidContext.bothGuidAbsent) {
            sojournerContext.setContextFlag(eBayFlags.GlobalFlags.BOTH_GUID_ABSNT, true);
        } else {
            guidContext.persistentGuidAbsent &&
                sojournerContext.setContextFlag(eBayFlags.GlobalFlags.PERSTNT_GUID_ABSNT, true);
            guidContext.sessionGuidAbsent &&
                sojournerContext.setContextFlag(eBayFlags.GlobalFlags.SESSION_GUID_ABSNT, true);
        }

        var currentGuid = guidContext.guid;
        sojournerContext.addContext(eBayTags.G, currentGuid);
        if (currentGuid.length > 7) {
            var hashStr = currentGuid.substring(6, 8);
            var d0 = parseInt(hashStr.charAt(0), 16);
            var d1 = parseInt(hashStr.charAt(1), 16);
            if (d0 >= 0 && d1 >= 0) { // don't use junk guid data
                sojournerContext.addContext(eBayTags.H, hashStr);
            }
        }
    },
    handleCGuid: function (req) {
        var cookies = req.ebay.cookies;
        var _tracking = req.ebay.tracking;
        var sojournerContext = _tracking.sojournerContext;
        var cookielets = cookies.Cookies;
        var cguid = cookies.getCookieValue(cookielets.DC_CORRELATION_GUID);
        if (cguid && cguid.length === 32) {//to-do, use the the cookielet metadata for max length
            sojournerContext.addData(SojournerEventEnum.APPLICATION, eBayTags.CORRELATION_GUID, cguid);
        }
    },
    handleClickCount: function (req) {
        var cookies = req.ebay.cookies;
        var _tracking = req.ebay.tracking;
        var sojournerContext = _tracking.sojournerContext;
        var cookielets = cookies.Cookies;
        var clickCount = cookies.getCookieValue(cookielets.SOJOURNER_CLICKID);
        var currentClickCount = 0;
        if (clickCount) {
            var countInCookie = parseInt(clickCount, 10);
            currentClickCount = isNaN(countInCookie) ? 0 : countInCookie;
        }
        currentClickCount++;
        cookies.setCookieValue(cookielets.SOJOURNER_CLICKID, currentClickCount);
        sojournerContext.addContext(eBayTags.C, currentClickCount);
    },
    handleUserLogin: function (req) {
        var cookies = req.ebay.cookies;
        var _tracking = req.ebay.tracking;
        var sojournerContext = _tracking.sojournerContext;
        var cookielets = cookies.Cookies;
        var acct = cookies.getCookieValue(cookielets.ACCOUNT_ID);
        if (!acct) {
            var puac = cookies.getCookieValue(cookielets.PERSISTENT_USERIDANDCHECK);
            if (puac) {
                var sawc = cookies.getCookieValue(cookielets.SOJOURNER_SAW);
                if (sawc && sawc.indexOf('.') > 0 && sawc.length > 2) {
                    acct = sawc.substring(2);//format is 1.acctId
                }
            }
        }
        if (acct) {
            sojournerContext.addContext(eBayTags.U, acct);
        } else {
            var usageAccounts = cookies.getCookieValue(cookielets.USAGE);
            if (usageAccounts) {
                // Need to see how many userids are in the cookie
                // e.g '20836680/0;20903773/0;25359898/0;'
                // most recently logged in user is 25359898 which is at the end of the string
                var usersArray = usageAccounts.split(';');
                for (var i = usersArray.length - 1; i > -1; i--) {
                    if (usersArray[i].length > 0) {
                        var userArray = usersArray[i].split('/');
                        if (userArray.length === 2) {
                            acct = userArray[0];
                            break;
                        }
                    }
                }
            }
            if (acct) {
                sojournerContext.addContext(eBayTags.BU, acct);
            }
        }
    },
    handleUserReg: function (req) {
        // cookielets.REG is blacklisted, hence the whole function is disable
        // var cookies = req.ebay.cookies;
        // var _tracking = req.ebay.tracking;
        // var sojournerContext = _tracking.sojournerContext;
        // var cookielets = cookies.Cookies;
        // BLACKLISTED ->> var regCookie = cookies.getCookieValue(cookielets.REG);
        // var regCookielet = cookies.getCookieValue(cookielets.REGISTERED);
        // if (regCookie && regCookielet) {
        //     sojournerContext.addContext(eBayTags.REGU, '1');
        // }
    },
    handleJsLogging: function (req) {
        var cookies = req.ebay.cookies;
        var _tracking = req.ebay.tracking;
        var sojournerContext = _tracking.sojournerContext;
        var cookielets = cookies.Cookies;
        var jsCookielet = cookies.getCookieValue(cookielets.EBAY_JS);
        if (jsCookielet === '1') {
            sojournerContext.addContext(eBayTags.JS, '1');
        }
    }
};
