// @ebay/openfeature-ebay/src/openfeature-lib/types/FeatureFlagErrorInfo.ts
export class FeatureFlagErrorInfo {
  private errorEvents: Map<string, number>;
  private errorsLogged: string[];
  constructor() {
    this.errorEvents = new Map();
    this.errorsLogged = [];
  }

  /**
   * Counts up the total count of errors from each error name in errorEvents.
   * @returns
   */
  public getTotalErrorCount() {
    if (this.errorEvents.size == 0) {
      return 0;
    }
    let total = 0;
    for (const [name, count] of this.errorEvents) {
      total += count;
    }
    return total;
  }

  public addErrorEvent(name: string): void {
    this.countOne(name);
    try {
      this.getErrorsList().push(name);
    } catch (e) {
        // EXPRMT-17397: avoid ArrayIndexOutOfBoundsException throw - do nothing since it is not critical
    }
  }

  public getErrorsList(): string[] {
    return this.errorsLogged;
  }

  public getErrorEvents(): Map<string, number> {
    return this.errorEvents;
  }

  private countOne(key: string): void {
    const currentCount = this.errorEvents.get(key) || 0;
    this.errorEvents.set(key, currentCount + 1);
  }
}
