// @ebay/tracking-ebay/lib/index.js
'use strict';


exports.Tags = require('./entities/tags');
exports.Flags = require('./entities/flags');
exports.Modules = require('./entities/modules');
exports.SojournerEventEnum = require('./soj/sojourner-event-enums');
exports.VectorTypeEnum = require('./soj/vector-type-enums');
exports.warmup = require('./warmup');
exports.middleware = require('./middleware');
exports.generateSPTagValue = require('./utils').generateSPTagValue;

[
    'trackTag',
    'trackVectorTag',
    'trackContextFlag',
    'trackFlag',
    'trackSojournerFromSvc',
    'trackModuleImpression',
    'track',
    'trackRequestHeader',
    'trackResponseHeader'
].forEach(function(deprecatedMethodName) {
    exports[deprecatedMethodName] = function() {
        throw new Error('The require("@ebay/tracking-ebay").' + deprecatedMethodName + '(...) method is deprecated. Use req.ebay.tracking.' + deprecatedMethodName + '(...) instead.');
    };
});

exports.renderInlineTrackingCode = require('./utils').renderInlineTrackingCode;
exports.updateCookies = require('./taas/helper').updateCookies;
