// @ebay/tracking-ebay/lib/utils.js
'use strict';

var _ = require('underscore');
var logger = require('@ebay/logging-inc').logger('tracking-ebay:utils');
const modConfig = require('@ebay/module-config-inc');
const Hmv = require('./tracking2/header-multi-value');

function toHex(c) {
    return c.charCodeAt(0).toString(16).toUpperCase();
}

function initEncodeMap() {
    var chars = ['!', '~', '\'', '(', ')', ' '];
    var map = {};
    chars.forEach(function(c) {
        map[c] = '%' + toHex(c);
    });
    return map;
}

var ENCODE_MAP = initEncodeMap();

exports.generateSPTagValue = function(pageId, moduleId, linkId) {
    if (!pageId) {
        logger.error(new Error('Unknown Page Id, Unable to generate SP Tag'));
        return '';
    }
    var id = [];
    id.push('p' + pageId);
    if (moduleId) {
        id.push('m' + moduleId);
    }
    if (linkId) {
        id.push('l' + linkId);
    }
    return id.join('.');
};

exports.encode = function encode(value) {
    var a = value.toString().split('');
    return _.map(a, function(c) {
        return ENCODE_MAP[c] ? ENCODE_MAP[c] : encodeURIComponent(c);
    }).join('');
};

var decode = exports.decode = function decode(value) {

    return decodeURIComponent(value.toString().replace(/\+/g, '%20'));
};

exports.vectorValueFromStr = function(val) {
    var string = val.toString();
    if (string.length === 0) {
        return null;
    }
    //1,2,3 or
    //1|2,3|4,5 or
    //nike%35shoes,nike
    return _.map(string.split(','), function(v) {
        if (v.length === 0) {
            return v;
        }
        var va = v.split('|');
        if (va.length === 1) {
            return decode(va[0]);
        } else {
            return _.map(va, function(r) {
                return decode(r);
            });
        }
    });
};

var isVector = exports.isVector = function(name) {
    return name && (name.indexOf('!') === 0);
};

exports.getSOJEventPart = function getSOJEventPart(sojEvent, part) {
    if (sojEvent && part && sojEvent.indexOf(part) > -1) {
        var start;

        if (sojEvent.indexOf('&' + part) > -1) {
            start = sojEvent.indexOf('&' + part) + (part.length + 2); //Add +1 for '=', Ex: ciid=123
        } else {
            start = sojEvent.indexOf(part) + (part.length + 1); //Add +1 for '=', Ex: ciid=123
        }

        var result = sojEvent.substring(start);
        var end = result.indexOf('&');

        return end > 0 ? result.substring(0, end) : result;
    }
};

exports.parseSOJResponseHeaders = function(sojEvent, headers) {
    if (!headers || !sojEvent) {
        return;
    }

    var result = {};
    result[sojEvent] = {};
    headers.split('&').forEach(function(nv) {
        var nvArray = nv.split('=');
        if (nvArray.length === 2) {

            var tag = nvArray[0];
            if (isVector(tag)) {
                result[sojEvent].vector = exports.vectorValueFromStr(nvArray[1]) || [];
            } else if (exports.isAppFlag(tag)) {
                result[sojEvent].appFlag = result.appFlag || nvArray[1];
            } else if (exports.isContextFlag(tag)) {
                result[sojEvent].contextFlag = result.contextFlag || nvArray[1];
            } else {
                result[sojEvent][nvArray[0]] = decode(nvArray[1]);
            }
        }
    });
    return result[sojEvent];
};

exports.getVectorName = function(name) {
    if (!isVector(name)) {
        return name;
    }
    return name.replace(/^(\!\_|\!)/, '');
};

exports.isAppFlag = function(name) {
    return name && name === 'flgs';
};

exports.isContextFlag = function(name) {
    return name && name === 'cflgs';
};

//convert base64 to bytes array
exports.btoa = function(b, standard) {
    if (!standard) {
        b = b.replace(/\*/g, '=');
    }
    return _.map(new Buffer(b, 'base64').toString().split(''), function(c) {
        return c.charCodeAt(0);
    });
};

//convert bytes array to base64 encoding
exports.atob = function(b, standard) {
    var base64Str = new Buffer(b).toString('base64');
    if (!standard) {
        base64Str = base64Str.replace(/\=/g, '*');
    }
    return base64Str;
};

exports.isFloat = function(n) {
    return (n === +n) && (n !== (n | 0));
};

exports.isInteger = function(n) {
    return (n === +n) && (n === (n | 0));
};

exports.hashCode = function(str) {
    if (!_.isString(str)) {
        str = str.toString();
    }
    for (var ret = 0, i = 0, len = str.length; i < len; i++) {
        ret = (31 * ret + str.charCodeAt(i)) << 0;
    }
    return ret;
};

function getInlineContent(contentPropName, taasContext) {

    var taasContextResponse = taasContext.response;

    var content = taasContextResponse[contentPropName];
    if (!content) {
        return '';
    }

    if (Array.isArray(content)) {
        content = content.join('\n');
    }

    return content;
}

function generateJsResources(resources, context, nonce, jsId) {
    if (resources) {
        resources.forEach(function(resourceUrl) {
            context.write('<script id="' + jsId + '" type="text/javascript" src="' + resourceUrl + '"' + nonce +
                ' crossorigin ></script>');
        });
    }
}

exports.renderInlineTrackingCode = function renderInlineTrackingCode(context, taasContext, expSvcImpressId) {
    var tracking = taasContext.trackingContext;
    if (tracking && tracking.isExpSvc && !tracking.expSvcImpression) {
        // if experience service flow is detected, but impression is missing,
        // let the developer know that the tracking-body should be
        // synchronized with service call response
        var err = new Error(
            'Experience service flow has been detected, tracking-body tag should be rendered once experience service response is received to make sure psi is correct, please see https://github.corp.ebay.com/nodejs/tracking-ebay#experience-service-case or due to experience service error the ciid was not passed back.');

        logger.error(err);
    }

    // NOTE: The inline "header" code depends on some RaptorJS modules
    //       that are loaded in the body slot so we put everything right
    //       after the body slot.
    var nonce = '';
    var req;
    if (context.stream && context.stream.req) {
        req = context.stream.req;
        nonce = req && req.csp && req.csp.nonce ? ' nonce="' + req.csp.nonce + '"' : '';
    }

    req = req || tracking.req;
    // render X_EBAY_C_CORRELATION_SESSION info for client side to use
    if (req && req.ebay) {
        const correlationSessionHeader = req.ebay.tracking.getCorrelationSession(true).buildHeader();
        if (tracking.isCacheRequest()) {
            const header = Hmv.parse(correlationSessionHeader);
            // delete si as it will be dynamically added on browser side
            delete header.value.si;
            context.write(`<script type="text/javascript"${nonce}>(function() {var trackingInfo = ${JSON.stringify({ X_EBAY_C_CORRELATION_SESSION: header.toString() })};var separator = trackingInfo.X_EBAY_C_CORRELATION_SESSION ? ',' : '';window.trkCorrelationSessionInfo={};window.trkCorrelationSessionInfo.getTrackingInfo=function(){return {X_EBAY_C_CORRELATION_SESSION: this.getTrackingCorrelationSessionInfo()};};window.trkCorrelationSessionInfo.getTrackingCorrelationSessionInfo=function(){return (trackingInfo.X_EBAY_C_CORRELATION_SESSION || '') + separator + "si=" + (window.tracker && window.tracker.guid);};})();</script>`);
        }
        else {
            context.write(`<script type="text/javascript"${nonce}>(function(scope) {var trackingInfo = ${
                JSON.stringify({ X_EBAY_C_CORRELATION_SESSION: correlationSessionHeader })};scope.trkCorrelationSessionInfo={};scope.trkCorrelationSessionInfo.getTrackingInfo=function(){return trackingInfo;};scope.trkCorrelationSessionInfo.getTrackingCorrelationSessionInfo=function(){return trackingInfo.X_EBAY_C_CORRELATION_SESSION};})(window)</script>`);
        }
    }

    context.write('<script type="text/javascript"' + nonce +
        '>if(typeof raptor !== "undefined" && raptor.require){var Uri=raptor.require("ebay/legacy/utils/Uri");$uri=function(href){return new Uri(href);};window.raptor.extend(window.raptor, require("ebay/legacy/adaptor-utils"));}</script>');

    modConfig(module, function(err, config) {
        if (err) {
            logger.error('failed to load configuration', err);
        }
        let enableBrotli = config && config.get('optimizer-plugin-inc:enableBrotli');
        if (enableBrotli) {
            if (req && req.headers && req.headers['accept-encoding']) {
                if (req.headers['accept-encoding'].indexOf('br') > -1) {
                    changeToBrotliUrls(taasContext.response.headerResourceFileName);
                    changeToBrotliUrls(taasContext.response.footerResourceFileName);
                }
            }
        }

        generateJsResources(taasContext.response.headerResourceFileName, context, nonce, 'taasHeaderRes');

        context.write('<script id="taasContent" type="text/javascript"' + nonce + '>try {' +
            replaceHeaderPSI(getInlineContent('headerContent', taasContext), expSvcImpressId) +
            getInlineContent('footerContent', taasContext) +
            '} catch (err) { console && console.log && console.log(err); }</script>');

        generateJsResources(taasContext.response.footerResourceFileName, context, nonce, 'taasFooterRes');
    });
};

function changeToBrotliUrls(arr) {
    //If client/browser supports brotli encoding, modify the url to
    for (let i in arr) {
        let ele = arr[i];
        arr[i] = ele.substring(0, ele.lastIndexOf('/') + 1) + 'br' +
            ele.substring(ele.lastIndexOf('/'));    //change http://example.com/a/b/c/ss.js to http://example.com/a/b/c/br/ss.js
    }
}

function replaceHeaderPSI(content, ciid) {
    if (!ciid || ciid === null) {
        return content;
    }
    var input = content;

    var startPattern = '"psi":"';
    var startIdx = content ? content.indexOf(startPattern) + startPattern.length : -1;

    if (ciid && startIdx >= 0 && content && content.length >= startIdx) {
        input = content.replace(/("psi":")([^"]*)(")/, '$1' + ciid + '$3');
        return input;
    } else {
        return content;
    }
}

exports.tryRequire = function tryRequire(name) {
    try {
        return require(name);
    } catch (err) {
        logger.error('Failed to load module: ', name);
    }
};

exports.filterTrksid = function filterTrksid(_trksid) {
    if (Array.isArray(_trksid)) {
        //This is Array, has duplicate URL params
        return _trksid.filter(function(item) {
            return item && item.length > 0;
        })[0];
    } else {
        return _trksid;
    }
};
