// @ebay/experimentation-ebay/lib/api.js
'use strict';

var extend = require('extend');

var Helper = require('./helper');

var CHANNEL_ENUM = {
    WEB: 1,
    MOBILE_WEB: 2,
    TABLET_WEB: 3,
    ANDROID_APP: 4,
    IOS_APP: 5,
    EMAIL: 6,
    PLACEHOLDER1: 7,
    PLACEHOLDER2: 8,
    OTHER: 99
};

module.exports = {
    CHANNEL_ENUM: CHANNEL_ENUM,

    getChannelId: function getChannelId(deviceInfo) {
        if (!deviceInfo) {
            return CHANNEL_ENUM.OTHER;
        }

        if (deviceInfo.isFSOM) {
            return CHANNEL_ENUM.WEB;
        }

        if (deviceInfo.requestIsWeb && deviceInfo.requestIsDesktop) {
            return CHANNEL_ENUM.WEB;
        }

        if (deviceInfo.requestIsWeb && deviceInfo.requestIsMobileWeb) {
            return CHANNEL_ENUM.MOBILE_WEB;
        }

        if (deviceInfo.requestIsWeb && /ios/i.test(deviceInfo.deviceOS)) {
            return CHANNEL_ENUM.WEB;
        }

        if (deviceInfo.requestIsWeb && /android/i.test(deviceInfo.deviceOS)) {
            return CHANNEL_ENUM.WEB;
        }

        if (!deviceInfo.requestIsWeb && /ios/i.test(deviceInfo.deviceOS)) {
            return CHANNEL_ENUM.IOS_APP;
        }

        if (!deviceInfo.requestIsWeb && /android/i.test(deviceInfo.deviceOS)) {
            return CHANNEL_ENUM.ANDROID_APP;
        }

        return CHANNEL_ENUM.OTHER;
    },

    getExperimentationContext: function getExperimentationContext(params, callback) {

        var defaultParams = {
            client: {
                siteId: 0,
                channelId: CHANNEL_ENUM.OTHER
            },
            userIdentity: {},
            sendQtags: true,

            contextParams: {},
            filterCriteria: {}
        };

        params = extend({}, defaultParams, params);

        if (params.deviceInfo) {
            params.client.channelId = Helper.getChannelId(params.deviceInfo);
            delete params.deviceInfo;
        }

        if (params.filterCriteria.needFactorInfo !== false) {
            params.filterCriteria.needFactorInfo = true;
        }

        Helper.invokeExperimentationSvc(params, callback);
    }
};
