// @ebay/experimentation-ebay/lib/apool.js
'use strict';

var NodeUtils = require('util');
var URL = require('url');

var extend = require('extend');
var Hoek = require('hoek');

var Cal = require('@ebay/cal');
var logger = require('@ebay/logging-inc').logger('experimentation-ebay/apool');
var httpunch = require("@ebay/httpunch");
var debug = require('debuglog')('@ebay/experimentation-ebay/apool');

var CLIENT_REDIRECT_TYPES = {
    REDIRECT_301: 301,
    REDIRECT_302: 302
};

/*
 * Returns true if the task has been executed successfuly
*/
module.exports = function apoolHandler(epContext, callback) {
    var requestContext = epContext.context;
    var bPoolTreatment = epContext.getBPoolTreatment();
    var redirectURL = bPoolTreatment.redirectUrl;
    var redirectType = bPoolTreatment.redirectType;
    var experimentId = bPoolTreatment.experimentId;
    var treatmentId = bPoolTreatment.treatmentId;

    if (requestContext.isBOTRequest() && !requestContext.isBotAllowed()) {
        logger.warn('EP Filter... This is a BOT request. Proxying terminated as experiment',
        bPoolTreatment.experimentId, 'does not allow Bots');
        // continue normal flow

        return callback();
	}

    logger.debug('EP Filter: Redirecting to destination host: %s', redirectURL);

    Cal.createTransaction('EPProxy',
        NodeUtils.format('EPProxy_E%sT%s', experimentId, treatmentId),
        function onTx(err, tx) {

            var onComplete = Hoek.once(function onComplete(status) {
                tx.complete(status);
            });

            var redirectCode = CLIENT_REDIRECT_TYPES[redirectType];

            if (redirectCode) {
                logger.info('EP Filter: %d Client Side Redirect initiated to: %s',
                    redirectCode, redirectURL);

                onComplete();
                requestContext.redirect(redirectCode, redirectURL);
                return;
            }

            var proxy = module.exports.createProxy(requestContext);

            proxy(bPoolTreatment)
                .on('error', Hoek.once(function onError(err) {
                    logger.error('EP Filter: Proxy failed', err);

                    onComplete(1);
                    if (err.code === 'ETIMEDOUT' || err.code === 'ECONNRESET') {
                        // continue old experience
                        logger.info('EP Filter: continue old epxerience');
                        err = undefined;
                    }

                    requestContext.trackFailedTreatment(bPoolTreatment);

                    callback(err);
                }))
                .once('close', onComplete)
                .once('end', onComplete);
        });
};

module.exports.createProxy = function createProxy(requestContext) {
    var options = requestContext.getAPoolConfig();
    var req = requestContext.req;
    var res = requestContext.res;

    // TODO: add metrics
    return function proxyTreatment(bpoolTreatment) {

        var requestOptions = URL.parse(bpoolTreatment.redirectUrl);
        requestOptions.socketTimeout = options.socketTimeout || 3000;
        requestOptions.connectTimeout = options.connectTimeout || 500;
        requestOptions.method = requestContext.getRequestMethod();

        // merge request headers with proxy headers
        requestOptions.headers = extend({}, req.headers, requestContext.getHeaders(), {
            'X-eBay-Client-IP': requestContext.getRemoteAddr(),
            'X-eBay-Original-URL': requestContext.getOriginalUrl(),
            'X-eBay-EP-Proxy-Treatment': bpoolTreatment.treatmentId,
            'X-eBay-EP-Proxy-Experiment': bpoolTreatment.experimentId,
            'X-eBay-EP-Proxy-Request': true,  // indicate this is Proxy request
            'X-eBay-Tracking-GUID-Flags': requestContext.serializeTrackingHeader()
        });

        if (requestContext.isPersistentGuidAbsent() || requestContext.isBothGuidAbsent()) {
            requestOptions.headers['X-eBay-New-GUID'] = requestContext.getGuid();
        }

        if (req.query) {
            requestOptions.qs = req.query;
        }

        if (Object.keys(req.body).length) {
            requestOptions.body = JSON.stringify(req.body);
        }

        debug('a-pool proxy to b-pool request options:', requestOptions);
        var proxyRequest = httpunch.request(requestOptions);

        proxyRequest.once('response', function(proxyResponse) {
            debug('proxy response headers:', proxyResponse.headers);

            res.status(proxyResponse.statusCode);
            res.set(proxyResponse.headers);

            proxyResponse.pipe(res);
        });

        req.pipe(proxyRequest);

        return proxyRequest;
    };
};
