// @ebay/experimentation-ebay/lib/ExperimentationContext.js
'use strict';

var ANY_ID_TREATMENT = 9;

function ExperimentationContext(context, result) {
    this.context = context;
    this._rawData = result || {};

    this._anyIDTreatments = null;
    this._treatmentLookup = null;
    this._treatmentsByFactorLookup = null;
    this._encodedEPQualificationResult = undefined;

    if (result.logId) {
        result.logId = decodeURIComponent(result.logId);
    }

    if (result.epQualificationResult) {
        this._encodedEPQualificationResult = result.epQualificationResult;
        result.epQualificationResult = decodeURIComponent(result.epQualificationResult);
    }
}

ExperimentationContext.prototype = {
    getRawData : function getRawData () {
        return this._rawData;
    },
    getEncodedEPQualificationResult: function getEncodedEPQualificationResult() {
        return this._encodedEPQualificationResult;
    },
    getAllTreatments : function getAllTreatments () {
        return this._rawData.treatments;
    },
    _getTreatmentLookup: function _getTreatmentLookup () {
        if (!this._treatmentLookup) {
            var lookup = this._treatmentLookup = {};

            var treatments = this._rawData.treatments;
            if (treatments) {
                for (var i = 0, len = treatments.length; i < len; i++) {
                    var treatment = treatments[i];
                    lookup[treatment.treatmentName] = treatment;
                    lookup[treatment.treatmentId] = treatment;
                }
            }
        }

        return this._treatmentLookup;
    },
    _getTreatmentsByFactorLookup: function _getTreatmentsByFactorLookup() {
        if (!this._treatmentsByFactorLookup) {
            var lookup = this._treatmentsByFactorLookup = {};

            var treatments = this._rawData.treatments;
            if (treatments) {
                for (var i = 0, len = treatments.length; i < len; i++) {
                    var treatment = treatments[i];
                    var factors = treatment.factors;
                    if (factors) {
                        for (var j = 0, len2 = factors.length; j < len2; j++) {
                            var factor = factors[j];
                            var factorIdTreatments = lookup[factor.id] || (lookup[factor.id] = []);
                            var factorNameTreatments = lookup[factor.name] || (lookup[factor.name] = []);
                            factorIdTreatments.push(treatment);
                            factorNameTreatments.push(treatment);
                        }
                    }
                }
            }
        }

        return this._treatmentsByFactorLookup;
    },
    _getFactorLookup: function _getFactorLookup() {
        if (!this._factorLookup) {
            var lookup = this._factorLookup = {};

            var treatments = this._rawData.treatments;
            if (treatments) {
                for (var i = 0, len = treatments.length; i < len; i++) {
                    var treatment = treatments[i];
                    var factors = treatment.factors;
                    if (factors) {
                        for (var j = 0, len2 = factors.length; j < len2; j++) {
                            var factor = factors[j];
                            lookup[factor.id] = factor;
                            lookup[factor.name] = factor;
                        }
                    }
                }
            }
        }

        return this._factorLookup;
    },

    _getAnyIDTreatments: function _getAnyIDTreatments() {

        if (this._anyIDTreatments) {
            return this._anyIDTreatments;
        }

        if (!this._rawData.treatments) {
            return undefined;
        }

       this._anyIDTreatments = [];
       var treatments = this._rawData.treatments;

       for (var i = 0, len = treatments.length; i < len; i++) {
           if (treatments[i].experimentType === ANY_ID_TREATMENT) {
               this._anyIDTreatments.push(treatments[i]);
           }
       }

       return this._anyIDTreatments;
    },

    getTreatment : function getTreatment(treatmentNameOrId) {
        var lookup = this._getTreatmentLookup();
        return lookup[treatmentNameOrId];
    },
    hasTreatment : function hasTreatment(treatmentNameOrId) {
        var lookup = this._getTreatmentLookup();
        return lookup.hasOwnProperty(treatmentNameOrId);
    },
    hasTreatmentsWithFactor : function hasTreatmentsWithFactor(factorNameOrId) {
        var lookup = this._getTreatmentsByFactorLookup();
        var treatments = lookup[factorNameOrId];
        return treatments !== undefined && treatments.length !== 0;
    },
    getTreatmentsWithFactor : function getTreatmentsWithFactor(factorNameOrId) {
        var lookup = this._getTreatmentsByFactorLookup();
        return lookup[factorNameOrId] || [];
    },
    getFactor : function getFactor(factorNameOrId) {
        var lookup = this._getFactorLookup();
        return lookup[factorNameOrId];
    },
    hasFactor : function hasFactor(factorNameOrId) {
        var lookup = this._getFactorLookup();
        return lookup.hasOwnProperty(factorNameOrId);
    },
    getAnyIDTreatments : function getAnyIDTreatments() {
        return this._getAnyIDTreatments();
    },

    trackServedTreatment: function trackServedTreatment(treatment) {
        if (!treatment) {
            throw new Error('Invalid treatment');
        }
        if (typeof treatment === 'string') {
            treatment = this.getTreatment(treatment);
        }

        this.context.trackServedTreatment(treatment);
    },

    isAPoolRequest: function isAPoolRequest() {
        return !!this._rawData.bPoolTreatment;
    },

    getBPoolTreatment: function getBPoolTreatment() {
        return this._rawData.bPoolTreatment;
    },

    getClientSideExperiments: function getClientSideExperiments() {
        const response = {};

        if (this._rawData.qTags) {
            response.qTags = this._rawData.qTags;
        }

        if (this._rawData.treatments) {
            const clientSideEps = [];
            for(const element of this._rawData.treatments) {
                if (!element.clientSideEp) {
                    continue;
                }
                let exp = {
                    'factors': element.factors,
                    'treatmentId': element.treatmentId
                };
                clientSideEps.push(exp);
            }
            response.clientSideEps = clientSideEps;
        }
        response.siteId = this.context.getSiteId();
        
        return response;
    }
};

module.exports = ExperimentationContext;
