// @ebay/experimentation-ebay/lib/middleware.js
'use strict';

var Url = require('url');
var modConfig = require('@ebay/module-config-inc');

var logger = require('@ebay/logging-inc').logger('experimentation-ebay/middleware');

var ContextFactory = require('./request-context');

var apool = require('./apool');
var bpool = require('./bpool');
var helper = require('./helper');

module.exports = function experimentation(req, res, next) {

    modConfig(module, function (err, config) {
        if (err) {
            throw new Error('Unable to initialize the experimentation-ebay module. Error: ' + (err.stack || err));
        }

        if (!config.get('experimentation:enabled')) {
            logger.info('experimentation is disabled');
            return next();
        }

        var routeConfig = req.route && req.route.config && req.route.config.ep;
        if (routeConfig && routeConfig.enabled === false) {
            logger.info('experimentation is disabled for this route');
            return next();
        }

        // TODO: handle COS and SSE service context
        //https://github.corp.ebay.com/experimentation-platform/ep_for_raptor/blob/dev-2.4.x-master/RaptorExperimentationParent/RaptorExperimentationAdaptor/src/main/java/com/ebay/raptor/experimentation/handler/WebExperimentationHandler.java#L120

        var requestContext = ContextFactory.create(req, res,
                config.get('experimentation'));

        logEpMeta(requestContext);

        var taskFn;
        if (requestContext.isBPoolRequest()) {
            logger.debug('B-pool request detected.',
                requestContext.isBPoolEnabled() ? '' : 'But it is disabled in config');

            requestContext.setResponseHeader('b-rlogid',
                requestContext.ebay && requestContext.ebay.getRlogId ?
                    requestContext.ebay.getRlogId() :
                    'unknown');

            if (requestContext.isBPoolEnabled()) {
                taskFn = bpool.bind(null, requestContext);
            }
        }
        else if (requestContext.isAPoolEnabled()) {
            taskFn = function aPool(next) {
                requestContext.getContext(function onContext(err, epContext) {

                    if (err) {
                        logger.error('Failed to get EP context', err);
                        return next();
                    }
                    if (!epContext.isAPoolRequest()) {
                        return next();
                    }

                    logger.debug('A-pool request detected');
                    // when apool proxy request, the response ends here
                    // otherwise next will be called
                    apool(epContext, next);
                });
                return;
            };

        }

        if (taskFn) {
            return taskFn(next);
        }
        return next();
    });
};

function logEpMeta(ep) {
    if (logger.isLogLevelEnabled('debug')) {
        logger.debug('EP Filter: URL command: %s', ep.getCommand());
        var urlParse = tryParse(ep.getRequestURL());
        if (urlParse) {
            logger.debug('EP Filter: URL Hostname: %s', urlParse.hostname);
            logger.debug('EP Filter: URL queryString: %s', urlParse.query);
        }
        else {
            logger.debug('EP Filter: URL Hostname: Malformed URL - Hostname could not be retrieved');
        }

        var params = ep.getRequestParams();
        if (params) {
            Object.keys(params).forEach(function forEach(name) {
                logger.debug('EP Filter: URL param name: %s valie: %s',
                    name, params[name]);
            });
        } else {
            logger.debug('EP Filter: Incoming URL has no paramters');
        }

        var headers = ep.getRequestHeaders();
        if (headers) {
            logger.debug('EP Filter: A-Pool received the following headers');
            Object.keys(headers).forEach(function forEach(name) {
                logger.debug('\tHeader ===> %s value=%s',
                    name, 
                    helper.isCookieHeader(name) ? '*masked*' : headers[name]);
            });
        }
    }
}

function tryParse(url) {
    try {
        return Url.parse(url);
    }
    catch (err) {

    }
}
