// @ebay/tracking-ebay/lib/services/tracking-wrappers.js
'use strict';

let agentList = require('../../config/tevBotAgents.json');
let ipList = require('../../config/tevBotIps.json');
var _ = require('underscore');

const Provider = require('@ebay/service-client-ebay');

const ROOT_NAMES = {
    getTrackingProperties: 'trackingPropertyList',
    getModules: 'moduleList',
    getTrackingFlags: 'flagList',
    getTrackingFlagsets: 'flagsetList',
    getPageFlagset: 'pageFlagsetList'
};

const REQUEST_ROOT_NAMES = {
    getTrackingProperties: 'trackingProperty',
    getModules: 'module',
    getTrackingFlags: 'trackingFlag',
    getTrackingFlagsets: 'trackingFlagset',
    getPageFlagset: 'pageFlagset'
};

const LIFE_CYCLE_STATUS = {
    New: 1,
    Deployed: 3,
    Deprecated: 4,
    Retired: 5
};

const LIFE_CYCLE_STATUS_NUM = {
    1: 'New',
    3: 'Deployed',
    4: 'Deprecated',
    5: 'Retired'
};

function remapPuslarMetaToTEV(data, rootName) {
    const list = data[rootName];
    delete data[rootName];
    switch (rootName) {
        case 'flagList':
            var flagListTransformedData = {
                ack: data.ack.charAt(0).toUpperCase() + data.ack.slice(1).toLowerCase(),
                [rootName]: list.map(it => ({
                    attributes: {
                        id: it.id.toString(),
                    },
                    name: it.name,
                    description: it.description,
                    lifeCycleState: LIFE_CYCLE_STATUS_NUM[it.status],
                    flagsetId: it.flagsetId.toString(),
                    bit: it.bitPosition.toString(),
                    authoringTool: it.authoringTool
                })),
            };
            removeEmptyValue(flagListTransformedData.flagList);
            return flagListTransformedData;

        case 'trackingPropertyList':
            var propertyListTransformedData = {
                ack: data.ack.charAt(0).toUpperCase() + data.ack.slice(1).toLowerCase(),
                [rootName]: list.map(it => ({
                    attributes: {
                        id: it.id.toString(),
                    },
                    name: it.name,
                    owner: it.ownerEmail,
                    description: it.description,
                    lifeCycleState: LIFE_CYCLE_STATUS_NUM[it.status],
                    sojournerName: it.sojName,
                    dataType: it.dataType,
                    authoringTool: it.authoringTool
                })),
                totalCount: data.totalCount,
                listCount: data.listCount
            };

            removeEmptyValue(propertyListTransformedData.trackingPropertyList);
            return propertyListTransformedData;

        case 'pageFlagsetList':
            return{
                ack: data.ack.charAt(0).toUpperCase() + data.ack.slice(1).toLowerCase(),
                [rootName]: list.map(it => ({
                    pageId: it.pageId.toString(),
                    flagsetId: it.flagsetId.toString(),
                    authoringTool: it.authoringTool,
                    lifeCycleState: LIFE_CYCLE_STATUS_NUM[it.status]
                })),
                totalCount: data.totalCount,
                listCount: data.listCount

            };

        case 'flagsetList':
            var flagsetListTransformedData = {
                ack: data.ack.charAt(0).toUpperCase() + data.ack.slice(1).toLowerCase(),
                [rootName]: list.map(it => ({
                    attributes: {
                        id: it.id.toString(),
                    },
                    name: it.name,
                    owner: it.ownerEmail,
                    description: it.description,
                    lifeCycleState: LIFE_CYCLE_STATUS_NUM[it.status],
                    authoringTool: it.authoringTool
                })),
                totalCount: data.totalCount,
                listCount: data.listCount
            };

            removeEmptyValue(flagsetListTransformedData.flagsetList);
            return flagsetListTransformedData;

        case 'moduleList':
            var moduleTransformedData = {
                ack: data.ack.charAt(0).toUpperCase() + data.ack.slice(1).toLowerCase(),
                [rootName]: list.map(it => ({
                    attributes: {
                        id: it.id.toString(),
                    },
                    name: it.name,
                    description: it.description,
                    lifeCycleState: LIFE_CYCLE_STATUS_NUM[it.status],
                    authoringTool: it.authoringTool
                })),
                totalCount: data.totalCount,
                listCount: data.listCount
            };
            removeEmptyValue(moduleTransformedData.moduleList);
            return moduleTransformedData;

         default:
            throw new Error(`Unknown transform type not supported ${data} `);
}
}

function removeEmptyValue(TransformedDataObj) {
    for (let item of TransformedDataObj) {
        for (let [key, value] of Object.entries(item)) {
            if (value === null || value === undefined){
                delete item[key];
            }
        }
    }
}

function invokeTrackingEventValidationService(operation, request, callback) {
    return invokePulsarMetadataService(operation, request, (err, data) => {
        if (err) {
            return callback(err);
        }
        if (!/success/i.test(data.ack)) {
            return callback(new Error(`${data.errorMessage || `Failure to fetch pulsar metadata`}, ack: "${data.ack}"`));
        }
        const rootName = ROOT_NAMES[operation];
        callback(null, remapPuslarMetaToTEV(data, rootName));
    });
}

function invokePulsarMetadataService(operation, request, callback) {
    callback = arguments[arguments.length - 1];

    const opRequest = request[REQUEST_ROOT_NAMES[operation]];
    opRequest.status = LIFE_CYCLE_STATUS[opRequest.status];
    opRequest.id = opRequest.$id;

    delete opRequest.lifeCycleState;
    delete opRequest.$id;

    var params = {
        body: JSON.stringify(request)
    };

    Provider
        .context({})
        .getClient('tracking20metadataservice')
        .post(params)
        .path(`/spf/${operation}`)
        .end((err, response) => callback(err, response && response.body));

}

module.exports.invokeTrackingEventValidationService = invokeTrackingEventValidationService;

module.exports.getTrackingFlags = invokeTrackingEventValidationService.bind(null, 'getTrackingFlags');

module.exports.getTrackingFlagsets = invokeTrackingEventValidationService.bind(null, 'getTrackingFlagsets');

module.exports.getPageFlagset = invokeTrackingEventValidationService.bind(null, 'getPageFlagset');

module.exports.getTrackingProperties = invokeTrackingEventValidationService.bind(null, 'getTrackingProperties');

module.exports.getModules = invokeTrackingEventValidationService.bind(null, 'getModules');

// PMSVC spf APIs
module.exports.invokePulsarMetadataService = invokePulsarMetadataService;

module.exports.getPulsarTrackingFlags = invokeTrackingEventValidationService.bind(null, 'getTrackingFlags');

module.exports.getPulsarTrackingFlagsets = invokeTrackingEventValidationService.bind(null, 'getTrackingFlagsets');

module.exports.getPulsarPageFlagset = invokeTrackingEventValidationService.bind(null, 'getPageFlagset');

module.exports.getPulsarTrackingProperties = invokeTrackingEventValidationService.bind(null, 'getTrackingProperties');

module.exports.getPulsarModules = invokeTrackingEventValidationService.bind(null, 'getModules');


module.exports.getTEVList = function getTEVList(callback) {
    var listValueExtractor = function (element) {
        return element.value || '';
    };

    var agents = [];
    var ips = [];
    agents = _.map(agentList.list.listElement, listValueExtractor);
    ips = _.map(ipList.list.listElement, listValueExtractor);

    callback(null, {
        'agents': agents,
        'ips': ips
    });
};

module.exports.getClientTrackingCtx = function getClientTrackingCtx(request, headers, callback) {
    var params = {
        operation: 'getClientTrackingCtx',
        body: request
    };

    const req = Provider
        .context({})
        .getClient('ebaytrackingservice')
        .post(params);

    if (headers) {
        Object.keys(headers).forEach(name => req.set(name, headers[name]));
    }

    req.end((err, response) =>
        callback(err, response && response.body));
};

module.exports.remapPuslarMetaToTEV = remapPuslarMetaToTEV;
