// @ebay/hostip-checker-ebay/lib/qualyscan.js
'use strict';

var ClientProvider = require('@ebay/service-client-ebay');
var moduleConfig = require('@ebay/module-config-inc');
var RaptorCache = require('raptor-cache');

var DEFAULT_REFRESH_TIMEOUT = 1000 * 60 * 60 * 24;

var logger = require('@ebay/logging-inc').logger('@ebay/hostip-checker-ebay/qualyscan');

var coldCache = createDiskCache();
var lastRefreshIP = 0;
var lastRefreshHeader = 0;
var qualysIPs;
var qualysHeader;
var config;

function loadIPs(callback) {
    var svc = ClientProvider.context({}).getClient('qualyIpSvc');
    svc.get().end(function handleResponse(err, res) {
        if (err) {
            return callback(err);
        }

        var ips;
        if (res.body && Array.isArray(res.body.qualysIPs)) {
            ips =  res.body.qualysIPs.reduce(function reduce(memo, ip) {
                memo[ip] = true;
                return memo;
            }, {});
        }
        else {
            err = new Error('Cannot recognize structure for qualyscan IP list ' +
                (Buffer.isBuffer(res.body) ? res.body.toString() : res.body));
        }

        callback(err, ips);
    });
}

function loadHeader(callback) {
    var svc = ClientProvider.context({}).getClient('qualyHeaderSvc');
    svc.get().end(function handleResponse(err, res) {
        if (err) {
            return callback(err);
        }
        var headerMap;
        if (res.body) {
            headerMap = res.body;
        }
        else {
            err = new Error('Cannot recognize structure for qualyscan header response ' +
                (Buffer.isBuffer(res.body) ? res.body.toString() : res.body));
        }
        callback(err, headerMap);
    });
}

function isEnabled() {
    return config && config.get('qualyscan:enabled');
}

function refreshCheckIP(callback) {
    callback = callback || function noop() {};
    if (coldCache && !qualysIPs) {
        // prevent other requests from loading cache, only one is allowed
        qualysIPs = {};
        // reload from the disk
        coldCache.get('qualysIPs', function onCacheEntry(err, data) {
            if (data) {
                qualysIPs = data;
            }
        }); 
    }

    moduleConfig(module, function handleConfig(err, _config) {
        if (err) {
            return callback(err);
        }

        config = _config;

        var refreshTimeout = _config.get('qualyscan:refreshTimeout') || DEFAULT_REFRESH_TIMEOUT;
        if (!lastRefreshIP || (Date.now() - lastRefreshIP) >  refreshTimeout) {
            // prevent other triggering refresh

            lastRefreshIP = Date.now();
            setImmediate(function doRefresh() {
                loadIPs(function (err, ips) {
                    if (err) {
                        // let's retry soon
                        return callback(err);
                    }
                    coldCache.put('qualysIPs', ips);
                    qualysIPs =  ips;
                    callback(null, ips);
                });
            });
            return;
        }
        callback();
    });
}

function refreshCheckHeader(callback) {
    callback = callback || function noop() {};
    if(coldCache && !qualysHeader) {
        // prevent other requests from loading cache, only one is allowed
        qualysHeader = {};
        // reload from the disk
        coldCache.get('qualysHeader', function onCacheEntry(err, data) {
            if (data) {
                qualysHeader = data;
            }
        });
    }

    moduleConfig(module, function handleConfig(err, _config) {
        if (err) {
            return callback(err);
        }

        config = _config;

        var refreshTimeout = _config.get('qualyscan:refreshTimeout') || DEFAULT_REFRESH_TIMEOUT;
        if (!lastRefreshHeader || (Date.now() - lastRefreshHeader) >  refreshTimeout) {
            // prevent other triggering refresh

            lastRefreshHeader = Date.now();
            setImmediate(function doRefresh() {
                loadHeader(function (err, headerMap) {
                    if(err) {
                        // let's retry soon
                        return callback(err);
                    }
                    coldCache.put('qualysHeader', headerMap);
                    qualysHeader =  headerMap;
                    callback(null, headerMap);
                });
            });
            return;
        }
        callback();
    });
}

function isQualyscanIP(ip) {
    // initiate update if any needed
    refreshCheckIP();
    // check ip
    if (isEnabled() && ip && qualysIPs) {
        return !!qualysIPs[ip];
    }
    return false;
}

function isQualysRequest(request) {
    // initiate update if any needed
    refreshCheckHeader();
    // look for the qualys header
    if(isEnabled() && request && qualysHeader) {
        for(const [key, value] of Object.entries(qualysHeader)) {
            if(request.headers && request.headers[key] && value == request.headers[key]) {// jshint ignore:line
                return true;
            }
        }
    }
    return false;
}

function deserialize(reader, callback) {
    var json = '';

    reader()
        .on('data', function(data) {
            json += data;
        })
        .on('end', function() {
            try {
                callback(null, JSON.parse(json));
            }
            catch (err) {
                callback(err);
            }
        })
        .on('error', function(err) {
            callback(err);
        });
}

function createDiskCache() {
    var nodeEnv = (process.env.NODE_ENV || 'development').toLowerCase();
    return RaptorCache.createDiskCache({
        dir: '.cache/hostip-checker-ebay-' + nodeEnv,
        serialize: JSON.stringify,
        deserialize: deserialize
    });
}

function reset() {
    qualysIPs = undefined;
    qualysHeader = undefined;
    coldCache = createDiskCache();
    lastRefreshIP = 0;
    lastRefreshHeader = 0;
    config = undefined;
}

exports.refreshCheckIP = refreshCheckIP;
exports.refreshCheckHeader = refreshCheckHeader;
exports.loadIPs = loadIPs;
exports.loadHeader = loadHeader;
exports.isQualyscanIP = isQualyscanIP;
exports.isQualysRequest = isQualysRequest;
exports.reset = reset;
Object.defineProperty(exports, 'config', {
    get: function getConfig() {
        return config;
    }
});
