// @ebay/openfeature-ebay/dist/provider/OpenFeatureProvider.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const js_sdk_1 = require("@openfeature/js-sdk");
const FFlagClient_1 = require("../FFlagClient");
function wrongTypeResult(value) {
    return {
        value,
        reason: js_sdk_1.StandardResolutionReasons.ERROR,
        errorCode: js_sdk_1.ErrorCode.TYPE_MISMATCH,
    };
}
function flagNotFoundResult(value) {
    return {
        value,
        reason: js_sdk_1.StandardResolutionReasons.DEFAULT,
        errorCode: js_sdk_1.ErrorCode.FLAG_NOT_FOUND,
    };
}
function flagDeferredResult(value) {
    return {
        value,
        reason: 'DEFERRED',
    };
}
class OpenFeatureProvider {
    constructor(client) {
        this.metadata = {
            name: 'openfeature-ebay Provider',
        };
        this.client = client;
    }
    async resolveBooleanEvaluation(flagKey, defaultValue, context) {
        var _a, _b;
        const { evaluationResults } = context;
        const result = await this.client.evaluateFlag(flagKey, context, evaluationResults);
        if (!result || ((_a = result === null || result === void 0 ? void 0 : result.details) === null || _a === void 0 ? void 0 : _a.errorCode) == js_sdk_1.ErrorCode.FLAG_NOT_FOUND) {
            return flagNotFoundResult(defaultValue);
        }
        // firstly check that it is deferred at ep aaff level (target audience). The expected behavior is to return default value
        if ((_b = result === null || result === void 0 ? void 0 : result.details) === null || _b === void 0 ? void 0 : _b.deferredForEpAaffLevel) {
            return flagDeferredResult(defaultValue);
        }
        result.featureFlagValue = (0, FFlagClient_1.convertToBoolean)(result.featureFlagValue);
        const translatedResult = this.translateResult(result, 'boolean');
        if (translatedResult != null) {
            return translatedResult;
        }
        return wrongTypeResult(defaultValue);
    }
    async resolveStringEvaluation(flagKey, defaultValue, context) {
        var _a, _b;
        const { evaluationResults } = context;
        const result = await this.client.evaluateFlag(flagKey, context, evaluationResults);
        if (!result || ((_a = result === null || result === void 0 ? void 0 : result.details) === null || _a === void 0 ? void 0 : _a.errorCode) == js_sdk_1.ErrorCode.FLAG_NOT_FOUND) {
            return flagNotFoundResult(defaultValue);
        }
        // firstly check that it is deferred at ep aaff level (target audience). The expected behavior is to return default value
        if ((_b = result === null || result === void 0 ? void 0 : result.details) === null || _b === void 0 ? void 0 : _b.deferredForEpAaffLevel) {
            return flagDeferredResult(defaultValue);
        }
        const translatedResult = this.translateResult(result, 'string');
        if (translatedResult != null) {
            return translatedResult;
        }
        return wrongTypeResult(defaultValue);
    }
    async resolveNumberEvaluation(flagKey, defaultValue, context) {
        var _a, _b;
        const { evaluationResults } = context;
        const result = await this.client.evaluateFlag(flagKey, context, evaluationResults);
        if (!result || ((_a = result === null || result === void 0 ? void 0 : result.details) === null || _a === void 0 ? void 0 : _a.errorCode) == js_sdk_1.ErrorCode.FLAG_NOT_FOUND) {
            return flagNotFoundResult(defaultValue);
        }
        // firstly check that it is deferred at ep aaff level (target audience). The expected behavior is to return default value
        if ((_b = result === null || result === void 0 ? void 0 : result.details) === null || _b === void 0 ? void 0 : _b.deferredForEpAaffLevel) {
            return flagDeferredResult(defaultValue);
        }
        result.featureFlagValue = Number(result.featureFlagValue);
        const translatedResult = this.translateResult(result, 'number');
        if (translatedResult !== undefined && !isNaN(translatedResult.value)) {
            return translatedResult;
        }
        return wrongTypeResult(defaultValue);
    }
    async resolveObjectEvaluation(flagKey, defaultValue, context) {
        var _a, _b;
        const { evaluationResults } = context;
        const result = await this.client.evaluateFlag(flagKey, context, evaluationResults);
        if (!result || ((_a = result === null || result === void 0 ? void 0 : result.details) === null || _a === void 0 ? void 0 : _a.errorCode) == js_sdk_1.ErrorCode.FLAG_NOT_FOUND) {
            return flagNotFoundResult(defaultValue);
        }
        // firstly check that it is deferred at ep aaff level (target audience). The expected behavior is to return default value
        if ((_b = result === null || result === void 0 ? void 0 : result.details) === null || _b === void 0 ? void 0 : _b.deferredForEpAaffLevel) {
            return flagDeferredResult(defaultValue);
        }
        try {
            result.featureFlagValue = this.tryParseJsonString(result.featureFlagValue);
            const translatedResult = this.translateResult(result, 'object');
            if (translatedResult != null) {
                return translatedResult;
            }
            else {
                return wrongTypeResult(defaultValue);
            }
        }
        catch (e) {
            return wrongTypeResult(defaultValue);
        }
    }
    translateResult(result, dataType) {
        let translatedResult;
        if (typeof result.featureFlagValue === dataType) {
            translatedResult = (0, FFlagClient_1.translateResult)(result);
            return translatedResult;
        }
    }
    tryParseJsonString(str) {
        let parsedString;
        try {
            parsedString = JSON.parse(str);
        }
        catch (e) {
            return str;
        }
        return parsedString;
    }
}
exports.default = OpenFeatureProvider;
