// @ebay/service-ebay/lib/configure.js
'use strict';

const Assert = require('assert');
const Fs = require('fs');
const Path = require('path');
const confit = require('confit');
const Async = require('async');
const Objutil = require('objutil');
const Handlers = require('shortstop-handlers');

module.exports = async function configure(options = {}) {
    options.protocols = options.protocols || {};

    options.onconfig = options.onconfig || function (configs) {
        return Promise.resolve(configs);
    };

    const protocols = Object.assign({}, {
        regexp: value => new RegExp(value)
    }, options.protocols);

    const configOptions = Object.assign({}, options, {
        protocols
    });

    let configs = await loadAppConfig(configOptions);
    configs = await loadDefaults(configs, configOptions);
    // allow app to configure apps
    return await options.onconfig(configs);
};

/**
 * Loads service app config into defaults.
 * @param config
 * @param done
 */
function loadSvcAppConfig(defaultSvcConfig, defaultConfig, options, next) {
    let baseDir = require.resolve(defaultSvcConfig.module);
    Assert.ok(baseDir, 'Cannot find application service module: ', defaultSvcConfig.module);
    baseDir = findModuleRoot(baseDir);
    Assert.ok(baseDir, 'Cannot find application service module: ', defaultSvcConfig.module);

    const tasks = [];
    const initPath = Path.resolve(baseDir, 'init.js');
    const initModule = Fs.existsSync(initPath) && require(initPath);
    if (typeof initModule === 'function') {
        tasks.push(initModule);
    }

    Async.waterfall(tasks, setupAppConfig);

    function setupAppConfig() {
        const protocols = Objutil.merge({
            path: Handlers.path(baseDir),
            require: Handlers.require(baseDir)
        }, options.protocols, {});

        const factory = confit({
            protocols: protocols,
            basedir: Path.join(baseDir || options.basedir, 'config')
        });

        factory.addOverride(defaultConfig);
        factory.addDefault(defaultSvcConfig);

        factory.create((error, config) => {
            if (error) {
                next(error);
                return;
            }

            config.set('app-name', config.get('app-name') || options.appName);
            next(null, config);
        });
    }
}

function findModuleRoot(fpath) {
    const appRoot = process.cwd();

    const cfgPath = Path.resolve(fpath, 'config');
    if (Fs.existsSync(cfgPath)) {
        return fpath;
    }
    const _fpath = Path.dirname(fpath);
    if (_fpath === fpath || _fpath === '/' || _fpath === appRoot) {
        return appRoot;
    }

    return findModuleRoot(_fpath);
}

/**
 * Loads hosting app config into defaults.
 * @param config
 * @param done
 */
function loadAppConfig(options) {
    const configRoot = Path.join(options.basedir, 'config');

    const factory = confit({
        protocols: Object.assign({}, {
            path: Handlers.path(configRoot),
            require: Handlers.require(configRoot)
        }, options.protocols),
        basedir: configRoot
    });

    return new Promise((resolve, reject) => {
        factory.create((error, config) => {
            if (error) {
                reject(error);
                return;
            }

            if (config.get('single-app-mode')) {
                return resolve([config]);
            }

            const applications = config.get('applications') || options.applications;
            if (!applications) {
                return reject(new Error('Cannot find service application descriptors in config.json'));
            }

            config = config._store;
            // delete apps list
            delete config.applications;

            Async.map(applications, (svcConfig, done) => {
                try {
                    loadSvcAppConfig(svcConfig, config, options, done);
                }
                catch (err) {
                    done(err);
                }
            }, (err, configs) => {
                if (err) {
                    return reject(err);
                }

                resolve(configs);
            });
        });
    });
}

/**
 * Loads defaults into config.
 * @param config
 * @param done
 */
function loadDefaults(configs, options) {
    const basedir = Path.join(__dirname, '../config');

    const protocols = Objutil.merge({
        path: Handlers.path(basedir),
        require: Handlers.require(basedir)
    }, options.protocols, {});

    return new Promise((resolve, reject) => {
        Async.map(configs, (config, cb) => {
            const factory = confit({
                protocols,
                basedir
            });

            factory.addOverride(config._store);

            factory.create((error, cfg) => {
                if (error) {
                    cb(error);
                    return;
                }

                cb(null, cfg);
            });
        }, (err, cfgs) => {
            if (err) {
                return reject(err);
            }
            resolve(cfgs);
        });
    });
}

module.exports.$internals = {
    loadAppConfig,
    loadDefaults,
    loadSvcAppConfig,
    findModuleRoot
};
