// @ebay/fidelius-client-ebay/lib/index.js
'use strict';

const provider = require('@ebay/service-client-ebay');
const trustFabric = require('@ebay/trust-fabric-ebay');
const assert = require('assert');
const ono = require('@jsdevtools/ono');
const logger = require('@ebay/logging-inc').logger('fidelius-client-ebay');

class FideliusClient {
    
    deleteSecret(path,  headers, callback) {
        const options = { method: 'DELETE' };
        this._makeRequest(path, headers, options, callback, 'fidelius-delete-objects');
    }

    setSecret(path, value, headers, callback) {
        const body = JSON.stringify({ value });
        const options = { method: 'PUT', body };
        this._makeRequest(path, headers, options, callback, 'fidelius-upsert-objects');
    }

    getSecret(path, headers, callback) {
        const options = { method: 'GET' };
        this._makeRequest(path, headers, options, (err, body) => {
            if (body) {
                callback(null, body.value, body);
            } else {
                callback(err, null);
            }
        }, 'fidelius-read-objects');
    }

    setAclNode(path, permissions, headers, callback) {
        const body = JSON.stringify({ permissions });
        const options = { method: 'PUT', body };
        this._makeRequest(path, headers, options, callback, 'fidelius-set-acls');
    }

    getAclNode(path, headers, callback) {
        const options = { method: 'GET' };
        this._makeRequest(path, headers, options, callback, 'fidelius-get-acls');
    }

    _makeRequest(path, headers, options, callback, serviceName) {
        assert.ok(path, 'Path is required');

        trustFabric.getToken((error, token) => {
            if (error) {
                return callback(new Error('Unable to generate token for calling Fidelius ' + error));
            }

            const client = provider.context({}).getClient(serviceName);
            options.headers = headers;
            client.request(options)
                .path(path)
                .set('Authorization', 'Bearer '+token)
                .end((err, res) => {

                    if (err) {
                        return callback(ono(err, 'Error while making secret request to Fidelius'), null);
                    }

                    if (res.statusCode >= 400) {
                        const message = res.statusCode === 404 ?
                            `Error while making secret request to Fidelius - Key ${path} doesn't exist` :
                            `Error while making secret request to Fidelius - Key ${path}, the app does not have permission to access it`;
                        const customError = new Error(message);
                        customError.code = res.statusCode;
                        const errorType = res.statusCode === 404 ? 'warn' : 'error';
                        logger[errorType](customError);
                        return callback(customError, null);
                    }
                    callback(null, res.body);
                });
        });
    }

    isAppWhiteListed(appName, callback) {
        assert.ok(appName, 'App Name is required');

        var client = provider.context({}).getClient('fidelius-whitelist');
    
        client.get()
        .path(`${appName}/config`)
        .end((err, res)=> {
            if(err || !res || !res.body) {
                return callback(new Error('Error while reading whitelist and fallbackEnabled from Fidelius config' + err), false);
            }
            
            callback(null, res.body.fideliusEnabled, res.body.fideliusFallbackEnabled);
        });
    }
}

module.exports = new FideliusClient();
module.exports.logger = logger;
