// @ebay/openfeature-ebay/src/openfeature-lib/provider/OpenFeatureProvider.ts
import {
    EvaluationContext,
    ErrorCode,
    JsonValue,
    Provider,
    ResolutionDetails,
    ProviderMetadata,
    StandardResolutionReasons,
} from '@openfeature/js-sdk';
import { FFlagClient, translateResult, convertToBoolean } from '../FFlagClient';
import { FlagEvalDetail } from '../types/FlagEvalDetail';

function wrongTypeResult<T>(value: T): ResolutionDetails<T> {
    return {
        value,
        reason: StandardResolutionReasons.ERROR,
        errorCode: ErrorCode.TYPE_MISMATCH,
    };
}

function flagNotFoundResult<T>(value: T): ResolutionDetails<T> {
    return {
        value,
        reason: StandardResolutionReasons.DEFAULT,
        errorCode: ErrorCode.FLAG_NOT_FOUND,
    };
}

function flagDeferredResult<T>(value: T): ResolutionDetails<T> {
    return {
        value,
        reason: 'DEFERRED',
    };
}

export default class OpenFeatureProvider implements Provider {
    client: FFlagClient;
    constructor(client: FFlagClient) {
        this.client = client;
    }
    readonly metadata = {
        name: 'openfeature-ebay Provider',
    } as ProviderMetadata;

    async resolveBooleanEvaluation(
        flagKey: string,
        defaultValue: boolean,
        context: EvaluationContext,
    ): Promise<ResolutionDetails<boolean>> {
        const { evaluationResults } = context;
        const result = await this.client.evaluateFlag(
            flagKey,
            context,
            evaluationResults as unknown as Map<string, FlagEvalDetail>,
        );

        if (!result || result?.details?.errorCode == ErrorCode.FLAG_NOT_FOUND) {
            return flagNotFoundResult(defaultValue);
        }

        // firstly check that it is deferred at ep aaff level (target audience). The expected behavior is to return default value
        if (result?.details?.deferredForEpAaffLevel) {
            return flagDeferredResult(defaultValue);
        }

        result.featureFlagValue = convertToBoolean(result.featureFlagValue);
        const translatedResult = this.translateResult(result, 'boolean');
        if (translatedResult != null) {
            return translatedResult;
        }
        return wrongTypeResult(defaultValue);
    }

    async resolveStringEvaluation(
        flagKey: string,
        defaultValue: string,
        context: EvaluationContext,
    ): Promise<ResolutionDetails<string>> {
        const { evaluationResults } = context;
        const result = await this.client.evaluateFlag(
            flagKey,
            context,
            evaluationResults as any,
        );

        if (!result || result?.details?.errorCode == ErrorCode.FLAG_NOT_FOUND) {
            return flagNotFoundResult(defaultValue);
        }
        // firstly check that it is deferred at ep aaff level (target audience). The expected behavior is to return default value
        if (result?.details?.deferredForEpAaffLevel) {
            return flagDeferredResult(defaultValue);
        }

        const translatedResult = this.translateResult(result, 'string');
        if (translatedResult != null) {
            return translatedResult;
        }
        return wrongTypeResult(defaultValue);
    }

    async resolveNumberEvaluation(
        flagKey: string,
        defaultValue: number,
        context: EvaluationContext,
    ): Promise<ResolutionDetails<number>> {
        const { evaluationResults } = context;
        const result = await this.client.evaluateFlag(
            flagKey,
            context,
            evaluationResults as any,
        );

        if (!result || result?.details?.errorCode == ErrorCode.FLAG_NOT_FOUND) {
            return flagNotFoundResult(defaultValue);
        }

        // firstly check that it is deferred at ep aaff level (target audience). The expected behavior is to return default value
        if (result?.details?.deferredForEpAaffLevel) {
            return flagDeferredResult(defaultValue);
        }

        result.featureFlagValue = Number(result.featureFlagValue);
        const translatedResult = this.translateResult(result, 'number');
        if (translatedResult !== undefined && !isNaN(translatedResult.value)) {
            return translatedResult;
        }
        return wrongTypeResult(defaultValue);
    }

    async resolveObjectEvaluation<T extends JsonValue>(
        flagKey: string,
        defaultValue: T,
        context: EvaluationContext,
    ): Promise<ResolutionDetails<T>> {
        const { evaluationResults } = context;
        const result = await this.client.evaluateFlag(
            flagKey,
            context,
            evaluationResults as any,
        );

        if (!result || result?.details?.errorCode == ErrorCode.FLAG_NOT_FOUND) {
            return flagNotFoundResult(defaultValue);
        }

        // firstly check that it is deferred at ep aaff level (target audience). The expected behavior is to return default value
        if (result?.details?.deferredForEpAaffLevel) {
            return flagDeferredResult(defaultValue);
        }

        try {
            result.featureFlagValue = this.tryParseJsonString(
                result.featureFlagValue,
            );
            const translatedResult = this.translateResult(result, 'object');
            if (translatedResult != null) {
                return translatedResult;
            } else {
                return wrongTypeResult(defaultValue);
            }
        } catch (e) {
            return wrongTypeResult(defaultValue);
        }
    }

    private translateResult(result: FlagEvalDetail, dataType: string) {
        let translatedResult;
        if (typeof result.featureFlagValue === dataType) {
            translatedResult = translateResult(result);
            return translatedResult;
        }
    }

    private tryParseJsonString(str) {
        let parsedString: any;
        try {
            parsedString = JSON.parse(str);
        } catch (e) {
            return str;
        }
        return parsedString;
    }
}
