// @ebay/kernel-services-ebay/lib/index.js
/* jslint node: true */
'use strict';

var querystring = require('querystring'),
    Cache = require('lru-backup-cache'),
    Url = require('url'),
    RequestLocal = require('request-local'),
    kernelService = require('./services/index');

const RATE_MAX_AGE = 4 * 60 * 60 * 1000; // 4 hours

var serviceCache = new Cache({ max: 1000, maxAge: 120 * 60 * 1000 }),
    CurrencyListCache;

const logger = require('@ebay/logging-inc').logger('kernel-services-ebay/index');

function getSiteInfo (url, siteoverride, callback) {
    var args = [].slice.call(arguments);
    url = args.shift();
    callback = args.pop();
    siteoverride = args.shift();

    var req = tryRequest();
    if (siteoverride === undefined) {
        siteoverride = req && req.query ? req.query.siteoverride : undefined;
    }
    if (!req || !req.ebay || !req.ebay.isInternalHost || siteoverride !== undefined && !req.ebay.isInternalHost()) {
        siteoverride = undefined;
    }

    var hostname = url && Url.parse(url).host || url;
    var cacheKey = 'SI_' + hostname;
    var info = serviceCache.get(cacheKey);

    if (info && siteoverride === undefined) {
        return callback(null, info);
    }
    kernelService.getSiteInfo(url, siteoverride, function(err, siteInfo) {
        if (err || !siteInfo) {
            siteInfo = serviceCache.getBackup(cacheKey);
            return callback(err, siteInfo);
        }
        if (siteoverride === undefined) {
            serviceCache.set(cacheKey, siteInfo);
        }
        return callback(null, siteInfo);
    });
}

function getCurrencyList (callback) {
    if (CurrencyListCache) {
        return callback(null, CurrencyListCache);
    }
    kernelService.getCurrencyList(function(err, list) {
        if (err || !list) {
            return callback(err, list);
        }
        CurrencyListCache = list;
        return callback(null, CurrencyListCache);
    });

}

function getSignInLink (options, callback) {
    if (!options || !options.returnUrl) {
        return callback(new Error('Please provide returnUrl for options argument!'), null);
    }
    var key = querystring.stringify(options);
    var cacheKey = 'Link_' + key;
    var cachedLink = serviceCache.get(cacheKey);
    if (cachedLink) {
        return callback(null, cachedLink);
    }
    kernelService.getSignInLink(options, function(err, link) {
        if (err || !link) {
            link = serviceCache.getBackup(cacheKey);
            return callback(err, null);
        }
        serviceCache.set(cacheKey, link);
        return callback(null, link);
    });
}

function getExchangeRate(from, to, when) {
    const noFallback = !!when;
    when = when || new Date();

    // cache key will be the same during RATE_MAX_AGE period and increment afterwards
    const cacheKey = `Rate_${from}>${to}@${parseInt(when/internals.getRateMaxAge())}`;
    const cached = internals.rateCache.get(cacheKey);
    if (cached) {
        return Promise.resolve(cached);
    }

    return kernelService.getExchangeRate(from, to, when)
        .then(rate => {
            internals.rateCache.set(cacheKey, rate);
            return rate;
        })
        .catch(err => {
            if (noFallback) {
                throw err;
            }
            const backupCacheKey = `Rate_${from}>${to}@${parseInt((when - internals.getRateMaxAge())/internals.getRateMaxAge())}`;
            const rate = internals.rateCache.get(backupCacheKey);
            if (!rate) {
                throw err;
            }
            logger.error(`Failed to get exchange rate for ${from}>${to}, using backup version`);
            return rate;
        });
}

function tryRequest() {
    try {
        return RequestLocal.data.request;
    }
    catch (err) {}
}

const internals = {
    rateCache: new Cache({ max: 300, maxAge: 2 * RATE_MAX_AGE }),
    reset: () => {
        this.rateMaxAge = undefined;
        internals.rateCache = new Cache({ max: 300, maxAge: 2 * internals.getRateMaxAge() });
    },
    setRateMaxAge(value) {
        this.rateMaxAge = value;
    },
    getRateMaxAge() {
        return this.rateMaxAge || RATE_MAX_AGE;
    }
};

module.exports = {
    getSiteInfo: getSiteInfo,
    getCurrencyList: getCurrencyList,
    getSignInLink: getSignInLink,
    getExchangeRate: getExchangeRate,
    $internals: internals
};
