// @ebay/kernel-services-ebay/lib/utils.js
'use strict';

const Url = require('url');

/*
    This is a initial version of eBay URL Parser/Formater

    Possible ebay host variations:
    www.ebay.com
    www.ebay.ca
    www.cafr.ebay.ca

    www.qa.ebay.com
    www.ca.paradise.ebay.com
    www.cafr.paradise.qa.ebay.com
*/

module.exports = {
    formatUrl(options) {
        const hostOptions = options.hostOptions;
        if (hostOptions) {
            const tokens = [];
            tokens.unshift(hostOptions.baseHost);
            // qa env
            if (hostOptions.type === 'qa') {
                tokens.unshift('qa');
                // qa intl environment
                if (hostOptions.intl && hostOptions.fp) {
                    tokens.unshift(hostOptions.fp);
                }
            }
            if(hostOptions.tail) {
                tokens.unshift(hostOptions.tail);
            }
            if (hostOptions.intl) {
                if (hostOptions.mini) {
                    tokens.unshift(hostOptions.mini);
                }
                else if (hostOptions.type === 'qa' && hostOptions.country !== 'us') {
                    tokens.unshift(hostOptions.country);
                }
            }
            if(hostOptions.app) {
                tokens.unshift(hostOptions.app);
            }
            options.hostname = tokens.join('.');
            options.host = options.port ? options.hostname + ':' + options.port : options.hostname;
        }

        return options.protocol ? Url.format(options) : options.hostname;
    },

    parseUrl(urlOrHost) {
        let urlParts = {};
        let host = urlOrHost;
        if (/^http/.test(urlOrHost)) {
            urlParts = Url.parse(urlOrHost);
            host = urlParts.hostname;
        }
        host = host || 'www.ebay.com';

        const tokens = host.split('.');
        const baseHost = [];
        let qa;
        let intlDomain;
        let token;
        let app;
        let mini;
        let fp;
        let country;
        const countryDomain = [];

        // get base host
        while(true) {
            token = tokens.pop();
            baseHost.unshift(token);
            if (!token) {
                return urlParts;
            }
            if (token === 'ebay') {
                break;
            }
            countryDomain.unshift(token);
        }

        // get qa
        token = tokens.pop();
        if (token === 'qa') {
            qa = token;
        }
        else {
            tokens.push(token);
        }

        // get intl domain (paradise)
        if (qa) {
            token = tokens.pop();
            if (token === 'paradise') {
                intlDomain = token;
                fp = token;
            }
            else {
                tokens.push(token);
            }
        }

        // get app domain
        token = tokens.shift();
        app = token;
        token = tokens.shift();
        // get intl country or mini
        if (token) {
            if (token.length === 4) {
                mini = token;
                country = mini.substring(0, 2);
            }
            else if (token.length === 2) {
                country = token;
            }
        }

        if (countryDomain.length === 2 || countryDomain[0] !== 'com') {
            country = countryDomain.pop();
        }

        urlParts.hostOptions = {
            baseHost: baseHost.join('.'),
            type: qa || 'production',
            intl: token === 'paradise' || country !== undefined,
            app,
            country: country || 'us',
            mini: mini
        };

        if (tokens.length) {
            urlParts.hostOptions.tail = tokens.join('.');
        }

        if (fp) {
            urlParts.hostOptions.fp = fp;
        }

        return urlParts;
    }
};
