// @ebay/service-client-ebay/lib/browser/index.js
/*jslint browser: true*/
'use strict';

var axios = require("axios");

var Trooba = require('trooba');
var httpfy = require('@ebay/service-instruments-ebay/transports/http-api');
var Legacy = require('@ebay/service-instruments-ebay/lib/legacy');
var Utils = require('./utils');

// TODO: delete deprecated use case
var xhr = require('trooba-xhr-transport');
var sifr = require('@ebay/service-instruments-ebay/transports/sifr-transport');
var XhrUtils = xhr.Utils;
XhrUtils.mixinMvHeaders = require('./utils').mixinMvHeaders;

function transport(pipe, config) {
    pipe.on('request', function onRequest(request) {

        if (!request.bridge) {
            request = XhrUtils.mixin(config,
                request, {});

            Legacy.updateRequest(request);

            var tr = xhr;

            if (request.sifr || !(request.xhr || isCors())) {
                tr = sifr;
            }

            tr.invoke(request, function onResponse(err, response) {
                if (err) {
                    return pipe.throw(err);
                }
                if (response) {
                    XhrUtils.deserializeResponseHeaders(response);
                }
                pipe.respond(response);
            });
            return;
        }

        Legacy.updateRequest(request);

        if (request.withCredentials === undefined) {
            request.withCredentials = true; // by default we require cookies
        }

        request.headers = request.headers || {};
        request.headers['x-ebay-client-id'] = pipe.context.clientId;
        if (request.bridgeMode) {
            request.headers['x-ebay-bridge-mode'] = request.bridgeMode;
        }
        if (request.path) {
            request.headers['x-ebay-client-target-path'] = request.path;
        }
        if (request.headers['cache-control'] === undefined && request['cache-control'] !== false) {
            request.headers['cache-control'] = 'no-cache'; // due to bridge being reused by multiple clients
            // this can be dropped if each client would use its own bridge
        }
        request.url = request.bridgePath || 'ajax-service-bridge';

        if (request.body) {
            request.data = request.body;
        }

        if (request.search) {
            if (typeof request.search === 'string') {
                request.params = {};
                const kvps = request.search.split(/&/g);
                for (const kvp of kvps) {
                    var pair = kvp.split('=');
                    request.params[decodeURIComponent(pair[0])] = decodeURIComponent(pair[1]);
                }    
            }
            else {
                request.params = request.search;
            }
            delete request.search;
        }

        axios.request(request)
            .then(function (res) {
                pipe.respond({
                    statusCode: res.status,
                    body: res.data,
                    headers: res.headers
                });
            })
            .catch(function(err) {
                pipe.throw(err);
            });
    });

    httpfy(pipe);
}
module.exports = transport;

var pipe = Trooba
    .use(config) // <<< attach configuration handler that would load the rest
    .use(transport)
    .build();

/**
 * Setup context for a request, if not provided a default will be used.
 * @param ctx is a context similar to http request or mock request
*/
module.exports = {
    context: function context(ctx) {
        ctx = ctx || {};

        return {
            getClient: function getClient(id) {
                ctx.clientId = id;

                return pipe.create(ctx, 'client:default');
            }
        };
    },

    getClient: function getClient(id) {
        console.error('getClient method is deprecated, ' +
            'please switch to require(\'service-client-ebay\').context(req).getClient(\'' +
            id + '\')');

        return this.context().getClient(id);
    }
};

function config(pipe) {
    var clientId = pipe.context.clientId;

    pipe.on('request', function onRequest(request, next) {
        var configs = window.$serviceConfigs;
        var config = configs[clientId];
        if (!config) {
            return pipe.throw(new Error('Cannot find client configuration, id: ' +
                clientId +
                ', please make sure you enabled it using service-use tag'));
        }

        // merge multivalue headers
        request.headers = Utils.mixinMvHeaders(request.headers,
            config.headers, request.mvheaders);

        request = Utils.mixin(config, request, {});
        next(request);
    });
}

function isCors() {
    return 'withCredentials' in new window.XMLHttpRequest();
}
