// @ebay/client-hints-ebay/lib/timer.js
'use strict';

const provider = require('@ebay/service-client-ebay');
const configbean = require('@ebay/config-bean-ebay');
const { promisify } = require('util');
const modConfig = require('@ebay/module-config-inc');
const loadConfig = promisify(modConfig);
const logger = require('@ebay/logging-inc').logger('client-hints-ebay:timer');
const appCtx = require('@ebay/app-context-ebay');

const DEFAULT_INTERVAL = 60 * 60 * 1000; // 30 min timeout

class Timer {
    constructor() {
        this.isRunning = false;
        this.startTimer();
        this.timer = undefined;
    }

    startTimer(refreshInterval) {
        let ddsConfig;
        // eslint-disable-next-line
        const fetchAndUpdate = async () => {
            // clean up context to avoid memory leak            
            while (process.domain) {
                process.domain.exit();
            }
            try {
                ddsConfig = await loadConfig(module).then(cfg => cfg.get('client-hints-ebay'));
                const data = await this.getCHConfig();
                this.persistToBeans(data);
            }
            catch (err) {
                logger.error('Failed to fetch config from DDS', err);
            }
            finally {
                const interval = refreshInterval || (ddsConfig && ddsConfig.configRefreshIntervel) || DEFAULT_INTERVAL;
                this.timer = setTimeout(fetchAndUpdate, interval).unref();
            }
        };

        if (!this.isRunning) {
            this.isRunning = true;
            setImmediate(fetchAndUpdate);
        }
    }

    clearTimer() {
        this.timer && clearTimeout(this.timer);
        this.timer = undefined;
    }

    getCHConfig() {
        return new Promise((resolve, reject) => {
            const client = provider.context({}).getClient('client-hints-config');
            const params = {app: appCtx.appName};
            client.get(params).path('/config/feature').end((err, res) => {
                if (err) {
                    return reject(err);
                }
                return resolve(res.body);
            });
        });
    }

    persistToBeans(config) {
        const bean = configbean.getBeanById('nodejs.config.client-hints-ebay.common');
        const clientHint = config.clientHint;
        
        if (bean && clientHint) {
            if (typeof clientHint.enable === 'string') {
                clientHint.enable = (clientHint.enable.toLowerCase() === 'true');
            }
            const val = clientHint.enable || false;
            bean.setPersist('client-hints-ebay:clientHint:enable', val);
            bean.setPersist('client-hints-ebay:clientHint:requestHeaders', clientHint.requestHeaders);
            bean.setPersist('client-hints-ebay:clientHint:chainHeaders', clientHint.chainHeaders);
            bean.setPersist('client-hints-ebay:clientHint:acceptCHs', clientHint.acceptCHs);
            bean.setPersist('client-hints-ebay:configRefreshIntervel', clientHint.interval);
        }
    }
}

module.exports = new Timer();
