// @ebay/httpunch/lib/metrics.js
'use strict';

var Hoek = require('hoek');

module.exports.hook = function hook(req) {
    var onFinish = Hoek.once(_onFinish);

    req.metrics = createMetrics();
    req.once('socket', onSocket);
    req.once('upgrade', onUpgrade);
    req.once('response', onReponse);
    req.once('information', onInformation);
    req.once('continue', onContinue);
    req.once('error', onFinish);
    req.once('finish', onUpload);
    var _res;
    var _socket;

    function _onFinish() {
        req.metrics.add('end');

        // cleanup
        req.removeListener('socket', onSocket);
        req.removeListener('upgrade', onUpgrade);
        req.removeListener('response', onReponse);
        req.removeListener('information', onInformation);
        req.removeListener('continue', onContinue);
        req.removeListener('error', onFinish);
        req.removeListener('finish', onUpload);

        if (_res) {
            _res.removeListener('readable', onReadable);
            _res.removeListener('socket', onFinish);
            _res.removeListener('error', onFinish);
        }

        if (_socket) {
            _socket.removeListener('lookup', onLookup);
            _socket.removeListener('connect', onConnect);
            _socket.removeListener('secureConnect', onSecureConnect);
        }
    }

    function onReadable() {
        req.metrics.add('firstByteAt');
    }

    function onConnect() {
        req.metrics.add('connect');
    }

    function onReponse(res) {
        _res = res;
        req.metrics.add('response');
        res.once('readable', onReadable);
        res.once('end', onFinish);
        res.once('error', onFinish);
    }

    function onSocket(socket) {
        _socket = socket;
        req.metrics.add('socket');
        socket.setNoDelay(true);
        socket.once('lookup', onLookup);
        socket.once('connect', onConnect);
        socket.once('secureConnect', onSecureConnect);
    }

    function onLookup() {
        req.metrics.add('lookup');
    }

    function onSecureConnect() {
        req.metrics.add('secureConnect');
    }

    function onUpgrade() {
        req.metrics.add('upgrade');
    }

    function onInformation() {
        req.metrics.add('information');
    }

    function onContinue() {
        req.metrics.add('continue');
    }

    function onUpload() {
        req.metrics.add('upload');
    }
};


function RequestMetrics() {
    this.events = [];
    this.add('start');
}

var proto = RequestMetrics.prototype;

proto.add = function add(name) {
    var last = this.events[this.events.length - 1];
    var start = last ? last.end : Date.now();
    var end = Date.now();
    var duration = end - start;
    this.events.push({
        name: name,
        start: start,
        end: end,
        duration: duration
    });
};

proto.toMap = function toMap() {
    return this.events.reduce(function reduce(memo, evt) {
        memo[evt.name] = evt;
        return memo;
    }, {});
};

var createMetrics = module.exports.createMetrics = function createMetrics() {
    return new RequestMetrics();
};
