// @ebay/locale-ebay/lib/CtxGeoConfig.js
'use strict';

var Objutil = require('objutil');
var Util = require('util');
var GeoConfig = require('./GeoConfig');
var utils = require('./utils');

function CtxGeoConfig(req, geoConfig) {
    // copy object properties instead of initing
    Object.keys(geoConfig).forEach(function forEach(key) {
        this[key] = geoConfig[key];
    }.bind(this));
    this._locality = req.locality;
    this._pageName = req.ebay.getPageName();
    this._userAgent = req.ebay.getUserAgent();
}

Util.inherits(CtxGeoConfig, GeoConfig);

CtxGeoConfig.prototype.isGeoEnabled = function isGeoEnabled() {
    return CtxGeoConfig.super_.prototype.isGeoEnabled.call(this, this._locality.isGeoCountry);
};

CtxGeoConfig.prototype.isLanguageSelectionEnabled = function isLanguageSelectionEnabled() {
    return CtxGeoConfig.super_.prototype.isLanguageSelectionEnabled.call(this,
        this._pageName, this._locality.country);
};

CtxGeoConfig.prototype.isGeoEnabledForCountry = function isGeoEnabledForCountry() {
    return CtxGeoConfig.super_.prototype.isGeoEnabledForCountry.call(this, this._locality.country,
        this._locality.siteId);
};

CtxGeoConfig.prototype.isCommandAndCountryEnabledForLanguageSupport =
        function isCommandAndCountryEnabledForLanguageSupport(pageName) {

    return CtxGeoConfig.super_.prototype.isCommandAndCountryEnabledForLanguageSupport.call(this, pageName,
        this._locality.country);
};

CtxGeoConfig.prototype.isSiteExcluded = function isSiteExcluded(client) {
    return CtxGeoConfig.super_.prototype.isSiteExcluded.call(this, client, this._locality.siteId);
};

CtxGeoConfig.prototype.getGroupingSiteIdForClient = function getGroupingSiteIdForClient(client) {
    return CtxGeoConfig.super_.prototype.getGroupingSiteIdForClient.call(this, client, this._locality.siteId);
},

CtxGeoConfig.prototype.getGroupedUserLocationSite = function getGroupedUserLocationSite(client) {
    return CtxGeoConfig.super_.prototype.getGroupedUserLocationSite.call(this, client, this._locality.siteId);
};

CtxGeoConfig.prototype.isMPMachineTranslationEnabled = function isMPMachineTranslationEnabled() {
    return CtxGeoConfig.super_.prototype.isMPMachineTranslationEnabled.call(this, this._locality.country,
        this._locality.locale);
};

CtxGeoConfig.prototype.getUserAgentMapping = function getUserAgentMapping() {
    return CtxGeoConfig.super_.prototype.getUserAgentMapping.call(this, this._userAgent);
};

CtxGeoConfig.prototype.isSiteEnabled = function isSiteEnabled() {
    return CtxGeoConfig.super_.prototype.isSiteEnabled.call(this, this._locality.siteId);
};

CtxGeoConfig.prototype.isCountryEnabled = function isCountryEnabled() {
    return CtxGeoConfig.super_.prototype.isCountryEnabled.call(this, this._locality.country);
};

CtxGeoConfig.prototype.isBlacklistedCommand = function isBlacklistedCommand() {
    return CtxGeoConfig.super_.prototype.isBlacklistedCommand.call(this, this._pageName);
};

CtxGeoConfig.prototype.hasDefaultLanguageForCountry = function hasDefaultLanguageForCountry() {
    return CtxGeoConfig.super_.prototype.hasDefaultLanguageForCountry.call(this, this._locality.country);
};

CtxGeoConfig.prototype.isCountrySupported = function isCountrySupported() {
    return CtxGeoConfig.super_.prototype.isCountrySupported.call(this, this._locality.country, this._locality.locale);
};

CtxGeoConfig.prototype.isCountryEnabledForLanguage = function isCountryEnabledForLanguage() {
    return CtxGeoConfig.super_.prototype.isCountryEnabledForLanguage.call(this, this._locality.country);
};

CtxGeoConfig.prototype.isCommandAndCountryEnabledForLanguageSupport = function isCommandAndCountryEnabledForLanguageSupport() {
    return CtxGeoConfig.super_.prototype.isCommandAndCountryEnabledForLanguageSupport.call(this,
        this._pageName, this._locality.country);
};

module.exports = CtxGeoConfig;
