// @ebay/locale-ebay/lib/GeoConfig.js
'use strict';

var utils = require('./utils');

var GBH_SITE_ID = 900;

function createGroupedSiteIdMap(client, mapping, groupedSiteIdByClientMap) {
    groupedSiteIdByClientMap[client] = mapping;
    return groupedSiteIdByClientMap;
}


function GeoConfig(moduleConfig, localeConfig) {
    if (!moduleConfig) {
        throw new Error('moduleConfig is required');
    }
    this._moduleConfig = moduleConfig;
    this._localeConfig = localeConfig;

    this._enabledSiteIds = utils.arrayToSet(moduleConfig.EnabledSiteIdList);
    this._enabledCountries = utils.arrayToSet(moduleConfig.GeoCountries);
    this._geoExpansionFeatureEnabled = !!moduleConfig.Enable_Geo_Expansion;
    this._blacklistedCommands = utils.normalizeSet(moduleConfig.BlacklistedCommands);
    this._languageEnabledCommands = utils.normalizeSet(moduleConfig.LanguageEnabledCommands);
    this._gbhExclusionSiteIds = moduleConfig.GBH_EXCLUSION_SITE_IDS || {};
    this._gbhSubdomainExclusionSiteIds = moduleConfig.GBH_SUBDOMAIN_EXCLUSION_SITE_IDS || {};
    this._supportedCountries = moduleConfig.GEO_COUNTRIES_ENABLED || {};
    this._exclusionSiteIdsByClient = {};
    this._supportedLocalesByCountry = {};
    this._userAgentMappings = utils.parseUserAgentMappings(moduleConfig.UserAgentMapping);
    this._languageEnabledCommandsV2 = moduleConfig['LanguageEnabledCommands.v2'];
    this._groupedSiteIdsByClientMap = createGroupedSiteIdMap('GBH_HEADER', moduleConfig['GBH_HEADER.GROUP'], {});
    this._languageEnabledCountries = moduleConfig.LanguageEnabledGeoCountries;

    //service related
    this.geoLocationServiceBaseURI = moduleConfig.MarketPlaceGeoLocationServiceBaseURI;
    this.geoLocationServiceTimeout = moduleConfig.MarketPlaceGeoLocationServiceTimeout || 3000;
    this.geoLocationServiceRetry = moduleConfig.MarketPlaceGeoLocationServiceRetry || 1;
    this.geoLocationServiceMarkdownThreshold = moduleConfig.MarketPlaceGeoLocationServiceMarkdownThreashold || 500;
    this.geoLocationServiceMaxSockets = moduleConfig.MarketPlaceGeoLocationServiceMaxsockets || 30;
}

GeoConfig.prototype = {
    isGeoEnabled: function (isGeoCountry) {
        return isGeoCountry &&
            this.isGeoExpansionFeatureEnabled();
    },
    isGeoEnabledForCountry: function (alpha2IsoCountryCode, siteId) {
        return this.isSiteEnabled(siteId) &&
            this.isCountryEnabled(alpha2IsoCountryCode) &&
            this.isGeoExpansionFeatureEnabled();
    },
    getGBHSiteId: function () {
        return GBH_SITE_ID;
    },
    isLanguageSelectionEnabled: function (pageName, country) {
        return this.isGeoEnabled() &&
            this.isCountryEnabledForLanguage(country) &&
            this.isCommandAndCountryEnabledForLanguageSupport(pageName, country);
    },
    getGroupedUserLocationSite: function (client, siteId) {
        if (this.isSiteExcluded(client, siteId)) {
            return siteId;
        }
        return this.getGroupingSiteIdForClient(client, siteId);
    },
    // This method is deprecated
    isMPMachineTranslationEnabled: function (country, locale) {
        return this.isCountrySupported(country, locale);
    },
    header: function (key, value) {
        this.serviceHeaders[key] = value;
    },

    getUserAgentMapping: function (userAgent) {
        if (!userAgent) {
            return null;
        }
        var mappings = this._userAgentMappings;
        if (mappings) {
            for (var i = 0, len = mappings.length; i < len; i++) {
                var mapping = mappings[i];
                if (mapping.pattern.test(userAgent)) {
                    return mapping.target;
                }
            }
        }
        return null;
    },
    isSiteEnabled: function (siteId) {
        return this._enabledSiteIds && (this._enabledSiteIds['ALL'] === true || this._enabledSiteIds[siteId] === true);
    },
    isCountryEnabled: function (alpha2IsoCountryCode) {
        return this._enabledCountries[alpha2IsoCountryCode] === true;
    },
    isGeoExpansionFeatureEnabled: function () {
        return this._geoExpansionFeatureEnabled;
    },
    isBlacklistedCommand: function (pageName) {
        return this._blacklistedCommands[pageName] === true;
    },
    hasDefaultLanguageForCountry: function (alpha2IsoCountryCode) {
        var languages = this._moduleConfig[alpha2IsoCountryCode + '.language'];//list
        return languages && languages.length ? true : false;
    },
    isCommandEnabledForLanguageSupport: function (pageName) {
        return this._languageEnabledCommands[pageName + '.Language'] === true;
    },
    isSiteExcluded: function (client, siteId) {
        var exclusionSet = this._exclusionSiteIdsByClient[client];
        if (exclusionSet === undefined) {
            var exclusionList = null;
            if(this._localeConfig.subdomainEnabled) {
                exclusionList = this._gbhSubdomainExclusionSiteIds[client];
            } else {
                exclusionList = this._gbhExclusionSiteIds[client];
            }
            if (!exclusionList) {
                return false;
            }

            exclusionList = exclusionList.split('|');
            exclusionSet = this._exclusionSiteIdsByClient[client] = utils.arrayToSet(exclusionList);
        }

        return exclusionSet[siteId] === true;
    },
    isCountrySupported: function (country, localeString) {
        var localeSet = this._supportedLocalesByCountry[country];
        if (localeSet === undefined) {
            var countryLocaleList = this._supportedCountries[country];
            if (!countryLocaleList) {
                return false;
            }
            countryLocaleList = countryLocaleList.split(',');
            localeSet = this._supportedLocalesByCountry[country] = utils.arrayToSet(countryLocaleList);
        }
        return localeSet[localeString] === true;
    },
    isCountryEnabledForLanguage: function (country) {
        this._enabledCountriesSet = this._enabledCountriesSet ||
            this._languageEnabledCountries ? utils.arrayToSet(this._languageEnabledCountries.split(',')) : undefined;
        return this._enabledCountriesSet && this._enabledCountriesSet[country] === true;
    },
    getGroupingSiteIdForClient: function (client, siteId) {
        if (this._groupedSiteIdsByClientMap.length === 0 || !(this._groupedSiteIdsByClientMap.hasOwnProperty(client))) {
            this._groupedSiteIdsByClientMap[client] = this._moduleConfig[client + '.GROUP'];
        }

        var groupedSiteIdsMapping = this._groupedSiteIdsByClientMap[client];
        if (!groupedSiteIdsMapping) {
            return GBH_SITE_ID;
        }

        for (var groupingSiteId in groupedSiteIdsMapping) {
            var siteIds = groupedSiteIdsMapping[groupingSiteId];
            if (!siteIds) {
                continue;
            }
            var siteIdSet = utils.arrayToSet(siteIds.split(','));
            if (siteIdSet[siteId] === true) {
                return parseInt(groupingSiteId, 10);
            }
        }

        return GBH_SITE_ID;
    },
    isCommandAndCountryEnabledForLanguageSupport: function (pageName, country) {
        var countryList = this._languageEnabledCommandsV2[pageName + '.Language'];
        if (!countryList) {
            return false;
        }

        countryList = countryList.split(',');
        var countrySet = utils.arrayToSet(countryList);
        return (countrySet[country] === true);
    }
};

module.exports = GeoConfig;
