// @ebay/site-meta-resources-ebay/dist/index.js
'use strict';
Object.defineProperty(exports, "__esModule", { value: true });
exports.EbaySiteIdEnum = exports.EbaySiteNameEnum = void 0;
var EbaySiteNameEnum;
(function (EbaySiteNameEnum) {
    EbaySiteNameEnum["EBAY_MAIN"] = "EBAY_MAIN";
    EbaySiteNameEnum["EBAY_US"] = "EBAY_US";
    EbaySiteNameEnum["EBAY_CA"] = "EBAY_CA";
    EbaySiteNameEnum["EBAY_UK"] = "EBAY_UK";
    EbaySiteNameEnum["EBAY_AU"] = "EBAY_AU";
    EbaySiteNameEnum["EBAY_AT"] = "EBAY_AT";
    EbaySiteNameEnum["EBAY_BE_FR"] = "EBAY_BE_FR";
    EbaySiteNameEnum["EBAY_FR"] = "EBAY_FR";
    EbaySiteNameEnum["EBAY_DE"] = "EBAY_DE";
    EbaySiteNameEnum["EBAY_AUTOS"] = "EBAY_AUTOS";
    EbaySiteNameEnum["EBAY_IT"] = "EBAY_IT";
    EbaySiteNameEnum["EBAY_JP"] = "EBAY_JP";
    EbaySiteNameEnum["EBAY_BE_NL"] = "EBAY_BE_NL";
    EbaySiteNameEnum["EBAY_NL"] = "EBAY_NL";
    EbaySiteNameEnum["EBAY_ES"] = "EBAY_ES";
    EbaySiteNameEnum["EBAY_CH"] = "EBAY_CH";
    EbaySiteNameEnum["EBAY_TW"] = "EBAY_TW";
    EbaySiteNameEnum["EBAY_HALFCOM"] = "EBAY_HALFCOM";
    EbaySiteNameEnum["EBAY_CZ"] = "EBAY_CZ";
    EbaySiteNameEnum["EBAY_DK"] = "EBAY_DK";
    EbaySiteNameEnum["EBAY_FI"] = "EBAY_FI";
    EbaySiteNameEnum["EBAY_GR"] = "EBAY_GR";
    EbaySiteNameEnum["EBAY_HK"] = "EBAY_HK";
    EbaySiteNameEnum["EBAY_HU"] = "EBAY_HU";
    EbaySiteNameEnum["EBAY_IN"] = "EBAY_IN";
    EbaySiteNameEnum["EBAY_ID"] = "EBAY_ID";
    EbaySiteNameEnum["EBAY_IE"] = "EBAY_IE";
    EbaySiteNameEnum["EBAY_IL"] = "EBAY_IL";
    EbaySiteNameEnum["EBAY_MY"] = "EBAY_MY";
    EbaySiteNameEnum["EBAY_NZ"] = "EBAY_NZ";
    EbaySiteNameEnum["EBAY_NO"] = "EBAY_NO";
    EbaySiteNameEnum["EBAY_QUEBEC"] = "EBAY_QUEBEC";
    EbaySiteNameEnum["EBAY_PH"] = "EBAY_PH";
    EbaySiteNameEnum["EBAY_PL"] = "EBAY_PL";
    EbaySiteNameEnum["EBAY_PT"] = "EBAY_PT";
    EbaySiteNameEnum["EBAY_PR"] = "EBAY_PR";
    EbaySiteNameEnum["EBAY_RU"] = "EBAY_RU";
    EbaySiteNameEnum["EBAY_SG"] = "EBAY_SG";
    EbaySiteNameEnum["EBAY_ZA"] = "EBAY_ZA";
    EbaySiteNameEnum["EBAY_SE"] = "EBAY_SE";
    EbaySiteNameEnum["EBAY_TH"] = "EBAY_TH";
    EbaySiteNameEnum["EBAY_VN"] = "EBAY_VN";
    EbaySiteNameEnum["EBAY_US1"] = "EBAY_US1";
    EbaySiteNameEnum["EBAY_CN"] = "EBAY_CN";
    EbaySiteNameEnum["EBAY_ROW_JP"] = "EBAY_ROW_JP";
    EbaySiteNameEnum["EBAY_PE"] = "EBAY_PE";
    EbaySiteNameEnum["EBAY_KR"] = "EBAY_KR";
    EbaySiteNameEnum["EBAY_IS"] = "EBAY_IS";
    EbaySiteNameEnum["EBAY_VAT"] = "EBAY_VAT";
    EbaySiteNameEnum["EBAY_NULL"] = "EBAY_NULL";
    EbaySiteNameEnum["EBAY_AR"] = "EBAY_AR";
    EbaySiteNameEnum["EBAY_BR"] = "EBAY_BR";
    EbaySiteNameEnum["EBAY_CL"] = "EBAY_CL";
    EbaySiteNameEnum["EBAY_CO"] = "EBAY_CO";
    EbaySiteNameEnum["EBAY_MX"] = "EBAY_MX";
    EbaySiteNameEnum["EBAY_RO"] = "EBAY_RO";
    EbaySiteNameEnum["EBAY_SA"] = "EBAY_SA";
    EbaySiteNameEnum["EBAY_SK"] = "EBAY_SK";
    EbaySiteNameEnum["EBAY_TR"] = "EBAY_TR";
    EbaySiteNameEnum["EBAY_UA"] = "EBAY_UA";
    EbaySiteNameEnum["EBAY_AE"] = "EBAY_AE";
    EbaySiteNameEnum["EBAY_GBH"] = "EBAY_GBH";
    EbaySiteNameEnum["EBAY_AF"] = "EBAY_AF";
    EbaySiteNameEnum["EBAY_AL"] = "EBAY_AL";
    EbaySiteNameEnum["EBAY_DZ"] = "EBAY_DZ";
    EbaySiteNameEnum["EBAY_AS"] = "EBAY_AS";
    EbaySiteNameEnum["EBAY_AD"] = "EBAY_AD";
    EbaySiteNameEnum["EBAY_AO"] = "EBAY_AO";
    EbaySiteNameEnum["EBAY_AI"] = "EBAY_AI";
    EbaySiteNameEnum["EBAY_AG"] = "EBAY_AG";
    EbaySiteNameEnum["EBAY_AM"] = "EBAY_AM";
    EbaySiteNameEnum["EBAY_AW"] = "EBAY_AW";
    EbaySiteNameEnum["EBAY_AZ"] = "EBAY_AZ";
    EbaySiteNameEnum["EBAY_BS"] = "EBAY_BS";
    EbaySiteNameEnum["EBAY_BH"] = "EBAY_BH";
    EbaySiteNameEnum["EBAY_BD"] = "EBAY_BD";
    EbaySiteNameEnum["EBAY_BB"] = "EBAY_BB";
    EbaySiteNameEnum["EBAY_BY"] = "EBAY_BY";
    EbaySiteNameEnum["EBAY_BZ"] = "EBAY_BZ";
    EbaySiteNameEnum["EBAY_BJ"] = "EBAY_BJ";
    EbaySiteNameEnum["EBAY_BM"] = "EBAY_BM";
    EbaySiteNameEnum["EBAY_BT"] = "EBAY_BT";
    EbaySiteNameEnum["EBAY_BO"] = "EBAY_BO";
    EbaySiteNameEnum["EBAY_BA"] = "EBAY_BA";
    EbaySiteNameEnum["EBAY_BW"] = "EBAY_BW";
    EbaySiteNameEnum["EBAY_VG"] = "EBAY_VG";
    EbaySiteNameEnum["EBAY_BN"] = "EBAY_BN";
    EbaySiteNameEnum["EBAY_BG"] = "EBAY_BG";
    EbaySiteNameEnum["EBAY_BF"] = "EBAY_BF";
    EbaySiteNameEnum["EBAY_BI"] = "EBAY_BI";
    EbaySiteNameEnum["EBAY_KH"] = "EBAY_KH";
    EbaySiteNameEnum["EBAY_CM"] = "EBAY_CM";
    EbaySiteNameEnum["EBAY_CV"] = "EBAY_CV";
    EbaySiteNameEnum["EBAY_KY"] = "EBAY_KY";
    EbaySiteNameEnum["EBAY_CF"] = "EBAY_CF";
    EbaySiteNameEnum["EBAY_TD"] = "EBAY_TD";
    EbaySiteNameEnum["EBAY_KM"] = "EBAY_KM";
    EbaySiteNameEnum["EBAY_ZR"] = "EBAY_ZR";
    EbaySiteNameEnum["EBAY_CG"] = "EBAY_CG";
    EbaySiteNameEnum["EBAY_CK"] = "EBAY_CK";
    EbaySiteNameEnum["EBAY_CR"] = "EBAY_CR";
    EbaySiteNameEnum["EBAY_CI"] = "EBAY_CI";
    EbaySiteNameEnum["EBAY_HR"] = "EBAY_HR";
    EbaySiteNameEnum["EBAY_CY"] = "EBAY_CY";
    EbaySiteNameEnum["EBAY_DJ"] = "EBAY_DJ";
    EbaySiteNameEnum["EBAY_DM"] = "EBAY_DM";
    EbaySiteNameEnum["EBAY_DO"] = "EBAY_DO";
    EbaySiteNameEnum["EBAY_EC"] = "EBAY_EC";
    EbaySiteNameEnum["EBAY_EG"] = "EBAY_EG";
    EbaySiteNameEnum["EBAY_SV"] = "EBAY_SV";
    EbaySiteNameEnum["EBAY_GQ"] = "EBAY_GQ";
    EbaySiteNameEnum["EBAY_QQ"] = "EBAY_QQ";
    EbaySiteNameEnum["EBAY_EE"] = "EBAY_EE";
    EbaySiteNameEnum["EBAY_ET"] = "EBAY_ET";
    EbaySiteNameEnum["EBAY_FK"] = "EBAY_FK";
    EbaySiteNameEnum["EBAY_FJ"] = "EBAY_FJ";
    EbaySiteNameEnum["EBAY_GF"] = "EBAY_GF";
    EbaySiteNameEnum["EBAY_PF"] = "EBAY_PF";
    EbaySiteNameEnum["EBAY_GA"] = "EBAY_GA";
    EbaySiteNameEnum["EBAY_GM"] = "EBAY_GM";
    EbaySiteNameEnum["EBAY_GE"] = "EBAY_GE";
    EbaySiteNameEnum["EBAY_GH"] = "EBAY_GH";
    EbaySiteNameEnum["EBAY_GI"] = "EBAY_GI";
    EbaySiteNameEnum["EBAY_GL"] = "EBAY_GL";
    EbaySiteNameEnum["EBAY_GD"] = "EBAY_GD";
    EbaySiteNameEnum["EBAY_GP"] = "EBAY_GP";
    EbaySiteNameEnum["EBAY_GU"] = "EBAY_GU";
    EbaySiteNameEnum["EBAY_GT"] = "EBAY_GT";
    EbaySiteNameEnum["EBAY_GG"] = "EBAY_GG";
    EbaySiteNameEnum["EBAY_GN"] = "EBAY_GN";
    EbaySiteNameEnum["EBAY_GW"] = "EBAY_GW";
    EbaySiteNameEnum["EBAY_GY"] = "EBAY_GY";
    EbaySiteNameEnum["EBAY_HT"] = "EBAY_HT";
    EbaySiteNameEnum["EBAY_HN"] = "EBAY_HN";
    EbaySiteNameEnum["EBAY_JM"] = "EBAY_JM";
    EbaySiteNameEnum["EBAY_SJ"] = "EBAY_SJ";
    EbaySiteNameEnum["EBAY_JE"] = "EBAY_JE";
    EbaySiteNameEnum["EBAY_JO"] = "EBAY_JO";
    EbaySiteNameEnum["EBAY_KZ"] = "EBAY_KZ";
    EbaySiteNameEnum["EBAY_KE"] = "EBAY_KE";
    EbaySiteNameEnum["EBAY_KI"] = "EBAY_KI";
    EbaySiteNameEnum["EBAY_KW"] = "EBAY_KW";
    EbaySiteNameEnum["EBAY_KG"] = "EBAY_KG";
    EbaySiteNameEnum["EBAY_LA"] = "EBAY_LA";
    EbaySiteNameEnum["EBAY_LV"] = "EBAY_LV";
    EbaySiteNameEnum["EBAY_LB"] = "EBAY_LB";
    EbaySiteNameEnum["EBAY_LR"] = "EBAY_LR";
    EbaySiteNameEnum["EBAY_LI"] = "EBAY_LI";
    EbaySiteNameEnum["EBAY_LT"] = "EBAY_LT";
    EbaySiteNameEnum["EBAY_LU"] = "EBAY_LU";
    EbaySiteNameEnum["EBAY_MO"] = "EBAY_MO";
    EbaySiteNameEnum["EBAY_MK"] = "EBAY_MK";
    EbaySiteNameEnum["EBAY_MG"] = "EBAY_MG";
    EbaySiteNameEnum["EBAY_MW"] = "EBAY_MW";
    EbaySiteNameEnum["EBAY_MV"] = "EBAY_MV";
    EbaySiteNameEnum["EBAY_ML"] = "EBAY_ML";
    EbaySiteNameEnum["EBAY_MT"] = "EBAY_MT";
    EbaySiteNameEnum["EBAY_MH"] = "EBAY_MH";
    EbaySiteNameEnum["EBAY_MQ"] = "EBAY_MQ";
    EbaySiteNameEnum["EBAY_MR"] = "EBAY_MR";
    EbaySiteNameEnum["EBAY_MU"] = "EBAY_MU";
    EbaySiteNameEnum["EBAY_YT"] = "EBAY_YT";
    EbaySiteNameEnum["EBAY_MI"] = "EBAY_MI";
    EbaySiteNameEnum["EBAY_MD"] = "EBAY_MD";
    EbaySiteNameEnum["EBAY_MC"] = "EBAY_MC";
    EbaySiteNameEnum["EBAY_MN"] = "EBAY_MN";
    EbaySiteNameEnum["EBAY_ME"] = "EBAY_ME";
    EbaySiteNameEnum["EBAY_MS"] = "EBAY_MS";
    EbaySiteNameEnum["EBAY_MA"] = "EBAY_MA";
    EbaySiteNameEnum["EBAY_MZ"] = "EBAY_MZ";
    EbaySiteNameEnum["EBAY_NA"] = "EBAY_NA";
    EbaySiteNameEnum["EBAY_NR"] = "EBAY_NR";
    EbaySiteNameEnum["EBAY_NP"] = "EBAY_NP";
    EbaySiteNameEnum["EBAY_AN"] = "EBAY_AN";
    EbaySiteNameEnum["EBAY_NC"] = "EBAY_NC";
    EbaySiteNameEnum["EBAY_NI"] = "EBAY_NI";
    EbaySiteNameEnum["EBAY_NE"] = "EBAY_NE";
    EbaySiteNameEnum["EBAY_NG"] = "EBAY_NG";
    EbaySiteNameEnum["EBAY_NU"] = "EBAY_NU";
    EbaySiteNameEnum["EBAY_OM"] = "EBAY_OM";
    EbaySiteNameEnum["EBAY_PK"] = "EBAY_PK";
    EbaySiteNameEnum["EBAY_PW"] = "EBAY_PW";
    EbaySiteNameEnum["EBAY_PA"] = "EBAY_PA";
    EbaySiteNameEnum["EBAY_PG"] = "EBAY_PG";
    EbaySiteNameEnum["EBAY_PY"] = "EBAY_PY";
    EbaySiteNameEnum["EBAY_QA"] = "EBAY_QA";
    EbaySiteNameEnum["EBAY_RE"] = "EBAY_RE";
    EbaySiteNameEnum["EBAY_RW"] = "EBAY_RW";
    EbaySiteNameEnum["EBAY_SH"] = "EBAY_SH";
    EbaySiteNameEnum["EBAY_KN"] = "EBAY_KN";
    EbaySiteNameEnum["EBAY_LC"] = "EBAY_LC";
    EbaySiteNameEnum["EBAY_PM"] = "EBAY_PM";
    EbaySiteNameEnum["EBAY_VC"] = "EBAY_VC";
    EbaySiteNameEnum["EBAY_SM"] = "EBAY_SM";
    EbaySiteNameEnum["EBAY_SN"] = "EBAY_SN";
    EbaySiteNameEnum["EBAY_RS"] = "EBAY_RS";
    EbaySiteNameEnum["EBAY_SC"] = "EBAY_SC";
    EbaySiteNameEnum["EBAY_SL"] = "EBAY_SL";
    EbaySiteNameEnum["EBAY_SI"] = "EBAY_SI";
    EbaySiteNameEnum["EBAY_SB"] = "EBAY_SB";
    EbaySiteNameEnum["EBAY_SO"] = "EBAY_SO";
    EbaySiteNameEnum["EBAY_LK"] = "EBAY_LK";
    EbaySiteNameEnum["EBAY_SR"] = "EBAY_SR";
    EbaySiteNameEnum["EBAY_SVL"] = "EBAY_SVL";
    EbaySiteNameEnum["EBAY_SZ"] = "EBAY_SZ";
    EbaySiteNameEnum["EBAY_TA"] = "EBAY_TA";
    EbaySiteNameEnum["EBAY_TJ"] = "EBAY_TJ";
    EbaySiteNameEnum["EBAY_TZ"] = "EBAY_TZ";
    EbaySiteNameEnum["EBAY_TG"] = "EBAY_TG";
    EbaySiteNameEnum["EBAY_TO"] = "EBAY_TO";
    EbaySiteNameEnum["EBAY_TT"] = "EBAY_TT";
    EbaySiteNameEnum["EBAY_TN"] = "EBAY_TN";
    EbaySiteNameEnum["EBAY_TM"] = "EBAY_TM";
    EbaySiteNameEnum["EBAY_TC"] = "EBAY_TC";
    EbaySiteNameEnum["EBAY_TV"] = "EBAY_TV";
    EbaySiteNameEnum["EBAY_UG"] = "EBAY_UG";
    EbaySiteNameEnum["EBAY_UY"] = "EBAY_UY";
    EbaySiteNameEnum["EBAY_UZ"] = "EBAY_UZ";
    EbaySiteNameEnum["EBAY_VU"] = "EBAY_VU";
    EbaySiteNameEnum["EBAY_VA"] = "EBAY_VA";
    EbaySiteNameEnum["EBAY_VE"] = "EBAY_VE";
    EbaySiteNameEnum["EBAY_VI"] = "EBAY_VI";
    EbaySiteNameEnum["EBAY_WF"] = "EBAY_WF";
    EbaySiteNameEnum["EBAY_EH"] = "EBAY_EH";
    EbaySiteNameEnum["EBAY_WS"] = "EBAY_WS";
    EbaySiteNameEnum["EBAY_YE"] = "EBAY_YE";
    EbaySiteNameEnum["EBAY_ZM"] = "EBAY_ZM";
    EbaySiteNameEnum["EBAY_ZW"] = "EBAY_ZW";
    EbaySiteNameEnum["EBAY_LS"] = "EBAY_LS";
    EbaySiteNameEnum["EBAY_YU"] = "EBAY_YU";
})(EbaySiteNameEnum = exports.EbaySiteNameEnum || (exports.EbaySiteNameEnum = {}));
;
var EbaySiteIdEnum;
(function (EbaySiteIdEnum) {
    EbaySiteIdEnum[EbaySiteIdEnum["EBAY_MAIN"] = 0] = "EBAY_MAIN";
    EbaySiteIdEnum[EbaySiteIdEnum["EBAY_US"] = 1] = "EBAY_US";
    EbaySiteIdEnum[EbaySiteIdEnum["EBAY_CA"] = 2] = "EBAY_CA";
    EbaySiteIdEnum[EbaySiteIdEnum["EBAY_UK"] = 3] = "EBAY_UK";
    EbaySiteIdEnum[EbaySiteIdEnum["EBAY_AU"] = 15] = "EBAY_AU";
    EbaySiteIdEnum[EbaySiteIdEnum["EBAY_AT"] = 16] = "EBAY_AT";
    EbaySiteIdEnum[EbaySiteIdEnum["EBAY_BE_FR"] = 23] = "EBAY_BE_FR";
    EbaySiteIdEnum[EbaySiteIdEnum["EBAY_FR"] = 71] = "EBAY_FR";
    EbaySiteIdEnum[EbaySiteIdEnum["EBAY_DE"] = 77] = "EBAY_DE";
    EbaySiteIdEnum[EbaySiteIdEnum["EBAY_AUTOS"] = 100] = "EBAY_AUTOS";
    EbaySiteIdEnum[EbaySiteIdEnum["EBAY_IT"] = 101] = "EBAY_IT";
    EbaySiteIdEnum[EbaySiteIdEnum["EBAY_JP"] = 104] = "EBAY_JP";
    EbaySiteIdEnum[EbaySiteIdEnum["EBAY_BE_NL"] = 123] = "EBAY_BE_NL";
    EbaySiteIdEnum[EbaySiteIdEnum["EBAY_NL"] = 146] = "EBAY_NL";
    EbaySiteIdEnum[EbaySiteIdEnum["EBAY_ES"] = 186] = "EBAY_ES";
    EbaySiteIdEnum[EbaySiteIdEnum["EBAY_CH"] = 193] = "EBAY_CH";
    EbaySiteIdEnum[EbaySiteIdEnum["EBAY_TW"] = 196] = "EBAY_TW";
    EbaySiteIdEnum[EbaySiteIdEnum["EBAY_HALFCOM"] = 20] = "EBAY_HALFCOM";
    EbaySiteIdEnum[EbaySiteIdEnum["EBAY_CZ"] = 197] = "EBAY_CZ";
    EbaySiteIdEnum[EbaySiteIdEnum["EBAY_DK"] = 198] = "EBAY_DK";
    EbaySiteIdEnum[EbaySiteIdEnum["EBAY_FI"] = 199] = "EBAY_FI";
    EbaySiteIdEnum[EbaySiteIdEnum["EBAY_GR"] = 200] = "EBAY_GR";
    EbaySiteIdEnum[EbaySiteIdEnum["EBAY_HK"] = 201] = "EBAY_HK";
    EbaySiteIdEnum[EbaySiteIdEnum["EBAY_HU"] = 202] = "EBAY_HU";
    EbaySiteIdEnum[EbaySiteIdEnum["EBAY_IN"] = 203] = "EBAY_IN";
    EbaySiteIdEnum[EbaySiteIdEnum["EBAY_ID"] = 204] = "EBAY_ID";
    EbaySiteIdEnum[EbaySiteIdEnum["EBAY_IE"] = 205] = "EBAY_IE";
    EbaySiteIdEnum[EbaySiteIdEnum["EBAY_IL"] = 206] = "EBAY_IL";
    EbaySiteIdEnum[EbaySiteIdEnum["EBAY_MY"] = 207] = "EBAY_MY";
    EbaySiteIdEnum[EbaySiteIdEnum["EBAY_NZ"] = 208] = "EBAY_NZ";
    EbaySiteIdEnum[EbaySiteIdEnum["EBAY_NO"] = 209] = "EBAY_NO";
    EbaySiteIdEnum[EbaySiteIdEnum["EBAY_QUEBEC"] = 210] = "EBAY_QUEBEC";
    EbaySiteIdEnum[EbaySiteIdEnum["EBAY_PH"] = 211] = "EBAY_PH";
    EbaySiteIdEnum[EbaySiteIdEnum["EBAY_PL"] = 212] = "EBAY_PL";
    EbaySiteIdEnum[EbaySiteIdEnum["EBAY_PT"] = 213] = "EBAY_PT";
    EbaySiteIdEnum[EbaySiteIdEnum["EBAY_PR"] = 214] = "EBAY_PR";
    EbaySiteIdEnum[EbaySiteIdEnum["EBAY_RU"] = 215] = "EBAY_RU";
    EbaySiteIdEnum[EbaySiteIdEnum["EBAY_SG"] = 216] = "EBAY_SG";
    EbaySiteIdEnum[EbaySiteIdEnum["EBAY_ZA"] = 217] = "EBAY_ZA";
    EbaySiteIdEnum[EbaySiteIdEnum["EBAY_SE"] = 218] = "EBAY_SE";
    EbaySiteIdEnum[EbaySiteIdEnum["EBAY_TH"] = 219] = "EBAY_TH";
    EbaySiteIdEnum[EbaySiteIdEnum["EBAY_VN"] = 220] = "EBAY_VN";
    EbaySiteIdEnum[EbaySiteIdEnum["EBAY_US1"] = 221] = "EBAY_US1";
    EbaySiteIdEnum[EbaySiteIdEnum["EBAY_CN"] = 223] = "EBAY_CN";
    EbaySiteIdEnum[EbaySiteIdEnum["EBAY_ROW_JP"] = 224] = "EBAY_ROW_JP";
    EbaySiteIdEnum[EbaySiteIdEnum["EBAY_PE"] = 225] = "EBAY_PE";
    EbaySiteIdEnum[EbaySiteIdEnum["EBAY_KR"] = 226] = "EBAY_KR";
    EbaySiteIdEnum[EbaySiteIdEnum["EBAY_IS"] = 227] = "EBAY_IS";
    EbaySiteIdEnum[EbaySiteIdEnum["EBAY_VAT"] = 999] = "EBAY_VAT";
    EbaySiteIdEnum[EbaySiteIdEnum["EBAY_NULL"] = 1000] = "EBAY_NULL";
    EbaySiteIdEnum[EbaySiteIdEnum["EBAY_AR"] = 12] = "EBAY_AR";
    EbaySiteIdEnum[EbaySiteIdEnum["EBAY_BR"] = 31] = "EBAY_BR";
    EbaySiteIdEnum[EbaySiteIdEnum["EBAY_CL"] = 44] = "EBAY_CL";
    EbaySiteIdEnum[EbaySiteIdEnum["EBAY_CO"] = 46] = "EBAY_CO";
    EbaySiteIdEnum[EbaySiteIdEnum["EBAY_MX"] = 136] = "EBAY_MX";
    EbaySiteIdEnum[EbaySiteIdEnum["EBAY_RO"] = 167] = "EBAY_RO";
    EbaySiteIdEnum[EbaySiteIdEnum["EBAY_SA"] = 176] = "EBAY_SA";
    EbaySiteIdEnum[EbaySiteIdEnum["EBAY_SK"] = 181] = "EBAY_SK";
    EbaySiteIdEnum[EbaySiteIdEnum["EBAY_TR"] = 231] = "EBAY_TR";
    EbaySiteIdEnum[EbaySiteIdEnum["EBAY_UA"] = 232] = "EBAY_UA";
    EbaySiteIdEnum[EbaySiteIdEnum["EBAY_AE"] = 233] = "EBAY_AE";
    EbaySiteIdEnum[EbaySiteIdEnum["EBAY_GBH"] = 900] = "EBAY_GBH";
    EbaySiteIdEnum[EbaySiteIdEnum["EBAY_AF"] = 4] = "EBAY_AF";
    EbaySiteIdEnum[EbaySiteIdEnum["EBAY_AL"] = 5] = "EBAY_AL";
    EbaySiteIdEnum[EbaySiteIdEnum["EBAY_DZ"] = 6] = "EBAY_DZ";
    EbaySiteIdEnum[EbaySiteIdEnum["EBAY_AS"] = 7] = "EBAY_AS";
    EbaySiteIdEnum[EbaySiteIdEnum["EBAY_AD"] = 8] = "EBAY_AD";
    EbaySiteIdEnum[EbaySiteIdEnum["EBAY_AO"] = 9] = "EBAY_AO";
    EbaySiteIdEnum[EbaySiteIdEnum["EBAY_AI"] = 10] = "EBAY_AI";
    EbaySiteIdEnum[EbaySiteIdEnum["EBAY_AG"] = 11] = "EBAY_AG";
    EbaySiteIdEnum[EbaySiteIdEnum["EBAY_AM"] = 13] = "EBAY_AM";
    EbaySiteIdEnum[EbaySiteIdEnum["EBAY_AW"] = 14] = "EBAY_AW";
    EbaySiteIdEnum[EbaySiteIdEnum["EBAY_AZ"] = 17] = "EBAY_AZ";
    EbaySiteIdEnum[EbaySiteIdEnum["EBAY_BS"] = 18] = "EBAY_BS";
    EbaySiteIdEnum[EbaySiteIdEnum["EBAY_BH"] = 19] = "EBAY_BH";
    EbaySiteIdEnum[EbaySiteIdEnum["EBAY_BD"] = 234] = "EBAY_BD";
    EbaySiteIdEnum[EbaySiteIdEnum["EBAY_BB"] = 21] = "EBAY_BB";
    EbaySiteIdEnum[EbaySiteIdEnum["EBAY_BY"] = 22] = "EBAY_BY";
    EbaySiteIdEnum[EbaySiteIdEnum["EBAY_BZ"] = 24] = "EBAY_BZ";
    EbaySiteIdEnum[EbaySiteIdEnum["EBAY_BJ"] = 25] = "EBAY_BJ";
    EbaySiteIdEnum[EbaySiteIdEnum["EBAY_BM"] = 26] = "EBAY_BM";
    EbaySiteIdEnum[EbaySiteIdEnum["EBAY_BT"] = 27] = "EBAY_BT";
    EbaySiteIdEnum[EbaySiteIdEnum["EBAY_BO"] = 28] = "EBAY_BO";
    EbaySiteIdEnum[EbaySiteIdEnum["EBAY_BA"] = 29] = "EBAY_BA";
    EbaySiteIdEnum[EbaySiteIdEnum["EBAY_BW"] = 30] = "EBAY_BW";
    EbaySiteIdEnum[EbaySiteIdEnum["EBAY_VG"] = 32] = "EBAY_VG";
    EbaySiteIdEnum[EbaySiteIdEnum["EBAY_BN"] = 33] = "EBAY_BN";
    EbaySiteIdEnum[EbaySiteIdEnum["EBAY_BG"] = 34] = "EBAY_BG";
    EbaySiteIdEnum[EbaySiteIdEnum["EBAY_BF"] = 35] = "EBAY_BF";
    EbaySiteIdEnum[EbaySiteIdEnum["EBAY_BI"] = 37] = "EBAY_BI";
    EbaySiteIdEnum[EbaySiteIdEnum["EBAY_KH"] = 38] = "EBAY_KH";
    EbaySiteIdEnum[EbaySiteIdEnum["EBAY_CM"] = 39] = "EBAY_CM";
    EbaySiteIdEnum[EbaySiteIdEnum["EBAY_CV"] = 40] = "EBAY_CV";
    EbaySiteIdEnum[EbaySiteIdEnum["EBAY_KY"] = 41] = "EBAY_KY";
    EbaySiteIdEnum[EbaySiteIdEnum["EBAY_CF"] = 42] = "EBAY_CF";
    EbaySiteIdEnum[EbaySiteIdEnum["EBAY_TD"] = 43] = "EBAY_TD";
    EbaySiteIdEnum[EbaySiteIdEnum["EBAY_KM"] = 47] = "EBAY_KM";
    EbaySiteIdEnum[EbaySiteIdEnum["EBAY_ZR"] = 48] = "EBAY_ZR";
    EbaySiteIdEnum[EbaySiteIdEnum["EBAY_CG"] = 49] = "EBAY_CG";
    EbaySiteIdEnum[EbaySiteIdEnum["EBAY_CK"] = 50] = "EBAY_CK";
    EbaySiteIdEnum[EbaySiteIdEnum["EBAY_CR"] = 51] = "EBAY_CR";
    EbaySiteIdEnum[EbaySiteIdEnum["EBAY_CI"] = 52] = "EBAY_CI";
    EbaySiteIdEnum[EbaySiteIdEnum["EBAY_HR"] = 53] = "EBAY_HR";
    EbaySiteIdEnum[EbaySiteIdEnum["EBAY_CY"] = 55] = "EBAY_CY";
    EbaySiteIdEnum[EbaySiteIdEnum["EBAY_DJ"] = 58] = "EBAY_DJ";
    EbaySiteIdEnum[EbaySiteIdEnum["EBAY_DM"] = 59] = "EBAY_DM";
    EbaySiteIdEnum[EbaySiteIdEnum["EBAY_DO"] = 60] = "EBAY_DO";
    EbaySiteIdEnum[EbaySiteIdEnum["EBAY_EC"] = 61] = "EBAY_EC";
    EbaySiteIdEnum[EbaySiteIdEnum["EBAY_EG"] = 62] = "EBAY_EG";
    EbaySiteIdEnum[EbaySiteIdEnum["EBAY_SV"] = 63] = "EBAY_SV";
    EbaySiteIdEnum[EbaySiteIdEnum["EBAY_GQ"] = 64] = "EBAY_GQ";
    EbaySiteIdEnum[EbaySiteIdEnum["EBAY_QQ"] = 65] = "EBAY_QQ";
    EbaySiteIdEnum[EbaySiteIdEnum["EBAY_EE"] = 66] = "EBAY_EE";
    EbaySiteIdEnum[EbaySiteIdEnum["EBAY_ET"] = 67] = "EBAY_ET";
    EbaySiteIdEnum[EbaySiteIdEnum["EBAY_FK"] = 68] = "EBAY_FK";
    EbaySiteIdEnum[EbaySiteIdEnum["EBAY_FJ"] = 69] = "EBAY_FJ";
    EbaySiteIdEnum[EbaySiteIdEnum["EBAY_GF"] = 72] = "EBAY_GF";
    EbaySiteIdEnum[EbaySiteIdEnum["EBAY_PF"] = 73] = "EBAY_PF";
    EbaySiteIdEnum[EbaySiteIdEnum["EBAY_GA"] = 74] = "EBAY_GA";
    EbaySiteIdEnum[EbaySiteIdEnum["EBAY_GM"] = 75] = "EBAY_GM";
    EbaySiteIdEnum[EbaySiteIdEnum["EBAY_GE"] = 76] = "EBAY_GE";
    EbaySiteIdEnum[EbaySiteIdEnum["EBAY_GH"] = 78] = "EBAY_GH";
    EbaySiteIdEnum[EbaySiteIdEnum["EBAY_GI"] = 79] = "EBAY_GI";
    EbaySiteIdEnum[EbaySiteIdEnum["EBAY_GL"] = 81] = "EBAY_GL";
    EbaySiteIdEnum[EbaySiteIdEnum["EBAY_GD"] = 82] = "EBAY_GD";
    EbaySiteIdEnum[EbaySiteIdEnum["EBAY_GP"] = 83] = "EBAY_GP";
    EbaySiteIdEnum[EbaySiteIdEnum["EBAY_GU"] = 84] = "EBAY_GU";
    EbaySiteIdEnum[EbaySiteIdEnum["EBAY_GT"] = 85] = "EBAY_GT";
    EbaySiteIdEnum[EbaySiteIdEnum["EBAY_GG"] = 86] = "EBAY_GG";
    EbaySiteIdEnum[EbaySiteIdEnum["EBAY_GN"] = 87] = "EBAY_GN";
    EbaySiteIdEnum[EbaySiteIdEnum["EBAY_GW"] = 88] = "EBAY_GW";
    EbaySiteIdEnum[EbaySiteIdEnum["EBAY_GY"] = 89] = "EBAY_GY";
    EbaySiteIdEnum[EbaySiteIdEnum["EBAY_HT"] = 90] = "EBAY_HT";
    EbaySiteIdEnum[EbaySiteIdEnum["EBAY_HN"] = 91] = "EBAY_HN";
    EbaySiteIdEnum[EbaySiteIdEnum["EBAY_JM"] = 102] = "EBAY_JM";
    EbaySiteIdEnum[EbaySiteIdEnum["EBAY_SJ"] = 103] = "EBAY_SJ";
    EbaySiteIdEnum[EbaySiteIdEnum["EBAY_JE"] = 105] = "EBAY_JE";
    EbaySiteIdEnum[EbaySiteIdEnum["EBAY_JO"] = 106] = "EBAY_JO";
    EbaySiteIdEnum[EbaySiteIdEnum["EBAY_KZ"] = 107] = "EBAY_KZ";
    EbaySiteIdEnum[EbaySiteIdEnum["EBAY_KE"] = 108] = "EBAY_KE";
    EbaySiteIdEnum[EbaySiteIdEnum["EBAY_KI"] = 109] = "EBAY_KI";
    EbaySiteIdEnum[EbaySiteIdEnum["EBAY_KW"] = 112] = "EBAY_KW";
    EbaySiteIdEnum[EbaySiteIdEnum["EBAY_KG"] = 113] = "EBAY_KG";
    EbaySiteIdEnum[EbaySiteIdEnum["EBAY_LA"] = 114] = "EBAY_LA";
    EbaySiteIdEnum[EbaySiteIdEnum["EBAY_LV"] = 115] = "EBAY_LV";
    EbaySiteIdEnum[EbaySiteIdEnum["EBAY_LB"] = 116] = "EBAY_LB";
    EbaySiteIdEnum[EbaySiteIdEnum["EBAY_LR"] = 118] = "EBAY_LR";
    EbaySiteIdEnum[EbaySiteIdEnum["EBAY_LI"] = 120] = "EBAY_LI";
    EbaySiteIdEnum[EbaySiteIdEnum["EBAY_LT"] = 121] = "EBAY_LT";
    EbaySiteIdEnum[EbaySiteIdEnum["EBAY_LU"] = 122] = "EBAY_LU";
    EbaySiteIdEnum[EbaySiteIdEnum["EBAY_MO"] = 235] = "EBAY_MO";
    EbaySiteIdEnum[EbaySiteIdEnum["EBAY_MK"] = 124] = "EBAY_MK";
    EbaySiteIdEnum[EbaySiteIdEnum["EBAY_MG"] = 125] = "EBAY_MG";
    EbaySiteIdEnum[EbaySiteIdEnum["EBAY_MW"] = 126] = "EBAY_MW";
    EbaySiteIdEnum[EbaySiteIdEnum["EBAY_MV"] = 128] = "EBAY_MV";
    EbaySiteIdEnum[EbaySiteIdEnum["EBAY_ML"] = 129] = "EBAY_ML";
    EbaySiteIdEnum[EbaySiteIdEnum["EBAY_MT"] = 130] = "EBAY_MT";
    EbaySiteIdEnum[EbaySiteIdEnum["EBAY_MH"] = 131] = "EBAY_MH";
    EbaySiteIdEnum[EbaySiteIdEnum["EBAY_MQ"] = 132] = "EBAY_MQ";
    EbaySiteIdEnum[EbaySiteIdEnum["EBAY_MR"] = 133] = "EBAY_MR";
    EbaySiteIdEnum[EbaySiteIdEnum["EBAY_MU"] = 134] = "EBAY_MU";
    EbaySiteIdEnum[EbaySiteIdEnum["EBAY_YT"] = 135] = "EBAY_YT";
    EbaySiteIdEnum[EbaySiteIdEnum["EBAY_MI"] = 236] = "EBAY_MI";
    EbaySiteIdEnum[EbaySiteIdEnum["EBAY_MD"] = 137] = "EBAY_MD";
    EbaySiteIdEnum[EbaySiteIdEnum["EBAY_MC"] = 138] = "EBAY_MC";
    EbaySiteIdEnum[EbaySiteIdEnum["EBAY_MN"] = 139] = "EBAY_MN";
    EbaySiteIdEnum[EbaySiteIdEnum["EBAY_ME"] = 228] = "EBAY_ME";
    EbaySiteIdEnum[EbaySiteIdEnum["EBAY_MS"] = 140] = "EBAY_MS";
    EbaySiteIdEnum[EbaySiteIdEnum["EBAY_MA"] = 141] = "EBAY_MA";
    EbaySiteIdEnum[EbaySiteIdEnum["EBAY_MZ"] = 142] = "EBAY_MZ";
    EbaySiteIdEnum[EbaySiteIdEnum["EBAY_NA"] = 143] = "EBAY_NA";
    EbaySiteIdEnum[EbaySiteIdEnum["EBAY_NR"] = 144] = "EBAY_NR";
    EbaySiteIdEnum[EbaySiteIdEnum["EBAY_NP"] = 145] = "EBAY_NP";
    EbaySiteIdEnum[EbaySiteIdEnum["EBAY_AN"] = 147] = "EBAY_AN";
    EbaySiteIdEnum[EbaySiteIdEnum["EBAY_NC"] = 148] = "EBAY_NC";
    EbaySiteIdEnum[EbaySiteIdEnum["EBAY_NI"] = 150] = "EBAY_NI";
    EbaySiteIdEnum[EbaySiteIdEnum["EBAY_NE"] = 151] = "EBAY_NE";
    EbaySiteIdEnum[EbaySiteIdEnum["EBAY_NG"] = 152] = "EBAY_NG";
    EbaySiteIdEnum[EbaySiteIdEnum["EBAY_NU"] = 153] = "EBAY_NU";
    EbaySiteIdEnum[EbaySiteIdEnum["EBAY_OM"] = 155] = "EBAY_OM";
    EbaySiteIdEnum[EbaySiteIdEnum["EBAY_PK"] = 156] = "EBAY_PK";
    EbaySiteIdEnum[EbaySiteIdEnum["EBAY_PW"] = 157] = "EBAY_PW";
    EbaySiteIdEnum[EbaySiteIdEnum["EBAY_PA"] = 158] = "EBAY_PA";
    EbaySiteIdEnum[EbaySiteIdEnum["EBAY_PG"] = 159] = "EBAY_PG";
    EbaySiteIdEnum[EbaySiteIdEnum["EBAY_PY"] = 160] = "EBAY_PY";
    EbaySiteIdEnum[EbaySiteIdEnum["EBAY_QA"] = 166] = "EBAY_QA";
    EbaySiteIdEnum[EbaySiteIdEnum["EBAY_RE"] = 237] = "EBAY_RE";
    EbaySiteIdEnum[EbaySiteIdEnum["EBAY_RW"] = 169] = "EBAY_RW";
    EbaySiteIdEnum[EbaySiteIdEnum["EBAY_SH"] = 170] = "EBAY_SH";
    EbaySiteIdEnum[EbaySiteIdEnum["EBAY_KN"] = 171] = "EBAY_KN";
    EbaySiteIdEnum[EbaySiteIdEnum["EBAY_LC"] = 172] = "EBAY_LC";
    EbaySiteIdEnum[EbaySiteIdEnum["EBAY_PM"] = 173] = "EBAY_PM";
    EbaySiteIdEnum[EbaySiteIdEnum["EBAY_VC"] = 174] = "EBAY_VC";
    EbaySiteIdEnum[EbaySiteIdEnum["EBAY_SM"] = 175] = "EBAY_SM";
    EbaySiteIdEnum[EbaySiteIdEnum["EBAY_SN"] = 177] = "EBAY_SN";
    EbaySiteIdEnum[EbaySiteIdEnum["EBAY_RS"] = 229] = "EBAY_RS";
    EbaySiteIdEnum[EbaySiteIdEnum["EBAY_SC"] = 178] = "EBAY_SC";
    EbaySiteIdEnum[EbaySiteIdEnum["EBAY_SL"] = 179] = "EBAY_SL";
    EbaySiteIdEnum[EbaySiteIdEnum["EBAY_SI"] = 182] = "EBAY_SI";
    EbaySiteIdEnum[EbaySiteIdEnum["EBAY_SB"] = 183] = "EBAY_SB";
    EbaySiteIdEnum[EbaySiteIdEnum["EBAY_SO"] = 184] = "EBAY_SO";
    EbaySiteIdEnum[EbaySiteIdEnum["EBAY_LK"] = 187] = "EBAY_LK";
    EbaySiteIdEnum[EbaySiteIdEnum["EBAY_SR"] = 189] = "EBAY_SR";
    EbaySiteIdEnum[EbaySiteIdEnum["EBAY_SVL"] = 190] = "EBAY_SVL";
    EbaySiteIdEnum[EbaySiteIdEnum["EBAY_SZ"] = 191] = "EBAY_SZ";
    EbaySiteIdEnum[EbaySiteIdEnum["EBAY_TA"] = 195] = "EBAY_TA";
    EbaySiteIdEnum[EbaySiteIdEnum["EBAY_TJ"] = 238] = "EBAY_TJ";
    EbaySiteIdEnum[EbaySiteIdEnum["EBAY_TZ"] = 239] = "EBAY_TZ";
    EbaySiteIdEnum[EbaySiteIdEnum["EBAY_TG"] = 240] = "EBAY_TG";
    EbaySiteIdEnum[EbaySiteIdEnum["EBAY_TO"] = 241] = "EBAY_TO";
    EbaySiteIdEnum[EbaySiteIdEnum["EBAY_TT"] = 242] = "EBAY_TT";
    EbaySiteIdEnum[EbaySiteIdEnum["EBAY_TN"] = 243] = "EBAY_TN";
    EbaySiteIdEnum[EbaySiteIdEnum["EBAY_TM"] = 244] = "EBAY_TM";
    EbaySiteIdEnum[EbaySiteIdEnum["EBAY_TC"] = 245] = "EBAY_TC";
    EbaySiteIdEnum[EbaySiteIdEnum["EBAY_TV"] = 246] = "EBAY_TV";
    EbaySiteIdEnum[EbaySiteIdEnum["EBAY_UG"] = 247] = "EBAY_UG";
    EbaySiteIdEnum[EbaySiteIdEnum["EBAY_UY"] = 248] = "EBAY_UY";
    EbaySiteIdEnum[EbaySiteIdEnum["EBAY_UZ"] = 249] = "EBAY_UZ";
    EbaySiteIdEnum[EbaySiteIdEnum["EBAY_VU"] = 250] = "EBAY_VU";
    EbaySiteIdEnum[EbaySiteIdEnum["EBAY_VA"] = 251] = "EBAY_VA";
    EbaySiteIdEnum[EbaySiteIdEnum["EBAY_VE"] = 252] = "EBAY_VE";
    EbaySiteIdEnum[EbaySiteIdEnum["EBAY_VI"] = 253] = "EBAY_VI";
    EbaySiteIdEnum[EbaySiteIdEnum["EBAY_WF"] = 254] = "EBAY_WF";
    EbaySiteIdEnum[EbaySiteIdEnum["EBAY_EH"] = 255] = "EBAY_EH";
    EbaySiteIdEnum[EbaySiteIdEnum["EBAY_WS"] = 256] = "EBAY_WS";
    EbaySiteIdEnum[EbaySiteIdEnum["EBAY_YE"] = 257] = "EBAY_YE";
    EbaySiteIdEnum[EbaySiteIdEnum["EBAY_ZM"] = 258] = "EBAY_ZM";
    EbaySiteIdEnum[EbaySiteIdEnum["EBAY_ZW"] = 259] = "EBAY_ZW";
    EbaySiteIdEnum[EbaySiteIdEnum["EBAY_LS"] = 117] = "EBAY_LS";
    EbaySiteIdEnum[EbaySiteIdEnum["EBAY_YU"] = 222] = "EBAY_YU";
})(EbaySiteIdEnum = exports.EbaySiteIdEnum || (exports.EbaySiteIdEnum = {}));
;
