// @ebay/locale-formatter-ebay/lib/formatUtil.js
"use strict";

var minusChar = '-',
    radix = 10;


var currencyPatterns = {
    Nes : {
        pattern : "Nes",
        matcher : function (options) {
            return options.value + " " + options.symbol;
        }
    },
    Ns : {
        pattern : "Ns",
        matcher : function (options) {
            return options.value + options.symbol;
        }
    },
    Sen : {
        pattern : "Sen",
        matcher : function (options) {
            return (options.deferredMinus ? "-" : "") + options.symbol + " " + options.value;
        }
    },
    Sn : {
        pattern : "Sn",
        matcher : function (options) {
            return (options.deferredMinus ? "-" : "") + options.symbol + options.value;
        }
    },
    seN : {
        pattern : "seN",
        matcher : function (options) {
            return options.symbol + " " + options.value;
        }
    },
    sN : {
        pattern : "sN",
        matcher : function (options) {
            return options.symbol + options.value;
        }
    }
};

var datePatterns = {
    short : {
        pattern : "short",
        matcher : function (format) {
            return format.getShortDate();
        }
    },
    shortclean : {
        pattern : "shortclean",
        matcher : function (format) {
            return format.getShortCleanDate();
        }
    },
    medium : {
        pattern : "medium",
        matcher : function (format) {
            return format.getMediumDate();
        }
    },
    numeric : {
        pattern : "numeric",
        matcher : function (format) {
            return format.getNumericDate();
        }
    },
    numericnoday : {
        pattern : "numericnoday",
        matcher : function (format) {
            return format.getNumericNoDayDate();
        }
    },
    mediumnoday : {
        pattern : "mediumnoday",
        matcher : function (format) {
            return format.getMediumNoDayDate();
        }
    },
    long : {
        pattern : "long",
        matcher : function (format) {
            return format.getLongDate();
        }
    },
    full : {
        pattern : "full",
        matcher : function (format) {
            return format.getFullDate();
        }
    },
    fullComplete : {
        pattern : "short",
        matcher : function (format) {
            return format.getFullCompleteDate();
        }
    },
    fullnoyear : {
        pattern : "fullnoyear",
        matcher : function (format) {
            return format.getFullNoYearDate();
        }
    }
};

var timePatterns = {
    short : {
        pattern : "short",
        matcher : function (format) {
            return format.getShortTime();
        }
    },
    shortampm : {
        pattern : "shortampm",
        matcher : function (format) {
            return format.getShortAmPmTime();
        }
    },
    shortzone : {
        pattern : "shortzone",
        matcher : function (format) {
            return format.getShortZoneTime();
        }
    },
    shortampmzone : {
        pattern : "shortampmzone",
        matcher : function (format) {
            return format.getShortAmPmZoneTime();
        }
    },
    medium : {
        pattern : "medium",
        matcher : function (format) {
            return format.getMediumTime();
        }
    },

    long : {
        pattern : "long",
        matcher : function (format) {
            return format.getLongTime();
        }
    },
    longampm : {
        pattern : "longampm",
        matcher : function (format) {
            return format.getLongAmPmTime();
        }
    }
};

/**
* Internal method to perform long - decimal conversion, inserting decimal
* point and grouping characters as needed.
*
* @param val            the value
* @param dec            the number of decimal places
* @param options
*           groupingUsed   true if grouping used
*           grouping       the number of characters in group (usually 3)
*           groupingChar   the grouping separator (usually comma or space)
*           decimalChar    the decimal separator (usually period or comma)
*           signSuffix     true if sign is to be a suffix
* @return               the converted string
*/
function stringify(val, dec, options) {

    options = options || {};

    val = convertToInt(val, 0);
    dec = convertToInt(dec, 2);

    var grouping = convertToInt(options.grouping, 3),
        groupingUsed = options.groupingUsed || true,
        groupingChar = options.groupingChar || ',',
        decimalChar = options.decimalChar || '.',
        signSuffix = options.signSuffix || false,
        array = [],
        deferredMinus = false;

    if (val < 0) {
        if (signSuffix) {

            array.push(minusChar);
        } else {
            deferredMinus = true;
        }
    } else {
        val = -val;
    }

    if (dec > 0) {
        do {
            array.unshift(parseInt(-(val % radix), radix));
            val = parseInt(val / radix, radix);
        } while (--dec > 0);

        array.unshift(decimalChar);
    }

    var digits = 0;
    for (;;) {

        array.unshift(parseInt(-(val % radix), radix));
        val = parseInt(val / radix, radix);
        if (val === 0) {
            break;
        }
        if (groupingUsed && (++digits === grouping)) {

            array.unshift(groupingChar);
            digits = 0;
        }
    }
    if (deferredMinus) {

        array.unshift(minusChar);
    }
    return array.join('');
}


function isNumeric(num) {

    return !Array.isArray(num) && (num - parseFloat(num) + 1) >= 0;
}

function convertToInt(val, def) {
    try {
        var value =  parseInt(val, radix);
        if (!isNumeric(val) && isNaN(value)) {
            return def;
        } else {
            return value;
        }

    } catch (e) {
        return def;
    }
}

exports = module.exports = {
    stringify : stringify,
    isNumeric : isNumeric,
    currencyPatterns : currencyPatterns,
    datePatterns : datePatterns,
    timePatterns : timePatterns
};
