// @ebay/locale-formatter-ebay/lib/Money.js
"use strict";
var amplifier = 1000000, //to keep 6 decimal
	radix = 10,
	internalDecimal = 6,
	currencyConf = require('../config/currency/currency.json');

function Money(val, currencyCode, decimal) {
	//Fix the amount precision to 6
	val = val.toFixed(6);
	//Convert the value using the amplifier
	this._internalVal = this._convert(val);
	this._currencyCode = currencyCode;
	this._currency = currencyConf[currencyCode];
	this._decimal = decimal !== undefined ? decimal : this._currency.decimal;
}

Money.prototype = {

	_convert : function (val) {
		return this._round(val * amplifier);
	},
	_round : function (val) {
		return (val >= 0) ? Math.round(val) : (-Math.round(Math.abs(val)));
	},
	getCurrency : function () {
		return this._currency;
	},
	getDecimal : function () {
		return this._decimal;
	},
	getCurrencyCode : function () {
		return this._currencyCode;
	},
	getInternalVal : function () {
		return this._internalVal;
	},
	getInternalDecimal : function () {
		return internalDecimal;
	},
	getValueInMinorUnits : function () {
		var factor = amplifier;
		for (var i = this._decimal; i > 0; --i) {
			factor = parseInt(factor / radix, radix);
		}

		return this._round(this._internalVal/factor);
	},
	isNegative : function () {
		return this._internalVal < 0;
	},
	getRadix : function () {
		return radix;
	},
	getAmplifier : function () {
		return amplifier;
	}
};

exports = module.exports = Money;
