// @ebay/prometheus-ebay/middleware.js
'use strict';
const platformMetrics = require('./lib/platform');
const modConfig = require('@ebay/module-config-inc');
const { parseClientTLSVersionHeader } = require('@ebay/utils-ebay');

module.exports = function () {
    return (req, res, next) => {
        if (req.isQualyscan || isWarmUp(req)) {
            return next();
        }
        // this code block better be first as it records start time
        const startTime = Date.now();

        modConfig(module, (_, cfg) => {
            const promConfig = (cfg && cfg.get('prometheus')) || { enabled: true, interceptRequest: false };
            if (promConfig.enabled) {
                let onResponseFinish = [];
                if (promConfig.tlsClientInfo) {
                    // tls version metrics
                    const { version } = parseClientTLSVersionHeader(req);
                    onResponseFinish.push(() => {
                        const isSecureRequest = isSecure(req);
                        if (version && isSecureRequest) {
                            platformMetrics.c_clientTlsVersion.inc({
                                version,
                                '_namespace_': 'platform'
                            });
                        }
                        const proto = isSecureRequest && 'HTTPS' || 'HTTP';
                        platformMetrics.c_clientTlsVersion.inc({
                            version: proto,
                            '_namespace_': 'platform'
                        });
                    });
                }
                if (promConfig.interceptRequest) {
                    const config = req.route?.config;
                    const traceId = req.otel?.traceId;
                    const spanId = req.otel?.spanId;
                    const isRecording = req.otel?.isRecording;
                    onResponseFinish.push(() => {
                        if (config) {
                            const labels = {
                                method: req.method,
                                status: res.statusCode,
                                uri: config && config.path,
                                '_namespace_': 'platform',
                                commandName: config && (config.pageName || config.pagename) || 'Unknown'
                            };

                            const duration = Date.now() - startTime;
                            if (isRecording === true && traceId && spanId) {
                                platformMetrics.s_httpRequestHistogramDuration
                                    .observeWithExemplar({
                                        labels: labels,
                                        value: duration,
                                        exemplarLabels: { trace_id: traceId, span_id: spanId }
                                    });
                            }
                            else {
                                platformMetrics.s_httpRequestHistogramDuration
                                    .observeWithoutExemplar(labels, duration);
                            }
                        }
                    });
                }
                res.on('finish', finishResponse);
                res.on('close', finishResponse);
                res.on('error', finishResponse);
                // eslint-disable-next-line no-inner-declarations
                function finishResponse() {
                    onResponseFinish.forEach(fn => fn());
                    // act as once
                    onResponseFinish = [];
                }
            }

            next();
        });
    };
};

function isSecure(req) {
    const proto = req.headers['x-forwarded-proto'] || req.protocol;
    const port = getPort(req);
    const host = req.get && req.get('Host');
    return proto === 'https' || port === 8082 || /:8082$/.test(host);
}

function getPort(req) {
    return req.app && req.app.get('port') ||
        req.socket && req.socket.localPort || 0;
}

function isWarmUp(req) {
    return (req.query && req.query._warmup && req.query._warmup === 'true') || false;
}
