// @ebay/experimentation-ebay/lib/request-context/ServiceContext.js
'use strict';

const hmv = require('@ebay/header-multi-value-ebay');
const logger = require('@ebay/logging-inc').logger('experimentation-ebay/ServiceContext');

const CHANNEL = 'CHANNEL';
const OPTIN_FLAG = 'EP_OPTIN';
const USERSEGMENT_FLAG = 'USERSEGMENT';
const COOKILET_HEADER_OPTIN_FLAG = '321';

const Context = require('./Context');

class ServiceContext extends Context {
    constructor(req, res, config) {
        super(req, res, config);
        this.epCOSHdr = toHMV(this.req.get('x-ebay-c-exp'));
    }

    getOptedIntoFlag() {
         return this.optinFlag || this.epCOSHdr && this.epCOSHdr.get(OPTIN_FLAG);
    }

    isEmailChannel() {
        return this.epCOSHdr && this.epCOSHdr.get(CHANNEL) === 6;
    }
    
    setExperimentTreatements(experimentationCookieletHeader) {
        if (experimentationCookieletHeader && !this.res.headersSent) {
            this.optinFlag = toHMV(experimentationCookieletHeader).get(COOKILET_HEADER_OPTIN_FLAG);
            /* Send the X-EBAY-SVC-EP-COOKIELET to web app so it can persist in cookie */
            this.res.set('X-EBAY-SVC-EP-COOKIELET', experimentationCookieletHeader);
        }
    }

    getRemoteAddr() {
        return this.ebay.getEndUserContext().getIPAddress();
    }

    getReferer() {
        return this.ebay.getEndUserContext().getReferer();
    }

    getRequestURL() {
        return this.ebay.getEndUserContext().getApplicationURL();
    }

    getCookiesAndCookielets() {
        return null;
    }

    getUserSegment() {
        return this.epCOSHdr && this.epCOSHdr.get(USERSEGMENT_FLAG);
    }

    isAPoolEnabled() {
        return false;
    }

    createEPResponseHeader() {
        // TODO: https://github.corp.ebay.com/experimentation-platform/ep_for_raptor/blob/39e0d54ba9eeb3d29727d16cf7b9520dc7dcccb5/RaptorExperimentationParent/RaptorExperimentationAdaptor/src/main/java/com/ebay/ebox/ap/experimentation/pres/cmd/handler/v2/ExperimentationGingerServiceHandlerV2.java
    }
}

function toHMV(httpHeader) {
    httpHeader = httpHeader || '';
    try {
        return hmv.parse(httpHeader);
    } catch (e) {
        logger.error('Unable to parse Header into HMV objet: ', httpHeader, e);
        return hmv.parse('');
    }
}

module.exports = ServiceContext;
