// @ebay/experimentation-ebay/lib/request-context/WebContext.js
'use strict';

const hmv = require('@ebay/header-multi-value-ebay');
const logger = require('@ebay/logging-inc').logger('experimentation-ebay/ServiceContext');

const Context = require('./Context');
const OPTIN_FLAG = '321';
const GUEST_USER_FLAG = 'IS_GUEST_USER';
const ACCOUNT_ID_COOKIE = 'ACCOUNT_ID';
const BEST_GUESS_USER_COOKIE = 'USAGE';

class WebContext extends Context {
    constructor (req, res, config) {
        super (req, res, config);
    }

    getOptedIntoFlag() {
        return this.ebay && this.ebay.cookies && this.ebay.cookies.getCookieValue('EXPERIMENT_TREATMENTS');
    }

    setExperimentTreatements(experienceServiceCookieletHeader) {
        if (experienceServiceCookieletHeader) {
            this.ebay && this.ebay.cookies &&
                this.ebay.cookies.setCookieValue('EXPERIMENT_TREATMENTS',
                    extractOptedIntoFlag(experienceServiceCookieletHeader));
        }
    }

    getRemoteAddr() {
        return this.ebay.getRemoteAddr();
    }

    getReferer() {
        return this.req.headers && this.req.headers.referer && this.req.headers.referer.trim();
    }

    getRequestURL() {
        return this.ebay.requestUrl;
    }

    getCookiesAndCookielets() {
        const cookiesAndCookielets = {
            cookies: {},
            cookielets: {}
        };

        if (this.ebay.cookies && this.ebay.cookies.getCookieManager()) {
            const allCookies = this.ebay.cookies.getCookieManager().getAllCookies();
            allCookies && allCookies.reduce(reduce, cookiesAndCookielets);
        }

        cookiesAndCookielets.cookies = Object.keys(cookiesAndCookielets.cookies).reduce(function(acc, key) {
            const val = encodeURIComponent(cookiesAndCookielets.cookies[key]).replace(/\%20/g, '+');
            return acc +=  key + '=' +  val + ';';
        }, '');

        cookiesAndCookielets.cookielets = Object.keys(cookiesAndCookielets.cookielets).reduce(function(acc, key) {
            const val = encodeURIComponent(cookiesAndCookielets.cookielets[key]).replace(/\%20/g, '+');
            return acc +=  key + '=' +  val + ';';
        }, '');

        return cookiesAndCookielets;
    }

    getUserSegment() {
        const isGuest = this.ebay && this.ebay.cookies && this.ebay.cookies.hasFlag(GUEST_USER_FLAG);
        const isValidUserId = this.ebay && this.ebay.cookies && this.ebay.cookies.getCookieValue(ACCOUNT_ID_COOKIE);
        const isBestGuessUser = this.ebay && this.ebay.cookies && this.ebay.cookies.getCookieValue(BEST_GUESS_USER_COOKIE);

        if (isValidUserId && (isGuest && isGuest === true)) {
            return 'G';
        }

        if (isValidUserId && (!isGuest && isGuest === false)) {
            return 'U';
        }

       if (!isValidUserId && isBestGuessUser) {
           return 'BU';
       }

       return null;
    }

    isEmailChannel() {
        return this.req.emailChannel;
    }
}

function reduce(acc, cookie) {
    if (!cookie) {
        return acc;
    }

    if (cookie.meta && cookie.meta.unpacked) {
        acc.cookies[cookie.meta.name] = cookie.getUnpackedValue();
    } else if (cookie.meta && cookie.meta.cookielet) {
        acc.cookielets[cookie.meta.id] = cookie.getValue();
    } else { // must be a packed cookie
        var cookielets = cookie.getAllCookielets();
        cookielets && cookielets.reduce(reduce, acc);
    }

    return acc;
}


function extractOptedIntoFlag(experienceServiceCookieletHeader) {
    if (experienceServiceCookieletHeader) {
        try {
            var parsed = hmv.parse(experienceServiceCookieletHeader);
            return parsed && parsed.get(OPTIN_FLAG);
        } catch (e) {
            logger.error('Unable to get EP Cookielet Value from Header: ', experienceServiceCookieletHeader, e);
        }
    }
}


module.exports = WebContext;
