// @ebay/prometheus-ebay/lib/platform/index.js
'use strict';

const { Metrics } = require('../metrics');
const errors = require('@ebay/errors-ebay');
const modConfig = require('@ebay/module-config-inc');
const labelNames = ['method', 'status', 'uri', '_namespace_', 'commandName'];
const histogramBuckets = [5, 10, 20, 40, 80, 160, 320, 640, 1280, 2560, 5120, 10240];
let enabled = true;

class PlatformMetrics extends Metrics {
    constructor() {
        super('platform');

        modConfig(module, (err, cfg) => {
            if (!err) {
                this.histogramBuckets = cfg.get('prometheus:histogramBuckets') || histogramBuckets;
                enabled = cfg.get('prometheus:enabled');

                cfg && cfg.on('change', () => {
                    enabled = cfg.get('prometheus:enabled');
                });
            }
            this.createDefaultMetrics();

            process.on('monitorStats', (stats) => {
                if (enabled) {this.update(stats);}
            });
        });
    }

    isEnabled() {
        return enabled;
    }

    /**
     * if we (typically in a test) reset the platform metrics
     * a few of the metrics are fixed constants and only incremented at setup
     * After resetting all metrics, we re-calculate and set those constant values
     */
    resetMetrics() {
        super.resetMetrics();
        this.initConstants();
    }

    createDefaultMetrics() {
        this.promClient.collectDefaultMetrics({ register: this.registry });

        this.g_httpStatusCode = this.createGauge(
            'nodejs_http_status_code',
            'No. of HTTP Status Codes',
            ['status']
        );

        this.c_oomError = this.createCounter(
            'nodejs_oom_error_count_total',
            'Total number of Out of Memory Errors'
        );

        this.g_heapCrash = this.createGauge(
            'nodejs_heap_size_oom_trigger_bytes',
            'Node.js Heap Size which if surpassed will cause the process to crash'
        );

        this.g_noWorkers = this.createGauge(
            'nodejs_no_workers',
            'Number of Nodejs Workers running'
        );

        this.c_restarts = this.createCounter(
            'nodejs_process_restart_count_total',
            'Total Number of Node.js process Restarts'
        );

        this.g_transactionTime = this.createGauge(
            'nodejs_url_transaction_time',
            'URL Transaction Time'
        );

        this.g_uptime = this.createGauge(
            'nodejs_process_uptime',
            'Process Uptime'
        );

        this.g_errorCount = this.createGauge(
            'nodejs_error_transactions',
            'Number of error transactions'
        );

        this.g_tps = this.createGauge(
            'nodejs_tps',
            'Number of transactions'
        );

        this.g_gcCount = this.createGauge(
            'nodejs_gc_count',
            'Node.js GC Count'
        );

        this.g_gcOverhead = this.createGauge(
            'nodejs_gc_overhead',
            'Node.js GC Overhead'
        );

        this.g_heapPostGC = this.createGauge(
            'nodejs_heap_post_gc',
            'Node.js Heap Size Post GC'
        );

        this.s_httpRequestHistogramDuration = this.createHistogram(
            'http_server_requests_duration_ms',
            'Histogram duration of HTTP requests in ms',
            labelNames,
            this.histogramBuckets,
            true
        );

        this.c_clientTlsVersion = this.createCounter(
            'https_request_tls_version_total',
            'Client TLS version',
            ['version', '_namespace_'],
            false
        );

        this.g_startupTime = this.createGauge(
            'app_startup_time_ms',
            'Nodejs app init time metrics '
        );

        this.initConstants();
    }

    initConstants() {
        // read the restarts log file using errors-ebay lib
        this.c_restarts.inc(errors.getRestartLog().getRestarts());
        // read the oom log file using errors-ebay lib
        this.c_oomError.inc(errors.getOomLog().getOom());
        // we are a worker starting up, so we are 1 worker :)
        this.g_noWorkers.inc(1);
        // read the max heap size using errors-ebay lib
        this.g_heapCrash.set(errors.getMaxHeap());
    }

    update(stats = {}) {
        this.g_httpStatusCode.set({ status: '200' }, stats.http2XX || 0);
        this.g_httpStatusCode.set({ status: '300' }, stats.http3XX || 0);
        this.g_httpStatusCode.set({ status: '400' }, stats.http4XX || 0);
        this.g_httpStatusCode.set({ status: '500' }, stats.http5XX || 0);

        this.g_transactionTime.set(stats.urlTime || 0);
        this.g_uptime.set(stats.uptime || 0);
        this.g_errorCount.set(stats.errorCount || 0);
        this.g_tps.set(stats.tps || 0);

        this.g_gcCount.set(stats.gc_count || 0);
        this.g_gcOverhead.set(Math.round((stats.gcInterval || 0) * 100));
        this.g_heapPostGC.set(Math.round((stats.heapSizePostGC || 0) * 100));
    }
}

module.exports = new PlatformMetrics();
