// @ebay/openfeature-ebay/src/openfeature-lib/hooks/MetricsPublisherHook.ts
import {
    EvaluationDetails,
    FlagValue,
    Hook,
    HookContext,
} from '@openfeature/js-sdk';
import { MetricsPublisher } from '../metrics/MetricsPublisher';
import { getLogger, LogTransaction, Logger } from '@ebay/logging-inc';
/**
 * A hook that is used to set the EvaluationContext with required values
 * from request before flag evaluation.
 */
const NANOSECONDS_PER_SECOND: number = 1000000000;

export class MetricsPublisherHook implements Hook {
    startTimeInNanoSeconds: bigint;
    transaction: LogTransaction;
    logger: Logger;

    constructor(private readonly metricsPublisher: MetricsPublisher) {}
    before(): void {
        this.startTimeInNanoSeconds = process.hrtime.bigint();
        this.logger = getLogger('openfeature-ebay');
    }

    after(
        hookContext: HookContext,
        evaluationDetails: EvaluationDetails<FlagValue>,
    ): void {
        this.metricsPublisher.usageCounter.inc(
            { apiName: 'evaluation', status: 'success', techStack: 'nodejs' },
            1,
        );

        const currentTimeInNanoSeconds = process.hrtime.bigint(); // returns time in nanoseconds
        const nsDuration =
            currentTimeInNanoSeconds - this.startTimeInNanoSeconds;
        // converting to seconds to make it same as raptor and raptorio
        const secondsDuration: number =
            Number(nsDuration) / NANOSECONDS_PER_SECOND;
        this.metricsPublisher.epEvaluationLatencySla
            .labels('nodejs')
            .observe(secondsDuration);
    }

    finally(): void {}
}
