// @ebay/openfeature-ebay/dist/Api.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.getChannelId = exports.CHANNEL_ENUM = void 0;
exports.CHANNEL_ENUM = {
    WEB: 1,
    MOBILE_WEB: 2,
    TABLET_WEB: 3,
    ANDROID_APP: 4,
    IOS_APP: 5,
    EMAIL: 6,
    PLACEHOLDER1: 7,
    PLACEHOLDER2: 8,
    OTHER: 99,
};
function getChannelId(deviceInfo) {
    if (!deviceInfo) {
        return exports.CHANNEL_ENUM.OTHER;
    }
    if (deviceInfo.isFSOM) {
        return exports.CHANNEL_ENUM.WEB;
    }
    if (deviceInfo.requestIsWeb && deviceInfo.requestIsDesktop) {
        return exports.CHANNEL_ENUM.WEB;
    }
    if (deviceInfo.requestIsWeb && deviceInfo.requestIsMobileWeb) {
        return exports.CHANNEL_ENUM.MOBILE_WEB;
    }
    if (deviceInfo.requestIsWeb && /ios/i.test(deviceInfo.deviceOS)) {
        return exports.CHANNEL_ENUM.WEB;
    }
    if (deviceInfo.requestIsWeb && /android/i.test(deviceInfo.deviceOS)) {
        return exports.CHANNEL_ENUM.WEB;
    }
    if (!deviceInfo.requestIsWeb && /ios/i.test(deviceInfo.deviceOS)) {
        return exports.CHANNEL_ENUM.IOS_APP;
    }
    if (!deviceInfo.requestIsWeb && /android/i.test(deviceInfo.deviceOS)) {
        return exports.CHANNEL_ENUM.ANDROID_APP;
    }
    return exports.CHANNEL_ENUM.OTHER;
}
exports.getChannelId = getChannelId;
