// @ebay/cookies-ebay/lib/cookies/Cookie.js
'use strict';

function Cookie(name, value) {
    this._name = name;
    this._value = value;
    this._domain = null;
    this._path = '/';   //Default to the root path
    this._maxAge = -1;  //Default to cookies that last the length of the session
    this._secure = false;
    this._httpOnly = false;
    this._deleted = false;
    this._modified = false;
    this.serialize = null;
    this._sameSite = null;
    this._partitioned = null;
}
Cookie.prototype = {
    hasValue: function () {
        return this._value !== null && this._value !== undefined;
    },
    setName: function (name) {
        if (this._name !== undefined) {
            throw new Error('The name of a cookie cannot be change after it is created');
        }
        this._name = name;
    },
    setValue: function (value) {
        if (!value) {
            value = '';
        }
        if (this._value !== value) {
            this._value = value;
            this._modified = true;
        }
    },
    setDomain: function (domain) {
        this._domain = domain;
    },
    setPath: function (path) {
        this._path = path;
    },
    setMaxAge: function (maxAge) {
        this._maxAge = maxAge;
    },
    setMaxAgeDays: function (days) {
        var maxAge = days * 24 * 60 * 60;
        this.setMaxAge(maxAge);
    },
    setSameSite: function (sameSite) {
        this._sameSite = sameSite;
    },
    setSecure: function (secure) {
        this._secure = secure;
    },
    setHttpOnly: function (httpOnly) {
        this._httpOnly = httpOnly;
    },
    setPartitioned: function (partitioned) {
        this._partitioned = partitioned;
    },
    setModified: function (modified) {
        this._modified = modified;
    },
    setDeleted: function (deleted) {
        if (this._deleted !== deleted) {
            this._deleted = deleted;
            this._modified = true;
        }
    },
    getName: function () {
        return this._name;
    },
    getValue: function () {
        return this._value;
    },
    getDomain: function () {
        return this._domain;
    },
    getPath: function () {
        return this._path;
    },
    getMaxAge: function () {
        return this._maxAge;
    },
    getSameSite: function () {
        return this._sameSite;
    },
    isSecure: function () {
        return this._secure;
    },    
    isHttpOnly: function () {
        return this._httpOnly;
    },
    isPartitioned: function () {
        return this._partitioned === true;
    },
    isModified: function () {
        return this._modified === true;
    },
    isDeleted: function () {
        return this._deleted;
    }
};
module.exports = Cookie;
