// @ebay/openfeature-ebay/dist/metrics/MetricsPublisher.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.MetricsPublisher = void 0;
const prometheus_ebay_1 = require("@ebay/prometheus-ebay");
const EP_PROMETHEUS_NAMESPACE = 'epadaptor';
const EP_EVALUATION_REQUESTS_DURATION_METRIC_NAME = 'ep_evaluate_flag_hook_duration_ns_seconds';
const EP_API_BENCHMARK_DURATION_METRIC_NAME = 'ep_api_benchmark_duration_ns_seconds';
const EP_API_QUALIFICATION_DURATION_METRIC_NAME = 'ep_api_qualification_duration_ns_seconds';
const EP_API_CALL_TOTAL_METRIC_NAME = 'api_call_total';
const TECHSTACK_LABEL_KEY = 'techStack';
const API_NAME_LABEL_KEY = 'apiName';
const STATUS_LABEL_KEY = 'status';
const namespaces = require('@ebay/prometheus-ebay').namespaces[EP_PROMETHEUS_NAMESPACE];
// See Raptor-experimentation-component MicrometerExperimentationClientMetrics for reference
const MED_BUCKETS = [
    0.00001, 0.00005, 0.0001, 0.0005, 0.001, 0.005, 0.01, 0.05, 0.1, 0.5, 1.0,
    10.0, 100.0, 1000.0,
];
const BENCHMARK1 = [
    1.0, 5.0, 10.0, 25.0, 40.0, 50.0, 75.0, 90.0, 100.0, 150.0, 200.0, 250.0, 300.0, 350.0, 400.0, 450.0, 500.0, 750.0, 1000.0,
];
const BENCHMARK2 = [
    0.01, 0.025, 0.05, 0.075, 0.1, 0.25, 0.5, 0.75, 1.0, 5.0, 10.0, 50.0, 100.0,
    1000.0,
];
class MetricsPublisher {
    constructor() {
        if (namespaces) {
            this.metrics = namespaces;
        }
        else {
            this.metrics = new prometheus_ebay_1.Metrics(EP_PROMETHEUS_NAMESPACE);
        }
        this.usageCounter = this.metrics.createCounter(EP_API_CALL_TOTAL_METRIC_NAME, 'Total experimentation API calls', [API_NAME_LABEL_KEY, STATUS_LABEL_KEY, TECHSTACK_LABEL_KEY]);
        let benchmark1Bucket = convertBucketsToNS(BENCHMARK1);
        let benchmark2Bucket = convertBucketsToNS(BENCHMARK2);
        this.epEvaluationLatencySla = this.metrics.createHistogram(EP_EVALUATION_REQUESTS_DURATION_METRIC_NAME, 'ep_evaluate_flag_hook_duration_ns_seconds histogram', [TECHSTACK_LABEL_KEY], benchmark2Bucket);
        this.apiBenchmarkLatencySla = this.metrics.createHistogram(EP_API_BENCHMARK_DURATION_METRIC_NAME, 'ep_api_benchmark_duration_ns_seconds histogram', [TECHSTACK_LABEL_KEY], benchmark1Bucket);
        this.apiQualificationLatencySla = this.metrics.createHistogram(EP_API_QUALIFICATION_DURATION_METRIC_NAME, 'ep_api_qualification_duration_ns_seconds histogram', [TECHSTACK_LABEL_KEY], benchmark1Bucket);
    }
}
exports.MetricsPublisher = MetricsPublisher;
function convertBucketsToNS(bucket) {
    var convertedBucket = new Array(bucket.length);
    for (var val in bucket) {
        convertedBucket[val] = Number(bucket[val]) / 1000;
    }
    return convertedBucket;
}
