// @ebay/logging-inc/lib/config.js
'use strict';

var events = require('events');
var objutil = require('objutil');
var modConfig = require('@ebay/module-config-inc');
module.exports = new events.EventEmitter();
var LOG_LEVELS = module.exports.LOG_LEVELS = ['debug', 'info', 'warn', 'transaction', 'error', 'fatal'];

let config;
let configLoadInitiated = false;

//here, we set default value for 'warn', 'transaction', 'error', 'fatal'
const defaultSettings = { skips: [], names: [ /^.*?$/ ] };
var loglevels  = {
    warn: defaultSettings,
    transaction: defaultSettings,
    error: defaultSettings,
    fatal: defaultSettings
};

function loadConfig() {
    if (configLoadInitiated) {
        return;
    }
    configLoadInitiated = true;
    modConfig(module, (err, cfg) => {
        //for err handling, skiping as no error will be thrown 
        cfg.on('change', () => {
            configure();
        });

        config = cfg;
        configure();
    });
}

var configure = module.exports.configure = function configure(merge) {
    var configObject = config ? config.get('logging-inc') : {};
    var configuration = merge && objutil.mixin(merge, configObject) || configObject;
    loglevels = refreshLogLevels(configuration);
    module.exports.emit('change');
};

module.exports.setupLogLevels = name => {
    loadConfig();
    var levelSettings = {};
    LOG_LEVELS.forEach(function (level) {
        var rules = loglevels ? loglevels[level] : null;
        if (rules) {
            var match = rules.skips.some(function (re) {
                return !!re.test(name);
            });
            if (match) {
                return false;
            }
            levelSettings[level] = rules.names.some(function (re) {
                return !!re.test(name);
            });
        }
    });
    return levelSettings;
};

//export logLevels for unit test purpose
module.exports[Symbol.for('internals')] = {
    set logLevels(value) {
        loglevels = value;
    },
    get logLevels() {
        return loglevels;
    }
};

function parseLogLevel(levelValue) {
    var skips = [];
    var names = [];
    (levelValue || '')
        .split(/[\s,]+/)
        .forEach(function (name) {
            name = name.replace('*', '.*?');
            if (name[0] === '-') {
                skips.push(new RegExp('^' + name.substr(1) + '$'));
            } else {
                names.push(new RegExp('^' + name + '$'));
            }
        });

    return {
        skips: skips,
        names: names
    };
}

function refreshLogLevels(config) {
    var loglevel = {};
    config = config && Object.keys(config.loglevel || {}).length && config.loglevel || {
        warn: '*'
    };
    LOG_LEVELS.forEach(function (level) {
        var value = config[level] || '';
        if (process.env.LOG_LEVEL) {
            value += ' ' + process.env[level];
        }
        if (!value) {
            return;
        }
        loglevel[level] = parseLogLevel(String(value));
    });
    return loglevel;
}
