// @ebay/commons-ebay/lib/user-mixins/cookies.js
/**
 *
 * eBay user for Node.js
 *
 */
'use strict';
var CookieConstants = require('@ebay/cookies-ebay').Constants;
var logger = require('@ebay/logging-inc').getLogger('commons-ebay/user-mixins/user-ebay-cookies');

function decryptId(id) {
    /*
        bring 10 entities back into an array and base64 encode the result (makes result printable).

        exlusive or (xor) each 32 bit value with a 32 bit int starting at 0xa0b29120, incrementing by 0x04
            e.g. first xor value = 0xa0b29120, second = 0xa0b29124, third = 0xa0b29128, etc.

        rotate 10 entities right 2.

        remove leading '=' and convert to int
    */
    var bytes = new Buffer(id, 'base64');

    // exclusive or
    var xor = new Buffer([0xa0, 0xb2, 0x91, 0x20]);

    var cnt = 0;
    for (var i = 0; i < 10; i++)
    {
        for (var j = 0; j < 4; j++)
        {
            bytes[cnt] = (bytes[cnt] ^ xor[j]);
            cnt++;
        }
        xor[3] += 4;
    }
    // rotate right 2 entities
    var buffer = new Buffer(8);
    // buf.copy(targetBuffer, [targetStart], [sourceStart], [sourceEnd])
    // System.arraycopy(Object src, int srcPos, Object dest, int destPos, int length)
    bytes.copy(buffer, 0, 32, 40);
    bytes.copy(bytes, 8, 0, 32);
    buffer.copy(bytes, 0, 0, 8);

    var str = bytes.toString();
    var pos = str.lastIndexOf('=');
    str = str.substring(pos + 1);
    return parseInt(str, 10);
}

module.exports = {
    getCookie: function getCookie(cookieId) {
        var cookies = this.cookies;
        return cookies && cookies.getCookieValue(cookieId);
    },

    getPersistentAccountId: function getPersistentAccountId() {
        var accountId;

        if (this.user?.accountId && !Number.isNaN(parseInt(this.user.accountId, 10))) {
            return parseInt(this.user.accountId, 10);
        }

        var cookie = this.getCookie('PERSISTENT_ACCOUNT_ID');

        if (cookie) {
            // use conversion logic in EIAS
            try {
                accountId = decryptId(cookie);
            }
            catch (e) {
                logger.warn('parse persistent account id cookie failed: ', cookie);
            }
        }
        return accountId;
    },

    /**
     * This cookie holds Account Id for Visitor's e.g Light Weight(LW) accounts.
     * In Addition to this, IS_EMAIL_VISITOR cookie flag will have flag 
     */
    getPersistentVisitorAccountId: function getPersistentVisitorAccountId() {
        var accountId;
        var cookie = this.getCookie('ACTOR_ID');

        if (cookie && this.isEmailVisitor()) {
            // use conversion logic in EIAS
            try {
                accountId = decryptId(cookie);
            }
            catch (e) {
                logger.warn('parse persistent Visitor account id cookie failed: ', cookie);
            }
        }
        return accountId;
    },

    /**
     * Returns true/false is the user is email visitor(Light Weight account)
     */
    isEmailVisitor: function isEmailVisitor() {
        var cookies = this.cookies;
        return cookies && cookies.hasFlag('IS_EMAIL_VISITOR');
    },

    /**
     * getPersistentUserId - get user id from persistent user id cookie
     *
     * @return user ID, null value indicating no user id can be found
     **/
    getPersistentUserId: function getPersistentUserId() {
        if (this.user?.userId) {
            return this.user.userId;
        }

        return this.getCookie('PERSISTENT_USERID');
    },

    getSecurityTokenAccountId() {
        // first check if security token is required for this route
        // by checking req.ebay.security context (injected by auth-ebay module) is presnt
        return this.security && this.security.accountId;
    },

    getSecurityTokenUserId() {
        // first check if security token is required for this route
        // by checking req.ebay.security context (injected by auth-ebay module) is presnt
        return this.security && this.security.userId;
    },

    getAccountId: function getAccountId() {
        if (this.getSecurityTokenAccountId()) {
            return this.getSecurityTokenAccountId();
        }

        if (this.user?.accountId && !Number.isNaN(parseInt(this.user.accountId, 10))) {
            return parseInt(this.user.accountId, 10);
        }

        var accountIdStr = this.getCookie('ACCOUNT_ID'),
            accountId;
        try {
            accountId = accountIdStr ? parseInt(accountIdStr, 10) : null;
        }
        catch (e) {
            logger.warn('getAccountId() error: ', e);
        }
        // should never use origAccountId due to https://jirap.corp.ebay.com/browse/NODE-2864
        return accountId;
    },

    getDelegatorAccountId: function getDelegatorAccountId() {
        return this.multiAccountLoginCtx && this.multiAccountLoginCtx.getAcctId();
    },

    getDelegatorAccountName: function getDelegatorAccountName() {
        return this.multiAccountLoginCtx && this.multiAccountLoginCtx.getAcctName();
    },

    getLoginName: function getLoginName() {
        return this.multiAccountLoginCtx && this.multiAccountLoginCtx.getLoginName();
    },

    getLoginId: function getLoginId() {
        return this.multiAccountLoginCtx && this.multiAccountLoginCtx.getLoginId();
    },

    getOriginalAccountId: function getOriginalAccountId() {
        var ouiCtx = this.getEndUserContext && this.getEndUserContext().getOriginalUserIdContext();
        return ouiCtx ? ouiCtx.accountId : undefined;
    },


    /**
     * getLevel2UserIdUnstripped - get user id from level 2 user id cookie
     * including admin user identifier
     *
     * @return user ID including admin user id if any,
     * null value indicating no user id can be found
     */
    getLevel2UserIdUnstripped: function getLevel2UserIdUnstripped() {

        var userId = null;
        var cookie = this.getCookie('LEVEL2');
        //format is like "00567dbao6"
        if (cookie && cookie.length > 5) {
            userId = cookie.substring(5);
        }
        return userId;
    },

    /**
     * Answers whether user has admin userid
     * @return true if user has admin userid
     */
    hasAdminUserId: function hasAdminUserId(userId) {
        if (arguments.length === 0) {
            userId = this.getCookie('USERID');
        }

        return userId ? (userId.indexOf(CookieConstants.PIGGYBACK_SEPARATOR) > -1) : false;
    },

    /**
     * utility function to extract the admin userId from given string
     */
    getAdminUserId: function getAdminUserId(userId) {
        if (arguments.length === 0) {
            userId = this.getCookie('USERID');
        }

        if (userId) {
            var pos = userId.indexOf(CookieConstants.PIGGYBACK_SEPARATOR);
            if (pos > -1) {
                return userId.substring(pos + 1);
            }
        }
        return null;
    },

    /**
     * Answers if user is logged in as a guest
     * @return true if user id logged in as a guest
     */
    isLoggedInAsGuest: function isLoggedInAsGuest() {
        return !this.hasUserId() && this.hasGuestCookie();
    },

    /**
     * Check if user has the guest cookielet within the session cookie
     *
     */
    hasGuestCookie: function hasGuestCookie() {
        var cookie = this.getCookie('LEVEL0');
        return !!(cookie && parseInt(cookie, 10) === 1);
    },

    /**
     * hasUserId - check if there exists user id in cookie
     *
     * @return boolean
     */
    hasUserId: function hasUserId() {
        return !!this.getLevel1UserId();
    },

    /**
     * Check if there exists user id for level one page in cookie
     * @return true or false
     */
    hasLevel1UserId: function hasLevel1UserId() {
        if (this.user?.userId) {
            return true;
        }

        return this.hasUserId();
    },

    getLevel1UserId: function getLevel1UserId() {
        if (this.getSecurityTokenUserId()) {
            return this.getSecurityTokenUserId();
        }

        if (this.user?.userId) {
            return this.user.userId;
        }

        var userId = null;
        var userIdPersistent = null;
        userId = this.getCookie('USERID');
        userIdPersistent = this.getCookie('PERSISTENT_USERIDANDCHECK');

        if (userIdPersistent) {
            // strip of the checksum at the end
            userIdPersistent = userIdPersistent.substring(0,
                    userIdPersistent.length - 1);
        }

        // check to see if the session and nonsession users match
        // user id must match
        if (userId && userIdPersistent) {
            return userId.toLowerCase() === userIdPersistent.toLowerCase() ? userId : null;
        }

        if (userIdPersistent && !userId) {
            // no session users found return nonsession userid
            return userIdPersistent;
        }
        userId = userId || this.getPersistentUserId();

        // should never use origUserId due to https://jirap.corp.ebay.com/browse/NODE-2864
        return userId; // return session user if found
    },

    getOriginalUserId: function getOriginalUserId() {
        var ouiCtx = this.getEndUserContext && this.getEndUserContext().getOriginalUserIdContext();
        return ouiCtx ? ouiCtx.userId : undefined;
    },

    /**
     * Check if there exists user id for level two page in cookie
     * @return true or false
     */
    hasLevel2UserId: function hasLevel2UserId() {
        if (this.user?.userId) {
            return true;
        }

        return !!this.getLevel2UserId();
    },

    getLevel2UserId: function getLevel2UserId(raw) {
        if (this.user?.userId) {
            return this.user.userId;
        }

        var userId = null;
        var cookie = this.getCookie('LEVEL2');
        // format is like "00567dbao6"
        if (cookie && cookie.length > 5) {
            userId = cookie.substring(5);
            // Strip off admin user Id if present.
            userId = raw ? userId : this._getPiggybackUserId(userId);
        }

        return userId;
    },

    /**
     * getPresentationUserId - get user id for UI presentation use, such as
     * header personalized greeting, etc.
     *
     * The user id is looked for in following order;
     * 1 - Level one page cookie (session user id)
     * 2 - Sign in cookie (persistent user id)
     * 3 - JS persistent cookie user id
     *
     * @return user ID, null value indicating no user id can be found
     **/
    getPresentationUserId: function getPresentationUserId() {
        if (this.user?.userId) {
            return this.user.userId;
        }

        // Get  userid from level1 page
        // minus any piggybacked admin user id
        var userId = this.getUserId();
        if (!userId) {
            // if not found, try to get it from signin cookie
            userId = this.getUserIdSignIn();

            if (!userId) {
                // if not found, try to get it from JS persistent cookie
                userId = this.getJSPersistentUserId();
            }
        }
        return userId;
    },

    /**
     * getUserId - get user id from cookie for level one page
     *
     * @return user ID, null value indicating no user id can be found
     **/
    getUserId: function getUserId() {
        var userId = this.getLevel1UserId();

        if (!userId || userId.trim().length === 0) {
            return null;
        }

        return this._getPiggybackUserId(userId); // strip out admin if exists
    },

    /**
     * getPersistentUserId - get user id from JS persistent user id cookie
     *
     * @return user ID, null value indicating no user id can be found
     **/
    getJSPersistentUserId: function getJSPersistentUserId() {
        return this.getCookie('PERSISTENT_JS_USERID');
    },

    /**
     * Returns the signed in userid
     * @return sting the signed in userid
     */
    getUserIdSignIn: function getUserIdSignIn() {
        return this.getCookie('USERID_SIGNIN');
    },

    /**
     * Returns the user first name
     * @return string user's first name
     */
    getUserFirstName: function getUserFirstName() {
        return this.getCookie('PH_FIRST_NAME');
    },

    /**
     * Returns the level1 external user id
     * @return string
     */
    getLevel1ExtUserId: function getLevel1ExtUserId() {

        var fullCookie = this.getCookie('LEVEL1_EXTERNAL');
        if (!fullCookie) {
            return null;
        }

        var expiryIndex = fullCookie.indexOf(CookieConstants.L1_EXTERNAL_SEPARATOR);

        if (expiryIndex === -1) {
            return null;
        }

        return fullCookie.substring(0, expiryIndex);
    },

    /**
     * getLevel1CheckSum
     *
     * @return String one character long checksum if we're using the persistent cookie to identify the person.
     *   Null if it's not the persistant cookie being used
     **/
    getLevel1Checksum: function getLevel1Checksum() {
        var checksum = null;
        var userId = null;
        userId = this.getCookie('PERSISTENT_USERIDANDCHECK');
        if (userId && userId.length > 0) {
            // get the checksum
            checksum = userId.substring(userId.length - 1, userId.length);
        }
        return checksum;
    },

    getSignInWelcomeUsed: function getSignInWelcomeUsed() {
        var signInWelcome = this.getCookie('SIGNIN_WELCOME');
        return '1' === signInWelcome;
    },

    /**
     * getLevel2PageType - get page type from level 2 cookie
     *
     * @return PageType, -1 indicating no valid page type can be found
     **/
    getLevel2PageType: function getLevel2PageType() {
        var pageType = -1;
        var cookie = this.getCookie('LEVEL2');

        if (!cookie) {
            return pageType;
        }

        try {
            pageType = parseInt(cookie.substring(0, 5), 10);
        }
        catch (e) {
            logger.warn('getLevel2PageType() error: ', e);
        }

        return pageType;
    },

    /**
    * getSignInPrefFlags - get the flags which represent the user's signin prefs.
    *
    * @return long - sign in preferences
    **/
    getSignInPrefFlags: function getSignInPrefFlags() {
        var flags;

        var cookie = this.getCookie('SIGNINPREF');

        if (!cookie) {
            flags = 0;
        } else {
            try {
                flags = parseInt(cookie, 10);
            } catch (e) {
                flags = 0;
            }
        }

        return flags;
    },
    /**
     * utility function to extract the userId from given string
     */
    _getPiggybackUserId: function (userId) {
        var CookieConstants = this.cookies && this.cookies.Constants;
        if (userId && CookieConstants) {
            var pos = userId.indexOf(CookieConstants.PIGGYBACK_SEPARATOR);
            return (pos > -1) ? userId.substring(0, pos) : userId;
        }
        return null;
    },

    getBestUserId: function getBestUserId() {
        var acct = this.getRequest().headers.buid;
        if(acct) {
            return acct;
        }

        // first check if security token is required for this route
        // by checking req.ebay.security context (injected by auth-ebay module) is presnt
        if (this.getSecurityTokenAccountId()) {
            return this.getSecurityTokenAccountId();
        }

        acct = this.cookies.getCookieValue('ACCOUNT_ID');
        if (!acct) {
            var sawc = this.cookies.getCookieValue('SOJOURNER_SAW');
            if (sawc && sawc.indexOf('.') > 0 && sawc.length > 2) {
                acct = sawc.substring(2); //format is 1.acctId
            }
        }

        if (!acct) {
            /*
             * "Best Guess" always gets the first account in the usage cookie.
             * Assumptions is that the first account is the most recently active
             * account (which may or may not be true). Due to some bug in
             * sign-in code, number of days is always zero.
             */
            var usageAccounts = this.cookies.getCookieValue('USAGE');
            if (usageAccounts) {
                // format of the Usage cookie is "accountid/days; accountid/days;"
                acct = mostRecentlyLoggedinUser(usageAccounts);
            }
        }

        return acct;
    }
};

function mostRecentlyLoggedinUser(usageAccounts) {
    if(!usageAccounts) {
        return null;
    }
    if (usageAccounts.length === 0) {
        return null;
    }
    // Need to see how many userids are in the cookie
    // e.g "20836680/0;20903773/0;25359898/0;"
    // most recently logged in user is 25359898 which is at the end of the
    // string
    var arr = usageAccounts.split(';');
    var user = usageAccounts;
    for(var i=arr.length-1; i>=0; i--) {
        if(arr[i]) {
            user = arr[i];
            break;
        }
    }
    var users = user.split('/');
    return users[0];
}
