// @ebay/commons-ebay/lib/user-mixins/headers.js
/**
 *
 * eBay user for Node.js
 *
 */
'use strict';
var objPath = require('object-path');

module.exports = {
    /**
     * Answers whether user has admin userid
     * @return true if user has admin userid
     */
    hasAdminUserId() {
        return !!this.getAdminUserId();
    },

    /**
     * utility function to extract the admin userId from given string
     */
    getAdminUserId() {
        return this.getEndUserContext && this.getEndUserContext().getAdminUserId();
    },

    getBestUserId: function getBestUserId() {
        return this.getCorrelationTracking().getBestUserId() ||
            this.getAccountId();
    },

    getPublicUserId: function getPublicUserId() {
        var req = this.getRequest();
        var tokenAttributes = req.security && req.security.getTokenAttributes();
        const domain = objPath.get(tokenAttributes, 'domain');
        return domain === 'EBAYUSER' ? objPath.get(tokenAttributes, 'userId') : undefined;
    },

    getPublicAccountId: function getPublicAccountId() {
        var req = this.getRequest();
        var tokenAttributes = req.security && req.security.getTokenAttributes();
        const domain = objPath.get(tokenAttributes, 'domain');
        return domain === 'EBAYUSER' ? objPath.get(tokenAttributes, 'accountId') : undefined;
    },

    getPersistentAccountId: function getPersistentAccountId() {
        return this.getAccountId();
    },

    /**
     * getPersistentUserId - get public user id
     *
     * @return user ID, null value indicating no user id can be found
     **/
    getPersistentUserId: function getPersistentUserId() {
        return this.getUserId() || this.getPublicUserId();
    },

    /**
     * This cookie holds Account Id for Visitor's e.g Light Weight(LW) accounts.
     * In Addition to this, IS_EMAIL_VISITOR cookie flag will have flag 
     */
    getPersistentVisitorAccountId: function getPersistentVisitorAccountId() {
        var uiCtx = this.getEndUserContext && this.getEndUserContext().getUserIdContext();
        return uiCtx ? uiCtx.acctId : undefined;
    },

    /**
     * Returns true/false is the user is email visitor(Light Weight account)
     */
    isEmailVisitor: function isEmailVisitor() {
        var uiCtx = this.getEndUserContext && this.getEndUserContext().getUserIdContext();
        return uiCtx && uiCtx.acctRole;
    },

    getAccountId: function getAccountId() {
        return this.getOriginalAccountId() || this.getPublicAccountId();
    },

    getOriginalAccountId: function getOriginalAccountId(level1Only) {
        var ouiCtx = this.getEndUserContext && this.getEndUserContext().getOriginalUserIdContext();
        if (ouiCtx && ouiCtx.recognized && level1Only) {
            return undefined;
        }
        return ouiCtx ? ouiCtx.accountId : undefined;
    },

    /**
     * Answers if user is logged in as a guest
     * @return true if user id logged in as a guest
     */
    isLoggedInAsGuest: function isLoggedInAsGuest() {
        return !this.hasUserId();
    },

    /**
     * hasUserId - check if user id exists
     *
     * @return boolean
     */
    hasUserId: function hasUserId() {
        return !!this.getLevel1UserId(true);
    },

    /**
     * Check if there exists user id
     * @return true or false
     */
    hasLevel1UserId: function hasLevel1UserId() {
        return this.hasUserId();
    },

    getLevel1UserId: function getLevel1UserId() {
        return this.getOriginalUserId(true) || this.getPublicUserId();
    },

    getOriginalUserId: function getOriginalUserId(level1Only) {
        var ouiCtx = this.getEndUserContext && this.getEndUserContext().getOriginalUserIdContext();
        if (ouiCtx && ouiCtx.recognized && level1Only) {
            return undefined;
        }
        return ouiCtx ? ouiCtx.userId : undefined;
    },

    getDelegatorAccountId: function getDelegatorAccountId() {
        var uiCtx = this.getEndUserContext && this.getEndUserContext().getUserIdContext();
        return uiCtx ? uiCtx.acctId : undefined;
    },

    getDelegatorAccountName: function getDelegatorAccountName() {
        var uiCtx = this.getEndUserContext && this.getEndUserContext().getUserIdContext();
        return uiCtx ? uiCtx.acctName : undefined;
    },

    getLoginId: function getLoginId() {
        var uiCtx = this.getEndUserContext && this.getEndUserContext().getUserIdContext();
        return uiCtx ? uiCtx.loginId : undefined;
    },

    getLoginName: function getLoginName() {
        var uiCtx = this.getEndUserContext && this.getEndUserContext().getUserIdContext();
        return uiCtx ? uiCtx.loginName : undefined;
    },

    /**
     * getPresentationUserId - get user id for UI presentation use, such as
     * header personalized greeting, etc.
     *
     * The user id is looked for in following order;
     * 1 - get user id from origUserId context
     * 2 - get user id from scoped token
     *
     * @return user ID, null value indicating no user id can be found
     **/
    getPresentationUserId: function getPresentationUserId() {
        // Get  userid from level1 page
        return this.getUserId() || this.getJSPersistentUserId();
    },

    /**
     * getUserId - get user id from origUserId context
     *
     * @return user ID, null value indicating no user id can be found
     **/
    getUserId: function getUserId() {
        return this.getLevel1UserId(true);
    },

    /**
     * getPersistentUserId - get user id from public scoped user token
     *
     * @return user ID, null value indicating no user id can be found
     **/
    getJSPersistentUserId: function getJSPersistentUserId() {
        return this.getPublicUserId();
    }

};
