// @ebay/openfeature-ebay/src/openfeature-lib/openfeature-context/ServiceContext.ts
import hmv from '@ebay/header-multi-value-ebay';
import { Context } from './Context';

const CHANNEL = 'CHANNEL';
const OPTIN_FLAG = 'EP_OPTIN';
const COOKILET_HEADER_OPTIN_FLAG = '321';

export class ServiceContext extends Context {
    epCOSHdr: any;
    constructor(req, res, openfeatureGlobalCache) {
        super(req, res, openfeatureGlobalCache);
        this.epCOSHdr = this.toHMV(this.req.get('x-ebay-c-exp'));
    }

    getOptedIntoFlag() {
        return (
            this.optinFlag || (this.epCOSHdr && this.epCOSHdr.get(OPTIN_FLAG))
        );
    }

    isEmailChannel() {
        return this.epCOSHdr && this.epCOSHdr.get(CHANNEL) === 6;
    }

    setExperimentTreatements(experimentationCookieletHeader: string) {
        if (experimentationCookieletHeader && !this.res.headersSent) {
            this.optinFlag = this.toHMV(experimentationCookieletHeader).get(
                COOKILET_HEADER_OPTIN_FLAG,
            );
            /* Send the X-EBAY-SVC-EP-COOKIELET to web app so it can persist in cookie */
            this.res.set(
                'X-EBAY-SVC-EP-COOKIELET',
                experimentationCookieletHeader,
            );
        }
    }

    getClientIP() {
        return this.ebay.getEndUserContext().getIPAddress();
    }

    getReferer() {
        return this.ebay.getEndUserContext().getReferer();
    }

    // Not applicable for service ctx so providing a default function
    getCookiesAndCookielets() {
        return null;
    }

    toHMV(httpHeader) {
        httpHeader = httpHeader || '';
        try {
            return hmv.parse(httpHeader);
        } catch (e) {
            console.error(
                'Unable to parse Header into HMV objet: ',
                httpHeader,
                e,
            );
            return hmv.parse('');
        }
    }

    getCountryCode() {
        return this.ebay.getEndUserContext().getUserContextualLocation().getCountry();
    }

}
