// @ebay/openfeature-ebay/src/openfeature-lib/openfeature-context/WebContext.ts
import { Context } from './Context';
import { parse } from '@ebay/header-multi-value-ebay';

const OPTIN_FLAG = '321';

export class WebContext extends Context {
    constructor(req, res, openfeatureGlobalCache) {
        super(req, res, openfeatureGlobalCache);
    }

    getOptedIntoFlag() {
        return (
            this.ebay &&
            this.ebay.cookies &&
            this.ebay.cookies.getCookieValue('EXPERIMENT_TREATMENTS')
        );
    }

    setExperimentTreatements(experienceServiceCookieletHeader: string) {
        if (experienceServiceCookieletHeader) {
            this.ebay &&
                this.ebay.cookies &&
                this.ebay.cookies.setCookieValue(
                    'EXPERIMENT_TREATMENTS',
                    this.extractOptedIntoFlag(experienceServiceCookieletHeader),
                );
        }
    }

    getCookiesAndCookielets() {
        const cookiesAndCookielets = {
            cookies: {},
            cookielets: {},
        };

        cookiesAndCookielets.cookies = Object.keys(
            cookiesAndCookielets.cookies,
        ).reduce((acc, key) => {
            const val = encodeURIComponent(
                cookiesAndCookielets.cookies[key],
            ).replace(/\%20/g, '+');
            return (acc += key + '=' + val + ';');
        }, '');

        cookiesAndCookielets.cookielets = Object.keys(
            cookiesAndCookielets.cookielets,
        ).reduce((acc, key) => {
            const val = encodeURIComponent(
                cookiesAndCookielets.cookielets[key],
            ).replace(/\%20/g, '+');
            return (acc += key + '=' + val + ';');
        }, '');

        return cookiesAndCookielets;
    }

    getClientIP() {
        return this.ebay.getClientIP();
    }

    getReferer() {
        return (
            this.req.headers &&
            this.req.headers.referer &&
            this.req.headers.referer.trim()
        );
    }

    isEmailChannel() {
        return this.req.emailChannel;
    }

    extractOptedIntoFlag(experienceServiceCookieletHeader) {
        if (experienceServiceCookieletHeader) {
            try {
                const parsed = parse(experienceServiceCookieletHeader);
                return parsed && parsed.get(OPTIN_FLAG);
            } catch (e) {
                console.error(
                    'Unable to get EP Cookielet Value from Header: ',
                    experienceServiceCookieletHeader,
                    e,
                );
            }
        }
    }

    getCountryCode() {
        return this.req && this.req.locality && this.req.locality.country;
    }
}
