// @ebay/telemetry-ebay/lib/metrics.js
const os = require('os');
const path = require('path');
const ip = require('ip');
const fs = require('fs');
const httpunch = require('@ebay/httpunch');
const debug = require('debug')('telemetry')
const indexName = 'nodejs_metrics';

class Metrics {
    getCommonMetrics() {
        let pkg = {};
        let ownerAndTeamDL = [];
        let gitRepo, cachedBuild = true;
        try {
            pkg = require(path.resolve(process.cwd(), 'package.json'));
            gitRepo = pkg['repository'] && pkg['repository'].url || 'Unknown';
            let ebayProps = pkg.gpaas || pkg.ebay;
            if (ebayProps) {
                ebayProps.owner && ownerAndTeamDL.push(ebayProps.owner);
                ebayProps['team-dl'] && ownerAndTeamDL.push(ebayProps['team-dl']);
            }
        } catch(err) {
            debug(`Error in parsing: ${err}`);
        }

        if(fs.existsSync(path.join(process.cwd(), 'yarn.lock'))) {
            const timeStats = fs.statSync(path.join(process.cwd(), 'yarn.lock'));
            if(Date.now() - timeStats.birthtime < 5*60*1000) { // 5 min
                cachedBuild = false;
            }
        }

        const re = new Date().toString().match(/\((.+)\)/);
        const timeZone = re && re.length > 1 ? re[1] : 'Unknown';

        let metrics = {};
        metrics.hostName = os.hostname();
        metrics.os = `${os.type()} ${os.release()} ${os.arch()}`;
        metrics.nodeVersion = process.version;
        metrics.ip = ip.address();
        metrics.cachedBuild = cachedBuild;
        metrics.timeZone = timeZone;
        metrics.appName = pkg.name || 'Unknown';
        metrics.totalModules = 0;
        metrics.buildUrl = process.env.BUILD_URL;
        metrics.jobName = process.env.JOB_NAME;
        metrics.homeDir = process.cwd();
        metrics['env.ecdx.pipelineId'] = process.env.ECDX_PIPELINE_ID;
        metrics['env.ecdx.pipelineRunId'] = process.env.ECDX_RUN_ID;
        metrics['env.ecdx.phaseId'] = process.env.ECDX_PHASE_ID
        try {
            metrics.userid = os.userInfo().username;
        } catch (error) {
            //no op
            metrics.userid = 'root';
        }
        metrics.gitRepo = gitRepo;
        metrics.ownerAndTeamDL = ownerAndTeamDL;
        metrics.timestamp = new Date();
        return metrics;
    }

    publishToES(data) {
        const request = {
            url: `https://eszeus-dataslc.stratus.qa.ebay.com/${indexName}/_doc/${Date.now()}`,
            method: 'PUT',
            headers: {
                'content-type': 'application/json',
                'Authorization': 'Basic OWMyMjU4ZjFiZWEwNDM3OWE1YjQwOGNjNDZjZjYyYjI6Z2lZdlZuRUp3bzNPMkZSUjlCZkQ2OTBNeE1uaDJETlEyMWd1U09KNVFJTnlURExxS2V6S2ZnRnpPSEpoSmpBdg=='
            },
            body: JSON.stringify(data),
            socketTimeout: 3000,
            rejectUnauthorized: false
        };

        debug(`request: ${JSON.stringify(request)}`);

        httpunch.request(request, (err, res) => {
            if(err) {
                debug(err);
            } else {
                debug(res.statusCode);
            }
        });
    }
}

module.exports = Metrics;
