// @ebay/security-ebay/lib/index.js
'use strict';
var unique = require('array-unique');
var moduleconfig = require('@ebay/module-config-inc');
var url = require('url');
var qs = require('qs');
var defaultDomains = require('../config/domains').domains;
var defaulteBayDomains = require('../config/domains').eBayDomains || [];
var localWhiteList = require('../config/domains').WhitelistPerAppname;
var domains;
var eBayDomains = [];
var defaultNativeDomains = require('../config/native-domains');
var nativeDomains;
const VALID_DOMAINS_MAX_CACHE_SIZE = 1000;
const VALID_NATIVE_DOMAINS_MAX_CACHE_SIZE = 300;
const allowEbayDomainsRegex = new RegExp(
    `\\.ebay\\.(com|${eBayDomains.map(domain => domain.replace('.ebay.', '').replace('.', '\\.')).join('|')})$`,
    'gi'
);
let validatedDomains = new Set();
let validatedNativeDomains = new Set();

let config;

function getDomainListPerApp(appName,cfgWhitelistedAppDomains){
    let domainsPerAppList = [];
    if(appName in cfgWhitelistedAppDomains){
        domainsPerAppList = cfgWhitelistedAppDomains[appName].split(',');
        domainsPerAppList = domainsPerAppList.map(domainPerApp => domainPerApp.trim());
        return domainsPerAppList;
    }
    else{
        return null;
    }
}

function isSameOrigin(userHost, ajaxHost) {
    const ajaxDomain = ajaxHost.match(allowEbayDomainsRegex) || [];
    const userDomain = userHost.match(allowEbayDomainsRegex) || [];
    return ajaxDomain && userDomain && ajaxDomain[0] === userDomain[0];
}

/**
 * The function validate only native ebay domains that belong to ebay web domains, ex: www.ebay.com
 * @param {*} host
 * @returns
 */
function isValidNativeEbayDomain(host) {
    if (validatedNativeDomains.has(host)) {
        return true;
    }
    for (let index = 0; index < defaultNativeDomains.length; index++) {
        const domain = nativeDomains[index];
        if (host.indexOf(domain, host.length - domain.length) > -1) {
            // allow to cache VALID_NATIVE_DOMAINS_MAX_CACHE_SIZE hosts
            if (validatedNativeDomains.size > VALID_NATIVE_DOMAINS_MAX_CACHE_SIZE) {
                // delete the oldest inserted
                validatedNativeDomains.delete(validatedNativeDomains.keys().next().value);
            }
            validatedNativeDomains.add(host);
            return true;
        }
    }
    return false;
}
/**
 * The function validate only native ebay domains that belong to ebay tld domains
 * @param {*} host
 * @returns
 */

function isValidEbayDomain(host) {
    if (validatedDomains.has(host)) {
        return true;
    }
    for (let index = 0; index < domains.length; index++) {
        const domain = domains[index];
        if (host.indexOf(domain, host.length - domain.length) > -1) {
            // allow to cache VALID_DOMAINS_MAX_CACHE_SIZE hosts
            if (validatedDomains.size > VALID_DOMAINS_MAX_CACHE_SIZE) {
                // delete the oldest inserted
                validatedDomains.delete(validatedDomains.keys().next().value);
            }
            validatedDomains.add(host);
            return true;
        }
    }
    return false;
}

/**
 * Check if its a valid ebay site domain
 * @param { String } host
 * @returns Boolean
 */
function isValidEbaySiteDomain(host) {
    if(!host || eBayDomains.length === 0) {
        return false;
    }
    if (eBayDomains.includes(host)) {
        return true;
    } else {
        return eBayDomains.find(eBayDomain => {
            return host.endsWith(eBayDomain);
        });
    }
}


function updateDomainList() {
    return new Promise(resolve => {
        moduleconfig(module, (err, cfg) => {
            //for err handling, skiping as no error will be thrown
            config = cfg;
            config.on('change', refreshDomainList);
            refreshDomainList();
            resolve(config);

            function refreshDomainList() {
                var cfgDomains = config ? config.get('security-ebay:domains') : [];
                var cfgeBayDomains = config && config.get('security-ebay:eBayDomains') ? config.get('security-ebay:eBayDomains') : [];
                var cfgNativeDomains = config ? config.get('security-ebay:native-domains') : [];
                var cfgWhitelistedAppDomains = config ? config.get('security-ebay:WhitelistPerAppname') : [];
                const appName = require('@ebay/app-context-ebay').appName;
                domains = unique(defaultDomains.concat(cfgDomains));
                eBayDomains = unique(defaulteBayDomains.concat(cfgeBayDomains));
                nativeDomains = unique(defaultNativeDomains.concat(cfgNativeDomains));
                if(appName in localWhiteList){
                    domains = unique(domains.concat(localWhiteList[appName]));
                }
                let domainListPerApp = getDomainListPerApp(appName,cfgWhitelistedAppDomains);
                if(domainListPerApp !== null){
                    domains = unique(domains.concat(domainListPerApp));
                }
                // reset validated domains
                validatedDomains = new Set();
            }
        });
    });
}

const pendingLoad = updateDomainList();

var sensitiveParamsMap;

function tryParseUrl(url) {
    try {
        return new URL(url);
    }
    catch (err) {
        return url;
    }
}

function isValidRedirect(redirectUrl) {
    if (config && !config.get('security-ebay:enabled')) {
        return true;
    }
    if (redirectUrl !== undefined) {
        if (redirectUrl.startsWith('ebay://')) {
            return true;
        }
        // Check if the URL starts with a backslash or a forward slash followed by a backslash
        if (redirectUrl.startsWith('\\') || redirectUrl.startsWith('/\\')) {
            return false;
        }
        redirectUrl = redirectUrl.startsWith('//') ? 'http:' + redirectUrl : redirectUrl;
        let ruWOBackslash = redirectUrl.replace('\\', '');

        // Replace backslashes with forward slashes
        let rwWithBackslach = redirectUrl.replace(/\\/g, '/');

        return parseAndValidate(rwWithBackslach) && parseAndValidate(ruWOBackslash);

    }
    return true;
}

function parseAndValidate(redirectUrl) {
    const parsedUrl = tryParseUrl(redirectUrl);
    // If URL pattern includes username or password, it is not a valid URL
    // e.g http://username:password@host:port/path
    // We only expect http[s]://host:port/path
    // Fixed part of https://jirap.corp.ebay.com/browse/NODE-5424
    if(parsedUrl.username || parsedUrl.password){
        return false;
    }

    var host = parsedUrl.hostname;
    if (!host) {
        if (redirectUrl.startsWith('/')) {
            return true;
        } else {
            return false;
        }
    }

    for (let index = 0; index < domains.length; index++) {
        const domain = domains[index].startsWith('.') ? domains[index].substring(1) : domains[index];
        if (host === domain || host.endsWith('.' + domain)) {
            return true;
        }
    }
    return false;
}

function isValidDomain(redirectUrl) {
    console.log('isValidDomain will be deprecated in the next version, please use isValidRedirect');
        return isValidRedirect(redirectUrl);
}

function getMaskPayload(req) {
    var reqUrl = req.originalUrl || req.url || req.path;
    var payload = url.parse(reqUrl).pathname;

    if (!config || !req.query) {
        return payload;
    }

    payload = payload + (Object.keys(req.query).reduce(function reduce(memo, key) {
        memo = memo ? memo += '&' : '?';
        var val = req.query[key];
        if(isSensitiveParameter(key)) {
          val = val ? '***' : '';
        }
        return memo + key + '=' + val;
    }, ''));

    return payload;
}

function isSensitiveParameter(name) {
    if (!name) {
        return false;
    }
    sensitiveParamsMap = sensitiveParamsMap || config && arrayToMap(config.get('security-ebay:sensitive-paramaters'));
    return sensitiveParamsMap && sensitiveParamsMap[name] || false;
}

function maskObject(payload) {
    payload = payload || {};

    if (!config) {
        return qs.stringify(payload);
    }

    return (Object.keys(payload).reduce(function reduce(memo, key) {
        memo = memo ? memo += '&' : '';
        var val = payload[key];
        if(isSensitiveParameter(key)) {
          val = val ? '***' : '';
        }
        return memo + key + '=' + val;
    }, ''));
}

function arrayToMap(arr) {
    var map = arr.reduce(function(accumulator, element){
      accumulator[element] = true;
      return accumulator;
    }, {});
    return map;
}

module.exports.isValidDomain = isValidDomain;
module.exports.isValidNativeEbayDomain = isValidNativeEbayDomain;
module.exports.isValidRedirect = isValidRedirect;
module.exports.getMaskPayload = getMaskPayload;
module.exports.maskObject = maskObject;
module.exports.isSensitiveParameter = isSensitiveParameter;
module.exports.getDomainListPerApp = getDomainListPerApp;
module.exports.isValidEbayDomain = isValidEbayDomain;
module.exports.isValidEbaySiteDomain = isValidEbaySiteDomain;
module.exports.isSameOrigin = isSameOrigin;
Object.defineProperty(module.exports, 'validatedDomains', {
    get() {
        return validatedDomains;
    }
});
Object.defineProperty(module.exports, 'validatedNativeDomains', {
    get() {
        return validatedNativeDomains;
    }
});
module.exports[Symbol.for('internals')] = {
    pendingLoad
};
