// @ebay/iaf-ebay/lib/iaf.js
'use strict';

var assert = require('assert');

var services = require('./services');

/**
 * @param options:
 * {
 *    token: <token>,
 *    consumerId: <consumer id>,
 *    secret: <secret>
 * }
*/
module.exports.authenticateToken = function authenticateToken(options, callback) {

    assert.ok(options, 'options must be provided');
    assert.ok(typeof callback === 'function', 'callback must be provided');
    options.secret = options.appSecret || options.secret;
    assert.ok(options.token, 'token must be provided');
    assert.ok(options.consumerId, 'consumerId must be provided');
    assert.ok(options.secret, 'secret must be provided');

    services.login(options.consumerId,
        options.secret, function getClient(err, client) {
        if (err) {
            return callback(err);
        }
        client.authenticateToken(options, callback);
    });
};

/**
 * @param options:
 * {
 *    consumerId: <consumer id>,
 *    secret: <secret>
 * }
*/
module.exports.getAsacToken = function (options, callback) {
    this.getAppToken(Object.assign({
        clientId: 'asac-iaf-service'
    }, options), callback);
};

/**
 * @param options:
 * {
 *    consumerId: <consumer id>,
 *    secret: <secret>
 * }
*/
module.exports.getAppToken = function getAppToken(options, callback) {
    assert.ok(options, 'options must be provided');
    assert.ok(typeof callback === 'function', 'callback must be provided');
    options.secret = options.appSecret || options.secret;
    assert.ok(options.consumerId, 'consumerId must be provided');
    assert.ok(options.secret, 'secret must be provided');
    options.domain = options.domain || 'EBAYAPP';

    services.login(options.consumerId,
        options.secret, function getClient(err, client) {
        if (err) {
            return callback(err);
        }

        client.requestSecurityToken(options, callback);
    });
};

/**
 * @param options:
 * {
 *    consumerId: <consumer id>,
 *    secret: <secret>,
 *    userId: <user id>,
 *    accountId: <account id>,
 *    scope: <scope>,
 *    type: <type> // ajax or Hosted
 * }
*/
module.exports.getUserToken = function getUserToken(options, callback) {
    assert.ok(options, 'options must be provided');
    assert.ok(typeof callback === 'function', 'callback must be provided');
    options.secret = options.appSecret || options.secret;
    assert.ok(options.consumerId, 'consumerId must be provided');
    assert.ok(options.secret, 'secret must be provided');
    assert.ok(options.userId, 'userId must be provided');
    assert.ok(options.accountId, 'accountId must be provided');
    options.domain = options.domain || 'EBAYUSER';

    services.login(options.consumerId,
        options.secret, function getClient(err, client) {
        if (err) {
            return callback(err);
        }

        client.requestSecurityToken(options, callback);
    });
};

/**
 * @param options can be string reference or options:
 * {
 *    reference: <reference>,
 *    consumerId: <consumer id>,
 *    secret: <secret>
 * }
*/
module.exports.getSecurityTokenByRef = function getSecurityTokenByRef(options, callback) {
    assert.ok(options, 'options must be provided');
    assert.ok(typeof callback === 'function', 'callback must be provided');
    options.secret = options.appSecret || options.secret;
    assert.ok(options.consumerId, 'consumerId must be provided');
    assert.ok(options.secret, 'secret must be provided');
    assert.ok(options.reference, 'token reference must be provided');

    services.login(options.consumerId,
        options.secret, function getClient(err, client) {
        if (err) {
            return callback(err);
        }

        client.getSecurityTokenByRef(options.reference, callback);
    });
};

/**
 * @param options can be string reference or options:
 * {
 *    consumerId: <consumerId>,
 *    secret: <secret>,
 *    reference: <reference>,
 *    subject: <subject>,
 *    requester: <requester>
 * }
*/
module.exports.logout = function logout(options, callback) {
    assert.ok(options, 'options must be provided');
    assert.ok(typeof callback === 'function', 'callback must be provided');
    options.secret = options.appSecret || options.secret;
    assert.ok(options.consumerId, 'consumerId must be provided');
    assert.ok(options.secret, 'secret must be provided');
    assert.ok(options.requester, 'requester must be provided');
    assert.ok(options.reference, 'token reference must be provided');
    options.domain = options.domain || 'EBAYUSER';

    services.login(options.consumerId,
        options.secret, function getClient(err, client) {
        if (err) {
            return callback(err);
        }

        client.logout(options, callback);
    });
};

/**
 * @param options:
 * {
 *    consumerId: <consumer id>,
 *    secret: <secret>,
 *    userId: <user id>,
 *    userSecret: <userSecret>,
 *    scope: <scope>,
 *    type: <type> // ajax or Hosted
 * }
*/
module.exports.authenticateUser = function authenticateUser(options, callback) {
    assert.ok(options, 'options must be provided');
    assert.ok(typeof callback === 'function', 'callback must be provided');
    options.secret = options.appSecret || options.secret;
    assert.ok(options.consumerId, 'consumerId must be provided');
    assert.ok(options.secret, 'application secret must be provided');
    assert.ok(options.userId, 'userId must be provided');
    assert.ok(options.userSecret, 'user secret must be provided');
    options.domain = options.domain || 'EBAYUSER';
    services.login(options.consumerId,
        options.secret, function getClient(err, client) {
        if (err) {
            return callback(err);
        }

        client.authenticateUser(options, callback);
    });
};
