// @ebay/iaf-ebay/lib/index.js
'use strict';

var Iaf = require('./iaf-request').Iaf;

function getAppToken(req, options, callback) {
    var iaf = new Iaf(req);
    iaf.getAppToken(options, callback);
}

function getUserToken(req, options, callback) {
    var iaf = new Iaf(req);
    iaf.getUserToken(options, callback);
}

function validateUserToken(req, options, callback){
    var iaf = new Iaf(req);
    iaf.authenticateRequest(options, callback);
}

function authenticateToken(req, options, callback){
    var iaf = new Iaf(req);
    iaf.authenticateToken(options, callback);
}

function getSecurityTokenByRef(req, options, callback){
    var iaf = new Iaf(req);
    iaf.getSecurityTokenByRef(options, callback);
}

function logout(req, options, callback){
    var iaf = new Iaf(req);
    iaf.logout(callback);
}

function authenticateUser(req, options, callback) {
    var iaf = new Iaf(req);
    iaf.authenticateUser(options, callback);
}

module.exports = function middleware(options) {
    return function (req, res, next) {
        req.ebay = req.ebay || {};
        req.iaf = req.ebay.iaf = new Iaf(req, options);
        next();
    };
};

module.exports.iaf = require('./iaf');

module.exports.getAppToken = getAppToken;
module.exports.getUserToken = getUserToken;
module.exports.validateUserToken = validateUserToken;
module.exports.getSecurityTokenByRef = getSecurityTokenByRef;
module.exports.authenticateToken = authenticateToken;
module.exports.logout = logout;
module.exports.authenticateUser = authenticateUser;
module.exports.parse = require('./utils').parse;
