// @ebay/tracking-ebay/lib/entities/cache.js
'use strict';

//not loaded, loading, succeeded, failed, expired
var StateType = {
    INIT: 0,
    LOADING: 1,
    SUCCEED: 2,
    FAILED: 3,
    EXPIRED: 4
};

var MAX_RETRY = 64;
var BACKOFF = 1.7;
function State(s) {
    this.state = s;
    this.promise = null;
    this.retry = 0;
    this.retryDelay = 16;
    this.lastRetry = 0;
}

var GLOBAL_CACHE, PAGE_CACHE = {};

var helper = require('./entity-helper.js');
var async = require('async');

function init(state, loader, callback) {
    loader(function (err, data) {
        if (err) {
            state.state = StateType.FAILED;
            return callback(err);
        }
        state.state = StateType.SUCCEED;
        state.retryDelay = 16;
        state.lastRetry = 0;
        state.retry = 0;
        return callback();
    });
    state.state = StateType.LOADING;
    return state;
}

function getState(state, loader, setter, callback) {
    state = state || new State(StateType.INIT);
    setter(state);

    switch (state.state) {
    case StateType.EXPIRED:
    case StateType.INIT:
        return init(state, loader, callback);
    case StateType.LOADING:
    case StateType.SUCCEED:
        break;
    case StateType.FAILED:
        var now = Date.now();
        if ((state.retryDelay < (now - state.lastRetry)) && state.retry++ < MAX_RETRY) {
            state.lastRetry = now;
            state.retryDelay = state.retryDelay * BACKOFF;
            return init(state, loader, callback);
        }
        break;
    }
    callback();
}

module.exports = {
    get: function (pageId, callback) {
        var args = Array.prototype.slice.call(arguments);
        callback = args.pop();
        pageId = args.shift();

        function globalLoader(callback) {
            async.parallel([helper.getGlobalFlags, helper.getTags, helper.getModules], callback);
        }

        function pageLoader(callback) {
            helper.getFlags(pageId, callback);
        }

        function globalCacheSetter(s) {
            GLOBAL_CACHE = s;
        }

        function pageCacheSetter(s) {
            PAGE_CACHE[pageId] = s;
        }

        var funcs = [
            getState.bind(null, GLOBAL_CACHE, globalLoader, globalCacheSetter)
        ];
        if (pageId) {
            funcs.push(getState.bind(null, PAGE_CACHE[pageId], pageLoader, pageCacheSetter));
        }
        async.parallel(funcs, callback);
    },
    reset: function () {
        GLOBAL_CACHE = undefined;
        PAGE_CACHE = {};
    }
};
