// @ebay/tracking-ebay/lib/entities/entity-helper.js
'use strict';

var DEPLOYED = 'Deployed',
    Tags = require('./tags.js'),
    Flags = require('./flags.js'),
    Modules = require('./modules.js'),
    _ = require('underscore'),
    Async = require('async'),
    assert = require('assert'),
    client = require('../services/tracking-wrappers'),
    pathTools = require('path'),
    utils = require('../utils'),
    logger = require('@ebay/logging-inc').logger('tracking-ebay:entities');

var globalFlagsMeta = utils.tryRequire(pathTools.resolve(__dirname, '../../resources/global-flags.json'));
var tagsMeta = utils.tryRequire(pathTools.resolve(__dirname, '../../resources/tags.json'));
var modulesMeta = utils.tryRequire(pathTools.resolve(__dirname, '../../resources/modules.json'));

var tagsRefreshed;
var tagsRefreshed2; // add 2nd batch, avoid complex batch logic as this is going to be removed soon
var flagsRefreshed;
var modulesRefreshed;

function loadDefaults() {

    // 1nst batch to load tags from 10000 to 20000
    if (!tagsRefreshed) {
        tagsRefreshed = true;
        setImmediate(function () {
            module.exports.loadTags(0, function (err, results) {
                if (!err) {
                    initTags(results);
                }
                else {
                    tagsRefreshed = false;
                }
            });
        });
    }

    // 2nd batch to load tags from 10000 to 20000
    if (!tagsRefreshed2) {
        tagsRefreshed2 = true;
        setImmediate(function () {
            module.exports.loadTags(10000, function (err, results) {
                if (!err) {
                    initTags(results);
                }
                else {
                    tagsRefreshed2 = false;
                }
            });
        });
    }

    if (!flagsRefreshed) {
        flagsRefreshed = true;
        setImmediate(function () {
            module.exports.loadGlobalFlags(function (err, results) {
                if (!err) {
                    initFlags(results);
                }
                else {
                    flagsRefreshed = false;
                }
            });
        });
    }

    if (!modulesRefreshed) {
        modulesRefreshed = true;
        setImmediate(function () {
            module.exports.loadModules(function (err, results) {
                if (!err) {
                    initModules(results);
                }
                else {
                    modulesRefreshed = false;
                }
            });
        });
    }

}

function initFlags(results, pageId) {
    logger.debug(results);
    var flagSet = results[1];
    var flags = results[0];
    pageId = pageId || Flags.GLOBAL_PAGE_ID;
    if (!flagSet || !flagSet.flagsetList || !flagSet.flagsetList.length || !flagSet.flagsetList[0].name) {
        return Flags;
    }
    var newFlagSet = Flags.newFlagSet(flagSet.flagsetList[0].name, pageId);
    if (!flags || !flags.flagList) {
        return Flags;
    }
    if (_.isArray(flags.flagList)) {
        flags.flagList.forEach(function (f) {
            if (f.name) {
                newFlagSet.newFlag(f.name.toUpperCase(), f.bit);
            }
        });
    } else if (_.isObject(flags.flagList)) {
        var f = flags.flagList;
        if (f.name) {
            newFlagSet.newFlag(f.name.toUpperCase(), f.bit);
        }
    }

    return Flags;
}

function getFlags(flagSetId, pageId, callback) {
    var args = Array.prototype.slice.call(arguments);
    callback = args.pop();
    flagSetId = args.shift();
    pageId = args.shift() || Flags.GLOBAL_PAGE_ID; // page id can not be 0

    var flagsReq = {
        trackingFlag: {
            flagsetId: flagSetId,
            lifeCycleState: DEPLOYED
        },
        startOffset: 0,
        maxRecord: 10000
    };
    var flagsetsReq = {
        trackingFlagset: {
            '$id': flagSetId,
            lifeCycleState: DEPLOYED
        },
        startOffset: 0,
        maxRecord: 1
    };
    logger.debug(flagsReq);
    logger.debug(flagsetsReq);

    Async.parallel([
        next => client.getTrackingFlags(flagsReq, next),
        next => client.getTrackingFlagsets(flagsetsReq, next),
    ], function (err, results) {
        if (err) {
            return callback(err);
        }
        return callback(null, results);
    });
}


function getPageFlagset(pageId, callback) {
    var pageFlagSetReq = {
        pageFlagset: {
            pageId: pageId,
            lifeCycleState: DEPLOYED
        }
    };
    logger.debug(pageFlagSetReq);
    client.getPageFlagset(pageFlagSetReq, callback);
}

function getFlagsByFlagset(pageId, pageFlagSetResponse, callback) {
    logger.debug(pageFlagSetResponse);
    var result = pageFlagSetResponse.pageFlagsetList;
    assert.ok(result, 'getPageFlagset dose not return pageFlagsetList');
    if (_.isArray(result)) {
        result = result[0];
    }
    getFlags(result.flagsetId, pageId, callback);
}

function getTrackingProperties(...args) {
    const callback = args.pop();
    const offset = args.shift();
    var req = {
        trackingProperty: {
            lifeCycleState: DEPLOYED
        },
        startOffset: offset || 0,
        maxRecord: 10000
    };
    logger.debug(req);
    client.getTrackingProperties(req, function (err, result) {
        if (err) {
            return callback(err);
        }
        process.emit('tracking-ebay:tags:load:complete', result);
        callback(null, result);
    });
}

function initTags(propsResponse) {
    logger.debug(propsResponse);
    var propList = propsResponse.trackingPropertyList;
    if (propList) {
        propList.forEach(function (p) {
            Tags.newTag(p.name.toUpperCase(), p.sojournerName, p.dataType);
        });
    } else {
        logger.warn('Failed to load any tags metadata, response: ', propsResponse);
    }
    process.emit('tracking-ebay:tags:init:complete');
    return Tags;
}

function getModules(callback) {
    var req = {
        module: {
            lifeCycleState: DEPLOYED
        },
        startOffset: 0,
        maxRecord: 10000
    };
    logger.debug(req);
    client.getModules(req, function (err, result) {
        if (err) {
            return callback(err);
        }
        callback(null, result);
    });
}

function initModules(modulesResponse) {
    logger.debug(modulesResponse);

    var list = modulesResponse.moduleList || [];
    list.forEach(function (m) {
        if (!m.attributes || !m.attributes.id) {//will id == 0?
            return;
        }
        Modules.newModule(m.attributes.id, m.name.toUpperCase());
    });
    return Modules;
}

module.exports = {
    loadGlobalFlags: function (callback) {
        //global flagset id == 1
        getFlags(1, callback);
    },
    getGlobalFlags: function (callback) {
        //global flagset id == 1
        if (globalFlagsMeta) {
            loadDefaults();
            return callback(null, initFlags(globalFlagsMeta));
        }
        getFlags(1, function(err, data) {
            callback(err, !err && initFlags(data));
        });
    },
    getFlags: function (pageId, callback) {
        Async.waterfall([
            next => getPageFlagset(pageId, next),
            (pageFlagSetResponse, next) => getFlagsByFlagset(pageId, pageFlagSetResponse, next),
        ], function (err, results) {
            if (err) {
                logger.error('get flags failed: ', err);
                return callback(err);
            }
            callback(null, initFlags(results, pageId));
        });
    },
    loadTags: getTrackingProperties,
    getTags: function (callback) {
        if (tagsMeta) {
            loadDefaults();
            return callback(null, initTags(tagsMeta));
        }
        module.exports.loadTags(function (err, results) {
            callback(err, !err && initTags(results));
        });
    },
    loadModules: getModules,
    getModules: function (callback) {
        if (modulesMeta) {
            loadDefaults();
            return callback(null, initModules(modulesMeta));
        }
        module.exports.loadModules(function (err, data) {
            callback(err, !err && initModules(data));
        });
    }
};

loadDefaults();
