// @ebay/tracking-ebay/lib/entities/flags.js
'use strict';

var EbayFlags = function () {
    this.nameMap = {};
    this.pageMap = {};
    this.GLOBAL_PAGE_ID = -1;
};
var FlagSet = function (codeName) {
    this.name = codeName;
    this.bitMap = [];
};

var _instance = new EbayFlags();

FlagSet.prototype.newFlag = function (codeName, bit) {
    var flag = this[codeName] = this.bitMap[bit] = new Flag(codeName, bit, this.name);
    return flag;
};
FlagSet.prototype.getFlag = function (bit) {
    return this.bitMap[bit];
};
var Flag = function (codeName, bit, setName) {
    this.name = codeName;
    this.bit = bit;
    this.setName = setName;
};
Flag.prototype.toString = function () {
    return this.name;
};
EbayFlags.prototype.newFlagSet = function (codeName, pageId) {
    var flagSet = this[codeName] = this.nameMap[codeName] = this.pageMap[pageId] = new FlagSet(codeName);
    return flagSet;
};
EbayFlags.prototype.getAppFlag = function (bit, pageId) {
    return this.pageMap[pageId] && this.pageMap[pageId].getFlag(bit);
};

EbayFlags.prototype.getContextFlag = function (bit) {
    return this.getAppFlag(bit, this.GLOBAL_PAGE_ID);
};

module.exports = _instance;
module.exports.reset = function () {
    _instance = new EbayFlags();
};
