// @ebay/tracking-ebay/lib/entities/tags.js
'use strict';
const utils = require('../utils.js');
const logger = require('@ebay/logging-inc').logger('tracking-ebay/entities/tagsr');
const TEN_SEXTILLION = 10000000000000000;

var Tag = function (codeName, sojournerName, dataType) {
    this.name = codeName;
    this.sojournerName = sojournerName;
    this.dataType = dataType;
};
Tag.prototype.toString = function () {
    return this.sojournerName;
};
Tag.prototype.equals = function (x) {
    if (!x || !x.sojournerName) {
        return false;
    }
    return (x.sojournerName === this.sojournerName);
};
Tag.prototype.isValidData = function (value) {
    if (!this.dataType) {
        return true;
    }
    if (!validateTagExtra(this.sojournerName, value)) {
        logger.warn(new Error(`Tag ${this.sojournerName} has invalid format for ${value}`));
        return false;
    }
    var d = this.dataType.toString().toLowerCase();
    if (d === 'java.lang.string') {
        return (value !== null) && (value !== undefined);
    } else if (d === 'java.lang.integer') {
        return utils.isInteger(parseFloat(value));
    } else if (d === 'java.lang.float') {
        return utils.isInteger(parseFloat(value)) || utils.isFloat(parseFloat(value));
    } else if (d === 'java.lang.boolean') {
        return (value === 'true') || (value === 'false');
    }
    return true;
};
var EbayTags = function () {
    this.sojMap = [];
};

EbayTags.prototype.newTag = function (codeName, sojournerName, dataType) {
    if (this[codeName]) {
        return this[codeName];
    }
    var tag = this[codeName] = this.sojMap[sojournerName] = new Tag(codeName, sojournerName, dataType);
    return tag;

};
EbayTags.prototype.getTagByName = function (name) {
    return this.sojMap[name];
};

var _instance = new EbayTags();

module.exports = _instance;

const PAT = /^[A-Za-z0-9]+$/;

// logic source: https://github.corp.ebay.com/ejan/TrackingComponents/blob/master/Tracking/src/main/java/com/ebay/kernel/presentation/sojourner/SojournerContextDataEnum.java

const sojournerContextDataRange = {
    /**
	 * A two character hash key used to help route published messages.
	 * The key is a hexadecimal number in string notation in the range
	 * "00" to "FF".  (case insensitive)
	 */
    h: [2, PAT],       // HASH
    /**
	 * A globally unique identifier for a client.
	 */
    g: [32, PAT],      // GUID
    /**
	 * A cguid.
	 */
    n: [32, PAT],      // CGUID
    /**
	 * The user account id, long integer as string.
	 */
    u: [10, /^\d+$/],       // USER_ACCT_ID
    /**
	 * Id of a request web page.
	 */
    p: 7,       // PAGE_ID
    t: 3,       // SITE_ID
    /**
	 * Command status. s=1 indicates an error.  s=0 or omitted indicates no error.
	 * In this usage, error generally means unhandled exception.
	 */
    s: 1,       // COMMAND_STATUS
    /**
	 * rd=1 indicates that the command redirected to another URL.
	 */
    rdt: 1,     // REDIRECT_INDICATOR
    /**
	 * regU=1 indicates that the reg cookie or dp1.reg cookielet was received
	 * on the request, indicating that this is a registered user.
	 */
    regU: 1,    // REGISTERED
    /**
	 * Best Guess User. When user is not logged in, user oracle id can be
	 * inferred from persistence cookie.  This is logged into
	 * SOJ as bu tag.
	 */
    bu: [10, /^\d+$/],     // BEST_GUESS_USER_ACCT_ID
    mid: 7,     // MODULE_ID
    uaid: [40, PAT],   // USER_ACTION_ID
};

function validateTagExtra(key, valueStr) {
    const range = sojournerContextDataRange[key];
    const [size, pattern] = Array.isArray(range) && range || [range];
    
    return validateSize() && validatePattern();

    function validateSize() {
        if (!size) {
            return true;
        }
        if (key === 'g') {
            return valueStr.length === size;
        }
        return valueStr.length <= size;
    }

    function validatePattern() {
        if (key === 'u' || key === 'bu') {
            const id = parseInt(valueStr);
            if (!id || id > TEN_SEXTILLION) {
                return false;
            }
        }
        return !pattern ||
            pattern.test(valueStr);
    }
}
