// @ebay/tracking-ebay/lib/taas/helper.js
'use strict';

var SojournerEventEnum = require('../soj/sojourner-event-enums.js');
var Flags = require('../entities/flags.js');
var Tags = require('../entities/tags.js');

require('../services/tracking-wrappers');
var utils = require('../utils');
var client = require('../services/tracking-wrappers');
var logger = require('@ebay/logging-inc').logger('tracking-ebay:taas');
var DataHolder = require('raptor-async/DataHolder');

/**
 * Call TAAS(eBayTrackingService) to fetch the static resources and impression id, cookies
 * @param {SojournerContext} sojournerContext the sojourner context for this request
 * @param {CookieManager} cookies the ebay cookie manager
 * @param {Cookies} cookielets ebay cookielets enums
 * @param {express request} request express request
 * @param {express response} response express response
 * @return {TaasContext} : {
    response:{
        headerContent:['new (raptor.require(\'raptor.tracking.core.Tracker\'))({"psi":"A3dy3oMU*","rover":{"clk":"/roverclk/0/0/9","imp":"/roverimp/0/0/9","uri":"rover.qa.ebay.com"},"pid":"p6028"});','var RoverSyncDropped = true;'],
        headerURL:'http://rover.qa.ebay.com/roversync/?site=0&stg=0&mpt=1372709575863',
        footerContent:'TaaSIdMapTrackerObj.roverService("http://rover.qa.ebay.com/idmap/0?footer");',
        headerResourceFileName:'http://include.qa.ebaystatic.com/js/v/us/tracking_RaptorheaderJS.js',
        footerResourceFileName:'http://include.qa.ebaystatic.com/js/v/us/tracking_RaptorfooterJS.js',
        TrackingStateInfo:{
            name:'',
            value:''
        },
        cookieData:[
            {
                cookieletID:'',
                cookieletValue:'',
                actionType:''
            }
        ],
        sojData:{

        }
    },
    trackingTrafficSource: '',
    ciid:''
 }
 */

function translateEventName(eventName) {
    if (eventName === 'click') {
        return SojournerEventEnum.CLICK;
    } else if (eventName === 'cookie') {
        return SojournerEventEnum.COOKIE;
    } else if (eventName === 'application') {
        return SojournerEventEnum.APPLICATION;
    } else {
        return null;
    }
}

function getFullUrl(request) {
    var protocol = request.secure ? 'https' : 'http';
    var host = request.hostname;
    var url = request.originalUrl;
    return protocol + '://' + host + url;
}

exports.track = function track(req, res, callback) {
    var ebay = req.ebay;
    var cookies = ebay.cookies;
    var cookielets = cookies.Cookies;
    var tracking = ebay.tracking;
    var sojournerContext = tracking.sojournerContext;
    var framework = 'nodeJS';
    var isPulsarJS = false;

    var prepareRequest = function (pageName, siteCatalystPageName, pageId) {
        var trackingRequest = {};
        // mandatory
        trackingRequest.mandatoryParams = {
            client: 'ebay',
            cobrand: ebay.getPartnerId(),
            pageId: pageId,
            siteId: ebay.getSiteId(),
            referrer: req.get('referrer'),
            framework: framework,
            contentVersion: 'v1',
            currentPageURL: getFullUrl(req)
        };
        //optional
        trackingRequest.otherParams = {
            secure: typeof req.secure !== 'undefined' ? req.secure : 'false',
            pageName: pageName,
            siteCatalystPageName: siteCatalystPageName
        };
        //copy cookies
        trackingRequest.cookieData = [{
            cookieletID: cookielets.PAGE_SOURCE_INFO.id,
            cookieValue: cookies.getCookieValue(cookielets.PAGE_SOURCE_INFO)
        }, {
            cookieletID: cookielets.SOJOURNER_TRACKING.id,
            cookieValue: cookies.getCookieValue(cookielets.SOJOURNER_TRACKING)
        }, {
            cookieletID: cookielets.TRACKING_ROVER_PIM.id,
            cookieValue: cookies.getCookieValue(cookielets.TRACKING_ROVER_PIM)
        }, {
            cookieletID: cookielets.TRACKING_GUID.id,
            cookieValue: cookies.getCookieValue(cookielets.TRACKING_GUID)
        }, {
            cookieletID: cookielets.DC_CORRELATION_GUID.id,
            cookieValue: cookies.getCookieValue(cookielets.DC_CORRELATION_GUID)
        }, {
            cookieletID: cookielets.SOJOURNER_GUID.id,
            cookieValue: cookies.getCookieValue(cookielets.SOJOURNER_GUID)
        }, {
            cookieletID: cookielets.FLAGREG.id,
            cookieValue: cookies.getCookieValue(cookielets.FLAGREG)
        }, {
            cookieletID: cookielets.EBAY_SIGNIN.id,
            cookieValue: cookies.getCookieValue(cookielets.EBAY_SIGNIN)
        }];
        if (Flags.GlobalFlags && Flags.GlobalFlags.DETECTED_BOT) {
            //copy soj data
            trackingRequest.sojData = [{
                type: 'GlobalFlags',
                name: Flags.GlobalFlags.DETECTED_BOT.name,
                value: (sojournerContext.isAppFlagSet(Flags.GlobalFlags.DETECTED_BOT) ||
                    sojournerContext.isContextFlagSet(Flags.GlobalFlags.DETECTED_BOT)) ? '1' : '0'
            }];
        }

        //Remove duplicate _trksid from URL params
        if(req.query._trksid) {
            req.query._trksid = utils.filterTrksid(req.query._trksid);
        }


        //copy url params
        trackingRequest.requestParam = [{
            name: 'sid',
            value: req.query._trksid ? req.query._trksid : ''
        }];

        var ciid = tracking && tracking.getCiid(); 
        
        if (ciid) {
            trackingRequest.requestParam.push({
                'name': 'generatedCiid',
                'value' : ciid
            });    
        }
           

        trackingRequest.responseFields = {
            pulsarJS: isPulsarJS
        };
        logger.debug(trackingRequest);

        return trackingRequest;
    };

    function processResponse(response) {
        var taasContext = {};
        var expSvcImpressionId = sojournerContext && sojournerContext.getCurrentImpression(SojournerEventEnum.APPLICATION);

        logger.debug(response);
        //update soj
        function processSojData(soj) {
            if (soj.type === 'GlobalTags') {
                var eventEnum = translateEventName(soj.event.toLowerCase());
                var keyEnum = Tags.getTagByName(soj.name);
                sojournerContext.addData(eventEnum, keyEnum, soj.value);
            }
            if (soj.name.toLowerCase() === 'ciid') {
                //set impression
                taasContext.ciid = soj.value;
            }
        }
        var sojData = response.sojData;
        if (sojData) {
            (Array.isArray(sojData) ? sojData : [sojData]).forEach(function (soj) {
                if (soj && soj.name === 'ciid') {
                    soj.value = expSvcImpressionId || soj.value;
                }
                processSojData(soj);
            });
        }
        //set traffic source
        var trackingStateInfo = response.trackingStateInfo;
        if (trackingStateInfo) {
            var ts = -1;
            trackingStateInfo.forEach(function (info) {
                if (info.name.toLowerCase() === 'ts') {
                    ts = info.value;
                }
            });
            taasContext.trackingTrafficSource = ts;
        }
        if (response.headerContent) {
            var headerContent = response.headerContent;
            if (!Array.isArray(headerContent)) {
                response.headerContent = [headerContent];
            }
        }
        if (response.footerContent) {
            var footerContent = response.footerContent;
            if (!Array.isArray(footerContent)) {
                response.footerContent = [footerContent];
            }
        }
        taasContext.response = response;
        return taasContext;
    }

    var pageId = tracking.pageId || ebay.getPageId();
    var pageName = tracking.pageName || ebay.getPageName();
    var siteCatalystPageName = tracking.siteCatalystPageName || pageName;
    var pending = new DataHolder();

    if (ebay.expSvc && ebay.expSvc.operationId) {
        pageId = ebay.expSvc.operationId;
        pageName = ebay.expSvc.operation || pageName;
        isPulsarJS = true;
    }

    if (req.route && req.route.config && req.route.config.tracking &&
        req.route.config.tracking.type === 'pulsar') {
        isPulsarJS = true;
    }

    const headers = tracking.isCacheRequest() ? {
        'x-ebay-c-tracking': req.ebay.getCorrelationTracking().buildHeader()
    } : undefined;

    client.getClientTrackingCtx(
    prepareRequest(pageName, siteCatalystPageName, pageId), headers, function (e, result) {
        if (e) {
            logger.error('failed to get taas response because of: ', e);
            e.message = 'failed to get taas response because of:' + e.message;
            pending.reject(e);
            return;
        }
        var taasContext = processResponse(result);
        taasContext.trackingContext = tracking;
        updateCookies(req, res, taasContext);
        pending.resolve(taasContext);
    });

    if (callback) {
        pending.done(callback);
    }
    tracking.getTaasContext = function getTaasContext(callback) {
        pending.done(callback);
    };
};


var updateCookies = exports.updateCookies = function updateCookies(req, res, taasContext) {
    if (!req.ebay || !req.ebay.cookies) {
        return false;
    }

    var cookieData = taasContext.response.cookieData;
    /*
    cookieData: [
        { cookieletID: '352', cookieValue: {}, actionType: 'D' },
        { cookieletID: '344', cookieValue: {}, actionType: 'D' },
        { cookieletID: '999', cookieValue: '3065243', actionType: 'A' }
    ],
     */

    if (!cookieData) {
        return false;
    }
    var cookies = req.ebay.cookies;
    var Cookies = cookies.Cookies;

    logger.debug(cookieData);

    if (!Array.isArray(cookieData)) {
        cookieData = [cookieData];
    }

    cookieData.forEach(function (cookieInstruction) {

        var action = cookieInstruction.actionType;
        var cookieletID = cookieInstruction.cookieletID;

        if ((res.statusCode === 300 || res.statusCode === 301 || res.statusCode === 302) &&
            (Cookies.PAGE_SOURCE_INFO.id.toString() === cookieletID ||
                Cookies.SOJOURNER_TRACKING.id.toString() === cookieletID)) {
            return;
        }

        if (action === 'D') {
            cookies.removeCookie(cookieletID);
        } else if (action === 'A' || action === 'I' || action === 'U') {
            cookies.setCookieValue(
                cookieletID,
                cookieInstruction.cookieValue);
        }
    });

    return true;
};
