// @ebay/monitor-inc/lib/cpu.js
'use strict';

const os = require('os');
const PERCENTILE_SUFFIX = 'Percentile';

function calculatePercentile(stats) {
    Object.keys(stats).forEach(name => {
        if (name === 'total') {
            return;
        }
        stats[name + PERCENTILE_SUFFIX] = Math.round(stats[name]/(stats.total || 1) * 100 * 100)/100;
    });
    return stats;
}

/*
    Native monitor produces stats in microseconds
*/
class NativeMonitor {
    get() {
        const currentCpuUsage = process.cpuUsage();
        const currentCpuTime = process.hrtime();

        if (!this.state) {
            // init phase
            this.state = {
                lastCpuUsage: currentCpuUsage,
                lastCpuTime: currentCpuTime
            };

            return calculatePercentile({
                sys: 0,
                idle: 0,
                user: 0,
                total: 0,
                usage: 0
            });
        }

        const cpuTimeDelta = process.hrtime(this.state.lastCpuTime);

        const state = this.state;


        // total cpu time in microseconds
        let totalCpuTime = hrtimeToMicroseconds(cpuTimeDelta);

        // calculate stats
        const stats = {
            user: (currentCpuUsage.user - state.lastCpuUsage.user),
            sys: (currentCpuUsage.system - state.lastCpuUsage.system),
            total: totalCpuTime
        };
        stats.usage = stats.user + stats.sys;
        stats.idle = totalCpuTime - stats.usage;

        this.state.lastCpuUsage = currentCpuUsage;
        this.state.lastCpuTime = currentCpuTime;

        return calculatePercentile(stats);
    }
}

function hrtimeToMicroseconds (hrtime) {
  return parseInt(hrtime[0] * 1000000 + hrtime[1]/1000);
}

/*
    Native monitor produces stats in milliseconds
*/
class SystemMonitor {
    get() {
        const currentCpuTime = os.cpus().map(cpu => {
            return cpu.times;
        });

        if (!this.state) {
            this.state = currentCpuTime;
        }

        const last = this.state;
        const stats = currentCpuTime.reduce((memo, current, i) => {
            return {
                idle: (memo.idle || 0) + current.idle - last[i].idle,
                sys: (memo.sys || 0) + current.sys - last[i].sys,
                user: (memo.user || 0) + current.user - last[i].user
            };
        }, {});

        stats.usage = stats.user + stats.sys;
        stats.total = stats.idle + stats.user + stats.sys;

        this.state = currentCpuTime;

        return calculatePercentile(stats);
    }
}

const nativeMonitor = new NativeMonitor();
const systemMonitor = new SystemMonitor();

module.exports = {
    NativeMonitor: NativeMonitor,
    SystemMonitor: SystemMonitor,
    nativeMonitor: nativeMonitor,
    systemMonitor: systemMonitor,
    /*
        Native monitor produces stats in microseconds
    */
    usage() {
        return nativeMonitor.get();
    },
    /*
        Native monitor produces stats in milliseconds
    */
    system() {
        return systemMonitor.get();
    }
};
