// @ebay/monitor-inc/lib/restarts.js
'use strict';

/* getClusterId returns a small integer number to use with
 * metrics reporting.
 *
 * getRestarts(options, callback) increments the restart count on disk
 * and returns the cumulative count.
 *
 */
var fs = require('fs'),
    deployEnv = require('@ebay/environment-ebay'),
    cal = require('@ebay/cal'),
    path = require('path'),
    utils = require('./utils'),
    assert = require('assert'),
    errors = require('@ebay/errors-ebay');

// Called when starting up and so also gets called on a restart
// which is when we collect the number of restarts to report and
// bump the number of restarts done.
function getRestarts(options, callback) {
    callback = arguments[arguments.length - 1];
    assert.ok(callback, 'callback required');
    assert.equal(typeof callback === 'function', true, 'callback should be function');
    var Restart = errors.getRestartLog();
    var restartCount = Restart.getRestarts();
    options = options && options.monitor || {};
    var toAddr = deployEnv.isProd() ? options.toAddr : options.stageToAddr;
    var reason = null;
    if (!deployEnv.isDev()) {
        if(restartCount > 0) {
            if(isDeployRestart()) {
                //Don't send email on first deployment
                Restart.sendEmail({subject:'Deploy Restart on ', to: toAddr, text: utils.deployEmailBody()});
            } else if(!Restart.getEmailStatus()) {
                const restartReason = Restart.getReason();
                //Send email ONLY when previous email status is false
                var emailBody = restartReason || utils.buildEmailBody({maxLogLines: options.maxLogLines});
                Restart.sendEmail({to: toAddr, text: emailBody});
                reason = restartReason;
            }
        }

        Restart.incrementRestarts()
           .execute(function(err, info){
               console.log('incrementRestarts() Command Executed ', err);
               reset();
           });
    }
    var health = {
        noWorkers: 1,
        restarts: restartCount,
        reason: reason
    };
    callback(null, health);
}

function reset() {
    var Restart = errors.getRestartLog();
    Restart.reset()
            .execute(function(){
                console.log('Reset() Command Executed ');
            });
}

function isDeployRestart() {
    var status = false;
    var deployLogPath = path.join(process.cwd(), 'logs/deploy.log');
    if(fs.existsSync(deployLogPath)) {
        var fileStat = fs.statSync(deployLogPath);
        var deployTime = new Date(fileStat.mtime);
        var diffInSecs = (Date.now() - deployTime)/1000;
        //If Deploy Timestamp is less than 2 mins
        if(diffInSecs <= 120) {
            return true;
        }
    }

    return status;
}

module.exports = {
    getRestarts: getRestarts
};
