// @ebay/esams-consumer-ebay/lib/legacy-esams-consumer.js
'use strict';

const logger = require('@ebay/logging-inc').logger('@ebay/ebay-esams-consumer');

const cache = require('./cache'),
    service = require('./service.js');

const {
    getIdentityFilePaths,
    getDefaultIdentityFile,
    getIdentity,
    reset
} = require('./identity-utils');

let consumers = {};

function EsamsConsumer(identityFile, identity, options) {
    options = options || {};

    this._options = options;
    this._identity = identity;
    this._identityFile = identityFile;
}

EsamsConsumer.prototype = {
    context: function(ctx) {
        var wrapper = Object.create(this);
        wrapper.ctx = ctx;
        return wrapper;
    },

    /**
     * @param {string} channelName
     * @param {string} channelVersion put -1 to return the latest.
     * @return {promise or object}
     */
    getSecretKey: function getSecretKey(channelName, channelVersion, callback) {
        this._get('getSecretKey', channelName, channelVersion, callback);
    },

    /**
     * @param {string} channelName
     * @param {string} channelVersion put -1 to return the latest.
     * @return {promise or object}
     */
    getNonKey: function getNonKey(channelName, channelVersion, callback) {
        this._get('getNonKey', channelName, channelVersion, callback);
    },

    /**
     * @param {string} channelName
     * @param {string} channelVersion put -1 to return the latest.
     * @return {promise or object}
     */
    getCertificate: function getCertificate(channelName, channelVersion, callback) {
        this._get('getCertificate', channelName, channelVersion, callback);
    },

    /**
     * @param {string} channelName
     * @param {string} channelVersion put -1 to return the latest.
     * @return {promise or object}
     */
    getKeyPair: function getKeyPair(channelName, channelVersion, callback) {
        this._get('getKeyPair', channelName, channelVersion, callback);
    },

    removeFromCache: function removeFromCache(channelName, channelVersion) {
        cache.removeKey(this._identityFile, channelName, channelVersion || -1);
    },

    _get: function _get(operation, channelName, channelVersion, callback) {
        var valueFromCache;

        channelVersion = channelVersion || -1;
        valueFromCache = cache.getKey(this._identityFile, channelName, channelVersion);
        if (valueFromCache) {
            return callback(null, Object.create(valueFromCache));
        }

        this._invokeService(operation, channelName, channelVersion, callback);
    },

    _invokeService: function _invokeService(operation, channelName, channelVersion, callback) {
        var self = this;
        var headers = {
            'X-EBAY-DELIVERABLE-IDENTITY': self._identity
        };

        // Avoid multiple service invocation when one is pending.
        if (cache.getAndsetFetchInProgress(self._identityFile, channelName, channelVersion, callback)) {
            return;
        }

        service[operation](self.ctx, channelName, channelVersion, headers, self._options, function (err, result) {
            if (err) {
                var valueFromCache = cache.getKey(self._identityFile, channelName, channelVersion, true);

                if (valueFromCache) {
                    cache.extendExpiry(self._identityFile, channelName, channelVersion);
                    logger.warn('ESAMS service invocation ' + operation + ' failed returning cached value');
                    result = valueFromCache;
                }
            } else {
                cache.putKey(self._identityFile, channelName, channelVersion, result);
            }

            var callbacks = cache.clearFetchInProgress(self._identityFile, channelName, channelVersion);
            for (var i=0; i < callbacks.length; i++) {
                callbacks[i]((result ? null : err), result);
            }
        });
    }
};

function getEsamsConsumer(identityFile, options) {
    if (!identityFile) {
        identityFile = getDefaultIdentityFile();
    }

    if (!identityFile) {
        throw new Error('ESAMS identity file(s) not found: ' + getIdentityFilePaths());
    }

    return consumers[identityFile] || (consumers[identityFile] = new EsamsConsumer(identityFile, getIdentity(identityFile), options));
}

function getIdentityFileContents(identityFile) {

    if (identityFile) {
        return getIdentity(identityFile);
    }

    var identityFilePath = getDefaultIdentityFile();

    if (!identityFilePath) {
        throw new Error('Unable to find an identity file.');
    }
    return getIdentity(identityFilePath);
}

module.exports = {
    /**
     * export the module
     * @return {EsamsConsumer}
     */
    getEsamsConsumer: getEsamsConsumer,
    getIdentityFileContents : getIdentityFileContents,

    // For testing
    __reset: function() {
        reset();
        consumers = {};
        cache.clean();
    }
};
