// @ebay/service-instruments-ebay/graphql/factory.js
const { Kind } = require('graphql');
const print = require('graphql/language/printer');
const extractor = require('extract-files');
const utils = require('./utils');

function getOperationName(query) {
    if (!query?.definitions) {
        return undefined;
    }
    for (const node of query.definitions) {
        if (node.kind === Kind.OPERATION_DEFINITION && node.name) {
            return node.name.value;
        }
    }
}

function createOperation({ query, extensions, variables }) {
    return {
        query: query ? print.print(query) : undefined,
        extensions,
        operationName: getOperationName(query),
        variables
    };
}

function createMultipartOperation({ query, extensions, variables }) {
    const { files, clone } = extractor.extractFiles(
        variables,
        '',
        utils.isExtractableFile
    );
    const operation = createOperation({ query, extensions, variables: clone });
    const result = {
        operations: operation,
        map: {}
    };
    const map = {};
    let i = 0;
    files.forEach((paths) => {
        map[i++] = paths.map((path) => `variables.${path}`);
    });
    result.map = map;

    i = 0;
    files.forEach((_, file) => {
        result[i++] = file;
    });

    return result;
}

module.exports = {
    createOperation,
    createMultipartOperation
};
