// @ebay/i18n-ebay/lib/ContentManager.js
var bundleLoader = require('./bundle-loader');
var extend = require('raptor-util/extend');
var async = require('async');

function ContentManager(lookupContext) {
    this.lookupContext = lookupContext;
    this.preloaded = {};
    this.defaultRoot = null;
}

ContentManager.prototype = {
    getTarget: function() {
        return this.lookupContext && this.lookupContext.target;
    },
    /*
     * Sync API is available only for browser side usage or when bundle has been already cached on server side
    **/
    getBundle: function (bundleName, from, callback) {
        if (typeof from === 'function') {
            callback = from;
            from = undefined;
        }

        var preloadedBundle = this.preloaded[bundleName];

        if (callback) {
            if (preloadedBundle) {
                return callback(null, preloadedBundle);
            }

            from = from || this.defaultRoot;
            bundleLoader.loadBundle(bundleName,
                this.lookupContext,
                from,
                callback);
        } else if (preloadedBundle) {
            return preloadedBundle;
        } else {
            throw new Error('Unable to return content bundle for "' + bundleName + '". callback is required unless a bundle has been pre-loaded.');
        }
    },

    _getBundles: function (bundleNames, from, callback) {
        if (typeof from === 'function') {
            callback = from;
            from = undefined;
        }

        async.map(bundleNames, function iter(bundleName, next) {
                this.getBundle(bundleName, from, next);
            }.bind(this), callback);
        }
    };

extend(ContentManager.prototype, require('./legacy'));

module.exports = ContentManager;
