// @ebay/i18n-ebay/lib/index.js
'use strict';

var fs = require('fs');
var path = require('path');
var extend = require('raptor-util/extend');
var managerProvider = require('./manager-provider');

function bundleNotFoundErr(name, hasContext) {
    var bundlenotfound_msg = 'Unable to find bundle "' + name + '"';
    var missingcontext = hasContext ? '' : ', Possible cause is missing request context.';
    return new Error(bundlenotfound_msg + missingcontext);
}

/**
Returns the root of the Module passed in the argument
The recursive search will stop when finds the folder containing package.json
Else will return the process.cwd();
**/
function noduleRoot(nodulename) {
    var appRoot = process.cwd();
    if (!nodulename) {
        return appRoot;
    } else {
        var nodulepkg = path.resolve(nodulename, 'package.json');

        if (fs.existsSync(nodulepkg)) {
            return path.dirname(nodulepkg);
        } else {
            var noduledir = path.dirname(nodulename);
            if (noduledir === nodulename || noduledir === '/' || noduledir === appRoot) {
                return appRoot;
            }
            return (noduleRoot(noduledir));
        }
    }
}

exports.getContentManager = function getContentManager(reqOrRenderContext) {
    if (!reqOrRenderContext) {
        return managerProvider.getDefault();
    } else if (reqOrRenderContext.url) {
        return managerProvider.fromRequest(reqOrRenderContext);
    } else if (reqOrRenderContext.attributes) {
        return managerProvider.fromRenderContext(reqOrRenderContext);
    }
};

exports.use = function(nodule) {
    var basepath = noduleRoot(nodule ? nodule.filename : process.cwd());
    return {
        getBundle: function(bundle, callback) {
            var manager = managerProvider.getDefault();
            if (manager) {
                return manager.getBundle(bundle, basepath, callback);
            } else {
                return callback(bundleNotFoundErr(bundle, false));
            }
        },
        // loads the bundle with the locale information passed, doesn't expect req context
        context: function(locality) {
            var manager = managerProvider.fromLocality(locality);
            return {
                getBundle: function(bundle, callback) {
                    return manager.getBundle(bundle, basepath, callback);
                }
            };
        }
    };
};

exports.t = function(bundle, path, data, context) {
    var manager = managerProvider.getOptional();
    if (manager) {
        return manager
            .getBundle(bundle)
            .getText(path, data, context);
    }

    // If no manager is available then we will just return a string that can be
    // used for tests, storybook, etc.
    return `i18n.t("${bundle}", "${path}"${data ? `, ${JSON.stringify(data)}` : ''})`;
};

exports.middleware = require('./middleware');
extend(exports, require('./legacy'));
