// @ebay/i18n-ebay/lib/manager-provider.js
'use strict';

var ContentManager = require('./ContentManager');
var getRequest = require('@ebay/marko-context-ebay').getRequest;
var OUT_GLOBAL_KEY = 'i18nEbay';
var requestLocal;
try {
    requestLocal = require('request-local');
} catch (e) {
    //No ops
}

function getRequestFromRequestLocal() {
    try {
        return requestLocal.data.request;
    } catch (e) {
    }
}

function fromLocality(locality, acceptLanguages) {
    var lookupContext = {
        target: 'EBAY_' + (locality.country || 'US'),
        locale: locality.locale || 'en-US',
        // we need to strip locality off any runtime bindings
        // to avoid caching request context closures that would cause memory leaks
        locality: JSON.parse(JSON.stringify(locality))
    };

    if (acceptLanguages) {
        lookupContext.acceptLanguages = acceptLanguages;
    }

    if (!locality.isGeoCountry && !lookupContext.locale.includes(locality.country)) {
        lookupContext.target = 'EBAY_' + lookupContext.locale.split('-')[1];
    }

    if (
        locality.country === 'CA' && ['fr', 'en'].includes(locality.language) ||
        locality.country === 'BE' && ['fr', 'nl'].includes(locality.language)
    ) {
        //special case for @target in [BEFR, BENL, CAFR, CAEN]
        lookupContext.target = 'EBAY_' + locality.country + locality.language.toUpperCase();
    }

    return new ContentManager(lookupContext);
}

exports.fromLocality = fromLocality;

exports.fromRequest = function (req) {
    var ebay = req.ebay || (req.ebay = {});

    var i18n = req.ebay && req.ebay.i18n;

    if (!i18n) {
        const locality = req && req.locality || {};
        // only allow accept languages for service and webview only
        const acceptLanguages = (req.ebay.isService || req.ebay.webView) && req && req.ebay &&
            req.ebay.getAcceptLanguages && req.ebay.getAcceptLanguages();
        i18n = ebay.i18n = fromLocality(locality, acceptLanguages);
    }

    return i18n;
};

exports.fromRenderContext = function (out) {
    var global = out.global;
    var contentManager = global[OUT_GLOBAL_KEY];

    if (!contentManager) {
        var req = (out.global.platform ? out.global.platform.request : out.global.request) || getRequest(out);

        if (req) {
            contentManager = exports.fromRequest(req);
        } else {
            var locality = global.locality || {};
            contentManager = fromLocality(locality);
        }

        global[OUT_GLOBAL_KEY] = contentManager;
    }

    return contentManager;
};

exports.fromOptimizerContext = function (optimizerContext) {
    var i18n = optimizerContext.data.ebayI18n;
    if (!i18n) {
        var renderContext = optimizerContext.data.renderContext;

        if (renderContext) {
            i18n = exports.fromRenderContext(renderContext);
        } else {
            i18n = new ContentManager({
                target: 'EBAY_US',
                locale: 'en-US'
            });
        }

        optimizerContext.data.ebayI18n = i18n;
    }
    return i18n;
};

exports.getDefault = function () {
    var req = getRequestFromRequestLocal();
    if (!req) {
        throw new Error('Unable to resolve content manager using request local lookup');
    }
    return exports.fromRequest(req);
};

exports.getOptional = function () {
    var req = getRequestFromRequestLocal();
    return req && exports.fromRequest(req);
};
