// @ebay/chomsky-client/lib/models/index.js
// Source of model definitions https://wiki.corp.ebay.com/pages/viewpage.action?spaceKey=COREAI&title=Chomsky+Model+Enumeration
// Java Implementation https://github.corp.ebay.com/KHF/chomskygw/blob/master/src/main/java/com/ebay/app/model/Model.java
const MODELS =  {
    OPENAI : {
        CHAT : {
            AZURE_CHAT_COMPLETIONS_GPT35_TURBO_0125: "azure-chat-completions-gpt35-turbo-0125",
            AZURE_CHAT_COMPLETIONS_GPT_4_TURBO_2024_04_09: "azure-chat-completions-gpt-4-turbo-2024-04-09",
            AZURE_CHAT_COMPLETIONS_GPT_4O_2024_05_13: "azure-chat-completions-gpt-4o-2024-05-13",
            AZURE_CHAT_COMPLETIONS_GPT_4O_MINI_2024_07_18: "azure-chat-completions-gpt-4o-mini-2024-07-18",
            AZURE_CHAT_COMPLETIONS_GPT4_0314: "azure-chat-completions-gpt4-0314",
            AZURE_CHAT_COMPLETIONS_GPT_4_1_2025_04_14: "azure-chat-completions-gpt-4-1-2025-04-14",
            AZURE_CHAT_COMPLETIONS_GPT_4_1_NANO_2025_04_14: "azure-chat-completions-gpt-4-1-nano-2025-04-14",
            AZURE_CHAT_COMPLETIONS_GPT_4_1_MINI_2025_04_14: "azure-chat-completions-gpt-4-1-mini-2025-04-14",
            AZURE_CHAT_COMPLETIONS_GPT_5_2025_08_07: "azure-chat-completions-gpt-5-2025-08-07",
            AZURE_CHAT_COMPLETIONS_GPT_5_CHAT_2025_08_07: "azure-chat-completions-gpt-5-chat-2025-08-07",
            AZURE_CHAT_COMPLETIONS_GPT_5_MINI_2025_08_07: "azure-chat-completions-gpt-5-mini-2025-08-07",
            AZURE_CHAT_COMPLETIONS_GPT_5_NANO_2025_08_07: "azure-chat-completions-gpt-5-nano-2025-08-07",
            AZURE_CHAT_COMPLETIONS_O1_MINI_2024_09_12: "azure-chat-completions-o1-mini-2024-09-12",
            AZURE_CHAT_COMPLETIONS_O3_MINI_2025_01_31: "azure-chat-completions-o3-mini-2025-01-31",
            AZURE_CHAT_COMPLETIONS_O4_MINI_2025_04_16: "azure-chat-completions-o4-mini-2025-04-16",
            AZURE_CHAT_COMPLETIONS_GPT35_TURBO_0125_SANDBOX: "azure-chat-completions-gpt35-turbo-0125-sandbox",
            AZURE_CHAT_COMPLETIONS_GPT_4_TURBO_2024_04_09_SANDBOX: "azure-chat-completions-gpt-4-turbo-2024-04-09-sandbox",
            AZURE_CHAT_COMPLETIONS_GPT_4O_2024_05_13_SANDBOX: "azure-chat-completions-gpt-4o-2024-05-13-sandbox",
            AZURE_CHAT_COMPLETIONS_GPT_4O_MINI_2024_07_18_SANDBOX: "azure-chat-completions-gpt-4o-mini-2024-07-18-sandbox",
            AZURE_CHAT_COMPLETIONS_GPT_4_1_2025_04_14_SANDBOX: "azure-chat-completions-gpt-4-1-2025-04-14-sandbox",
            AZURE_CHAT_COMPLETIONS_GPT_4_1_NANO_2025_04_14_SANDBOX: "azure-chat-completions-gpt-4-1-nano-2025-04-14-sandbox",
            AZURE_CHAT_COMPLETIONS_GPT_4_1_MINI_2025_04_14_SANDBOX: "azure-chat-completions-gpt-4-1-mini-2025-04-14-sandbox",
            AZURE_CHAT_COMPLETIONS_GPT_5_2025_08_07_SANDBOX: "azure-chat-completions-gpt-5-2025-08-07-sandbox",
            AZURE_CHAT_COMPLETIONS_GPT_5_CHAT_2025_08_07_SANDBOX: "azure-chat-completions-gpt-5-chat-2025-08-07-sandbox",
            AZURE_CHAT_COMPLETIONS_GPT_5_MINI_2025_08_07_SANDBOX: "azure-chat-completions-gpt-5-mini-2025-08-07-sandbox",
            AZURE_CHAT_COMPLETIONS_GPT_5_NANO_2025_08_07_SANDBOX: "azure-chat-completions-gpt-5-nano-2025-08-07-sandbox",
            AZURE_CHAT_COMPLETIONS_O1_MINI_2024_09_12_SANDBOX: "azure-chat-completions-o1-mini-2024-09-12-sandbox",
            AZURE_CHAT_COMPLETIONS_O3_MINI_2025_01_31_SANDBOX: "azure-chat-completions-o3-mini-2025-01-31-sandbox",
            AZURE_CHAT_COMPLETIONS_O4_MINI_2025_04_16_SANDBOX: "azure-chat-completions-o4-mini-2025-04-16-sandbox",
        },
        TEXT: {
        },
        EMBEDDING: {
            AZURE_EMBEDDINGS_ADA_TEXT_SIMILARITY_2: "azure-embeddings-ada-text-similarity-2",
            AZURE_EMBEDDINGS_ADA_TEXT_SIMILARITY_2_SANDBOX: "azure-embeddings-ada-text-similarity-2-sandbox"
        },
        IMAGE: {
            AZURE_TEXT_TO_IMAGE_DALLE: "azure-text-to-image-dalle",
            AZURE_TEXT_TO_IMAGE_DALLE_IMAGE_DETAILS: "azure-text-to-image-dalle-image-details",
        }
    },
    GCP : {
        CHAT : {
            GCP_CHAT_COMPLETIONS_CHAT_GEMINI_1_5_PRO_002: "gcp-chat-completions-chat-gemini-1.5-pro-002",
            GCP_CHAT_COMPLETIONS_CHAT_GEMINI_1_5_FLASH_002: "gcp-chat-completions-chat-gemini-1.5-flash-002",
            GCP_CHAT_COMPLETIONS_CHAT_GEMINI_2_0_FLASH_001: "gcp-chat-completions-chat-gemini-2.0-flash-001",
            GCP_CHAT_COMPLETIONS_CHAT_GEMINI_2_0_FLASH_LITE_001: "gcp-chat-completions-chat-gemini-2.0-flash-lite-001",
            GCP_CHAT_COMPLETIONS_CHAT_GEMINI_2_5_FLASH: "gcp-chat-completions-chat-gemini-2.5-flash",
            GCP_CHAT_COMPLETIONS_CHAT_GEMINI_2_5_PRO: "gcp-chat-completions-chat-gemini-2.5-pro",
            GCP_CHAT_COMPLETIONS_ANTHROPIC_CLAUDE_3_5_SONNET_V2: "gcp-chat-completions-anthropic-claude-3.5-sonnet-v2",
            GCP_CHAT_COMPLETIONS_ANTHROPIC_CLAUDE_3_7_SONNET: "gcp-chat-completions-anthropic-claude-3.7-sonnet",
            GCP_CHAT_COMPLETIONS_ANTHROPIC_CLAUDE_SONNET_4: "gcp-chat-completions-anthropic-claude-sonnet-4",
            GCP_CHAT_COMPLETIONS_ANTHROPIC_CLAUDE_OPUS_4: "gcp-chat-completions-anthropic-claude-opus-4",
            GCP_CHAT_COMPLETIONS_CHAT_GEMINI_1_5_PRO_002_SANDBOX: "gcp-chat-completions-chat-gemini-1.5-pro-002-sandbox",
            GCP_CHAT_COMPLETIONS_CHAT_GEMINI_1_5_FLASH_002_SANDBOX: "gcp-chat-completions-chat-gemini-1.5-flash-002-sandbox",  
            GCP_CHAT_COMPLETIONS_CHAT_GEMINI_2_0_FLASH_001_SANDBOX: "gcp-chat-completions-chat-gemini-2.0-flash-001-sandbox",
            GCP_CHAT_COMPLETIONS_CHAT_GEMINI_2_0_FLASH_LITE_001_SANDBOX: "gcp-chat-completions-chat-gemini-2.0-flash-lite-001-sandbox",
            GCP_CHAT_COMPLETIONS_CHAT_GEMINI_2_5_FLASH_SANDBOX: "gcp-chat-completions-chat-gemini-2.5-flash-sandbox",
            GCP_CHAT_COMPLETIONS_CHAT_GEMINI_2_5_PRO_SANDBOX: "gcp-chat-completions-chat-gemini-2.5-pro-sandbox",
            GCP_CHAT_COMPLETIONS_ANTHROPIC_CLAUDE_3_5_SONNET_V2_SANDBOX: "gcp-chat-completions-anthropic-claude-3.5-sonnet-v2-sandbox",
            GCP_CHAT_COMPLETIONS_ANTHROPIC_CLAUDE_3_7_SONNET_SANDBOX: "gcp-chat-completions-anthropic-claude-3.7-sonnet-sandbox",
            GCP_CHAT_COMPLETIONS_ANTHROPIC_CLAUDE_SONNET_4_SANDBOX: "gcp-chat-completions-anthropic-claude-sonnet-4-sandbox",
            GCP_CHAT_COMPLETIONS_ANTHROPIC_CLAUDE_OPUS_4_SANDBOX: "gcp-chat-completions-anthropic-claude-opus-4-sandbox" 
        },
        TEXT: {
        },
        IMAGE: {
            GCP_TEXT_TO_IMAGE_STABLE_DIFFUSION: "gcp-text-to-image-stable-diffusion"
        }
    },
    EBAY : {
        CHAT : {
            EBAY_INTERNAL_CHAT_COMPLETIONS_LLAMA_70B_CHAT: "ebay-internal-chat-completions-llama-70b-chat",
            EBAY_INTERNAL_CHAT_COMPLETIONS_LLAMA_3_8B_INSTRUCT: "ebay-internal-chat-completions-llama-3-8b-instruct",
            EBAY_INTERNAL_CHAT_COMPLETIONS_LLAMA_3_70B_INSTRUCT: "ebay-internal-chat-completions-llama-3-70b-instruct",
            EBAY_INTERNAL_CHAT_COMPLETIONS_ELLAMA_3_8B_INSTRUCT: "ebay-internal-chat-completions-ellama-3-8b-instruct",
            EBAY_INTERNAL_CHAT_COMPLETIONS_ELLAMA_3_70B_INSTRUCT: "ebay-internal-chat-completions-ellama-3-70b-instruct",
            EBAY_INTERNAL_CHAT_COMPLETIONS_LILIUM_3_4B_INSTRUCT: "ebay-internal-chat-completions-lilium-3-4b-instruct",
            EBAY_INTERNAL_CHAT_COMPLETIONS_LILIUM_3_8B_INSTRUCT: "ebay-internal-chat-completions-lilium-3-8b-instruct",
            EBAY_INTERNAL_CHAT_COMPLETIONS_ELLAMA_VL_8B_PREVIEW: "ebay-internal-chat-completions-ellama-vl-8b-preview",
            EBAY_INTERNAL_CHAT_COMPLETIONS_QWEN_3_8B: "ebay-internal-chat-completions-qwen-3-8B",
            EBAY_INTERNAL_CHAT_COMPLETIONS_SANDBOX_LLAMA_3_8B_INSTRUCT: "ebay-internal-chat-completions-sandbox-llama-3-8b-instruct",
            EBAY_INTERNAL_CHAT_COMPLETIONS_SANDBOX_LLAMA_3_70B_INSTRUCT: "ebay-internal-chat-completions-sandbox-llama-3-70b-instruct",
            EBAY_INTERNAL_CHAT_COMPLETIONS_SANDBOX_ELLAMA_3_8B_INSTRUCT: "ebay-internal-chat-completions-sandbox-ellama-3-8b-instruct",
            EBAY_INTERNAL_CHAT_COMPLETIONS_SANDBOX_ELLAMA_3_70B_INSTRUCT: "ebay-internal-chat-completions-sandbox-ellama-3-70b-instruct",
            EBAY_INTERNAL_CHAT_COMPLETIONS_SANDBOX_LILIUM_3_4B_INSTRUCT: "ebay-internal-chat-completions-sandbox-lilium-3-4b-instruct",
            EBAY_INTERNAL_CHAT_COMPLETIONS_SANDBOX_LILIUM_3_8B_INSTRUCT: "ebay-internal-chat-completions-sandbox-lilium-3-8b-instruct",
            EBAY_INTERNAL_CHAT_COMPLETIONS_SANDBOX_ELLAMA_VL_8B_PREVIEW: "ebay-internal-chat-completions-sandbox-ellama-vl-8b-preview",
            EBAY_INTERNAL_CHAT_COMPLETIONS_SANDBOX_QWEN_3_8B: "ebay-internal-chat-completions-sandbox-qwen-3-8B"
        },
        TEXT: {
            EBAY_INTERNAL_TEXT_COMPLETIONS_LLAMA_70B_CHAT: "ebay-internal-text-completions-llama-70b-chat",
        },
        EMBEDDING: {
            EBAY_INTERNAL_VECTOR_PRIME: "ebay-internal-vector-prime",
            EBAY_INTERNAL_VECTOR_PRIME_2: "ebay-internal-vector-prime-2",
            EBAY_INTERNAL_SANDBOX_VECTOR_PRIME: "ebay-internal-sandbox-vector-prime",
            EBAY_INTERNAL_SANDBOX_VECTOR_PRIME_2: "ebay-internal-sandbox-vector-prime-2"
        }
    }
}


module.exports = MODELS;
