// @ebay/commons-inc/lib/utils.js
'use strict';

var path = require('path'),
    debug = require('debug')('@ebay/commons-inc:utils'),
    env = require('@ebay/environment-ebay'),
    appContext = require('@ebay/app-context-ebay'),
    tryRequire = require('try-require');

var debugInfo;

function getPkgJSON() {
    var root = process.cwd();
    return tryRequire(path.join(root, 'package'));
}

/*
 * Build Debug info
 * Read manifect.json for scm and version info
 * if this json file is missing (in local), using git commands build scm info
 */
function buidDebugInfo() {
    var root = process.cwd(),
        manifectJson = tryRequire(path.join(root, 'manifest.json'));

    if (!debugInfo) {
        if (manifectJson) {
            debugInfo = {
                scm : manifectJson.scm,
                version : manifectJson.version
            };
        }
    }
    return debugInfo;
}

// Function to remove userId and deviceId from a string
function removeUserIdDeviceIdAndIP(value) {
    var modifiedValue = value.replace(/(origAcctId|guessedUserId|deviceId|ip|buid)=[^,]+,?/g, '');
    // Remove UrlEncoded values for origAcctId like: %2CorigAcctId%3D1111111111%2C
    return modifiedValue.replace(/%2CorigAcctId%3D[^,]+%2C/g, '');
}

module.exports = {

    getPkgJSON: getPkgJSON,

    buidDebugInfo: buidDebugInfo,

    tryGetGuid: function tryGetGuid(req) {
        try {
            return req.ebay && req.ebay.getGuid();
        }
        catch (err) {
            debug(err);
        }
    },

    tryGetCorrId: function tryGetCorrId(req) {
        try {
            return req.ebay && req.ebay.getCorrelationRequestId() && req.ebay.getCorrelationRequestId().getCorrelationRequestId();
        }
        catch (err) {
            debug(err);
        }
    },

    tryGetNodeId: function tryGetNodeId(req) {
        try {
            return req.ebay && req.ebay.getCorrelationRequestId() && req.ebay.getCorrelationRequestId().getNodeId();
        }
        catch (err) {
            debug(err);
        }
    },

    tryGetRequestId: function tryGetRequestId(req) {
        try {
            return req.ebay && req.ebay.getCorrelationRequestId() && req.ebay.getCorrelationRequestId().getRequestId();
        }
        catch (err) {
            debug(err);
        }
    },

    tryGetRlogId: function tryGetRlogId(req) {
        try {
            return req.ebay && req.ebay.getRlogId();
        }
        catch (err) {
            debug(err);
        }
    },

    tryGetRcmdId: function tryGetRcmdId(req) {
        try {
            return req.ebay && req.ebay.getPageName();
        }
        catch (err) {
            debug(err);
        }
    },

    tryGetSiteId: function tryGetSiteId(req) {
        try {
            return req.ebay && req.ebay.getSiteId();
        }
        catch (err) {
            debug(err);
        }
    },

    tryGetEnvironment: function tryGetEnvironment(req) {
        return env._env();
    },

    tryGetAppName: function tryGetAppName(req) {
        return appContext.appName;
    },

    tryGetPageId: function tryGetPageId(req) {
        try {
            return req.ebay && req.ebay.getPageId();
        }
        catch (err) {
            debug(err);
        }
    },

    generateCosHeadersFunction(req, removeEndUserCtx) {
        if (req.ebay && req.ebay.buildOutboundContext && req.ebay.cookies && req.route?.config?.cosHeadersInfo !== false) {
            const outboundContext = req.ebay.buildOutboundContext();
            req.ebay.getCorrelationTracking().update(outboundContext);

            // Create a new object with modified values
            const modifiedOutboundContext = {};
            for (const key in outboundContext) {
                if (key === 'X-EBAY-C-TRACKING' || ( key === 'X-EBAY-C-ENDUSERCTX' && !removeEndUserCtx)) {
                    modifiedOutboundContext[key] = removeUserIdDeviceIdAndIP(outboundContext[key]);
                }
            }

            if (removeEndUserCtx) {
                return `function(scope){var CosHeaders=${
                    JSON.stringify({
                        X_EBAY_C_TRACKING: modifiedOutboundContext['X-EBAY-C-TRACKING']
                    })
                };scope.cosHeadersInfo={getCosHeaders:function(){return CosHeaders;},getTrackingHeaders:function(){return CosHeaders.X_EBAY_C_TRACKING;}};}`;
            }

            return `function(scope){var CosHeaders=${
                JSON.stringify({
                    X_EBAY_C_TRACKING: modifiedOutboundContext['X-EBAY-C-TRACKING'],
                    X_EBAY_C_ENDUSERCTX:modifiedOutboundContext['X-EBAY-C-ENDUSERCTX']
                })
            };scope.cosHeadersInfo={getCosHeaders:function(){return CosHeaders;},getTrackingHeaders:function(){return CosHeaders.X_EBAY_C_TRACKING;},getEndUserContext:function(){return CosHeaders.X_EBAY_C_ENDUSERCTX;}};}`;
        }
    }
};
