// @ebay/security-ebay/token/credentials.js
'use strict';

var pathNode = require('path');

var env = require('@ebay/environment-ebay');
var esams = require('@ebay/esams-consumer-ebay');

var RuntimeError = require('@ebay/errors-ebay/errors').RuntimeError;

var CHANNEL = 'site.iaf.nodejs.auth.consumer';

var esamsConsumer;

/*
    This esams functionality is used to manage access to validation API that requires
    explicit whitelisting in production, which is kind of a big friction point for any application
    thath would like to use security API.

    Mainly, this is used by nodejs service stack where each app is required to use OAuth2 security
    according to COS/DECO standard.

    This module is also used by Frontend applications via auth-ebay module
    to support BUSINESS_POLICY based on IAF security token.
*/

function getEsamsConsumer() {

    if (esamsConsumer) {
        return esamsConsumer;
    }

    var identifyFile = (env.isProd() || env.isPreProd() || env.isSandbox()) ? undefined :
        pathNode.join(__dirname, '../config/unp.xml');

    esamsConsumer = esams.getEsamsConsumer(identifyFile);

    return esamsConsumer;
}

module.exports.getEsamsConsumer = getEsamsConsumer;

module.exports.getSecret = function getSecret() {
    return new Promise((resolve, reject) => {
        getEsamsConsumer().getNonKey(CHANNEL, -1, function(err, nonKey) {
            var value;

            if (err) {
                return reject(new RuntimeError(err));
            }

            var envName = selectKeyEnv();
            try {
                value = JSON.parse(nonKey.value);
                value = value[envName];
            }
            catch (err) {
                return reject(new RuntimeError({
                    message: 'Failed to parse nonKey value' +
                    ' while getting secret, env: ' + envName,
                    error: err,
                }));
            }

            if (value) {
                return resolve(value);
            }

            reject(new RuntimeError('Failed to find secret key, env: ' + envName));
        });
    });
};

function getConsumerId() {
    // consumer id for unp nodejs stack app
    // this app is allowed to use IAF validation
    return 'urn:ebay-marketplace-consumerid:7e80615d-56f3-4796-a265-2298a5312ce2';
}

function getDomain() {
    return 'EBAYAPP';
}

module.exports.get = function () {
    return module.exports.getSecret().then(secret => {
        return {
            domain: getDomain(),
            consumerId: getConsumerId(),
            secret: secret
        };
    });
};

function selectKeyEnv() {
    if (env.isProd() || env.isPreProd()) {
        return 'production';
    }
    if (env.isSandbox()) {
        return 'sandbox';
    }
    return 'staging';
}
