// @ebay/security-ebay/token/index.js
'use strict';

const Credentials = require('./credentials');
const IafEbay = require('@ebay/iaf-ebay');
const Iaf = IafEbay.iaf;
const debug = require('debug')('security-ebay:token');
const parse = IafEbay.parse;

function loadToken(tokenValue) {
    const meta = parse(tokenValue);
    if (meta.isByReference()) {
        return loadTokenByRef(meta);
    }
    return Promise.resolve(meta.token);
}

function loadTokenByRef(meta) {
    return Credentials.get()
    .then(creds => {
        debug('loading token by ref');
        return new Promise((resolve, reject) => {
            Iaf.getSecurityTokenByRef({
                consumerId: creds.consumerId,
                secret: creds.secret,
                reference: meta.token
            }, (err, token) => {
                if (err) {
                    return reject(err);
                }
                resolve(token);
            });
        });
    });
}

/**
 * @callback returns token
*/
function validateTokenAndGetAttributes(token) {
    return Credentials.get()
    .then(creds => {
        return authenticateToken({
            consumerId: creds.consumerId,
            secret: creds.secret,
            targetService: 'ScopedToken',
            token: token
        })
        .catch(function (err) {
            if (err.code === 'bad-token') {
                return authenticateToken({
                    consumerId: creds.consumerId,
                    secret: creds.secret,
                    targetService: '',
                    token: token
                });
            }
            throw err;
        });
    });
}

function authenticateToken(options) {
    return new Promise((resolve, reject) => {
        Iaf.authenticateToken(options, (err, token, attributes) => {
            if (err) {
                return reject(err);
            }
            resolve({
                token: token,
                attributes: attributes
            });
        });
    });
}

function openToken(tokenValue) {
    return loadToken(tokenValue)
    .then(validateTokenAndGetAttributes);
}

module.exports = {
    loadToken,
    validateTokenAndGetAttributes,
    openToken
};
