// @ebay/site-meta-ebay/lib/site-meta-ebay/index.js
'use strict';

var metadata = require('./metadata');

var ebaySiteEnum = {
    getSiteByName: function(key) {
        return metadata.getSiteByName(key);
    },
    getCountries: function() {
        return metadata.getCountries();
    },
    getSiteByNameOrElse: function(key, elseReturn) {
        if (metadata.getSiteByName(key)) {
            return metadata.getSiteByName(key);
        } else {
            return elseReturn;
        }
    },
    getCurrencies: function() {
        return metadata.getCurrencies();
    },
    getCountriesBySite: function(site) {
        if (site) {
            return metadata.getCountriesBySite(site.id);
        }
    },
    getNationalitiesBySite: function(site) {
        if (site) {
            return metadata.getNationalitiesBySite(site.id);
        }
    },
    getSiteById: function(siteId) {
        return metadata.getSiteById(siteId);
    },
    getStatesBySite: function(site) {
        if (site) {
            return metadata.getStatesBySite(site.id);
        }
    },
    getStatesBySiteId: function(siteId) {
        var site = metadata.getSiteById(siteId);

        if (site) {
            return metadata.getStatesBySite(site.id);
        }
        return null;
    },
    getStatesListByName: function(name) {
        return metadata.getStatesBySite(name);
    }
};

module.exports = ebaySiteEnum;
