// @ebay/site-meta-ebay/lib/site-meta-ebay/metadata.js
'use strict';

var util = require('util');
var metadataParser = require('./metadata-parser');
var ebaySites = require('@ebay/site-meta-resources-ebay/resources/EbaySites');
var localizedListIndex = require('@ebay/site-meta-resources-ebay/resources/localized-lists/index');

var siteEnumsByName = {};
var localizedListBasePath = '@ebay/site-meta-resources-ebay/resources/localized-lists';

var metadata = module.exports = {
    SiteEnum: {},
    Countries: {},
    Currencies: {},
    LocalizedLists: {},
    addSiteByName: function(siteEnum) {
        this.SiteEnum[siteEnum.id] = siteEnum;

    },
    addCountries: function(countries) {
        this.Countries = countries;
    },
    addCurrencies: function(currencies) {
        this.Currencies = currencies;
    },
    setLocalizedLists: function(siteName, dataArgs) {
        if (!this.getSiteByName(siteName.id).localizedList) {
            this.getSiteByName(siteName.id).localizedList = {};
        }
        this.getSiteByName(siteName.id).localizedList[dataArgs.key] = dataArgs.value;
    },
    getSiteByName: function(id) {
        return this.SiteEnum[id];
    },
    getSiteById: function(siteId) {
        var sitename = ebaySites.EbaySiteIds[siteId];
        return this.getSiteByName(sitename);
    },
    getCountries: function() {
        return this.Countries;
    },
    getCurrencies: function() {
        return this.Currencies;
    },
    getStatesBySite: function(siteName) {

        if (siteName === 'EBAY_US') {
            siteName = 'EBAY_MAIN';
        }

        var site = this.getSiteByName(siteName);
        if (site) {
            if (!site.localizedList) {
                site.localizedList = getLocalizedList(site);
            }
            return site.localizedList ? site.localizedList.states : site.localizedList;
        } else {
            return null;
        }

    },
    getCountriesBySite: function(siteName) {

        if (siteName === 'EBAY_US') {
            siteName = 'EBAY_MAIN';
        }

        var site = this.getSiteByName(siteName);
        if (site) {
            if (!site.localizedList) {
                site.localizedList = getLocalizedList(site);
            }
            return site.localizedList ? site.localizedList.countries : site.localizedList;
        } else {
            return null;
        }

    },

    getNationalitiesBySite: function(siteName) {

        if (siteName === 'EBAY_US') {
            siteName = 'EBAY_MAIN';
        }

        var site = this.getSiteByName(siteName);
        if (site) {
            if (!site.localizedList) {
                site.localizedList = getLocalizedList(site);
            }
            return site.localizedList ? site.localizedList.nationalities : site.localizedList;
        } else {
            return null;
        }

    }

};


function getLocalizedList(site) {
    if (site && ((localizedListIndex && localizedListIndex[site.value]))) {
        var datapath = util.format('%s/%s/%s.json', localizedListBasePath, site.id, site.id);
        return require(datapath);
    }
}

// lazy loading of SiteEnum metadata
var metaLoaded = false;

function loadMetadataSync() {
    if (metaLoaded) {
        return;
    }
    metaLoaded = true;
    // Delete the old getters
    delete metadata.SiteEnum;
    metadata.SiteEnum = {};
    metadataParser.parse(metadata); // Load the metadata and register it with this module
}

function lazyMetaProp(name) {
    Object.defineProperty(metadata, name, {
        configurable: true,
        get: function() {
            if (!metaLoaded) {
                loadMetadataSync();
            }
            return metadata[name];
        }
    });
}

lazyMetaProp('SiteEnum');

// Preload the SiteEnum metadata during the next CPU cycle
// NOTE: We do this so that we don't slow the startup if
//       if the SiteEnum metadata is not immediately needed
setTimeout(loadMetadataSync, 100);
