// @ebay/service-security-ebay/lib/token-utils/errors.js
'use strict';

const Enum = require('@ebay/common-types-ebay/enum');

class ErrorId {
    constructor(message, longMessage, cosErrId, httpStatus) {
        this.message = message;
        this.longMessage = longMessage;
        this.cosErrId = cosErrId;
        this.httpStatus = httpStatus;
    }
}

const Status = Enum.define({
    UNAUTHORIZED: 401,
    BAD_REQUEST: 400,
    INTERNAL_SERVER_ERROR: 500,
    FORBIDDEN: 403
});

const ErrorIds = Enum.define({
    INVALID_TOKEN: new ErrorId('Invalid access token',
        'Invalid access token. Check the value of the Authorization HTTP request header.', 1001, Status.UNAUTHORIZED),
    SERIALIZE_ERROR: new ErrorId('Invalid access token',
        'Invalid access token. Check the value of the Authorization HTTP request header.', 1001, Status.UNAUTHORIZED),
    INVALID_DOMAIN: new ErrorId('Invalid access token',
        'Invalid access token. Check the value of the Authorization HTTP request header.', 1001, Status.UNAUTHORIZED),
    MISSING_TOKEN: new ErrorId('Missing access token',
        'Access token is missing in the Authorization HTTP request header.', 1002,
        Status.BAD_REQUEST),
    MISSING_HEADER: new ErrorId('Missing access token',
        'Access token is missing in the Authorization HTTP request header.', 1002,
        Status.BAD_REQUEST),
    INVALID_TOKEN_TYPE: new ErrorId('Token type in the Authorization header is invalid:',
        'Token type in the Authorization header is invalid:', 1003, Status.BAD_REQUEST),
    SYSTEM_ERROR: new ErrorId('Internal error',
        'Error processing the access token.', 1004, Status.INTERNAL_SERVER_ERROR),
    CONNECTION_ERROR: new ErrorId('Internal error',
        'Error processing the access token.', 1004, Status.INTERNAL_SERVER_ERROR),
    INVALID_SESSION: new ErrorId('Invalid access token',
        'Invalid access token. Check the value of the Authorization HTTP request header.', 1001, Status.UNAUTHORIZED),
    INVALID_REQUEST_SIGNATURE: new ErrorId('Invalid token shield header',
        'Token shield header has invalid signature', 1001, Status.UNAUTHORIZED),
    REQUEST_SIGNATURE_ARTIFACT_NOT_EXISTS: new ErrorId('Invalid token shield header',
        'Token shield header is not present', 1001, Status.FORBIDDEN),
    INVALID_REQUEST_SIGNATURE_ARTIFACT: new ErrorId('Invalid token shield header',
        'Token shield header request signature does not match token', 1001, Status.UNAUTHORIZED)
});

module.exports = {
    valueOf(errorMessage) {
        try {
            return ErrorIds.valueOf(errorMessage).value;
        }
        catch (err) {
            return ErrorIds.SYSTEM_ERROR.value;
        }
    },
    ErrorIds,
    Status
};
