// @ebay/service-security-ebay/lib/token-utils/esams.js
'use strict';

const pathNode = require('path');

const env = require('@ebay/environment-ebay');
const esams = require('@ebay/esams-consumer-ebay');

const RuntimeError = require('../errors').RuntimeError;

const CHANNEL = 'site.iaf.nodejs.auth.consumer';

let esamsConsumer;

function getEsamsConsumer() {
    if (esamsConsumer) {
        return esamsConsumer;
    }

    const identifyFile = (env.isProd() || env.isPreProd() || env.isSandbox()) ? undefined :
        pathNode.join(__dirname, '../../config/unp.xml');

    esamsConsumer = esams.getEsamsConsumer(identifyFile);

    return esamsConsumer;
}

module.exports.getSecret = function getSecret() {
    return new Promise((resolve, reject) => {
        getEsamsConsumer().getNonKey(CHANNEL, -1, (err, nonKey) => {
            let value;

            if (err) {
                return reject(new RuntimeError(err));
            }

            const envName = selectKeyEnv();
            try {
                value = JSON.parse(nonKey.value);
                value = value[envName];
            }
            catch (err) {
                return reject(new RuntimeError({
                    message: `Failed to parse nonKey value' +
                    ' while getting secret, env: ${envName}`,
                    error: err
                }));
            }

            if (value) {
                return resolve(value);
            }

            reject(new RuntimeError(`Failed to find secret key, env: ${envName}`));
        });
    });
};

module.exports.getCertificate = async (channel, version = -1) => {
    const keyPairHolder = await esams.getNameService().getKeyPair(channel, version);
    return keyPairHolder?.certificates?.[0];
};

module.exports.getPrivateKey = async (channel, version = -1) => {
    const keyPairHolder = await esams.getNameService().getKeyPair(channel, version);
    return keyPairHolder?.privateKey;
};

function getConsumerId() {
    // consumer id for unp app
    return 'urn:ebay-marketplace-consumerid:7e80615d-56f3-4796-a265-2298a5312ce2';
}

function getDomain() {
    return 'EBAYAPP';
}

module.exports.getConfig = async function getConfig() {
    const secret = await module.exports.getSecret();
    return {
        domain: getDomain(),
        consumerId: getConsumerId(),
        secret: secret
    };
};

function selectKeyEnv() {
    if (env.isProd() || env.isPreProd()) {
        return 'production';
    }
    if (env.isSandbox()) {
        return 'sandbox';
    }
    return 'staging';
}
