// @ebay/service-security-ebay/lib/token-utils/load-token.js
'use strict';

const Esams = require('./esams');
const iafEbay = require('@ebay/iaf-ebay');
const iaf = iafEbay.iaf;
const debug = require('debug')('service-security-ebay:token-utils:load-token');
const parse = iafEbay.parse;
const predicate = require('./utils').predicate;

module.exports = function load(token) {
    return Promise.resolve(parse(token))
        .then(
            predicate(isReference)
                .then(loadTokenByRef)
                .otherwise(meta => meta.token)
        );
};

function isReference(meta) {
    return meta.isByReference();
}

function loadTokenByRef(meta) {
    return Esams.getConfig()
        .then(
            config => {
                debug('loading token by ref');
                return new Promise((resolve, reject) => iaf.getSecurityTokenByRef({
                    consumerId: config.consumerId,
                    secret: config.secret,
                    reference: meta.token
                }, (err, token) => {
                    if (err) {
                        return reject(err);
                    }
                    resolve(token);
                }));
            }
        );
}
