// @ebay/service-security-ebay/lib/token-utils/utils.js
'use strict';

const crypto = require('crypto');
const tryRequire = require('try-require');
const Path = require('path');

module.exports.predicate = function predicate(conditionFn) {
    const choose = function choose() {
        const args = [].slice.call(arguments);
        return conditionFn(...args) ?
            this.thenFn(...args) :
            this.otherwiseFn && this.otherwiseFn(...args) || args[0];
    };

    return {
        then: function (fn) {
            this.thenFn = fn;
            return this;
        },
        otherwise: function (fn) {
            this.otherwiseFn = fn;
            return choose.bind(this);
        }
    };
};

module.exports.checksum = function checksum(data) {
    const shasum = crypto.createHash('sha1');
    shasum.update(data);
    return shasum.digest('hex');
};

module.exports.getDefaultConsumerId = function getDefaultConsumerId() {
    const pkg = tryRequire(Path.resolve(process.cwd(), 'package.json'));
    let consumerId = pkg && pkg.ebay && pkg.ebay['consumer-id'];
    consumerId = consumerId || pkg && pkg.gpaas && pkg.gpaas['consumer-id'];
    return consumerId;
};
