// @ebay/service-security-ebay/lib/token-utils/validate-user-token.js
'use strict';

const Provider = require('@ebay/service-client-ebay');
const AppContext = require('@ebay/app-context-ebay');
const Errors = require('./errors');

function extractClient(req) {
    const ua = req.ebay.getUserAgent();
    if (ua) {
        const backslashIndex = ua.indexOf('/');
        if (backslashIndex > 0) {
            return ua.substring(0, backslashIndex);
        }
    }
}

function extractAppVersion(req) {
    const ua = req.ebay.getUserAgent();
    if (ua) {
        const backslashIndex = ua.indexOf('/');
        if (backslashIndex > 0) {
            return ua.substring(backslashIndex + 1);
        }
    }
}

function extractToken(req) {
    return req && req.headers && req.headers.authorization &&
        req.headers.authorization.split(/\s+/).pop();
}

function extractSignatureHeader(req) {
    return req && req.headers && req.headers &&
        req.headers['x-ebay-c-request-auth'];
}

function extractIntentHeader(req) {
    return req && req.headers && req.headers &&
        req.headers['ebay-ets-api-intent'];
}

function isSecurityCriticalRequest(req) {
    if (req && req.route && req.route.config && req.route.config.isCritical) {
        return true;
    }
    return false;
}

function validateRequestUserToken(req) {
    const requestToken = extractToken(req);
    const client = extractClient(req);
    const appVersion = extractAppVersion(req);
    const requestSignatureHeader = extractSignatureHeader(req);
    const isCritical = isSecurityCriticalRequest(req);

    return new Promise((resolve, reject) => {
        const requestBody = {
            consumerId: AppContext.consumerId,
            requestToken,
            appName: AppContext.appName,
            client,
            appVersion,
            method: req.method,
            requestUri: `${req.baseUrl || ''}${req.path}`,
            allowedDomains: 'EBAYUSER,EBAYAPP',
            siteId: req.ebay.getSiteId(),
            requestSignatureHeader
        };

        const intent = extractIntentHeader(req);
        if (intent === 'foreground') {
            requestBody.foreground = true;
        }

        // Any request to usertokensessionmgmt with a flag critical=true enables MFA checks
        if (isCritical) {
            requestBody.critical = true;
        }

        Provider
            .context(req)
            .getClient('usertokensessionmgmt')
            .post(JSON.stringify(requestBody))
            .options({
                'accept-language': req.ebay.getAcceptLanguages().join(',')
            })
            .end((err, res) => {
                if (err) {
                    if (err.body && err.body.errors && err.body.errors.length) {
                        const errorId = Errors.valueOf(err.body.errors[0].message);
                        err.status = errorId.httpStatus.value;
                    }
                    return reject(err);
                }
                // make it compatible with other attributes
                const attrs = res.body;
                const subject = attrs.subject || {};

                const verification = attrs.tokenShieldVerificationTime !== undefined ? {
                    tokenShieldVerificationTime: attrs.tokenShieldVerificationTime,
                    tokenShieldVerificationStatus: attrs.tokenShieldVerificationStatus
                } : {};

                resolve({
                    token: `${subject.domain}/${subject.id}`,
                    attributes: {
                        expirationDate: attrs.expires,
                        domain: subject.domain,
                        accountId: attrs.userId,
                        userId: subject.id,
                        attributes: attrs.attributes,
                        ...verification
                    }
                });
            });
    });
}

module.exports = {
    validateRequestUserToken,
    extractToken,
    extractAppVersion,
    extractClient,
    isSecurityCriticalRequest
};
