// @ebay/errors-ebay/lib/mailer.js
'use strict';

var nodemailer = require('nodemailer');
var _ = require('underscore');
var InstanceDetector = require('@ebay/instance-detect-ebay');
var coloDetect = require('@ebay/colo-detect-ebay');
var appContext = require('@ebay/app-context-ebay');
var assert = require('assert');
var os = require('os');
var modConfig = require('@ebay/module-config-inc');

var DEFAULTS = {
    smtphost: coloDetect.coloMxDetect() || 'atom.corp.ebay.com',
    subject: 'Uncaught exception occurred on ',
    from: 'website@corp.ebay.com',
    to: 'DL-eBay-NodeJS-LiveMonitoring@corp.ebay.com',
    text: 'No information provided',
    subjectTail: ''
};

var instanceId = InstanceDetector.getInstanceId();

function sendEmail(mailOptions, callback) {
    callback = arguments[arguments.length - 1];
    mailOptions = mailOptions || {};

    assert.ok(callback, 'callback required');
    assert.equal(typeof callback === 'function', true, 'callback should be function');

    modConfig(module, function(err, config){
        mailOptions = buildEmailOptions(mailOptions);
        var subject = mailOptions.subject;
        if(!config.get('errors-ebay:email:enabled') ||
            (isOOM(subject) && !config.get('errors-ebay:email:sendOomEmail')) ||
            (isDeploy(subject) && !config.get('errors-ebay:email:sendDeployEmail')) ||
            (isUncaught(subject) && !config.get('errors-ebay:email:sendUncaughtEmail'))) {
                console.log('Skipping send email due to config override.', config.get('errors-ebay:email'));
                return callback(null, {});
        }

        // create reusable transporter object using SMTP transport
        var transporter = nodemailer.createTransport({
            host: mailOptions.smtphost
        });
        var now = Date.now();

        // send mail with defined transport object
        transporter.sendMail(mailOptions, function(error, info) {
            if (error) {
                console.error('Error Sending email: ', error);
            } else {
                info = info || {};
                info.mailOptions = mailOptions;
                console.log('Message Sent: ', info.response);
            }
            callback(error, info);
        });
    });
}

function buildEmailOptions(options) {
    var mailOptions = _.defaults(options||{}, _.omit(DEFAULTS, 'to'));

    //Subject:
    var subjectTail = appContext.appName + '('+ process.env.NODE_ENV +') ' + os.hostname() + ':' + instanceId + mailOptions.subjectTail;

    //To:
    var to = [].concat(appContext.teamDL);
    to.push(DEFAULTS.to);
    options.to && to.push(options.to);

    //Populate email options
    mailOptions.subject = getEmailCategory(mailOptions.subject) + mailOptions.subject + subjectTail;
    mailOptions.to = to;

    return mailOptions;
}

function isOOM(subject) {
    return subject.indexOf('Out of Memory') > -1;
}

function isDeploy(subject) {
    return subject.indexOf('Deploy') > -1;
}

function isUncaught(subject) {
    return subject.indexOf('Uncaught') > -1;
}

function getEmailCategory(subject) {
    if(isOOM(subject)) {
        return '[WARN] ';
    } else if(isDeploy(subject)) {
        return '[INFO] ';
    } else if(isUncaught(subject)) {
        return '[CRITICAL] ';
    } else {
        return '';
    }
}

module.exports.sendEmail = sendEmail;
