// @ebay/errors-ebay/lib/oom.js
'use strict';

const fs = require('fs');
const path = require('path');
const getPm2Id = require('@ebay/instance-detect-ebay').getInstanceId;
const MAX_HEAP = 1.2e9; // 1.2GB

module.exports.instance = undefined; // for testing purposes

function Oom() {
    const pm2id = getPm2Id();
    const logsFolder = process.env.LOG_HOME || './logs';
    ensureLogsFolderExists(logsFolder);
    this._oomFile = path.resolve(logsFolder + '/oom_' + pm2id);
    this._oom = readOomFile(this._oomFile);
}

/**
 * Get the OOM error count
 * @returns the number of OOM errors that have happened on this process
 */
Oom.prototype.getOom = function getOom() {
    return this._oom;
};

/**
 * Updates the OOM error count and stores it to disk
 * @param {Number} oomCount the new number of OOM errors to record
 */
Oom.prototype.setOom = function setOom(oomCount) {
    this._oom = oomCount;
    fs.writeFileSync(this._oomFile, oomCount.toString());
};

/**
 * Add 1 to the number of OOM error counts and write it to disk
 */
Oom.prototype.incOom = function incOom() {
    this.setOom(this._oom + 1);
};

function ensureLogsFolderExists(folder) {
    if (!fs.existsSync(folder)) {
        fs.mkdirSync(folder, { recursive: true });
    }
}

function readOomFile(file) {
    if (fs.existsSync(file)) {
        const contents = fs.readFileSync(file, 'utf8');
        return (contents && Number(contents)) || 0;
    }
    return 0; // default
}

/**
 * Gets the max heap size in bytes. There is no enforcement here.
 * This is just a number that will be enforced by another library (monitor-ebay currently)
 * @returns The maximum heap memory we allow
 */
module.exports.getMaxHeap = function getMaxHeap() {
    return MAX_HEAP;
};

module.exports.log = function createOom() {
    // only ever need one instance of the class to prevent inconsistencies
    if (!module.exports.instance) {
        module.exports.instance = new Oom();
    }
    return module.exports.instance;
};
