// @ebay/cookies-ebay/lib/ebay-cookies/EbayCookielet.js
'use strict';

var metadata = require('./metadata');
var Constants = metadata.Constants;

function EbayCookielet(meta, value) {
    if (!meta) {
        throw new Error('cookielet meta argument is required');
    }
    this.meta = meta;
    this.value = value;
    this._errMsg = null;
    this._deleted = false;
    this._creationDate = Date.now(); // in milliseconds
    // set max age
    this._maxAge = meta.maxAge;
    if (this._maxAge <= 0) {
        if (meta.parent.maxAge < 0) {
            this._maxAge = Constants.DEFAULT_SESSION_MAX_AGE;
        }
        else {
            this._maxAge = Constants.DEFAULT_MAX_AGE;
        }
    }
    this._expiryTime = parseInt(this._creationDate/1000, 10) + this._maxAge;
}

function isNumber(n) {
    return !isNaN(parseFloat(n)) && isFinite(n);
}

EbayCookielet.prototype = {
    getValue: function () {
        return this.value;
    },
    getId: function () {
        return this.meta.id;
    },
    isExpired: function () {
        return this._expiryTime ? Date.now() > this._expiryTime * 1000 : false;
    },
    getExpiryTime: function () {
        return this._expiryTime;
    },
    setExpiryTime: function (expiryTime) {
        this._expiryTime = parseInt(expiryTime, 10);
        if (isNumber(this._expiryTime)) {
            this._creationDate = (this._expiryTime - this.getMaxAge()) * 1000;
        }
    },
    isValid: function () {
        var _isValid = !this._errMsg &&
                    !this.isExpired() &&
                    this.value &&
                    this.meta !== metadata.Cookies.UNUSED;

        return _isValid;
    },
    needsSecurity: function () {
        return this.meta.signed;
    },
    getMaxAge: function () {
        return this._maxAge;
    },
    refresh: function () {
        if (!this.isValid()) {
            return;
        }
        var meta = this.meta;
        if (meta.autoRefresh) {
            var now = Date.now();
            this._expiryTime = parseInt(now / 1000 + this.getMaxAge(), 10);
            this._creationDate = now;
        }
    },
    getCreationDate: function () {
        return this._creationDate;
    },
    getDelimitedName: function () {
        return this.meta.name || this.meta.id;
    },
    setDeleted: function setDeleted(value) {
        this._deleted = value;
    },
    isDeleted: function isDeleted() {
        return this._deleted;
    }
};

module.exports = EbayCookielet;
