// @ebay/cookies-ebay/lib/ebay-cookies/EbayCookie.js
'use strict';

var calUtil = require('../utils/calUtils');
var tryRequire = require('try-require');
var pathTools = require('path');
var serializers = {};

function trySerializer(format) {
    if (!format) {
        return;
    }

    // lazy load of the file, avoid loading it evey time for every request
    var serializer = serializers[format] ||
        (serializers[format] = tryRequire(pathTools.resolve(__dirname, '../serializer/serializer-' + format)));

    return serializer;
}

function EbayCookie(meta, rawCookie) {
    if (!meta || !rawCookie) {
        throw new Error('Illegal arguments');
    }
    this.meta = meta;
    this._rawCookie = rawCookie;
    this._deserialized = !rawCookie.hasValue();
    // Mark the cookie as deserialized
    // if the raw cookie does not have a value
    this._valid = true;
    if (meta.maxAge) {
        rawCookie.setMaxAge(meta.maxAge);
    }
    rawCookie.setSameSite(meta.sameSite);
    rawCookie.setSecure(meta.secure === true);
    rawCookie.setHttpOnly(meta.httpOnly === true);
    rawCookie.setPartitioned(meta.partitioned === true);

    rawCookie._ebaySerializer = this.serializer = trySerializer(meta.format);
    if (!this.serializer) {
        throw new Error('Failed to load serializer-' + meta.format + ' for cookie: ' + meta.name);
    }
    rawCookie._ebayCookie = this;
    rawCookie.serialize = this._serialize;
}

EbayCookie.prototype = {
    _enforceValidation: function () {
        if (!this.isValidated()) {
            calUtil.createErrorEvent(new Error('The cookie ' + this.meta.name +
                ' is being accessed before it has been validated. Please invoke ' +
                'EbayCookie.validate() first or add the cookie middleware before this call.'));
            return false;
        }
        return true;
    },
    isValid: function () {
        return this._valid;
    },
    setValid: function (valid) {
        this._valid = valid;
    },
    _serialize: function () {
        if (this._ebayCookie._enforceValidation() && this._ebayCookie.isValid()) {
            return this._ebaySerializer.serialize(this._ebayCookie);
        } else {
            return this.getValue();
        }
    },
    deserialize: function () {
        if (!this._deserialized) {
            this._deserialized = true;
            try {
                this.serializer.deserialize(this, this._rawCookie);
                // Since the cookie was just created we mark it as not being modified
                this._rawCookie.setModified(false);
            } catch (err) {
                calUtil.createErrorEvent(this.meta.name + ' cookie is corrupted, error during deserialization is ' + (err.message || err.stack));
                this.setValid(false);
            }
            // However, if cookielets expired or checksum checks failed
            // then we may have to mark the cookie as modified (to remove expired cookielets)
            // or deleted if top-level checksums failed
            this.afterDeserialize();
            if (!this.isValid()) {
                calUtil.createWarnEvent('Deserialized cookies is not valid: ' + this.meta.idName);
                this._rawCookie.setModified(true);
            }
        }
    },
    isSession: function () {
        return this.meta.maxAge < 0;
    },
    remove: function () {
        this._rawCookie.setDeleted(true);
    }
};

module.exports = EbayCookie;
