// @ebay/cookies-ebay/lib/ebay-cookies/metadata.js
'use strict';

var metadataLoader = require('./metadata-loader');

var cookieMetasByName = {};
var cookieletMetasByName = {};

var metadata = module.exports = {
    Cookies: {},
    Constants: {},
    Flags: {},
    addCookieMeta: function (cookieMeta) {
        this.Cookies[cookieMeta.id] = cookieMeta;
        this.Cookies[cookieMeta.idName] = cookieMeta;
        var isCookielet = cookieMeta.cookielet;
        var name = cookieMeta.name;
        if (name) {
            if (isCookielet) {
                cookieletMetasByName[name] = cookieMeta;
            } else {
                cookieMetasByName[name] = cookieMeta;
            }
            this.Cookies[name] = cookieMeta;
        }
    },
    getCookieMetaByName: function (name, isCookielet) {
        return (isCookielet ? cookieletMetasByName : cookieMetasByName)[name] || this.Cookies[name];
    },
    addConstant: function (constantDef) {
        this.Constants[constantDef.name] = constantDef.value;
    },
    addFlag: function (flagDef) {
        this.Flags[flagDef.name] = flagDef;
    }
};

// lazy loading of cookie metadata
var metaLoaded = false;
function loadMetadataSync() {
    if (metaLoaded) {
        return;
    }
    metaLoaded = true;
    // Delete the old getters
    delete metadata.Cookies;
    delete metadata.Constants;
    metadata.Cookies = {};
    metadata.Constants = {};
    metadataLoader.load(metadata);    // Load the metadata and register it with this module
}

function lazyMetaProp(name) {
    Object.defineProperty(metadata, name, {
        configurable: true,
        get: function () {
            if (!metaLoaded) {
                loadMetadataSync();
            }
            return metadata[name];
        }
    });
}

lazyMetaProp('Cookies');
lazyMetaProp('Constants');

// Preload the cookie metadata during the next CPU cycle
// NOTE: We do this so that we don't slow the startup if
//       if the cookie metadata is not immediately needed
setTimeout(loadMetadataSync, 100);
