// @ebay/tracking-ebay/lib/cos/correlation-session.js
'use strict';

/**
 * Test it asble to parse header or req.
 */

const HMV = require('@ebay/header-multi-value-ebay');
const util = require('../utils');
const SojWriter = require('../soj/sojourner-writer');
const UserActionID = require('../tracking2/user-action-id');
const { PageSourceInfoCookielet } = require('../tracking2/page-source-info');
const SourceIdBuilder = require('../tracking2/source-id-builder');

const CONST = {
    X_EBAY_C_CORRELATION_SESSION: 'X-EBAY-C-CORRELATION-SESSION',

    C: 'c',                         // Click ID - String
    SI: 'si',                       // Stream ID	- String
    GCI: 'gci',                     // Global stream ID - optional - String
    SID: 'sid',                     // Source ID - String
    SIID: 'siid',                   // Source impression ID	- String
    UAID: 'uaid',                   // Global unique id - optional - String
    TRKP: 'trkp',                   // Application tracking parameters - String
    TRKGFLAGS: 'trk-gflgs',         // Global Tracking flags - String
    OPERATION_ID: 'operationId',     // Operation ID same as page id - Integer
    SERVICE_CORRELATION_ID: 'serviceCorrelationId', // Service correlation ID - String
    PAGE_IMPRESSION_ID: 'corr_pg_inst_id', // Correlation page instance ID - String
};

function createULID() {
    const currPrng = () => Math.random();
    const TIME_LEN = 10;
    const RANDOM_LEN = 16;
    const ENCODING = "0123456789ABCDEFGHJKMNPQRSTVWXYZ"; // Crockford's Base32
    const ENCODING_LEN = ENCODING.length;

    const encodeTime = (now, len) => {
        let mod;
        let str = "";
        for (; len > 0; len--) {
            mod = now % ENCODING_LEN;
            str = ENCODING.charAt(mod) + str;
            now = (now - mod) / ENCODING_LEN;
        }
        return str;
    };

    const encodeRandom = (len, prng) => {
        let str = "";
        for (; len > 0; len--) {
            let rand = Math.floor(prng() * ENCODING_LEN);
            if (rand === ENCODING_LEN) {
                rand = ENCODING_LEN - 1;
            }
            str = ENCODING.charAt(rand) + str;
        }
        return str;
    };

    return encodeTime(Date.now(), TIME_LEN) + encodeRandom(RANDOM_LEN, currPrng);
}

class CorrelationSession{
    constructor(ctx) {
        this.c = ctx[CONST.C];
        this.si = ctx[CONST.SI];
        this.gci = ctx[CONST.GCI];
        this.sid = ctx[CONST.SID];
        this.uaid = ctx[CONST.UAID];
        this.siid = ctx[CONST.SIID];
        this.trkp = ctx[CONST.TRKP];
        this.trkgFlags = ctx[CONST.TRKGFLAGS];
        this.operationId = ctx[CONST.OPERATION_ID];
        this.serviceCorrelationId = ctx[CONST.SERVICE_CORRELATION_ID];
        this.pageImpressionId = ctx[CONST.PAGE_IMPRESSION_ID];
    }

    /* Do not cache this if UAID is present count must be incremented */
    buildHeader() {
        const headerVal = HMV.create();
        headerVal.set(CONST.C, this.c);
        this.si && headerVal.set(CONST.SI, this.si);
        this.gci && headerVal.set(CONST.GCI, this.gci);
        this.sid && headerVal.set(CONST.SID, this.sid);
        this.siid && headerVal.set(CONST.SIID, this.siid);
        this.trkp && headerVal.set(CONST.TRKP, this.trkp);
        this.trkgFlags && headerVal.set(CONST.TRKGFLAGS, this.trkgFlags);
        this.operationId && headerVal.set(CONST.OPERATION_ID, this.operationId);
        this.serviceCorrelationId && headerVal.set(CONST.SERVICE_CORRELATION_ID, this.serviceCorrelationId);
        this.pageImpressionId && headerVal.set(CONST.PAGE_IMPRESSION_ID, this.pageImpressionId);
        this.uaid && headerVal.set(CONST.UAID, this.uaid.toString());
        return headerVal.toString();
    }

    getSID() {
        return this.sid;
    }

    getCTag() {
        return this.c;
    }

    getUAID() {
        return this.uaid;
    }

    getSIID() {
        return this.siid;
    }

    getStreamId() {
        return this.si;
    }

    getTrackingParameters() {
        return this.trkp;
    }

    getTrackingGlobalFlags() {
        return this.trkgFlags;
    }

    getGlobalSessionCorrelationId() {
        return this.gci;
    }

    getOperationId() {
        return this.operationId;
    }

    setUAID(uaid) {
        this.uaid = uaid;
    }

    static create (req, clientSide) {
        const isService = req.ebay && req.ebay.isService;
        const hdr = req.headers[CONST.X_EBAY_C_CORRELATION_SESSION] || req.headers[CONST.X_EBAY_C_CORRELATION_SESSION.toLowerCase()];
        if (hdr || isService) {
            return isService ? CorrelationSession.fromHeader(hdr) : 
                CorrelationSession.fromHeader(hdr, CorrelationSession.fromWebRequest(req));

        }
        return CorrelationSession.fromWebRequest(req, clientSide);
    }

    static fromHeader(hdr, correlationSession) {
        const ctx = HMV.parse(hdr, false, true);
        correlationSession = correlationSession || {};
        
        correlationSession[CONST.C] = ctx.get(CONST.C) || correlationSession[CONST.C];
        correlationSession[CONST.SI] = ctx.get(CONST.SI) || correlationSession[CONST.SI];
        correlationSession[CONST.GCI] = ctx.get(CONST.GCI);
        correlationSession[CONST.SID] = ctx.get(CONST.SID) || correlationSession[CONST.SID];
        correlationSession[CONST.SIID] = ctx.get(CONST.SIID) || correlationSession[CONST.SIID];
        correlationSession[CONST.TRKP] = ctx.get(CONST.TRKP) || correlationSession[CONST.TRKP];
        correlationSession[CONST.TRKGFLAGS] = ctx.get(CONST.TRKGFLAGS);
        correlationSession[CONST.OPERATION_ID] = ctx.get(CONST.OPERATION_ID) || correlationSession[CONST.OPERATION_ID];
        correlationSession[CONST.UAID] = correlationSession[CONST.UAID] || UserActionID.parse(ctx.get(CONST.UAID));
        correlationSession[CONST.SERVICE_CORRELATION_ID] = ctx.get(CONST.SERVICE_CORRELATION_ID) || correlationSession[CONST.SERVICE_CORRELATION_ID] || createULID();
        correlationSession[CONST.PAGE_IMPRESSION_ID] = ctx.get(CONST.PAGE_IMPRESSION_ID) || correlationSession[CONST.PAGE_IMPRESSION_ID];

        return new CorrelationSession(correlationSession);
      }

    static fromWebRequest(req, clientSide) {
        const ctx = req.ebay;
        const correlationSession = {};

        correlationSession[CONST.SI] = ctx && ctx.getGuid();
        correlationSession[CONST.GCI] = ctx && ctx.getDCCorrelationGuid();
        correlationSession[CONST.SIID] = getSIID(ctx, 'PAGE_SOURCE_INFO');
        correlationSession[CONST.C] = getCookieValue(ctx, 'SOJOURNER_CLICKID');

        if (ctx && ctx.expSvc && ctx.expSvc.operationId || clientSide) {
            correlationSession[CONST.SID] = getSID(req);
            const operationId = ctx && ctx.expSvc && ctx.expSvc.operationId;
            if (operationId) {
                correlationSession[CONST.OPERATION_ID] = operationId;
            }
            else if (clientSide) {
                correlationSession[CONST.OPERATION_ID] = ctx.getPageId();
            }
            correlationSession[CONST.TRKP] = (req.query && req.query._trkparms) || (req.body && req.body._trkparms);
        }

        var pageImpressionId = ctx?.tracking?.getCiid?.();
        if (pageImpressionId) {
            correlationSession[CONST.PAGE_IMPRESSION_ID] = pageImpressionId;
        } 

        const sojournerContext = ctx && ctx.tracking && ctx.tracking.sojournerContext;

        /**
         * TODO: Check if this is too early to construct the flags
         */
        if (sojournerContext && ctx && ctx.tracking) {
            ctx.tracking.updateSojournerBookKeeping(req);
            correlationSession[CONST.TRKGFLAGS] = SojWriter.flagMapToString(sojournerContext.m_contextFlagMap);
        }

        return new CorrelationSession(correlationSession);
    }
}

module.exports = CorrelationSession;

module.exports.createULID = createULID;

function getSID(req) {
    if (req.query && req.query._trksid) {
        const trksid = util.filterTrksid(req.query._trksid);
        return trksid;
    }

    var cookieValue = req && getCookieValue(req.ebay, 'SOJOURNER_TRACKING');

    if (cookieValue) {
        var decodedValue = SourceIdBuilder.decode(cookieValue);
        cookieValue = decodedValue ? SourceIdBuilder.format(decodedValue) : cookieValue;
    }
    return cookieValue;
}


function getSIID(ctx) {
    let sourceID = getCookieValue(ctx, 'PAGE_SOURCE_INFO');

    if (!sourceID){
        return null;
    }

    const psiCookielet = new PageSourceInfoCookielet(sourceID);
    const siid = psiCookielet.getCookieletValueWithoutVersion();
    return siid ? siid : null;
}

function getCookieValue(ctx, key, queryParam) {
    if (queryParam) {
        return queryParam;
    } else {
        if (ctx && ctx.cookies) {
            var cookielet = ctx.cookies.Cookies && ctx.cookies.Cookies[key];
            return cookielet && ctx.cookies.getCookieValue(cookielet);
        }
    }
}
