// @ebay/tracking-ebay/optimizer/dependency-tracking.js
'use strict';
var requestLocal = require('request-local');
var ATTRIBUTES_KEY = 'trackingEbay';
module.exports = {
    // Declare which properties can be passed to the dependency type
    properties: {
        'path': 'string'
    },
    // Validation checks and initialization based on properties:
    init: function () {
    },
    // Read the resource:
    read: function (optimizerContext, callback) {
        var tracking = fromOptimizerContext(optimizerContext);
        var result = '(function (ebay) {' +
        'ebay.tracking = ebay.tracking || {};' +
        'ebay.tracking.pageId = '+tracking.pageId+';' +
        'ebay.tracking.pageName = "'+tracking.pageName+'";' +
    '}(window.$ebay || (window.$ebay = {})));';
        if (typeof callback === "function") {
            callback(null,
                result
            );
        } else {
            return Promise.resolve(result);
        }
    },
    getUnbundledTarget: function() {
        return 'content/' + this.path;
    },
    // // getSourceFile is optional and is only used to determine the last modified time
    // // stamp and to give the output file a reasonable name when bundling is disabled
    getDir: function () {
        return null;
    }
};
function getRequestFromRequestLocal() {
    try {
        return requestLocal.data.request;
    } catch(e) {}
}
function fromOptimizerContext(optimizerContext) {
    var tracking = optimizerContext.data.tracking;
    if (!tracking) {
        var renderContext = optimizerContext.data.renderContext;
        if (renderContext) {
            tracking = fromRenderContext(renderContext);
        } 
        optimizerContext.data.tracking = tracking;
    }
    return tracking;
}
function fromRenderContext(context) {
    var attributes = context.attributes;
    var tracking = attributes[ATTRIBUTES_KEY];
    if (!tracking) {
        var req = context.attributes.request;
        if (!req) {
            var res = context.stream;
            req = res && res.req;
        }
        if (!req) {
            // Finally, fall back to using request-local
            if (requestLocal) {
                req = getRequestFromRequestLocal();
            }
            if (!req) {
                throw new Error('Unable to resolve the request for the provided render context');
            }
        }
        tracking = attributes[ATTRIBUTES_KEY] = req.ebay.tracking;
    }
    return tracking;
}
