// @ebay/oauth-ebay/tags/auth-token/index.js
'use strict';
var logger = require('@ebay/logging-inc').logger('auth-token/renderer');

module.exports = {

    render: function (input, context) {
        var response = context.stream;
        var request = response && response.req;

        var inputAppName = input.clientId;
        var inputScope = input.scope;
        var inputType = input.type;
        var schema = {
            startsWith: /production$/.test(request.ebay.oauth.getEnvironment()) ? 'https:' : '*',
            endsWith: 'ebay.com'
        };
        var correlationSession = request.ebay.oauth.getCorrelationSession();

        function renderForm(input, context, token){
            input.token = token;
            var jsonStrToken = JSON.stringify(token);
            var nonce = request && request.csp && request.csp.nonce ? ' nonce="'+request.csp.nonce+'"' : '';
            context.write('<script type="text/javascript"'+nonce+'>');
            context.write('(function(){'+
                'var myEbayTokenValue = '+jsonStrToken+';'+
                'var myClientId='+'"'+input.clientId+'";'+
                'if(!window.ebayAuthTokens) window.ebayAuthTokens={};'+
                'if(myClientId && myEbayTokenValue) window.ebayAuthTokens[myClientId]=myEbayTokenValue;');
            context.write('window.ebayAllowedUrlScheme='+JSON.stringify(schema || {startsWith:'https:', endWith: 'ebay.com'})+';');
            if (correlationSession !== undefined) {
                context.write('window.correlation_session="'+correlationSession+'";');
            }
            context.write('})();');
            context.write('</script>');
        }
        var oauth = request.ebay.oauth;

        var getToken;

        if (inputType) {
            var err;
            switch(inputType.toLowerCase()) {
                case 'user':
                    if (!oauth.isSignIn()) {
                        err = new Error('Requesting user token token when user is not signed');
                        logger.error(err);
                        getToken = function terminateWithError() {
                            arguments[arguments.length - 1](err);
                        };
                    }
                    else {
                        getToken = oauth.getPublicUserToken.bind(oauth, inputAppName, inputScope);
                    }
                    break;
                case 'recognizeduser':
                    if (!oauth.isRecognizedUser() && !oauth.isSignIn()) {
                        err = new Error('Requesting recognized user token when user is not signed in or recognized');
                        logger.error(err);
                        getToken = function terminateWithError() {
                            arguments[arguments.length - 1](err);
                        };
                    }
                    else {
                        getToken = oauth.getPublicRecognizedUserToken.bind(oauth, inputAppName);
                    }
                    break;
                case 'app':
                    getToken = oauth.getAppToken.bind(oauth, inputAppName, inputScope, 'ajax');
                    break;
                default:
                    throw new Error('Unknown token type ' + inputType);
            }
        }
        else {
            getToken = oauth.isSignIn() && oauth.getPublicUserToken.bind(oauth, inputAppName, inputScope) ||
                oauth.isRecognizedUser() && oauth.getPublicRecognizedUserToken.bind(oauth, inputAppName) ||
                oauth.getAppToken.bind(oauth, inputAppName, inputScope, 'ajax');
        }

        var asyncContext = context.beginAsync();
        getToken(function (err, token) {
            if (err) {
                logger.error('Failed to obtain token: ', err);
                return asyncContext.end();
            }
            renderForm(input, asyncContext, token);
            asyncContext.end();
        });

    }
};
