// @ebay/service-instruments-ebay/lib/sse-utils.js
/** *
 name: utils/index.js
 desc: Simple Helpers to find if response is of certain type.
 ***/

'use strict';
const HDR_TYPE_SSE = 'sse-content-type';
const HDR_CONTENT_TYPE = 'content-type';
const CONTENT_TYPE_JSON = 'application/json';
const MULTIPART_JSON = 'multipart/json';
const TRACKING_DATA = 'TRACKING_DATA';
const EP_DATA = 'EP_DATA';
const TRACKING = 'TRACKING';
const EP = 'EP';

function isValidContentType(headers) {
    const contentType = headers[HDR_CONTENT_TYPE],
        multipartRegexp = new RegExp(`^${MULTIPART_JSON}(;|$)`, 'i');

    return CONTENT_TYPE_JSON === headers[HDR_TYPE_SSE] ||
        multipartRegexp.test(contentType);
}

module.exports.isChunkedResponse = function (res) {
    return res && res.headers && isValidContentType(res.headers);
};

module.exports.isTrackingResponse = function (chunk) {
    return chunk && (TRACKING_DATA === chunk.id || TRACKING === chunk.id);
};

module.exports.isEpResponse = function (chunk) {
    return chunk && (EP_DATA === chunk.id || EP === chunk.id);
};
