// @ebay/cal-publishing-inc/lib/appenders/console.js
'use strict';

var utils = require('../utils');
var NodeUtils = require('util');

function pushColor(array, push, txt, color) {
    color && push.call(array, '\u001b[9'+color+'m');
    push.call(array, txt);
    color && push.call(array, '\u001b[0m');
}

function colorWrap(enabled, txt, color) {
    return enabled && '\u001b[9'+color+'m' +
        txt +
        '\u001b[0m' || txt;
}

var colorMap = {
    warn: 3,
    error: 1,
    fatal: 5
};

function format(evt, colorEnabled) {
    var outArgs = [];
    var colorize = colorWrap.bind(null, colorEnabled);

    outArgs.push(evt.messageClass + utils.formatEventTime(new Date(evt.timestamp)));
    outArgs.push(colorize('[' + (evt.loglevel && evt.loglevel.toUpperCase() || 'TX') + ']',
        colorMap[evt.loglevel] || 7));
    outArgs.push('['+evt.name+']');
    outArgs.push(evt.type);
    evt.data = evt.data || '';
    if (evt.data.msg || evt.data.error) {
        outArgs.push(evt.data.msg && '"'+evt.data.msg+'"' || evt.data.error && evt.data.error.stack);
    }
    if (typeof evt.data === 'object') {
        for (var k in evt.data) {
            if (k === 'error' || k === 'msg') {
                continue;
            }
            var value = evt.data[k];
            outArgs.push('&' + k + '=' +
                (typeof value === 'string' ? value : NodeUtils.inspect(value, null, 2)));
        }
    }
    else {
        outArgs.push(evt.data);
    }
    if (evt.duration) {
        outArgs.push(colorize(utils.humanize(evt.duration), 5));
    }
    outArgs.push(colorize(evt.status, 7));

    return outArgs;
}

function consume(config, evt) {
    console.log.apply(console, format(evt, config.color));
}

module.exports.configure = function configure(config) {

    var onEventHandle = consume.bind(null, config);

    return onEventHandle;
};
