// @ebay/cal-publishing-inc/lib/appenders/debug.js
'use strict';

var debug = require('debug');
var utils = require('../utils');
var NodeUtils = require('util');

function getDebugger(instance, evt) {
    var parts = [];
    parts.push(evt.logger);
    parts.push(evt.loglevel && evt.loglevel.toUpperCase() || 'TX');
    parts.push(evt.type);
    parts.push(evt.name);
    parts.push(evt.parentEventId);
    var key = parts.join(':');
    instance._debugs = instance._debugs || {};
    return instance._debugs[key] = instance._debugs[key] || (function getLogger() {
        var dbg = debug(key);
        return function out(evt) {
            dbg(format(evt));
        };
    })();
}

function format(evt) {
    var outArgs = [];
    var data = typeof evt.data === 'object' && Object.create(evt.data) || evt.data;
    outArgs.push(evt.messageClass + utils.formatEventTime(new Date(evt.timestamp)));
    if (data.msg || data.error) {
        outArgs.push(data.msg && '"'+data.msg+'"' || data.error && data.error.stack);
        delete data.msg;
        delete data.error;
    }
    if (data) {
        for (var k in data) {
            var value = data[k];
            outArgs.push(k);
            outArgs.push('=');
            outArgs.push(typeof value === 'string' ? value : NodeUtils.inspect(value, null, 2));
        }
    }
    else {
        outArgs.push(data);
    }
    outArgs.push('\u001b[95m');
    if (evt.duration) {
        outArgs.push(utils.humanize(evt.duration));
    }
    outArgs.push(evt.status);
    outArgs.push('\u001b[0m');
    return outArgs;
}

function consume(instance, evt) {
    var out;
    if (process.env.DEBUG) {
        out = getDebugger(instance, evt);
    }

    out && out(evt);
}

module.exports.configure = function configure(config, globalConfig) {

    if (process.env.DEBUG) {
        globalConfig.loglevel = globalConfig.loglevel || {};
        globalConfig.loglevel.debug = globalConfig.loglevel.debug ?
            globalConfig.loglevel.debug + ',' + process.env.DEBUG :
            process.env.DEBUG;
    }

    var onEventHandle = consume.bind(null, {});

    return onEventHandle;
};
