// @ebay/openfeature-ebay/dist/tracking/utils.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.addTagToPulsarEventTracker = exports.addXtTagsToPulsarEventTracker = exports.addQtagsToPulsarEventTracker = exports.translateTags = exports.EP_TRACKING_EVENT_FAMILY = exports.EP_TRACKING_EVENT_ACTION = exports.EP_TRACKING_PAGEID = exports.tagTypeLookup = void 0;
const EpTagsEventMap_1 = require("../types/EpTagsEventMap");
exports.tagTypeLookup = {
    es: EpTagsEventMap_1.TrackingTagTypes.SCALAR,
    ec: EpTagsEventMap_1.TrackingTagTypes.SCALAR,
    os: EpTagsEventMap_1.TrackingTagTypes.VECTOR,
    ot: EpTagsEventMap_1.TrackingTagTypes.VECTOR,
    mdbreftime: EpTagsEventMap_1.TrackingTagTypes.SCALAR,
    epcalenv: EpTagsEventMap_1.TrackingTagTypes.SCALAR,
    eprlogid: EpTagsEventMap_1.TrackingTagTypes.SCALAR,
};
const EP_TRACKING_PAGEID = 2555220;
exports.EP_TRACKING_PAGEID = EP_TRACKING_PAGEID;
const EP_TRACKING_EVENT_ACTION = 'TAGXT';
exports.EP_TRACKING_EVENT_ACTION = EP_TRACKING_EVENT_ACTION;
const EP_TRACKING_EVENT_FAMILY = 'EXPRMT';
exports.EP_TRACKING_EVENT_FAMILY = EP_TRACKING_EVENT_FAMILY;
/**
 * Translates the string ep tags map from service to qTags data structure.
 *
 * @param epTagsMap
 * @returns
 */
function translateTags(epTagsMap) {
    const epTags = {
        qTags: [],
    };
    if (epTagsMap) {
        const epTagsEventMap = epTagsMap;
        const epTagsMapArray = epTagsEventMap.split('&');
        epTagsMapArray.forEach((str) => {
            const keyVal = str.split('=');
            if (keyVal != null && keyVal.length == 2) {
                const tagName = keyVal[0].replace('!', '');
                const tagValue = keyVal[1];
                //add only valid tag names to qTags
                if (exports.tagTypeLookup[tagName]) {
                    let tag = {
                        key: tagName,
                        value: tagValue,
                        type: exports.tagTypeLookup[tagName],
                    };
                    epTags.qTags.push(tag);
                }
            }
        });
    }
    return epTags;
}
exports.translateTags = translateTags;
function addQtagsToPulsarEventTracker(pulsarEventTracker, data) {
    const { qTags = [] } = data || {};
    if (pulsarEventTracker) {
        pulsarEventTracker.setEventType(EP_TRACKING_EVENT_ACTION, EP_TRACKING_EVENT_FAMILY, `1.0`);
        pulsarEventTracker.addData(`p`, EP_TRACKING_PAGEID);
        qTags.forEach((tag) => {
            if (exports.tagTypeLookup.hasOwnProperty(tag.key)) {
                pulsarEventTracker.addData(tag.key, tag.value);
            }
        });
    }
}
exports.addQtagsToPulsarEventTracker = addQtagsToPulsarEventTracker;
/**
 * This helper function adds the list of treatment ids that were triggered to the xt tag using pulsar event tracker
 * @param pulsarEventTracker
 * @param treatmentIds
 */
function addXtTagsToPulsarEventTracker(pulsarEventTracker, treatmentIds) {
    pulsarEventTracker.addData(`!xt`, treatmentIds);
}
exports.addXtTagsToPulsarEventTracker = addXtTagsToPulsarEventTracker;
/**
 * This helper function adds generic tag data to pulsarEventTracker
 * @param pulsarEventTracker
 * @param dataKey
 * @param dataValue
 */
function addTagToPulsarEventTracker(pulsarEventTracker, dataKey, dataValue) {
    pulsarEventTracker.addData(dataKey, dataValue);
}
exports.addTagToPulsarEventTracker = addTagToPulsarEventTracker;
