// @ebay/tracking-ebay/lib/bot/BotDetector.js
'use strict';

var trie = require('./trie');

var Services = require('../services/tracking-wrappers');
var modConfig = require('@ebay/module-config-inc');
var botDetector;
var refreshTimeout = 3600000; // 1 hour refresh period
var pathTools = require('path');
var utils = require('../utils');
var logger = require('@ebay/logging-inc').logger('tracking-ebay/BotDetector');
var agentsAndIpsMeta = utils.tryRequire(pathTools.resolve(__dirname, '../../resources/agents-and-ips.json'));

var BotDetector = module.exports.BotDetector = function (agents, ips) {
    var map = {};
    agents = (agents || []).reduce(function reduce(memo, agent) {
        if (!map[agent]) {
            map[agent] = true;
            agent = agent[agent.length - 1] === '*' ? agent.substring(0, agent.length - 1) : agent;
            memo.push(agent.toLowerCase());
        }
        return memo;
    }, []);
    this.botAgents = trie.createTrieFromArray(agents);
    this.botIps = trie.createTrieFromArray(ips || []);
    this._timestamp = Date.now();
};

BotDetector.prototype.setBotAgents = function (agents) {
    this.botAgents = trie.createTrieFromArray(agents || []);
    return this;
};

BotDetector.prototype.setBotIps = function (ips) {
    this.botIps = trie.createTrieFromArray(ips || []);
    return this;
};

BotDetector.prototype.isBotIp = function (ip) {
    return this.botIps.lookup(ip);
};

BotDetector.prototype.isBotAgent = function (userAgent) {
    return this.botAgents.match(userAgent && userAgent.toLowerCase());
};

BotDetector.prototype.toString = function () {
    return JSON.stringify({
        'agents' : this.botAgents.dumpJson(),
        'ips' : this.botIps.dumpJson()
    });
};

function loadAgentsAndIps(callback) {
    Services.getTEVList(callback);
}

function loadBotDetector(callback) {
    loadAgentsAndIps(function (err, agentsAndIps) {
        if (err) {
            return callback && callback(err);
        }
        botDetector = new BotDetector(agentsAndIps.agents, agentsAndIps.ips);
        return callback && callback(null, botDetector);
    });
}

function loadDefaults() {
    if (agentsAndIpsMeta) {
        botDetector = new BotDetector(agentsAndIpsMeta.agents, agentsAndIpsMeta.ips);
        botDetector._timestamp = 0; // mark it as need to reload
    }
}

module.exports.loadAgentsAndIps = loadAgentsAndIps;
module.exports.load = function load(callback) {
    modConfig(module, function (error, config) {
        if (!error) {
            refreshTimeout = config.get('bot-detector:refresh-timeout') || refreshTimeout;
        }

        if (botDetector) {
            callback(null, botDetector);
            // refresh the list
            if (Date.now() - botDetector._timestamp > refreshTimeout) {
                botDetector._timestamp = Date.now();
                setImmediate(function refresh() {
                    // try to reload only once
                    loadBotDetector(); // just refresh
                });
            }
            return;
        }

        logger.error(new Error('Loading BotDetector metadata from service possibly due to failure to load static metadata'));

        loadBotDetector(callback);
    });
};

loadDefaults();
