// @ebay/locale-formatter-ebay/lib/reader/formatsReader.js
"use strict";

var	path = require('path'),
	fs = require('fs'),
	cs = require('../constants'),
	Cache = require('lru-backup-cache'),
	extend = require('extend'),
	localeCache = new Cache({
		"enabled": true,
        "max": 20000,          //Cache size
        "maxAge": 1800000
	});

module.exports = exports = {
	load : load,
	$internals: {
		localeCache: localeCache
	}
};

function Formats(dir, locale) {

	this._store = {};
	this._extendStore(dir, locale);

}

Formats.prototype = {

	/**
	 * Load the Locale based formats
	 */
	_extendStore : function(dir, locale) {
		var defFmts = path.join(dir, cs.FORMATS_FILE_BASENAME + cs.JSON_EXT),
			//defFmtsContent = path.join(dir, cs.FORMATS_CONTENT_FILE_BASENAME + cs.JSON_EXT),
			fileExt = "_" + locale + cs.JSON_EXT,
			localeFmts = path.join(dir, cs.FORMATS_FILE_BASENAME + fileExt),
			//localeFmtsContent = path.join(dir, cs.FORMATS_CONTENT_FILE_BASENAME + fileExt),
			lang,
			localeLangFmts;
			//localeLangFmtsContent;

		if(locale && locale.split('_') && locale.split('_').length > 1) {

			lang = locale.split('_')[0];
			localeLangFmts = path.join(dir, cs.FORMATS_FILE_BASENAME + "_" + lang +  cs.JSON_EXT);
			//localeLangFmtsContent = path.join(dir, cs.FORMATS_CONTENT_FILE_BASENAME + "_" + lang +  cs.JSON_EXT);
		}

		if (fs.existsSync(defFmts)) {
			var defStore = require(defFmts);
			extend(true, this._store, defStore);
		}

		if (fs.existsSync(localeLangFmts)) {
			var langStore = require(localeLangFmts);
			extend(true, this._store, langStore);
		}

		if (fs.existsSync(localeFmts)) {			
			var localeStore = require(localeFmts);
			extend(true, this._store, localeStore);
		}
	},
	/**
	 * Decimal Formats
	 */
	getDecimalChar : function () {
		return this._store[cs.DECIMAL_CHAR];
	},
	getDecimalDigits : function () {

		return this._store[cs.DECIMAL_DIGITS];
	},
	getGroupingChar : function () {

		return this._store[cs.GROUPING_CHAR];
	},
	getGroupingDigits : function () {

		return this._store[cs.GROUPING_DIGITS];
	},
	/**
	 * Currency Formats
	 */
	getPositiveCurrencyPattern : function (currencyCode) {

		return this._store[cs.CURRENCY + currencyCode + cs.POSITIVE];
	},
	getNegativeCurrencyPattern : function (currencyCode) {

		return this._store[cs.CURRENCY + currencyCode + cs.NEGATIVE];
	},
	getStandardCurrencySymbol : function (currencyCode) {

		return this._store[cs.CURRENCY + currencyCode + cs.STANDARD];
	},
	getSimpleCurrencySymbol : function (currencyCode) {

		return this._store[cs.CURRENCY + currencyCode + cs.EMAIL];
	},
	getWorldCurrencySymbol : function (currencyCode) {

		return this._store[cs.CURRENCY + currencyCode + cs.SYMBOL];
	},
	getFancyCurrencySymbol : function (currencyCode) {

		return this._store[cs.CURRENCY + currencyCode + cs.FANCY];
	},
	getNumCurrencyDecPoints : function (currencyCode) {

		return this._store[cs.CURRENCY + currencyCode + cs.DECIMAL];
	},
	getMaxCurrencyVal : function (currencyCode) {

		return this._store[cs.CURRENCY + currencyCode + cs.MAX];
	},
	//From Content formats
	// getCurrencySingularName : function (currencyCode) {
	// 	return this._conf.get(cs.CURRENCY + currencyCode + cs.SINGULAR);
	// },
	// getCurrencyPluralName : function (currencyCode) {
	// 	return this._conf.get(cs.CURRENCY + currencyCode + cs.PLURAL);
	// },
	/**
	 * Date Formats
	 */
	getShortDateFormat : function () {

		return this._store[cs.SHORT_DATE];
	},
	getShortCleanDateFormat : function () {

		return this._store[cs.SHORT_CLEAN_DATE];
	},
	getMediumDateFormat : function () {

		return this._store[cs.MEDIUNM_DATE];
	},
	getNumericDateFormat : function () {

		return this._store[cs.NUMERIC_DATE];
	},
	getMediumNoDayDateFormat : function () {

		return this._store[cs.MEDIUM_DATE_NO_DAY];
	},
	getNumericNoDayDateFormat : function () {

		return this._store[cs.NUMERIC_NO_DAY];
	},
	getFullDateFormat : function () {

		return this._store[cs.LONG_DATE];
	},
	getFullNoYearDateFormat : function () {

		return this._store[cs.LONG_DATE_NO_YEAR];
	},
	/**
	 * Time Formats
	 */
	getShortTimeFormat : function () {

		return this._store[cs.SHORT_TIME];
	},
	getShortAmPmTimeFormat : function () {
		return this._store[cs.SHORT_TIME_AMPM];
	},
	getShortZoneTimeFormat : function () {

		return this._store[cs.SHORT_TIME_ZONE];
	},
	getShortAmPmZoneTimeFormat : function () {
		return this._store[cs.SHORT_TIME_ZONE_AMPM];
	},
	getMediumTimeFormat : function () {

		return this._store[cs.MEDIUM_TIME];
	},
	getLongTimeFormat : function () {

		return this._store[cs.LONG_TIME];
	},
	getLongAmPmTimeFormat : function () {
		return this._store[cs.LONG_AM_PM_TIME];
	}
};

function load(locale, options) {
	locale = locale && locale.replace('-', '_');

	options = options || {};
	options.confDir = options.confDir || cs.FORMATS_CONFIG_DIR;
	var localeObj = localeCache.get(locale);
	if (!localeObj) {
		localeObj = new Formats(options.confDir, locale);
		localeCache.set(locale, localeObj);
	}

	return localeObj;
}
