// @ebay/service-ebay/middleware/consumer-status.js
'use strict';

const Hoek = require('hoek');
const modConfig = require('@ebay/module-config-inc');
const vi = require('@ebay/validate-internals-ebay');
const X_EBAY_CONSUMER_ID = 'x-ebay-consumer-id';
const X_EBAY_CLIENT_APP_NAME = 'x-ebay-client-app-name';
const UNKNOWN = 'Unknown';

module.exports = function () {
    // if we find validate internal, we will create status component
    const statusManager = createConsumerStatusManager();

    return (req, res, next) => {
        modConfig(module, (err, cfg) => {
            // error will never happen
            const statusEnabled = cfg.get('service-ebay:consumer-status:enabled');
            if (statusEnabled) {
                const finishResponse = Hoek.once(err => {
                    const { consumerId, appName } = getConsumerAttributes();
                    const status = statusManager.getConsumerStatus(consumerId, appName);

                    status.total++;
                    if (err || res.statusCode >= 400 && res.statusCode < 600) {
                        status.failures++;
                    }
                });

                res.once('finish', finishResponse);
                res.once('close', finishResponse);
                res.once('error', finishResponse);
            }

            // to avoid module config catching error thrown in next and making a new callback
            setImmediate(() => next());
        });

        function getConsumerAttributes() {
            const headers = req.headers;
            return {
                consumerId: headers && headers[X_EBAY_CONSUMER_ID] || getConsumerIdFromToken(),
                appName: headers && headers[X_EBAY_CLIENT_APP_NAME]
            };
        }

        function getConsumerIdFromToken() {
            const attrs = req.security && req.security.getTokenAttributes();
            return attrs && attrs.attributes && attrs.attributes.AppId;
        }
    };
};

function createConsumerStatusManager() {
    // hold metadata in the manager instance
    const metadata = {};

    return {
        getConsumerStatus(consumerId = UNKNOWN, appName) {
            const status = metadata[consumerId];
            if (status) {
                return status;
            }
            const newStatus = metadata[consumerId] = createComponentStatus(consumerId, appName);
            return newStatus;
        }
    };

    function createComponentStatus(consumerId, appName) {
        // consumerId will more likely be available,
        // but if appName is available then consumerId is there for sure
        // try to use more meaninful name for component name first
        const componentName = `Inbound.Service.Call.${appName || consumerId}`;
        vi.register(componentName, () => {
            const status = metadata[consumerId];
            return {
                name: componentName,
                alias: componentName,
                properties: Object.keys(status).map(key => ({
                    name: key,
                    value: status[key]
                }))
            };
        });

        return {
            consumerId,
            appName: appName || UNKNOWN,
            total: 0,
            failures: 0
        };
    }
}

module.exports.createConsumerStatusManager = createConsumerStatusManager;
module.exports.X_EBAY_CONSUMER_ID = X_EBAY_CONSUMER_ID;
module.exports.X_EBAY_CLIENT_APP_NAME = X_EBAY_CLIENT_APP_NAME;
