// @ebay/tracking-ebay/lib/soj/SojournerVectorData.js
'use strict';

/**
 * This is a rewrite to com.ebay.globalenv.sojourner.SojournerVectorData
 */
var _ = require('underscore');
var CHAR_PIPE = '|';
var CHAR_COMMA = ',';
var VECTOR_TYPE = require('./vector-type-enums.js');
var utils = require('../utils.js');
var logger = require('@ebay/logging-inc').logger('tracking-ebay:SoujornerVectorData');
var SojournerVectorData = module.exports = function (type) {
    this.type = type || VECTOR_TYPE.STRING;
    this.MAX_VECTOR_SIZE = 100;
    this.NEW_MAX_VECTOR_SIZE = 200;
    this.m_data = [];
};

SojournerVectorData.prototype.toString = function () {
    return this.m_data.join(CHAR_COMMA);
};
SojournerVectorData.prototype.getSafeTagName = function (tag) {
    return this.type.prefix + tag;
};

SojournerVectorData.prototype.appendData = function (data) {
    if (this.m_data.length >= this.NEW_MAX_VECTOR_SIZE) {
        logger.error('The SojournerVectorData array is full, reaching the maximum size', this.m_data, data);
    }else{
        return this.m_data.push(this.getSafeData(data));
    }
};
SojournerVectorData.prototype.getSafeData = function (data) {
    if (_.isArray(data)) {
        return _.map(data, function (element) {
            return utils.encode(element);
        }).join(CHAR_PIPE);
    } else {
        return utils.encode(data);
    }
};
