// @ebay/device-detection-ebay/lib/DeviceInfo.js
'use strict';

const useragent = require('express-useragent');

const TV_PATTERN = /GoogleTV|SmartTV|Internet TV|NetCast|NETTV|AppleTV|boxee|Kylo|Roku|DLNADOC|CE\\-HTML|Xbox|PLAYSTATION 3|Wii/i;
const TABLET_PATTERN = /ip(a|ro)d|silk|xoom|playbook|tablet|kindle|Nexus 7|GT-P10|SC-01C|SHW-M180S|SGH-T849|SCH-I800|SHW-M180L|SPH-P100|SGH-I987|zt180|HTC( Flyer|_Flyer)|Sprint ATP51|ViewPad7|pandigital(sprnova|nova)|Ideos S7|Dell Streak 7|Advent Vega|A101IT|A70BHT|MID7015|Next2|nook|FOLIO|MB511.*RUTEM|Mac OS.*Silk/i;
const TOUCH_SCREEN_PATTERN = /iphone|ip(a|ro)d|android|xoom|playbook|tablet|bb10|lumia|windows .* touch|Silk/i;
const PROPERTIES = ['isTouchScreen', 'isTV', 'isTablet', 'isMobile', 'isSmall', 'isLarge', 'isRobot', 'osName', 'browserVersion', 'isDesktop', 'browser', 'isFSOM', 'fallback', 'isKindle'];

function parse(uaString) {
    const ua = useragent.parse(uaString || '');

    if (/Kindle/i.test(uaString)) {
        ua.isKindle = true;
        ua.isDesktop = false;
        ua.isMobile = true;
    }

    if (/Silk-Accelerated/i.test(uaString)) {
        ua.isDesktop = false;
        ua.isMobile = !ua.isDesktop;
        ua.isTablet = true;
    }
    return ua;
}

var DeviceInfo = module.exports = function DeviceInfo(uaString) {
    this._uaString = uaString;
    this._ua = parse(uaString);
    this._isFSOM = false;
};

DeviceInfo.prototype = {
    get isTouchScreen() {
        if (typeof this._isTouchScreen === 'undefined') {
            this._isTouchScreen = TOUCH_SCREEN_PATTERN.test(this._uaString);
        }
        return this._isTouchScreen;
    },

    get isTV() {
        if (typeof this._isTV === 'undefined') {
            this._isTV = TV_PATTERN.test(this._uaString);
        }
        return this._isTV;
    },

    get isTablet() {
        if (typeof this._isTablet === 'undefined') {
            this._isTablet = TABLET_PATTERN.test(this._uaString);
        }
        return this._isTablet;
    },

    get isMobile() {
        return this._ua.isMobile;
    },

    get isSmall() {
        return this.isMobile && !this.isTablet;
    },

    get isLarge() {
        return !this.isSmall;
    },

    get isRobot() {
        return this._ua.isBot;
    },

    get osName() {
        return this._ua.OS;
    },

    get browserVersion() {
        return this._ua.Version;
    },

    get isDesktop() {
        return this._ua.isDesktop && !this.isTV;
    },

    get browser() {
        return this._ua.Browser;
    },

    get fallback() {
        return true;
    },

    get isKindle() {
        return this._ua.isKindle;
    }
};

module.exports.fromUserAgent = function fromUserAgent(uaString) {
    const ua = new DeviceInfo(uaString);

    return PROPERTIES.reduce((memo, key) => {
        if (key.indexOf('_') !== 0) {
            memo[key] = ua[key];
        }
        return memo;
    }, {});
};
