// @ebay/device-detection-ebay/lib/service.js
'use strict';

const svcClient = require('@ebay/service-client-ebay');
const debug = require('debug')('device-detection-ebay:client');

function getDDSAPI(ddsVersion) {
    return ddsVersion === "v3" ? 'device-detection-ebay-v3' : 'device-detection-ebay'
}

exports = module.exports = {
    getDeviceInfo(ddsVersion, params, callback) {
        debug("Params : ", params);
        // V3 will not support GET request, so still default to v2
        const provider = svcClient.context({}).getClient('device-detection-ebay');
        debug("Provider : ", provider);

        provider.get(params).path('/deviceinfo').end(function (err, response) {
            if (err || !response || response.statusCode !== 200|| !response.body || !response.body.status || response.body.status.ack !== 'SUCCESS') {
                err = err || (response && response.body && response.body.status && response.body.status.message) || new Error('failed to get device info for (' + params + ')');
                debug("Error :", err);
                callback(err, null);
            } else {
                callback(err, response.body);
            }
        });
    },

    // Device Info for Client Hints
    getCHDeviceInfo(ddsVersion, body, callback) {
        debug("body : ", body);
        const provider = svcClient.context({}).getClient(getDDSAPI(ddsVersion));
        debug("Provider : ", provider);
        provider.post({
            body: JSON.stringify(body),
            path: '/deviceinfo',
            headers: {
                'Content-Type': 'application/json'
            }
        }).end(function (err, response) {
            if (err || !response || response.statusCode !== 200|| !response.body || !response.body.status || response.body.status.ack !== 'SUCCESS') {
                err = err || (response && response.body && response.body.status && response.body.status.message) || new Error('failed to get device info for (' + body + ')');
                debug("Error :", err);
                callback(err, null);
            } else {
                callback(err, response.body);
            }
        });
    }
};
