import localeEbay = require('.');
import { Request } from 'express';

(req: Request) => {
  // $ExpectType Locality
  req.locality;
  req.locality = {
    country: 'DE',
    countryId: 77,
    isGeoCountry: false,
    language: 'de',
    locale: 'de-DE',
    defaultLanguage: 'de',
    supportedLanguages: ['de'],
    timezone: 'Central Europe',
    directionality: 'ltr',
    siteId: 77,
    isoCurrencyCode: 'EUR',
    siteDefaultIsoCurrencyCode: 'EUR',
    timeZoneStd: 'CET',
    timeZoneSummer: 'CET'
  };
};

// $ExpectType RequestHandler<ParamsDictionary, any, any, ParsedQs, Record<string, any>>
localeEbay();

// $ExpectType Determiner
localeEbay.removeDeterminer('determiner');

// $ExpectError
localeEbay()();

// $ExpectError
localeEbay.registerDeterminer();

// $ExpectError
localeEbay.removeDeterminer();
