// @ebay/prometheus-ebay/lib/metrics.js
'use strict';

const promClient = require('prom-client');
const getPm2Id = require('@ebay/instance-detect-ebay').getInstanceId;

const namespaces = module.exports.namespaces = {};
class Metrics {
    constructor(namespace) {
        const processId = getPm2Id();

        this.registry = new promClient.Registry(promClient.Registry.OPENMETRICS_CONTENT_TYPE);
        this.registry.setDefaultLabels({ processId: processId });

        this.promClient = promClient;
        if (namespace) {
            namespaces[namespace] = this;
        }
    }

    async readMetrics(contentType) {
        this.registry.setContentType(contentType || promClient.Registry.PROMETHEUS_CONTENT_TYPE);
        const metrics = await this.registry.metrics();
        return {
            'contentType': this.registry.contentType,
            'metrics': metrics
        };
    }

    async readOpenMetrics() {
        return this.readMetrics(promClient.Registry.OPENMETRICS_CONTENT_TYPE);
    }

    async readPrometheusMetrics() {
        return this.readMetrics(promClient.Registry.PROMETHEUS_CONTENT_TYPE);
    }

    // reset the metrics
    resetMetrics(contentType) {
        this.registry.setContentType(contentType || promClient.Registry.PROMETHEUS_CONTENT_TYPE);
        this.registry.resetMetrics();
    }

    // clear the metrics with the labels
    clearMetrics(contentType) {
        this.registry.setContentType(contentType || promClient.Registry.PROMETHEUS_CONTENT_TYPE);
        this.registry.clear();
    }

    // this is a type, still keep it here to avoid conflicts
    // need to mark it as deprecated
    createGuage(name, description, labels = [], collect) {
        return new promClient.Gauge({
            name: name,
            labelNames: labels,
            help: description || `Help for ${name}`,
            registers: [this.registry],
            collect: collect
        });
    }

    createGauge(name, description, labels = [], collect) {
        return new promClient.Gauge({
            name: name,
            labelNames: labels,
            help: description || `Help for ${name}`,
            registers: [this.registry],
            collect: collect
        });
    }

    createCounter(name, description, labels = [], enableExemplars = false, collect) {
        return new promClient.Counter({
            name: name,
            labelNames: labels,
            help: description || `Help for ${name}`,
            registers: [this.registry],
            enableExemplars: enableExemplars,
            collect: collect
        });
    }

    // eslint-disable-next-line max-len
    createSummary(name, description, labels = [], percentiles, maxAgeSeconds, ageBuckets) {
        return new promClient.Summary({
            name: name,
            labelNames: labels,
            help: description || `Help for ${name}`,
            percentiles: percentiles,
            maxAgeSeconds: maxAgeSeconds,
            ageBuckets: ageBuckets,
            registers: [this.registry]
        });
    }

    createHistogram(name, description, labels = [], buckets, enableExemplars = false) {
        return new promClient.Histogram({
            name: name,
            labelNames: labels,
            help: description || `Help for ${name}`,
            buckets: buckets,
            registers: [this.registry],
            enableExemplars: enableExemplars
        });
    }
}

module.exports.Metrics = Metrics;
