// @ebay/raptor-config-ebay/lib/index.js
'use strict';

var _ = require('underscore'),
    util = require('util'),
    async = require('async'),
        EventEmitter = require('events').EventEmitter,
        caller = require('caller');

var modConfig;

var utils = require('./utils');

var moduleConfig = {
    forceUpdate: false,
    configBean: undefined,
    globalConfig: false,
    configChangeListenerRegistered: false,
};

var globalConfig;

/**
 * Loaded cache is a map of config ref as string key, and a loaded structure as value.
 * Loaded value looks like the following
 *  {
 *      "latest": "",
 *      "baseline": [],
 *      "changes": {
 *          "": [],
 *          "changeset-0": []
 *      },
 *      "listeners":[],
 *      "timer":Object
 *  }
 */

var _cache = {};

function forceUpdateConfigs() {
    if (globalConfig) {
        globalConfig.refreshManifest(true);
    }
    _.each(_cache, function (config) {
        if (config.reloadGlobalConfig) {
            config.reloadGlobalConfigTimer && clearImmediate(config.reloadGlobalConfigTimer);
            config.reloadGlobalConfigTimer = setImmediate(() => {
                config.reloadGlobalConfigTimer = undefined;
                config.reloadGlobalConfig();
            });
        }
    });
}

function onConfigChange() {
    var refreshconfigBean = moduleConfig.configBean.get('refreshconfig');

    if (moduleConfig.forceUpdate !== refreshconfigBean.forceUpdate) {
        moduleConfig.forceUpdate = refreshconfigBean.forceUpdate;
        forceUpdateConfigs();
    }
}

function RaptorConfig() {
    //create config bean for Raptor Config if not created already
    const configBean = require('@ebay/config-bean-ebay');
    if (!configBean.getBeanById(utils.configBeanId)) {
        utils.defineBean();
    }
}
util.inherits(RaptorConfig, EventEmitter);

/**
 * Initialize and cache config,
 * As this module contribute back to module-config-inc,
 * OK to cache it here
 * Lazy load module config and raptor config service
 */
RaptorConfig.prototype._init = function () {
    modConfig = modConfig || require('@ebay/module-config-inc');

    moduleConfig.configBean = modConfig(module);
    moduleConfig.globalConfig = true;

    // Cache config value and bind on change listener
    var refreshconfig = moduleConfig.configBean.get('refreshconfig');
    moduleConfig.forceUpdate = refreshconfig.forceUpdate;
    moduleConfig.configBean.on('change', onConfigChange);
    moduleConfig.configChangeListenerRegistered = true;
};

//main api, loading a config based on the ref defined in the given message;
RaptorConfig.prototype.load = function (configSpec) {
    if (!moduleConfig.configChangeListenerRegistered) {
        this._init();
    }
    var cacheKey;
    var config;
    // go with global config
    if (!globalConfig) {
        globalConfig = require('./globalConfig');
    }
    cacheKey = utils.buildLoadKey(configSpec);
    cacheKey += "&" + caller();
    config = _cache[cacheKey] || {};
    _cache[cacheKey] = config;
    var self = this;

    var gcfgCB = function (error, gcfg) {
        if (error) {
            if (error.logType === 'warn') {
                configSpec.errorMessage = error.message;
            }
            else {
                configSpec.error = error;
            }
            self.emitReadEvent(configSpec, []);
            return;
        }

        self.emitReadEvent(configSpec, gcfg);
    };

    if (!config.reloadGlobalConfig) {
        config.reloadGlobalConfig = _.bind(globalConfig.load, self, configSpec, gcfgCB);
    }
    globalConfig.load(configSpec, gcfgCB);
};

//emit the 'read' event
RaptorConfig.prototype.emitReadEvent = function (configSpec, latestConfig) {
    var response = {};

    if (configSpec) {
        _.extend(response, configSpec);
    }

    if (latestConfig) {
        _.extend(response, {
            'type': 'read',
            'message': _.extend({}, configSpec),
            'properties': latestConfig.properties,
            'validContexts': latestConfig.validContexts
        });
    }

    this.emit('read', response);
};

//shutdown all schedulers
RaptorConfig.shutdown = function shutdown(cacheKey) {
    _.each(_cache, function (config, key) {
        if (cacheKey && cacheKey !== key) {
            return;
        }
        config.reloadGlobalConfigTimer && clearImmediate(config.reloadGlobalConfigTimer);
        delete _cache[key];
    });

    if (!cacheKey) {
        _cache = {};
    }
    modConfig && modConfig.cache.clear();
    moduleConfig = {
        forceUpdate: false,
        configBean: undefined,
        globalConfig: false,
        configChangeListenerRegistered: false,
    };
};

module.exports = RaptorConfig;
Object.defineProperty(module.exports, 'stateCache', {
    get: function () {
        return Object.assign({}, _cache);
    }
});
Object.defineProperty(module.exports, 'globalConfig', {
    get: function () {
        if (globalConfig) {
            return globalConfig.refreshManifest();
        }
        return;
    }
});
