// @ebay/tracking-ebay/lib/webview/index.js
'use strict';

const Cal = require('@ebay/cal');
const modConfig = require('@ebay/module-config-inc');
const logger = require('@ebay/logging-inc').logger('@ebay/tracking-ebay/webview');
const Provider = require('@ebay/service-client-ebay');
//const { response } = require('../../test/requestMock');
const GUIDGenerator = require('@ebay/commons-ebay/lib/guid-generator');
const guidGenerator = new GUIDGenerator();

const GUID_LENGTH = 32;
const WEB_VIEW = 'webview';
const TRACKING_APP = 'trackingApp';
const TRACKING_SRC_APP_ID = 'trackingSrcAppId';
const TRACKING_NATIVE_APP_GUID = 'trackingNativeAppGuid';

const NATIVE_APP_VER = "appVer";
const CLIENT_ID = "clientId";
const SIGNATURE = "signature";
const ARTIFACT = "artifact";
const SIGNATURE_VERSION = "signatureVersion";

const TRACKING_NATIVE_APP_GUID_EMPTY_OR_NULL_WARN_MESSAGE =
    `${TRACKING_NATIVE_APP_GUID} in input request is empty or null`;
const TRACKING_SRC_APP_ID_EMPTY_OR_NULL_WARN_MESSAGE =
    `${TRACKING_SRC_APP_ID} in input request is empty or null`;

module.exports = (req, res, next) => {
    modConfig(module, async (err, config) => {
        if (config.get('tracking-ebay:webview:enable') &&
            req.query[TRACKING_APP] === WEB_VIEW) {
            // implementation of validateWebView does not fail with any expected error
            // so we do not need to catch errors
            const result = await validateWebView(req);
            if (result.appGuid && result.srcAppId) {
                const { srcAppId, appGuid, loggerHMAC } = result;
                const sojournerContext = req.ebay.tracking.sojournerContext;
                // Add a soj tag to indicate its a web view request
                req.ebay.tracking.trackTag(WEB_VIEW, true);
                // Add a soj tag to indicate its hmacTestable request
                if (loggerHMAC.hmacTestable) {
                    req.ebay.tracking.trackTag('hmacTestable', loggerHMAC.hmacTestable);
                    req.ebay.tracking.trackTag('isValidatedGuid', loggerHMAC.isValidatedGuid);
                    console.log('loggerHMAC.isValidatedGuid', loggerHMAC.isValidatedGuid);
                    if (loggerHMAC.invalidPassedGuid) {
                        req.ebay.tracking.trackTag('invalidPassedGuid', loggerHMAC.invalidPassedGuid);
                    }
                }
                // set the app id
                req.ebay.tracking.trackTag('app', srcAppId);
                sojournerContext.addContext('g', appGuid);
                // Setting the GUID into the response header cookie
                req.ebay.cookies.setCookieValue(
                    'SOJOURNER_GUID', appGuid);
                req.ebay.cookies.setCookieValue(
                    'SOJOURNER_SESSION_GUID', appGuid);
            }
        }

        next();
    });
};

function logCalWarning(inputRequestType, warnMessage) {
    Cal.createEvent(
        'Warn', `Invalid webview ${inputRequestType}`, 1, warnMessage
    ).complete();
}

const ERROR_CODES = {
    EMPTY_APP_ID: 1,
    INVALID_APP_ID: 2,
    EMPTY_APP_GUID: 3,
    INVALID_APP_GUID_LENTGH: 4,
    INVALID_APP_GUID_PATTERN: 5
};

async function validateWebView(webRequest) {
    const query = webRequest.query;
    const srcAppId = query[TRACKING_SRC_APP_ID];
    const appGuid = query[TRACKING_NATIVE_APP_GUID];

    if (!srcAppId) {
        logCalWarning(TRACKING_SRC_APP_ID,
            TRACKING_SRC_APP_ID_EMPTY_OR_NULL_WARN_MESSAGE);
        return ERROR_CODES.EMPTY_APP_ID;
    }

    if (!/^\d+$/.test(srcAppId)) {
        logCalWarning(TRACKING_SRC_APP_ID,
            `trackingSrcAppId - ("${srcAppId}") in input request is not numeric`);
        return ERROR_CODES.INVALID_APP_ID;
    }

    if (!appGuid || typeof appGuid !== 'string') {
        logCalWarning(TRACKING_NATIVE_APP_GUID,
            TRACKING_NATIVE_APP_GUID_EMPTY_OR_NULL_WARN_MESSAGE);
        return ERROR_CODES.EMPTY_APP_GUID;
    }

    if (appGuid.length !== GUID_LENGTH) {
        logCalWarning(TRACKING_NATIVE_APP_GUID,
            "trackingNativeAppGuid - (" + appGuid +
            ") length must be in " + GUID_LENGTH + ".");
        return ERROR_CODES.INVALID_APP_GUID_LENTGH;
    }

    if (!/^[A-Za-z0-9]+$/.test(appGuid)) {
        logCalWarning(TRACKING_NATIVE_APP_GUID,
            `trackingNativeAppGuid - ("${appGuid}") pattern must be of [A-Za-z0-9].`);
        return ERROR_CODES.INVALID_APP_GUID_PATTERN;
    }

    // implementation of validateGuid does not fail with any expected error
    const { appGuid: newGuid, loggerHMAC } = await validateGuid(webRequest);
    if (newGuid !== appGuid) {
        return { srcAppId, appGuid: newGuid, loggerHMAC };
    }
    return { srcAppId, appGuid, loggerHMAC };
}

async function getDMSVC() {
    return new Promise((resolve, reject) => {
        modConfig(module, (err, config) => {
            if (err) {
                reject(err);
            } else {
                resolve(config.get('tracking-ebay:hmacDmsSvc'));
            }
        });
    });
}

async function validateGuid(webRequest) {
    const query = webRequest.query;
    var appGuid = query[TRACKING_NATIVE_APP_GUID];
    const loggerHMAC = {};

    const dmsSvc = await getDMSVC();
    if (dmsSvc === true) { //config changes
        const appVer = query[NATIVE_APP_VER];
        const clientId = query[CLIENT_ID];
        const signature = query[SIGNATURE];
        const artifact = query[ARTIFACT];
        const signatureVersion = query[SIGNATURE_VERSION];

        if (appVer && clientId && signature && artifact && signatureVersion) {
            const validationRequest = {
                guid: appGuid,
                clientId: clientId,
                appVersion: appVer,
                signature: signature,
                artifact: artifact,
                signatureVersion: signatureVersion
            };
            loggerHMAC.hmacTestable = true;
            const response = await callGuidValidation(validationRequest)
                .catch(err => err);
            loggerHMAC.isValidatedGuid = true;
            // we do not through error to avoid handling catch case
            if (response instanceof Error) {
                loggerHMAC.isValidatedGuid = false;
                logger.warn('Failed to call guid validation service ', response);
            }
            //else
            if (response.statusCode !== 204) {
                loggerHMAC.isValidatedGuid = false;
                loggerHMAC.invalidPassedGuid = appGuid;
                appGuid = guidGenerator.nextPaddedGUID();
            }

        }
        else {
            logger.warn('One of the properties are missing for webView request: %s', [
                NATIVE_APP_VER,
                CLIENT_ID,
                SIGNATURE,
                ARTIFACT,
                SIGNATURE_VERSION,

            ].filter(name => query[name] === undefined));
        }
    }
    return { appGuid, loggerHMAC };
}
async function callGuidValidation(validationRequest) {
    return new Promise((resolve, reject) => {
        Provider.context({})
            .getClient('hmacDmsSvc')
            .post({
                body: JSON.stringify(validationRequest)

            })
            .end((error, response) => {
                if (error) {
                    reject(error);
                    return;
                }
                resolve(response);
            });
    });
}

module.exports.callGuidValidation = callGuidValidation;
module.exports.validateGuid = validateGuid;
module.exports.validateWebView = validateWebView;
module.exports.getDMSVC = getDMSVC;
module.exports.ERROR_CODES = ERROR_CODES;
module.exports.WEB_VIEW = WEB_VIEW;
module.exports.TRACKING_APP = TRACKING_APP;
module.exports.TRACKING_SRC_APP_ID = TRACKING_SRC_APP_ID;
module.exports.TRACKING_NATIVE_APP_GUID = TRACKING_NATIVE_APP_GUID;
module.exports.NATIVE_APP_VER = NATIVE_APP_VER;
module.exports.CLIENT_ID = CLIENT_ID;
module.exports.SIGNATURE = SIGNATURE;
module.exports.ARTIFACT = ARTIFACT;
module.exports.SIGNATURE_VERSION = SIGNATURE_VERSION;
