// @ebay/service-instruments-ebay/oauth.js
'use strict';

const logger = require('@ebay/logging-inc').logger('service-instruments-ebay/oauth');
const Objutil = require('objutil');

const errors = require('./lib/errors');
const RuntimeError = errors.RuntimeError;

let oldAuthModule;
const resolvers = {
    'user': function getUserToken(oauth, oauthConfig, options) {
        if (oauth.isSignIn()) {
            logger.info('Using getUserToken to get oauth token');
            return oauth[options.browser ? 'getPublicUserToken' :
                'getUserToken'].bind(oauth, oauthConfig['app-name'], oauthConfig.scope &&
                (oauthConfig.scope.user || oauthConfig.scope));
        }

        return new RuntimeError({
            message: 'User is not signed in, operation: getUserToken',
            code: 'UserNotSignedIn'
        });
    },

    'recognized-user': function getRecognizedUserToken(oauth, oauthConfig, options) {
        if (oauth.isRecognizedUser()) {
            logger.info('Using getRecognizedUserToken to get oauth token');

            const scopes = oauthConfig.scope &&
                (oauthConfig.scope.app || oauthConfig.scope) || undefined;

            if (oldAuthModule === undefined) {
                oldAuthModule = oauth.getRecognizedUserToken.toString()
                    .indexOf('function getRecognizedUserToken(callback)') !== -1;

                if (oldAuthModule) {
                    logger.error(new Error('Detected old oauth-ebay module version, please upgrade'));
                }
            }

            // check if oauth-ebay was upgraded
            if (options.browser) {
                return oldAuthModule ?
                    oauth.getPublicRecognizedUserToken.bind(oauth, oauthConfig['app-name']) :
                    oauth.getPublicRecognizedUserToken.bind(oauth, oauthConfig['app-name'], scopes);
            }

            return oldAuthModule ?
                oauth.getRecognizedUserToken.bind(oauth) :
                oauth.getRecognizedUserToken.bind(oauth, scopes);
        }

        return new RuntimeError({
            message: 'User is not recognized, operation: getUserToken',
            code: 'UserNotRecognized'
        });
    },

    'app': function getAppToken(oauth, oauthConfig, options) {
        logger.info('Using getAppToken to get oauth token');
        if (options.browser) {
            return oauth.getPublicAppToken.bind(oauth, oauthConfig['app-name'], oauthConfig.scope &&
                (oauthConfig.scope.app || oauthConfig.scope));
        }

        return oauth.getAppToken
            .bind(oauth, oauthConfig['app-name'], oauthConfig.scope &&
                (oauthConfig.scope.app || oauthConfig.scope), null, oauthConfig);
    }
};

module.exports = function oauth(pipe) {
    const app = pipe.context.app;
    const ebay = app && app.request && app.request.ebay;

    pipe.on('request', (request, next) => {
        logger.info('oauth handler', pipe.context.clientId);
        const options = request;
        const oauthConfig = options.oauth;
        if (!oauthConfig) {
            return next();
        }

        // add commerce OS headers
        if (ebay) {
            if (!ebay.oauth) {
                return pipe.throw(new RuntimeError(
                    'oauth requestContext was not found, please make sure you use oauth middleware'));
            }

            options.headers = Objutil.clone(options.headers) || {};

            // find what method to use
            const tokenOrder = oauthConfig.token || ['user', 'recognized-user', 'app'];
            // store first token as a type for downstream instruments to consume
            pipe.context.appTokenOnly = tokenOrder[0] === 'app';
            // get resolver
            const resolve = tokenOrder.reduce((memo, type) => {
                if (memo && !(memo instanceof Error)) {
                    return memo;
                }

                const resolver = resolvers[type];
                if (!resolver) {
                    throw new RuntimeError(`Unknown token type: ${type}`);
                }

                return resolver(ebay.oauth, oauthConfig, options);
            }, new RuntimeError('No resolvers provided'));

            // if we get error, return it
            const err = resolve;
            if (err instanceof Error) {
                if (options.tolerant) {
                    options.errors = options.errors || [];
                    options.errors.push(`Failed to get token due to ${err.message}`);
                    return next();
                }
                return pipe.throw(err);
            }

            // or call
            resolve((err, token) => {
                if (err) {
                    if (options.tolerant) {
                        logger.error('Failed to get token due to', err);
                        options.errors = options.errors || [];
                        options.errors.push(`Failed to get token, for more details see rlogid: ${ebay.getRlogId()}`);
                    }
                    else {
                        return pipe.throw(new RuntimeError({
                            message: 'Failed to get token',
                            error: err
                        }));
                    }
                }
                else {
                    options.headers.authorization = `Bearer ${token.value}`;
                }

                next();
            });
        }
        else {
            // eslint-disable-next-line max-len
            pipe.throw(new Error('oauth instrument has been requested, but ebay context is missing, make sure you have commons-ebay middlewares'));
        }
    });
};
