// @ebay/commons-ebay/lib/request-context/context.js
'use strict';

var Locale = require('../Locale');
var reqUtils = require('../request-utils');
var rlogidGenerator = require('../rlogid-generator');
var appContext = require('../app-context');
var pageUtils = require('../page-utils');
var commerceUtils = require('../cos/commerceos-utils');
var loggingInc = require('@ebay/logging-inc');
var eucTools = require('../cos/end-user-context');
var ucpTools = require('../cos/user-cultural-pref-context');
var shippingAwarenessTool = require('../cos/shipping-awareness');
var criTools = require('../cos/c-request-id');
var CTracking = require('../cos/c-tracking');
var GUIDGenerator = require('../guid-generator');
var utilsEbay = require('@ebay/utils-ebay');
const Edp = require('../cos/edp');
const deployEnv = require('@ebay/environment-ebay');

var guidGenerator = new GUIDGenerator();

var logger = loggingInc.logger('commons-ebay/ebay-request-context/context');

const BWECookielets = ['BANDWIDTH_EDGE_1', 'BANDWIDTH_EDGE_2', 'BANDWIDTH_EDGE_3', 'BANDWIDTH_EDGE_4'];
const BWOCookielets = ['BANDWIDTH_ORIGIN_1', 'BANDWIDTH_ORIGIN_2', 'BANDWIDTH_ORIGIN_3', 'BANDWIDTH_ORIGIN_4'];
//const BWBucket = {"LOW" : 1, "LOW_MEDIUM": 2, "MEDIUM_HIGH": 3, "HIGH": 4};

function getPort(req) {
    return req.app && req.app.get('port') ||
        req.socket && req.socket.localPort || 0;
}

function Context(req, res, options) {
    this.getRequest = function () {
        return req;
    };
    this.getResponse = function () {
        return res;
    };

    this.appContext = appContext;
    this.guidGenerator = guidGenerator;
    this.host = reqUtils.getServerName(req);
    this.clientIP = reqUtils.getClientIP(req);
    this.remoteAddrIP = reqUtils.getRemoteAddr(req);
    this._rlogid = rlogidGenerator.generate(req);
    // only allow isolate into the context if this is NOT a prod environment,
    if (deployEnv.isNotProd()) {
        const _isolate = reqUtils.getIsolate(req);
        // and only if it is a valid value (empty string, or 3-12 alphanumeric characters)
        if (
            _isolate === '' ||
            (_isolate && _isolate.match(/^[a-z0-9]{3,12}$/i))
        ) {
            this.isolate = _isolate;
        }
    }
    this.isIPFromAkamaiHeader = utilsEbay.isIPFromAkamaiHeader(req);

    // Try to be friendly to the V8 optimizer by always having the same properties
    this.appMeta = null;
    this.siteInfo = null;
    this.requestUrl = null;
    this.signInPageUrl = null;
    this.acceptLanguages = null;
    this.locale = null;
    this.cobrandId = 0;
    if (options && options.defaultCobrandId) {
        this.cobrandId = options.defaultCobrandId;
    }

    Object.defineProperty(this, 'locale', {
        get: function () {
            var locality = req.locality || {};
            var localeParts = (locality.locale || 'en-US').split('-');
            return new Locale(localeParts[0], localeParts[1]);
        }
    });
}

Context.prototype = {
    getEndUserContext: function getEndUserContext() {
        this.endUserContext = this.endUserContext || eucTools.create(this.getRequest());
        return this.endUserContext;
    },
    getEdp() {
        this.edp = this.edp || Edp.create(this.getRequest());
        return this.edp;
    },
    getShippingAwareness: function getShippingAwareness() {
        this.shippingAwareness = this.shippingAwareness || shippingAwarenessTool.create(this.getRequest());
        return this.shippingAwareness;
    },
    getUserCulturalPrefContext: function getUserCulturalPrefContext() {
        this.userCulturalPrefsContext = this.userCulturalPrefsContext || ucpTools.create(this.getRequest());
        return this.userCulturalPrefsContext;
    },
    getCorrelationRequestId: function getCorrelationRequestId() {
        this.corrRequestIdContext = this.corrRequestIdContext || criTools.create(this.getRequest());
        return this.corrRequestIdContext;
    },
    getCorrelationTracking: function getCorrelationTracking() {
        this.correlationTracking = this.correlationTracking || CTracking.create(this.getRequest());
        return this.correlationTracking;
    },
    getPartnerId: function getPartnerId() {
        return this.cobrandId;
    },
    getAreaId: function getAreaId() {
        return 1;
    },
    getCoBrandId: function getCoBrandId() {
        return 0;
    },
    isInternalHost: function isInternalHost() {
        if (!this._isInternalHost) {
            this._isInternalHost = reqUtils.isInternalHost(this.getRemoteAddr());
        }

        //If request is from Akamai, it should always external
       if(this.isIPFromAkamaiHeader && this._isInternalHost) {
            this._isInternalHost.isInternal = false;
            this._isInternalHost.hasUpdate = false;
        }
        return this._isInternalHost;
    },
    isInternalRequest: function isInternalRequest() {
        return (this.isInternalHost().isInternal === true);
    },
    isHostInNetwork: function isHostInNetwork() {
        return this.isInternalRequest();
    },

    getRlogId: function getRlogId() {
        return this._rlogid;
    },

    getClientIP: function getClientIP() {
        return this.clientIP;
    },
    getRemoteAddr: function getRemoteAddr() {
        return this.remoteAddrIP;
    },
    isSecure: function isSecure() {
        var req = this.getRequest();
        var proto = req.headers['x-forwarded-proto'] || req.protocol;
        var port = getPort(req);
        var host = req.get && req.get('Host');
        return proto === 'https' || port === 8082 || /:8082$/.test(host);
    },
    getSiteId: function getSiteId() {
        return reqUtils.getSiteId(this.getRequest());
    },
    getMarketplaceId: function getMarketplaceId() {
        return reqUtils.getMarketplaceId(this.getRequest());
    },
    getTerritoryId: function getTerritoryId() {
        return reqUtils.getTerritoryId(this.getRequest());
    },
    getGuid: function getGuid() {
        return this.getGuidContext().guid;
    },
    // @deprecated
    addGuid: function addGuid() {
        logger.warn(new Error('method addGuid is deprecated'));
        this.getGuid();
    },
    getMachineTagCookie: function getMachineTagCookie() {
        throw new Error('Not implemented');
    },
    isCookieSupportedEnabled: function isCookieSupportedEnabled() {
        throw new Error('Not implemented');
    },
    getPageId: function getPageId() {
        return pageUtils.getPageId(this.getRequest());
    },
    getPageName: function getPageName() {
        return pageUtils.getPageName(this.getRequest());
    },
    getPageNameFromRoute: function getPageNameFromRoute() {
        return pageUtils.getPageNameFromRoute(this.getRequest());
    },
    getExpSvcOperationId: function getExperienceService() {
        return pageUtils.getExperienceServiceOperation(this.getRequest());
    },
    getAcceptLanguages: function getAcceptLanguages() {
        return commerceUtils.getAcceptLanguages(this);
    },
    getSiteInfo: function getSiteInfo() {
        return this.siteInfo;
    },
    getServerName: function getServerName() {
        return this.host;
    },
    getLocale: function getLocale() {
        return this.locale;
    },
    logger: function logger(name) {
        return loggingInc.getLogger(name);
    },
    getTrackingGuid: function getTrackingGuid() {
        throw new Error('Not implemented');
    },
    getCorrelationGuid: function getCorrelationGuid() {
        throw new Error('Not implemented');
    },
    setCorrelationGuid: function setCorrelationGuid(guid) {
        throw new Error('Not implemented');
    },
    setSojGuid: function setSojGuid(guid) {
        throw new Error('Not implemented');
    },
    setMerchSojGuid: function setMerchSojGuid(guid) {
        throw new Error('Not implemented');
    },
    getRoverMediaplex: function getRoverMediaplex() {
        throw new Error('Not implemented');
    },
    getGuidContext: function getGuidContext(req) {
        var guidContext = this.guidContext;
        req = req || this.getRequest();

        if (!guidContext) {

            guidContext = {
                persistentGuidAbsent: false,
                bothGuidAbsent: false,
                sessionGuidAbsent: false,
                guid: null
            };

            var guid = this.getSojGuid();

            var hasSojournerGuid = false;
            if (guid) {
                hasSojournerGuid = true;
            } else {
                guid = this.getTrackingGuid();
            }

            if (!guid) {
                guidContext.persistentGuidAbsent = true;
            }

            var sessionGuid = this.getCorrelationGuid();

            if (guid && sessionGuid !== guid) {
                // Invalidate session guid if it doesn't match the other cookie
                sessionGuid = null;
            }

            var hasSojournerSessionGuid = false;
            if (sessionGuid) {
                hasSojournerSessionGuid = true;
                if (!guid) {
                    // Sync the persistent GUID with the session GUID
                    guid = sessionGuid;
                }
            } else {
                if (!guid) {
                    guidContext.bothGuidAbsent = true;
                }
                guidContext.sessionGuidAbsent = true;
            }

            if (!guid) {
                // check if A-pool provides one
                guid = req && req.headers && req.headers['X-eBay-New-GUID'.toLowerCase()];
            }

            // fallback to merch soj guid if exists
            if (!guid && !this.isService && this.getMerchSojGuid) {
                guid = this.getMerchSojGuid();
            }

            if (!guid) {
                guid = guidGenerator.nextPaddedGUID();
            }

            if (!hasSojournerGuid) {
                this.setSojGuid(guid);
            }
            if (!hasSojournerSessionGuid) {
                this.setCorrelationGuid(guid);
            }

            if (!this.isService && this.getMerchSojGuid() !== guid) {
                this.setMerchSojGuid(guid);
            }

            guidContext.guid = guid;

            this.guidContext = guidContext;
        }
        return guidContext;
    },

    getUserBandwidth: function() {
        return getDecimalFromBitFlags(this.getRequest(), BWECookielets);
    },

    getUserBandwidthToOrigin: function() {
        return getDecimalFromBitFlags(this.getRequest(), BWOCookielets);
    },

    isExpSvc: function isExpSvc(){
        return pageUtils.isExpSvc(this.getRequest());
    },

    isExpSvcModuleProvider: function isExpSvcModuleProvider() {
        return pageUtils.isExpSvcModuleProvider(this.getRequest());
    },

    getIsolate: function getIsolate() {
        return this.isolate;
    },

    setIsolateCookie: function setIsolateCookie() {
        // no-op by default, should be overriden in webcontext
        return;
    },

    buildOutboundContext(filter, useMarketplaceId){
        const ctx = {};
        const req = this.getRequest();
        if (req.headers && req.headers['trafficmirror.requestid']) {
            ctx['trafficmirror.requestid'] = req.headers['trafficmirror.requestid'];
        }
        if (req.headers && req.headers['x-ebay-c-trackable-id']) {
            ctx['x-ebay-c-trackable-id'] = req.headers['x-ebay-c-trackable-id'];
        }
        // this need to be rebuilt every time as some may change during request flow
        // we can later reset ctx in case of update
        ctx['X-EBAY-TERRITORY-ID'] = this.getTerritoryId();
        ctx['X-EBAY-C-MARKETPLACE-ID'] = this.getMarketplaceId(useMarketplaceId);
        this.getEndUserContext().update(ctx, filter);
        this.getUserCulturalPrefContext().update(ctx);
        this.getShippingAwareness().update(ctx);
        this.getEdp().update(ctx);
        ctx['Accept-Language'] = this.getAcceptLanguages().toString();

        // Propagate BOT header to ALL downstream request
        if(req.ebay && req.ebay.bot) {
            ctx['X-EBAY-C-BOT'] = encodeURIComponent(JSON.stringify(req.ebay.bot));
        }
        // Propagate Sec-GPC header to all downstream request,
        // https://jirap.corp.ebay.com/browse/NODE-4074
        if (req.headers && req.headers['sec-gpc']) {
            ctx['sec-gpc'] = req.headers['sec-gpc'];
        }

        // Propagate the X-EBAY-C-ISOLATE header to all downstream requests,
        // if it exists. Don't propagate if it is an empty string.
        const isolate = this.getIsolate();
        if (isolate) {
            ctx['X-EBAY-C-ISOLATE'] = isolate;
        }

        return ctx;
    }
};

class UserBandwidth {
    constructor(bucket) {
        this.bucket = bucket;
    }

    isLow() {
        return this.bucket >= 1 && this.bucket <=5;
    }

    isLowMedium() {
        return this.bucket >= 6 && this.bucket <=8;
    }

    isMediumHigh() {
        return this.bucket >= 9 && this.bucket <=12;
    }

    isHigh() {
        return this.bucket >= 13 && this.bucket <=14;
    }
}

//This method reads cookie flags as '1010' (binary format) and converts to decimal format as '10'
//Returns the bucket where bandwidth falls
function getDecimalFromBitFlags(req, bwCookies) {
    let cookies = req && req.ebay && req.ebay.cookies;
    if(!cookies) {
        return 0;
    }

    let bwFlags = bwCookies.reduce((accum, item) => {
            accum += cookies.hasFlag(item) ? 1 : 0;
            return accum;
        }, '');
    let decimalNum = parseInt(bwFlags, 2);

    return new UserBandwidth(decimalNum);
}

Object.defineProperty(Context.prototype, 'outboundContext', {
    get: function () {
        return this.buildOutboundContext();
    }
});

module.exports = Context;
