// @ebay/commons-ebay/lib/request-context/service-context.js
'use strict';

/**
 * Test end user context ip is returned.
*/
var util = require('util');

var extend = require('raptor-util/extend');
var userMixins = require('../user-mixins/headers');
var Context = require('./context');

function ServiceContext(req, res, options) {
    Context.call(this, req, res, options);
}

util.inherits(ServiceContext, Context);

var proto = ServiceContext.prototype;

proto.getSojGuid = function getSojGuid() {
    return this.getCorrelationTracking().guid;
};

proto.setSojGuid = function setSojGuid(guid) {
    this.getCorrelationTracking().guid = guid;
};

proto.getCorrelationGuid = function getCorrelationGuid() {
    return this.getCorrelationTracking().getCorrelationGuid();
};

proto.setCorrelationGuid = function setCorrelationGuid(guid) {
    if (!this.getCorrelationTracking().cguid) {
        this.getCorrelationTracking().cguid = guid;
    }
};

proto.getTrackingGuid = function getTrackingGuid() {
    return this.getCorrelationTracking().getTrackingGuid();
};

proto.getRoverMediaplex = function getRoverMediaplex() {
    return this.getCorrelationTracking().getRoverMediaplex();
};

proto.getMachineTagCookie = function getMachineTagCookie() {
    // return cookieUtils.getMachineTagCookie(this.getRequest());
};

proto.isCookieSupportedEnabled = function isCookieSupportedEnabled() {
        // return cookieUtils.isCookieSupportedEnabled(this.getRequest());
};

proto.getUserAgent = function getUserAgent() {
    return this.getEndUserContext().getUserAgent();
};

proto.getRemoteAddr = function getRemoteAddr() {
    return this.getEndUserContext().getIPAddress();
};

proto.isService = true;

// We add some additional methods related to reading user info from headers:
extend(ServiceContext.prototype, userMixins);

module.exports.ServiceContext = ServiceContext;
