// @ebay/commons-ebay/lib/request-context/web-context.js
'use strict';

var util = require('util');
var utilsEbay = require('@ebay/utils-ebay');
var reqUtils = require('../request-utils');

var extend = require('raptor-util/extend');
var userMixins = require('../user-mixins/cookies');
var Context = require('./context');

/**
 * The function extracts abd validate native app guid for webview mode
 */
function createWebViewContext(req) {
    const appGuid = req.query && req.query.trackingNativeAppGuid;

    if (appGuid && appGuid.length === 32 && /^[A-Za-z0-9]+$/.test(appGuid)) {
        return {
            appGuid
        };
    }

    return {};
}

function WebContext(req, res, options) {
    Context.call(this, req, res, options);
    this.remoteAddr = reqUtils.getRemoteAddr(req);
    this.webView = req.query && req.query.trackingApp === 'webview' ?
        createWebViewContext(req) : null;
    this._isolateCookieSet = false;
}

util.inherits(WebContext, Context);

var proto = WebContext.prototype;

proto.setIsolateCookie = function setIsolateCookie() {
    // protect against setting multiple isolate cookies
    if (this._isolateCookieSet) {
        return;
    }
    const req = this.getRequest();
    const res = this.getResponse();
    // if request or response is not available, then we can't set the cookie
    if (!req || !res || !res.setHeader) {
        return;
    }
    let hostHeader = req.headers && req.headers.host;
    // if we don't have a host header, then we can't set the cookie because we don't know the domain
    if (!hostHeader) {
        return;
    }
    let baseHost = utilsEbay.getBaseHost(hostHeader);
    // if we failed to get the base host, then we can't set the cookie
    if (!baseHost) {
        return;
    }
    const isolate = this.isolate;
    // isolate comes from parent Context constructor
    if (isolate === '') {
        // if isolate is empty, then we need to expire the cookie
        res.setHeader('Set-Cookie', `isolate=; Path=/; Domain=.${baseHost}; SameSite=none; Secure; Max-Age=0;`);
        this._isolateCookieSet = true;
    } else if (typeof isolate === 'string') {
        // if isolate exists, then we need to set the cookie
        res.setHeader('Set-Cookie', `isolate=${isolate}; Path=/; Domain=.${baseHost}; SameSite=none; Secure;`);
        this._isolateCookieSet = true;
    }
};

proto.getSojGuid = function getSojGuid() {
    return this.webView && this.webView.appGuid ||
        this.cookies.getCookieValue('SOJOURNER_GUID');
};

proto.setSojGuid = function setSojGuid(guid) {
    this.cookies.setCookieValue('SOJOURNER_GUID', guid);
};

proto.getMerchSojGuid = function getMerchSojGuid() {
    const req = this.getRequest();
    return this.webView && this.webView.appGuid ||
        this.cookies.getCookieValue('MERCH_SOJOURNER_GUID') ||
        req && req.headers && req.headers['x-ebay-merch-guid'];
};

proto.setMerchSojGuid = function setMerchSojGuid(guid) {
    this.cookies.setCookieValue('MERCH_SOJOURNER_GUID', guid);
};

proto.getCorrelationGuid = function getCorrelationGuid() {
    return this.webView && this.webView.appGuid ||
        this.cookies.getCookieValue('SOJOURNER_SESSION_GUID');
};

proto.getDCCorrelationGuid = function getDCCorrelationGuid(){
    return this.cookies.getCookieValue(this.cookies.Cookies.DC_CORRELATION_GUID);
};

proto.setCorrelationGuid = function setCorrelationGuid(guid) {
    this.cookies.setCookieValue('SOJOURNER_SESSION_GUID', guid);
};

proto.getTrackingGuid = function getTrackingGuid() {
    return this.cookies.getCookieValue('TRACKING_GUID');
};

proto.getRoverMediaplex = function getRoverMediaplex() {
    return this.cookies.getCookieValue('ROVER_MEDIAPLEX');
};

proto.getMachineTagCookie = function getMachineTagCookie() {
    var cookies = this.cookies;
    return cookies.getCookieValue('MACHINE_TAG');
};

proto.isCookieSupportedEnabled = function isCookieSupportedEnabled() {
    var cookies = this.cookies;
    if (cookies.getCookieValue('TESTBROWSER')) {
        return true;
    } else {
        return !!cookies.getCookieValue('SOJOURNER_SESSION_GUID');
    }
};

proto.getUserAgent = function getUserAgent() {
    return reqUtils.getUserAgent(this.getRequest());
};

proto.getRemoteAddr = function getRemoteAddr() {
    return this.remoteAddr;
};

proto.getClientHints = function getClientHints() {
    const req = this.getRequest();
    const cfg = req.clientHintsConfig;
    const chReqHeaders = cfg && cfg.getCHRequestHeaders() || [];
    const reqClientHintsMap = {};
    for(let elem of chReqHeaders) {
        if(req.headers[elem]) {
            reqClientHintsMap[elem] = req.headers[elem];
        }
    }
    return reqClientHintsMap;
};

proto.getMarketplaceId = function getMarketplaceId(commerceOsCompatible = false) {
    let mktId = Context.prototype.getMarketplaceId.call(this);
    if (!commerceOsCompatible) {
        return mktId;
    }
    const locale = this.getAcceptLanguages()[0];
    if (mktId === 'EBAY-FRCA' || mktId === 'EBAY-CA' && locale === 'fr-CA') {
        mktId = 'EBAY-CA';
    }
    else if (mktId === 'EBAY-ENCA') {
        mktId = 'EBAY-CA';
    }
    else if (/EBAY[-_]NLBE|EBAY[-_]BE[-_]NL/.test(mktId) || mktId === 'EBAY-BE' && locale === 'nl-BE') {
        mktId = 'EBAY-BE';
    }
    else if (/EBAY[-_]FRBE|EBAY[-_]BE[-_]FR/.test(mktId) || mktId === 'EBAY-BE' && locale === 'fr-BE') {
        mktId = 'EBAY-BE';
    }
    return mktId;
};

proto.isService = false;

// We add some additional methods related to reading user info from cookies:
extend(WebContext.prototype, userMixins);

module.exports.WebContext = WebContext;
