// @ebay/location-service-ebay/lib/LocationService.js
'use strict';

const { deprecate } = require('util');
const IP = require('ip');
var svcClient = require('@ebay/service-client-ebay');

const DEFAULTS = {
    alpha2ISOcountryCode: 'US',
    isoCountryCode2: 'US'
};

function svcClientGet(options, callback) {
    const client = (options.req ? svcClient.context(options.req) : svcClient).getClient('locationservice');
    return client.get(options.qs).path(options.path).headers(options.headers || {}).end(callback);
}

function createLocator(req) {
    return {
        /**
         * getLocationByIp:
         *      MarketPlace IPBased Service which takes IP Address as input and
         *      gives a json String as output
         * @param ip
         * @param headers
         * @param callback
         */
        getLocationByIp(ip, headers, callback) {
            // LBS Service returns 403 for Private IP's, which inturn marks down.
            // Hence the below check.
            if(IP.isPrivate(ip)) {
                return callback(null, DEFAULTS);
            }

            return svcClientGet({
                req,
                path: '/lookup',
                qs: {                
                    ipAddress: ip
                },
                headers: headers
            }, function (err, response) {
                if (err || !(response && response.body && response.body.allResults && response.body.allResults[0] && response.body.allResults[0].queryResult && response.body.allResults[0].queryResult[0] && response.body.allResults[0].httpStatus === 200)) {
                    err = err || new Error('fail to get locationByIP for ' + ip);
                    callback(err, null);
                } else {
                    let lbsResponse = response.body.allResults[0].queryResult[0];
                    //Added for backward compatible as alpha2ISOcountryCode is not available in v3
                    lbsResponse.alpha2ISOcountryCode = lbsResponse.isoCountryCode2; 
                    callback(err, lbsResponse);
                }
            });
        },

        /**
         * getLocationByPostalCode:
         *      Entry point for Zip based Location service that fetches
         *      the latitude and longitude for the zip centroid and
         *      other place related information
         * @param params {isoCountryCode2, postalCode}
         * @param headers
         * @param callback
         */
        getLocationByPostalCode(params, headers, callback) {        
            return svcClientGet({
                req,
                path: '/geocode',
                qs: params,
                headers: headers
            }, function (err, response) {
                if (err || !(response && response.body && response.body.allResults && response.body.allResults[0] && response.body.allResults[0].queryResult && response.body.allResults[0].queryResult[0] && response.body.allResults[0].httpStatus === 200)) {
                    err = err || new Error('fail to get locationByPostalCode for ' + params);
                    callback(err, null);
                } else {
                    let lbsResponse = response.body.allResults[0].queryResult[0];                  
                    callback(err, lbsResponse);
                }

            });
        }
    };
}

const deprecatedLocator = createLocator();

module.exports = {
    context(req) {
        return createLocator(req);
    },

    getLocationByIp: deprecate(deprecatedLocator.getLocationByIp,
        'getLocationByIp now requires context, please use new API "location.context(req).getLocationByIp()"'),
    getLocationByPostalCode: deprecate(deprecatedLocator.getLocationByPostalCode,
        'getLocationByPostalCode now requires context, please use new API "location.context(req).getLocationByPostalCode()"')
};
