// @ebay/commons-ebay/lib/header-multi-value.js
'use strict';

/**
 * Normalizes space encoding in values to handle double-encoding issues
 * between Raptor (form-URL encoding) and Node.js (percent encoding)
 * 
 * @param {string} value - The value to normalize
 * @param {string} key - The key name for context-specific normalization
 * @returns {string} - Normalized value
 */

var logger = require('@ebay/logging-inc').logger('commons-ebay/header-multi-value');
function normalizeSpaceEncoding(value, key) {
    if (typeof value !== 'string') {
        return value;
    }
    
    // Special handling for userAgent field to fix DDS library compatibility
    // Handle both case-sensitive and case-insensitive userAgent
    if (key === 'userAgent') {
        // Handle double-encoded spaces: %2B -> + -> space
        // This fixes the issue where Raptor encodes spaces as + and Node.js re-encodes + as %2B
        var normalized = value;
        
        // First, handle any %2B (encoded +) back to +
        normalized = normalized.replace(/%2[Bb]/g, '+');
        
        // Then normalize + to spaces (form-URL decoding)
        normalized = normalized.replace(/\+/g, ' ');
        
        return normalized;
    }
    
    return value;
}

function HeaderMultiValue(value, caseInsensitive, tolerant) {
    this.value = {};
    if (value && typeof value === 'object') {
        var self = this;
        Object.keys(value).forEach(function (key) {
            self.set(caseInsensitive ? key.toLowerCase() : key, value[key]);
        });
    }
    else if (value && typeof value === 'string') {
        var values = value.split(/ +, +| ?, +| +, ?| ?, ?/);

        this.value = values.reduce(function (memo, item) {
                var kvp = item.split('=');
                if (kvp.length < 2) {
                    if (tolerant) {
                        return memo;
                    }
                    logger.error('Header has invalid format, missing \'=\'', item);
                }
                
                var key = caseInsensitive ? kvp[0].toLowerCase() : kvp[0];
                var decodedValue = decodeURIComponent(kvp[1]);
                if (key === 'userAgent') {
                    var normalizedValue = normalizeSpaceEncoding(decodedValue, key);
                    memo[key] = normalizedValue;
                } else {
                    memo[key] = decodedValue;
                }
                return memo;
            }, {});
    }
}

HeaderMultiValue.prototype = {
    get: function (name) {
        return this.value[name];
    },

    set: function (name, value) {
        if (value !== undefined && value !== null) {
            this.value[name] = value;
        }
        return this;
    },

    names: function () {
        return Object.keys(this.value);
    },

    toString: function (filter) {
        var self = this;
        return this.names().reduce(function (memo, key) {
            if (!filter || filter(key)) {
                memo.push(key + '=' + encodeURIComponent(self.get(key)));
            }
            return memo;
        }, []).join(',');
    }
};

module.exports.parse = function (headerValue, caseInsensitive, tolerant) {
    return new HeaderMultiValue(headerValue, caseInsensitive, tolerant);
};

module.exports.create = function (kvps, caseInsensitive) {
    return new HeaderMultiValue(kvps, caseInsensitive);
};
