// @ebay/commons-ebay/lib/page-utils.js
'use strict';

var parseurl = require('parseurl');
var logger = require('@ebay/logging-inc').logger('commons-ebay/page-utils');

function getPageId(req) {
    var ebayRequestContext = req.ebay;
    var pageId = ebayRequestContext.pageId;
    if (!pageId) {
        var appMeta = ebayRequestContext.appMeta;
        var pageName = getPageName(req);
        pageId = appMeta && appMeta.getPageId(pageName);
        ebayRequestContext.pageId = pageId;
    }
    return pageId;
}

function getExperienceServiceOperation(req) {
    return req && req.ebay && req.ebay.expSvc && req.ebay.expSvc.operationId;
}

function getPageNameFromRoute(req) {
    var pageName;
    if (req) {
        var route = req.route;
        pageName = route && route.config && (route.config.pageName || route.config.pagename);
    }
    return pageName;
}

function getPageName(req) {
    var ebayRequestContext = req.ebay,
        pageName = ebayRequestContext && ebayRequestContext.pageName,
        appMeta = ebayRequestContext && ebayRequestContext.appMeta;

    if (pageName) {
        return pageName;
    }

    var route, reqUrl;

    if (req) {
        route = req.route;

        pageName = route && route.config && (route.config.pageName || route.config.pagename);

        if (!pageName) {
            reqUrl = parseurl(req).pathname;
            reqUrl = reqUrl.split('/');
            pageName = reqUrl.pop();
            if (!pageName) {
                pageName = reqUrl.pop();
            }
        }

        if (pageName) {
            var pageId = appMeta && appMeta.getPageId(pageName);
            if (!pageId) {
                if (route) {
                    logger.error('Failed to find pageId for page:', pageName, ', resetting it to DefaultPage', new Error('Page id is not found'));
                }

                // reset the pageName to DefaultPage if page is not registered
                pageName = 'DefaultPage';
            }
        }
    }

    if (!pageName) {
        pageName = 'DefaultPage';
    }

    ebayRequestContext && (ebayRequestContext.pageName = pageName);
    return pageName;
}

function isExpSvc(req) {
    let pageName = getPageName(req);

    if (!pageName){
        return false;
    }

    const pageMeta = req.ebay && req.ebay.appMeta && req.ebay.appMeta.getPageMeta(pageName);
    return pageMeta && pageMeta.paramsMap && pageMeta.paramsMap['ExpSvcRequest'] === 'true';
}

function isExpSvcModuleProvider(req) {
    let pageName = getPageName(req);

    if (!pageName){
        return false;
    }

    const pageMeta = req.ebay && req.ebay.appMeta && req.ebay.appMeta.getPageMeta(pageName);
    return pageMeta && pageMeta.paramsMap && pageMeta.paramsMap['ModuleProviderRequest'] === 'true';
}

module.exports = {
    isExpSvc: isExpSvc,
    getPageId: getPageId,
    getPageName: getPageName,
    getPageNameFromRoute: getPageNameFromRoute,
    isExpSvcModuleProvider: isExpSvcModuleProvider,
    getExperienceServiceOperation: getExperienceServiceOperation
};
