// @ebay/commons-ebay/lib/request-utils.js
'use strict';

var hostipChecker = require('@ebay/hostip-checker-ebay');
var X_EBAY_C_MARKETPLACE_ID = 'X-EBAY-C-MARKETPLACE-ID';
var X_EBAY_TERRITORY_ID = 'X-EBAY-TERRITORY-ID';
var X_EBAY_C_TERRITORY_ID = 'x-ebay-c-territory-id';
var appContext = require('./app-context');
var utilsebay = require('@ebay/utils-ebay');
let securityEbay = require('@ebay/security-ebay');

function getServerName(req) {
    return utilsebay.getServerName(req);
}

function getRequestURL(req) {
    var protocol,
        isSecure = req.ebay.isSecure();

    if (isSecure) {
        protocol = 'https';
    } else {
        protocol = req.protocol || 'http';
    }
    var postParams;
    var url = req.originalUrl || req.url || '/';

    var isTextHtml = req.headers && (!req.headers['content-type'] || req.headers['content-type'].toLowerCase().indexOf('text/html') >= 0);

    var post2Get = req.route && req.route.config && req.route.config.post2Get || false;

    var bodyParams = (isTextHtml || post2Get) && req.body && Object.getOwnPropertyNames(req.body) || [];

    if (bodyParams && bodyParams.length > 0) {
        postParams = securityEbay.maskObject(req.body);
        var queryparams = req.query ? Object.keys(req.query) : [];
        if (queryparams.length > 0) {
            url += "&" + postParams;
        } else {
            url += "?" + postParams;
        }
    }

    var hostName = req.headers.host || 'www.ebay.com';
    return protocol + '://' + hostName + url;
}

function getRemoteAddr(req) {
    return utilsebay.getRemoteAddr(req);
}

function getClientIP(req) {
    return utilsebay.getClientIP(req);
}

function isInternalHost(ip) {
    var checkIP = hostipChecker.checkIpSync(ip);
    return checkIP;
}

function getSiteId(req) {
    var siteInfo = req.ebay.siteInfo;
    return siteInfo && siteInfo['site-id'] || 0;
}

/*
 * Get it from siteInfo (kernel service)
 * else from headers
 * else default to EBAY-US
 */
function getMarketplaceId(req) {
    var siteInfo = req.ebay.siteInfo,
        marketplaceId;

    marketplaceId = req && req.headers && (req.headers[X_EBAY_C_MARKETPLACE_ID] ||
            req.headers[X_EBAY_C_MARKETPLACE_ID.toLowerCase()]) ||
        siteInfo && siteInfo['marketplace-id'] ||
        'EBAY-US';

    return marketplaceId;
}


function getTerritoryId(req) {
    var territoryId;

    territoryId = req && (req.headers &&
        (req.headers[X_EBAY_TERRITORY_ID] || req.headers[X_EBAY_TERRITORY_ID.toLowerCase()]) ||
        req.headers[X_EBAY_C_TERRITORY_ID] ||
        (req.locality && req.locality.country)) || 'US';

    return territoryId;
}

function getUserAgent(req) {
    var userAgent = req && req.headers && req.headers['user-agent'] || '';
    return userAgent;
}

function isExpSvcRoute(req) {
    return req && req.route && req.route.config && req.route.config.expSvcOperation;
}

function getExpSvcByRoute(req) {
    if (!isExpSvcRoute(req) || !appContext.defaultExpSvc) {
        return;
    }

    var expSvcOperationDef = req && req.route && req.route.config && req.route.config.expSvcOperation;
    var expSvcOperation = {};

    if (expSvcOperationDef) {
        if (typeof expSvcOperationDef === 'string' && expSvcOperationDef.indexOf(':') !== -1) {
            expSvcOperation.name = expSvcOperationDef.substring(0, expSvcOperationDef.indexOf(':'));
            expSvcOperation.operation = expSvcOperationDef.substring(expSvcOperationDef.indexOf(':') + 1);
        } else {
            expSvcOperation.name = appContext.defaultExpSvc;
            expSvcOperation.operation = expSvcOperationDef;
        }
    }
    return expSvcOperation;
}

function getMWebHost(req) {
    return (req.headers && req.headers.host) ? req.headers.host.replace(/(.+\.)?m\.((.+)?ebay\..+){1}/, '$1$2') : 'ebay.com';
}

function getHeaderValue(req, name) {
    return req.headers && (req.headers[name] || req.headers[name.toLowerCase()]);
}

function getIsolate(req) {
    // If the incoming request has an isolate querystring, use that.
    let isolate = req.query && req.query.isolate;
    if (typeof isolate === 'string') {
        return isolate;
    }

    // If the incoming request has an isolate header, use that.
    isolate = getHeaderValue(req, 'x-ebay-c-isolate');
    if (typeof isolate === 'string') {
        return isolate;
    }

    // If the incoming request has a cookie with an isolate value, use that.
    const cookie = getHeaderValue(req, 'cookie');
    isolate = cookie && cookie.match(/(?:^|;\s*)isolate=([^;]*)/);
    if (isolate && isolate.length === 2 && typeof isolate[1] === 'string') {
        return isolate[1];
    }

    // If there is an ISOLATE environment variable, use that.
    isolate = process && process.env && process.env.ISOLATE;
    if (typeof isolate === 'string') {
        return isolate;
    }

    return;
}

module.exports = {
    getServerName: getServerName,
    getRequestURL: getRequestURL,
    getRemoteAddr: getRemoteAddr,
    getClientIP: getClientIP,
    isInternalHost: isInternalHost,
    getUserAgent: getUserAgent,
    getSiteId: getSiteId,
    getMarketplaceId: getMarketplaceId,
    getTerritoryId: getTerritoryId,
    isExpSvcRoute: isExpSvcRoute,
    getExpSvcByRoute: getExpSvcByRoute,
    getMWebHost: getMWebHost,
    getHeaderValue: getHeaderValue,
    getIsolate: getIsolate
};
